<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en-US" xml:lang="en-US">
<head>
  <meta charset="utf-8" />
  <meta name="generator" content="pandoc" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes" />
  <meta name="author" content="Daniel S. Himmelstein" />
  <meta name="author" content="Michael Zietz" />
  <meta name="author" content="Vincent Rubinetti" />
  <meta name="author" content="Kyle Kloster" />
  <meta name="author" content="Benjamin J. Heil" />
  <meta name="author" content="Faisal Alquaddoomi" />
  <meta name="author" content="Dongbo Hu" />
  <meta name="author" content="David N. Nicholson" />
  <meta name="author" content="Yun Hao" />
  <meta name="author" content="Blair D. Sullivan" />
  <meta name="author" content="Michael W. Nagle" />
  <meta name="author" content="Casey S. Greene" />
  <meta name="dcterms.date" content="2023-07-07" />
  <meta name="keywords" content="hetnets, networks, connectivity, search, hetionet, knowledge graphs, algorithms, nodes, bioinformatics, biomedical informatics, path counts, matrix multiplication, DWPC" />
  <title>Hetnet connectivity search provides rapid insights into how two biomedical entities are related</title>
  <style>
    code{white-space: pre-wrap;}
    span.smallcaps{font-variant: small-caps;}
    div.columns{display: flex; gap: min(4vw, 1.5em);}
    div.column{flex: auto; overflow-x: auto;}
    div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
    ul.task-list{list-style: none;}
    ul.task-list li input[type="checkbox"] {
      width: 0.8em;
      margin: 0 0.8em 0.2em -1.6em;
      vertical-align: middle;
    }
    div.csl-bib-body { }
    div.csl-entry {
      clear: both;
    }
    .hanging div.csl-entry {
      margin-left:2em;
      text-indent:-2em;
    }
    div.csl-left-margin {
      min-width:2em;
      float:left;
    }
    div.csl-right-inline {
      margin-left:2em;
      padding-left:1em;
    }
    div.csl-indent {
      margin-left: 2em;
    }
  </style>
  <!--
  Manubot generated metadata rendered from header-includes-template.html.
  Suggest improvements at https://github.com/manubot/manubot/blob/main/manubot/process/header-includes-template.html
  -->
  <meta name="dc.format" content="text/html" />
  <meta property="og:type" content="article" />
  <meta name="dc.title" content="Hetnet connectivity search provides rapid insights into how two biomedical entities are related" />
  <meta name="citation_title" content="Hetnet connectivity search provides rapid insights into how two biomedical entities are related" />
  <meta property="og:title" content="Hetnet connectivity search provides rapid insights into how two biomedical entities are related" />
  <meta property="twitter:title" content="Hetnet connectivity search provides rapid insights into how two biomedical entities are related" />
  <meta name="dc.date" content="2023-07-07" />
  <meta name="citation_publication_date" content="2023-07-07" />
  <meta property="article:published_time" content="2023-07-07" />
  <meta name="dc.modified" content="2023-07-07T04:25:44+00:00" />
  <meta property="article:modified_time" content="2023-07-07T04:25:44+00:00" />
  <meta name="dc.language" content="en-US" />
  <meta name="citation_language" content="en-US" />
  <meta name="dc.relation.ispartof" content="Manubot" />
  <meta name="dc.publisher" content="Manubot" />
  <meta name="citation_journal_title" content="Manubot" />
  <meta name="citation_technical_report_institution" content="Manubot" />
  <meta name="citation_author" content="Daniel S. Himmelstein" />
  <meta name="citation_author_institution" content="Department of Systems Pharmacology and Translational Therapeutics, University of Pennsylvania, Philadelphia, PA 19104, USA" />
  <meta name="citation_author_institution" content="Related Sciences, Denver, CO 80202, USA" />
  <meta name="citation_author_orcid" content="0000-0002-3012-7446" />
  <meta name="twitter:creator" content="@dhimmel" />
  <meta name="citation_author" content="Michael Zietz" />
  <meta name="citation_author_institution" content="Department of Systems Pharmacology and Translational Therapeutics, University of Pennsylvania, Philadelphia, PA 19104, USA" />
  <meta name="citation_author_institution" content="Department of Biomedical Informatics, Columbia University, New York, NY 10032, USA" />
  <meta name="citation_author_orcid" content="0000-0003-0539-630X" />
  <meta name="twitter:creator" content="@ZietzMichael" />
  <meta name="citation_author" content="Vincent Rubinetti" />
  <meta name="citation_author_institution" content="Department of Systems Pharmacology and Translational Therapeutics, University of Pennsylvania, Philadelphia, PA 19104, USA" />
  <meta name="citation_author_institution" content="Center for Health AI, University of Colorado School of Medicine, Aurora, CO 80045, USA" />
  <meta name="citation_author_orcid" content="0000-0002-4655-3773" />
  <meta name="citation_author" content="Kyle Kloster" />
  <meta name="citation_author_institution" content="Carbon, Inc., Redwood City, CA 94063, USA" />
  <meta name="citation_author_institution" content="Department of Computer Science, North Carolina State University, Raleigh, NC 27606, USA" />
  <meta name="citation_author_orcid" content="0000-0001-5678-7197" />
  <meta name="twitter:creator" content="@kylekloster" />
  <meta name="citation_author" content="Benjamin J. Heil" />
  <meta name="citation_author_institution" content="Genomics and Computational Biology Graduate Group, Perelman School of Medicine, University of Pennsylvania, Philadelphia, PA 19104, USA" />
  <meta name="citation_author_orcid" content="0000-0002-2811-1031" />
  <meta name="twitter:creator" content="@autobencoder" />
  <meta name="citation_author" content="Faisal Alquaddoomi" />
  <meta name="citation_author_institution" content="Department of Biochemistry and Molecular Genetics, University of Colorado School of Medicine, Aurora, CO 80045, USA" />
  <meta name="citation_author_institution" content="Center for Health AI, University of Colorado School of Medicine, Aurora, CO 80045, USA" />
  <meta name="citation_author_orcid" content="0000-0003-4297-8747" />
  <meta name="citation_author" content="Dongbo Hu" />
  <meta name="citation_author_institution" content="Department of Pathology, Perelman School of Medicine University of Pennsylvania, Philadelphia, PA 19104, USA" />
  <meta name="citation_author_orcid" content="0000-0001-7870-8242" />
  <meta name="twitter:creator" content="@None" />
  <meta name="citation_author" content="David N. Nicholson" />
  <meta name="citation_author_institution" content="Department of Systems Pharmacology and Translational Therapeutics, University of Pennsylvania, Philadelphia, PA 19104, USA" />
  <meta name="citation_author_orcid" content="0000-0003-0002-5761" />
  <meta name="twitter:creator" content="@None" />
  <meta name="citation_author" content="Yun Hao" />
  <meta name="citation_author_institution" content="Genomics and Computational Biology Graduate Group, Perelman School of Medicine, University of Pennsylvania, Philadelphia, PA 19104, USA" />
  <meta name="citation_author_orcid" content="0000-0002-1684-0085" />
  <meta name="twitter:creator" content="@YhaoC" />
  <meta name="citation_author" content="Blair D. Sullivan" />
  <meta name="citation_author_institution" content="School of Computing, University of Utah, Salt Lake City, UT 84112, USA" />
  <meta name="citation_author_orcid" content="0000-0001-7720-6208" />
  <meta name="twitter:creator" content="@BlairDSullivan" />
  <meta name="citation_author" content="Michael W. Nagle" />
  <meta name="citation_author_institution" content="Integrative Biology, Internal Medicine Research Unit, Worldwide Research, Development, and Medicine, Pfizer Inc., Cambridge, MA 02139, USA" />
  <meta name="citation_author_institution" content="Human Biology Integration Foundation, Deep Human Biology Learning, Eisai Inc., Cambridge, MA 02140, USA" />
  <meta name="citation_author_orcid" content="0000-0002-4677-7582" />
  <meta name="twitter:creator" content="@mikenagle84" />
  <meta name="citation_author" content="Casey S. Greene" />
  <meta name="citation_author_institution" content="Department of Systems Pharmacology and Translational Therapeutics, University of Pennsylvania, Philadelphia, PA 19104, USA" />
  <meta name="citation_author_institution" content="Department of Biochemistry and Molecular Genetics, University of Colorado School of Medicine, Aurora, CO 80045, USA" />
  <meta name="citation_author_institution" content="Center for Health AI, University of Colorado School of Medicine, Aurora, CO 80045, USA" />
  <meta name="citation_author_orcid" content="0000-0001-8713-9213" />
  <meta name="twitter:creator" content="@GreeneScientist" />
  <link rel="canonical" href="https://greenelab.github.io/connectivity-search-manuscript/" />
  <meta property="og:url" content="https://greenelab.github.io/connectivity-search-manuscript/" />
  <meta property="twitter:url" content="https://greenelab.github.io/connectivity-search-manuscript/" />
  <meta name="citation_fulltext_html_url" content="https://greenelab.github.io/connectivity-search-manuscript/" />
  <meta name="citation_pdf_url" content="https://greenelab.github.io/connectivity-search-manuscript/manuscript.pdf" />
  <link rel="alternate" type="application/pdf" href="https://greenelab.github.io/connectivity-search-manuscript/manuscript.pdf" />
  <link rel="alternate" type="text/html" href="https://greenelab.github.io/connectivity-search-manuscript/v/918f6af1c06d5dc22a5502997955b48e44e0c2b8/" />
  <meta name="manubot_html_url_versioned" content="https://greenelab.github.io/connectivity-search-manuscript/v/918f6af1c06d5dc22a5502997955b48e44e0c2b8/" />
  <meta name="manubot_pdf_url_versioned" content="https://greenelab.github.io/connectivity-search-manuscript/v/918f6af1c06d5dc22a5502997955b48e44e0c2b8/manuscript.pdf" />
  <meta property="og:type" content="article" />
  <meta property="twitter:card" content="summary_large_image" />
  <meta property="og:image" content="https://github.com/hetio/het.io/raw/e1ca4fd591e0aa01a3767bbf5597a910528f6f86/explore/connectivity-search.png" />
  <meta property="twitter:image" content="https://github.com/hetio/het.io/raw/e1ca4fd591e0aa01a3767bbf5597a910528f6f86/explore/connectivity-search.png" />
  <link rel="icon" type="image/png" sizes="192x192" href="https://manubot.org/favicon-192x192.png" />
  <link rel="mask-icon" href="https://manubot.org/safari-pinned-tab.svg" color="#ad1457" />
  <meta name="theme-color" content="#ad1457" />
  <!-- end Manubot generated metadata -->
  <!--[if lt IE 9]>
    <script src="//cdnjs.cloudflare.com/ajax/libs/html5shiv/3.7.3/html5shiv-printshiv.min.js"></script>
  <![endif]-->
</head>
<body>
<header id="title-block-header">
<h1 class="title">Hetnet connectivity search provides rapid insights into how two biomedical entities are related</h1>
</header>
<p><em>A DOI-citable version of this manuscript is available at <a href="https://doi.org/10.1101/2023.01.05.522941" class="uri">https://doi.org/10.1101/2023.01.05.522941</a></em>.</p>
<p><small><em>
This manuscript
(<a href="https://greenelab.github.io/connectivity-search-manuscript/v/918f6af1c06d5dc22a5502997955b48e44e0c2b8/">permalink</a>)
was automatically generated
from <a href="https://github.com/greenelab/connectivity-search-manuscript/tree/918f6af1c06d5dc22a5502997955b48e44e0c2b8">greenelab/connectivity-search-manuscript@918f6af</a>
on July 7, 2023.
</em></small></p>
<h2 id="authors">Authors</h2>
<!-- include the Font Awesome library, per: https://fontawesome.com/start -->
<!-- <link rel="stylesheet" href="https://use.fontawesome.com/releases/v5.7.2/css/all.css"> -->
<ul>
<li><p><strong>Daniel S. Himmelstein</strong>
<br>
<img src="images/orcid.svg" class="inline_icon" width="16" height="16" alt="ORCID icon" />
<a href="https://orcid.org/0000-0002-3012-7446">0000-0002-3012-7446</a>
· <img src="images/github.svg" class="inline_icon" width="16" height="16" alt="GitHub icon" />
<a href="https://github.com/dhimmel">dhimmel</a>
· <img src="images/twitter.svg" class="inline_icon" width="16" height="16" alt="Twitter icon" />
<a href="https://twitter.com/dhimmel">dhimmel</a>
<br>
<small>
Department of Systems Pharmacology and Translational Therapeutics, University of Pennsylvania, Philadelphia, PA 19104, USA; Related Sciences, Denver, CO 80202, USA
· Funded by The Gordon and Betty Moore Foundation (GBMF4552); Pfizer Worldwide Research, Development, and Medical
</small></p></li>
<li><p><strong>Michael Zietz</strong>
<br>
<img src="images/orcid.svg" class="inline_icon" width="16" height="16" alt="ORCID icon" />
<a href="https://orcid.org/0000-0003-0539-630X">0000-0003-0539-630X</a>
· <img src="images/github.svg" class="inline_icon" width="16" height="16" alt="GitHub icon" />
<a href="https://github.com/zietzm">zietzm</a>
· <img src="images/twitter.svg" class="inline_icon" width="16" height="16" alt="Twitter icon" />
<a href="https://twitter.com/ZietzMichael">ZietzMichael</a>
<br>
<small>
Department of Systems Pharmacology and Translational Therapeutics, University of Pennsylvania, Philadelphia, PA 19104, USA; Department of Biomedical Informatics, Columbia University, New York, NY 10032, USA
</small></p></li>
<li><p><strong>Vincent Rubinetti</strong>
<br>
<img src="images/orcid.svg" class="inline_icon" width="16" height="16" alt="ORCID icon" />
<a href="https://orcid.org/0000-0002-4655-3773">0000-0002-4655-3773</a>
· <img src="images/github.svg" class="inline_icon" width="16" height="16" alt="GitHub icon" />
<a href="https://github.com/vincerubinetti">vincerubinetti</a>
<br>
<small>
Department of Systems Pharmacology and Translational Therapeutics, University of Pennsylvania, Philadelphia, PA 19104, USA; Center for Health AI, University of Colorado School of Medicine, Aurora, CO 80045, USA
</small></p></li>
<li><p><strong>Kyle Kloster</strong>
<br>
<img src="images/orcid.svg" class="inline_icon" width="16" height="16" alt="ORCID icon" />
<a href="https://orcid.org/0000-0001-5678-7197">0000-0001-5678-7197</a>
· <img src="images/github.svg" class="inline_icon" width="16" height="16" alt="GitHub icon" />
<a href="https://github.com/kkloste">kkloste</a>
· <img src="images/twitter.svg" class="inline_icon" width="16" height="16" alt="Twitter icon" />
<a href="https://twitter.com/kylekloster">kylekloster</a>
<br>
<small>
Carbon, Inc., Redwood City, CA 94063, USA; Department of Computer Science, North Carolina State University, Raleigh, NC 27606, USA
· Funded by The Gordon and Betty Moore Foundation (GBMF4560)
</small></p></li>
<li><p><strong>Benjamin J. Heil</strong>
<br>
<img src="images/orcid.svg" class="inline_icon" width="16" height="16" alt="ORCID icon" />
<a href="https://orcid.org/0000-0002-2811-1031">0000-0002-2811-1031</a>
· <img src="images/github.svg" class="inline_icon" width="16" height="16" alt="GitHub icon" />
<a href="https://github.com/ben-heil">ben-heil</a>
· <img src="images/twitter.svg" class="inline_icon" width="16" height="16" alt="Twitter icon" />
<a href="https://twitter.com/autobencoder">autobencoder</a>
<br>
<small>
Genomics and Computational Biology Graduate Group, Perelman School of Medicine, University of Pennsylvania, Philadelphia, PA 19104, USA
· Funded by The Gordon and Betty Moore Foundation (GBMF4552)
</small></p></li>
<li><p><strong>Faisal Alquaddoomi</strong>
<br>
<img src="images/orcid.svg" class="inline_icon" width="16" height="16" alt="ORCID icon" />
<a href="https://orcid.org/0000-0003-4297-8747">0000-0003-4297-8747</a>
· <img src="images/github.svg" class="inline_icon" width="16" height="16" alt="GitHub icon" />
<a href="https://github.com/falquaddoomi">falquaddoomi</a>
<br>
<small>
Department of Biochemistry and Molecular Genetics, University of Colorado School of Medicine, Aurora, CO 80045, USA; Center for Health AI, University of Colorado School of Medicine, Aurora, CO 80045, USA
</small></p></li>
<li><p><strong>Dongbo Hu</strong>
<br>
<img src="images/orcid.svg" class="inline_icon" width="16" height="16" alt="ORCID icon" />
<a href="https://orcid.org/0000-0001-7870-8242">0000-0001-7870-8242</a>
· <img src="images/github.svg" class="inline_icon" width="16" height="16" alt="GitHub icon" />
<a href="https://github.com/dongbohu">dongbohu</a>
<br>
<small>
Department of Pathology, Perelman School of Medicine University of Pennsylvania, Philadelphia, PA 19104, USA
· Funded by The Gordon and Betty Moore Foundation (GBMF4552)
</small></p></li>
<li><p><strong>David N. Nicholson</strong>
<br>
<img src="images/orcid.svg" class="inline_icon" width="16" height="16" alt="ORCID icon" />
<a href="https://orcid.org/0000-0003-0002-5761">0000-0003-0002-5761</a>
· <img src="images/github.svg" class="inline_icon" width="16" height="16" alt="GitHub icon" />
<a href="https://github.com/danich1">danich1</a>
<br>
<small>
Department of Systems Pharmacology and Translational Therapeutics, University of Pennsylvania, Philadelphia, PA 19104, USA
· Funded by The Gordon and Betty Moore Foundation (GBMF4552); The National Institutes of Health (T32 HG000046)
</small></p></li>
<li><p><strong>Yun Hao</strong>
<br>
<img src="images/orcid.svg" class="inline_icon" width="16" height="16" alt="ORCID icon" />
<a href="https://orcid.org/0000-0002-1684-0085">0000-0002-1684-0085</a>
· <img src="images/github.svg" class="inline_icon" width="16" height="16" alt="GitHub icon" />
<a href="https://github.com/yhao-compbio">yhao-compbio</a>
· <img src="images/twitter.svg" class="inline_icon" width="16" height="16" alt="Twitter icon" />
<a href="https://twitter.com/YhaoC">YhaoC</a>
<br>
<small>
Genomics and Computational Biology Graduate Group, Perelman School of Medicine, University of Pennsylvania, Philadelphia, PA 19104, USA
</small></p></li>
<li><p><strong>Blair D. Sullivan</strong>
<br>
<img src="images/orcid.svg" class="inline_icon" width="16" height="16" alt="ORCID icon" />
<a href="https://orcid.org/0000-0001-7720-6208">0000-0001-7720-6208</a>
· <img src="images/github.svg" class="inline_icon" width="16" height="16" alt="GitHub icon" />
<a href="https://github.com/bdsullivan">bdsullivan</a>
· <img src="images/twitter.svg" class="inline_icon" width="16" height="16" alt="Twitter icon" />
<a href="https://twitter.com/BlairDSullivan">BlairDSullivan</a>
<br>
<small>
School of Computing, University of Utah, Salt Lake City, UT 84112, USA
</small></p></li>
<li><p><strong>Michael W. Nagle</strong>
<br>
<img src="images/orcid.svg" class="inline_icon" width="16" height="16" alt="ORCID icon" />
<a href="https://orcid.org/0000-0002-4677-7582">0000-0002-4677-7582</a>
· <img src="images/github.svg" class="inline_icon" width="16" height="16" alt="GitHub icon" />
<a href="https://github.com/naglem">naglem</a>
· <img src="images/twitter.svg" class="inline_icon" width="16" height="16" alt="Twitter icon" />
<a href="https://twitter.com/mikenagle84">mikenagle84</a>
<br>
<small>
Integrative Biology, Internal Medicine Research Unit, Worldwide Research, Development, and Medicine, Pfizer Inc., Cambridge, MA 02139, USA; Human Biology Integration Foundation, Deep Human Biology Learning, Eisai Inc., Cambridge, MA 02140, USA
</small></p></li>
<li><p><strong>Casey S. Greene</strong>
<sup><a href="#correspondence">✉</a></sup><br>
<img src="images/orcid.svg" class="inline_icon" width="16" height="16" alt="ORCID icon" />
<a href="https://orcid.org/0000-0001-8713-9213">0000-0001-8713-9213</a>
· <img src="images/github.svg" class="inline_icon" width="16" height="16" alt="GitHub icon" />
<a href="https://github.com/cgreene">cgreene</a>
· <img src="images/twitter.svg" class="inline_icon" width="16" height="16" alt="Twitter icon" />
<a href="https://twitter.com/GreeneScientist">GreeneScientist</a>
<br>
<small>
Department of Systems Pharmacology and Translational Therapeutics, University of Pennsylvania, Philadelphia, PA 19104, USA; Department of Biochemistry and Molecular Genetics, University of Colorado School of Medicine, Aurora, CO 80045, USA; Center for Health AI, University of Colorado School of Medicine, Aurora, CO 80045, USA
· Funded by the National Human Genome Research Institute (R01 HG010067); the National Cancer Institute (R01 CA237170); the Gordon and Betty Moore Foundation (GBMF 4552); Pfizer Worldwide Research, Development, and Medical; the Eunice Kennedy Shriver National Institute of Child Health and Human Development (R01 HD109765)
</small></p></li>
</ul>
<div id="correspondence">
<p>✉ — Correspondence possible via <a href="https://github.com/greenelab/connectivity-search-manuscript/issues">GitHub Issues</a>
or email to
Casey S. Greene &lt;casey.s.greene@cuanschutz.edu&gt;.</p>
</div>
<h2 class="page_break_before" id="abstract">Abstract</h2>
<p><strong>Background</strong>
Hetnets, short for “heterogeneous networks”, contain multiple node and relationship types and offer a way to encode biomedical knowledge.
One such example, Hetionet connects 11 types of nodes
— including genes, diseases, drugs, pathways, and anatomical structures
— with over 2 million edges of 24 types.
Previous work has demonstrated that supervised machine learning methods applied to such networks can identify drug repurposing opportunities.
However, a training set of known relationships does not exist for many types of node pairs,
even when it would be useful to examine how nodes of those types are meaningfully connected.
For example, users may be curious about not only how metformin is related to breast cancer
but also how a given gene might be involved in insomnia.</p>
<p><strong>Findings</strong>
We developed a new procedure, termed hetnet connectivity search, that proposes important paths between any two nodes without requiring a supervised gold standard.
The algorithm behind connectivity search identifies types of paths that occur more frequently than would be expected by chance (based on node degree alone).
Several optimizations were required to precompute significant instances of node connectivity at the scale of large knowledge graphs.</p>
<p><strong>Conclusion</strong>
We implemented the method on Hetionet and provide an online interface at <a href="https://het.io/search" class="uri">https://het.io/search</a>.
We provide an open-source implementation of these methods in our new Python package named <a href="https://github.com/hetio/hetmatpy" title="Python package for matrix storage and operations on hetnets">hetmatpy</a>.</p>
<h2 class="page_break_before" id="introduction">Introduction</h2>
<p>A <em>network</em> (also known as a <a href="https://en.wikipedia.org/wiki/Graph_(discrete_mathematics)">graph</a>) is a conceptual representation of a group of entities — called <em>nodes</em> — and the relationships between them — called <em>edges</em>.
Typically, a network has only one type of node and one type of edge.
However, in many cases, it is necessary to be able to distinguish between different types of entities and relationships.</p>
<p>A <em>hetnet</em> (short for <strong>het</strong>erogeneous information <strong>net</strong>work <span class="citation" data-cites="QWTAHxFj">[<a href="#ref-QWTAHxFj" role="doc-biblioref">1</a>]</span>) is a network where nodes and edges have type.
The ability to differentiate between different types of entities and relationships allows a hetnet to describe more complex data accurately.
Hetnets are particularly useful in biomedicine,
where it is important to capture the conceptual distinctions between various entities,
such as genes and diseases, and linkages, such as upregulation and binding.</p>
<p>The types of nodes and edges in a hetnet are defined by a schema, referred to as a metagraph.
The metagraph consists of metanodes (types of nodes) and metaedges (types of edges).
Note that the prefix <em>meta</em> refers to the type (e.g. compound),
as opposed to a specific node/edge/path itself (e.g. acetaminophen).</p>
<p>One such network is Hetionet, which provides a foundation for building hetnet applications.
It unifies data from several different, disparate sources into a single, comprehensive, accessible, common-format network.
The database is publicly accessible without login at <a href="https://neo4j.het.io" class="uri">https://neo4j.het.io</a>.
The Neo4j graph database enables querying Hetionet using the Cypher language,
which was designed to interact with networks where nodes and edges have both types and properties.</p>
<p>The initial application of Hetionet, named Project Rephetio, focused on drug repurposing <span class="citation" data-cites="O21tn8vf">[<a href="#ref-O21tn8vf" role="doc-biblioref">2</a>]</span>.
The authors <a href="https://het.io/repurpose/">predicted</a> the probability of drug efficacy for 209,168 compound–disease pairs.
A supervised machine learning approach identified types of paths that occur more or less frequently between known treatments than non-treatments (Figure <a href="#fig:rephetio">1</a>B).
To train the model, the authors created <a href="https://doi.org/10.6084/m9.figshare.3103054">PharmacotherapyDB</a>,
a physician-curated catalog of 755 disease-modifying treatments <span class="citation" data-cites="5t7pfaPv">[<a href="#ref-5t7pfaPv" role="doc-biblioref">3</a>]</span>.</p>
<div id="fig:rephetio" class="fignos">
<figure>
<img src="https://github.com/greenelab/connectivity-search-manuscript/raw/f98c3470a8bf8f40f5f6aed2794c6ea66b93b14b/content/media/rephetio/metagraph-and-features.png" class="white" style="width:100.0%" alt="Figure 1: A. Hetionet v1.0 metagraph. The types of nodes and edges in Hetionet. B. Supervised machine learning approach from Project Rephetio. This figure visualizes the feature matrix used by Project Rephetio to make supervised predictions. Each row represents a compound–disease pair. The bottom half of rows correspond to known treatments (i.e. positives), while the top half correspond to non-treatments (i.e. negatives under a closed-world assumption, not known to be treatments in PharmacotherapyDB). Here, an equal number of treatments and non-treatments are shown, but in reality the problem is heavily imbalanced. Project Rephetio scaled models to assume a positive prevalence of 0.36% [2,4]. Each column represents a metapath, labeled with its abbreviation. Feature values are degree-weighted path counts (abbreviated DWPCs, transformed and standardized), which assess the connectivity along the specified metapath between the specific compound and disease. Green values indicate above-average connectivity, whereas blue values indicate below-average connectivity. In general, positives have greater connectivity for the selected metapaths than negatives. Rephetio used a logistic regression model to learn the effect of each type of connectivity (feature) on the likelihood that a compound treats a disease. The model predicts whether a compound–disease pair is a treatment based on its features, but requires supervision in the form of known treatments. " />
<figcaption aria-hidden="true"><span>Figure 1:</span> <strong>A. Hetionet v1.0 metagraph.</strong>
The types of nodes and edges in Hetionet.<br />
<strong>B. Supervised machine learning approach from Project Rephetio.</strong>
This figure visualizes the feature matrix used by Project Rephetio to make supervised predictions.
Each row represents a compound–disease pair.
The bottom half of rows correspond to known treatments (i.e. positives),
while the top half correspond to non-treatments
(i.e. negatives under a <em>closed-world assumption</em>, not known to be treatments in PharmacotherapyDB).
Here, an equal number of treatments and non-treatments are shown, but in reality the problem is heavily imbalanced.
Project Rephetio scaled models to assume a positive prevalence of 0.36% <span class="citation" data-cites="O21tn8vf vvdoVrc6">[<a href="#ref-O21tn8vf" role="doc-biblioref">2</a>,<a href="#ref-vvdoVrc6" role="doc-biblioref">4</a>]</span>.
Each column represents a metapath, labeled with its abbreviation.<br />
Feature values are degree-weighted path counts (abbreviated DWPCs, transformed and standardized),
which assess the connectivity along the specified metapath between the specific compound and disease.
Green values indicate above-average connectivity,
whereas blue values indicate below-average connectivity.
In general, positives have greater connectivity for the selected metapaths than negatives.
Rephetio used a logistic regression model to learn the effect of each type of connectivity (feature) on the likelihood that a compound treats a disease.
The model predicts whether a compound–disease pair is a treatment based on its features,
but requires supervision in the form of known treatments.
<!-- info on this figure's creation in https://github.com/greenelab/connectivity-search-manuscript/issues/11 --></figcaption>
</figure>
</div>
<p>Project Rephetio successfully predicted treatments,
including those under investigation by clinical trial.
However, two challenges limit the applicability of Rephetio.
First, Rephetio required known labels (i.e. treatment status) to train a model.
Hence, the approach cannot be applied to domains where training labels do not exist.
Second, the degree-weighted path count (DWPC) metric used to assess connectivity is sensitive to node degree.
The Rephetio approach was incapable of detecting whether a high DWPC score indicated meaningful connectivity above the level expected by the background network degrees.
Here we develop Hetnet connectivity search, which defines a null distribution for DWPCs that accounts for degree and enables detecting meaningful hetnet connectivity without training labels.</p>
<!--
Issues related to other works and important references:
https://github.com/greenelab/connectivity-search-analyses/issues/56
https://github.com/greenelab/connectivity-search-manuscript/issues/28
-->
<p>Existing research into methods for determining whether two nodes are related primarily focuses on homogeneous networks (without type).
Early approaches detected related nodes by measuring neighborhood overlap or path similarity between two nodes <span class="citation" data-cites="EfWvuSjX 7iZtCGcX">[<a href="#ref-EfWvuSjX" role="doc-biblioref">5</a>,<a href="#ref-7iZtCGcX" role="doc-biblioref">6</a>]</span>.
These approaches predicted node relatedness with success.
However, they are challenging to scale as a network grows in size or semantic richness (i.e. type) <span class="citation" data-cites="EfWvuSjX">[<a href="#ref-EfWvuSjX" role="doc-biblioref">5</a>]</span>.</p>
<p>More recently, focus has shifted to graph embeddings to determine if two nodes are related,
specifically in the context of knowledge graphs, which are often semantically rich and include type
<span class="citation" data-cites="otY29wFV NnOS86ev GYLacVTV 3gzmoW2V zHcAMCig">[<a href="#ref-otY29wFV" role="doc-biblioref">7</a>,<a href="#ref-NnOS86ev" role="doc-biblioref">8</a>,<a href="#ref-GYLacVTV" role="doc-biblioref">9</a>,<a href="#ref-3gzmoW2V" role="doc-biblioref">10</a>,<a href="#ref-zHcAMCig" role="doc-biblioref">11</a>]</span>.
These types of methods involve mapping nodes and sometimes edges to dense vectors via
neural network models <span class="citation" data-cites="U1Dfux5r 1AeZs6xaT jyhFBVzQ">[<a href="#ref-U1Dfux5r" role="doc-biblioref">12</a>,<a href="#ref-1AeZs6xaT" role="doc-biblioref">13</a>,<a href="#ref-jyhFBVzQ" role="doc-biblioref">14</a>]</span>,
matrix factorization <span class="citation" data-cites="xq3Nsf5M 1EeE9qtFA">[<a href="#ref-xq3Nsf5M" role="doc-biblioref">15</a>,<a href="#ref-1EeE9qtFA" role="doc-biblioref">16</a>]</span>,
or translational distance models <span class="citation" data-cites="cnCLKeU9">[<a href="#ref-cnCLKeU9" role="doc-biblioref">17</a>]</span>.
Bioteque creates node embeddings from the bipartite network of DWPCs for a given metapath <span class="citation" data-cites="vhsYX2y">[<a href="#ref-vhsYX2y" role="doc-biblioref">18</a>]</span>.
Once these dense vectors have been produced, quantitative scores that measure node relatedness can be generated via a machine learning model <span class="citation" data-cites="NnOS86ev taI1UUAE 17ioZ6znu">[<a href="#ref-NnOS86ev" role="doc-biblioref">8</a>,<a href="#ref-taI1UUAE" role="doc-biblioref">19</a>,<a href="#ref-17ioZ6znu" role="doc-biblioref">20</a>]</span> or by selected similarity metrics <span class="citation" data-cites="otY29wFV 1D9FTzRBg 6PISrkV5 GYLacVTV 19E33rJiu">[<a href="#ref-otY29wFV" role="doc-biblioref">7</a>,<a href="#ref-GYLacVTV" role="doc-biblioref">9</a>,<a href="#ref-1D9FTzRBg" role="doc-biblioref">21</a>,<a href="#ref-6PISrkV5" role="doc-biblioref">22</a>,<a href="#ref-19E33rJiu" role="doc-biblioref">23</a>]</span>.
These approaches have been quite successful in determining node relatedness.
Yet, they only state <em>whether</em> two nodes are related and fail to explain <em>why</em> two nodes are related.</p>
<p>Explaining why two nodes are related is a nontrivial task because approaches must output more than a simple similarity score.
The first group of approaches output a list of ranked paths that are most relevant between two nodes <span class="citation" data-cites="Dbie18xe GPBbIfq9 L3DO8oq1">[<a href="#ref-Dbie18xe" role="doc-biblioref">24</a>,<a href="#ref-GPBbIfq9" role="doc-biblioref">25</a>,<a href="#ref-L3DO8oq1" role="doc-biblioref">26</a>]</span>.
For example, the FAIRY framework explains for why items appear on a user’s social media feed based on a network of users and content classes (e.g. categories, user posts, songs) <span class="citation" data-cites="GPBbIfq9">[<a href="#ref-GPBbIfq9" role="doc-biblioref">25</a>]</span>.
ESPRESSO explains how two sets of nodes are related by returning subgraphs <span class="citation" data-cites="Hf24Hzrl">[<a href="#ref-Hf24Hzrl" role="doc-biblioref">27</a>]</span>.
Other approaches such as MetaExp return important metapaths rather than paths,
but require some form of supervision <span class="citation" data-cites="19H7j1WBd Vz9QIDKV">[<a href="#ref-19H7j1WBd" role="doc-biblioref">28</a>,<a href="#ref-Vz9QIDKV" role="doc-biblioref">29</a>]</span>.</p>
<p>MechRepoNet is a hetnet containing 250,035 nodes across 9 metanodes and 9,652,116 edges across 68 metaedges <span class="citation" data-cites="KQNSDt1L">[<a href="#ref-KQNSDt1L" role="doc-biblioref">30</a>]</span>.
The study trained a model using DWPCs as features to predict <em>Compound–treats–Disease</em> relationships,
which was able to select 89 metapaths with positive regression coefficients.
The authors also created <a href="https://sulab.github.io/DrugMechDB/">DrugMechDB</a> with a curated set of paths capturing known mechanisms of action for 123 compound–disease pairs <span class="citation" data-cites="KQNSDt1L">[<a href="#ref-KQNSDt1L" role="doc-biblioref">30</a>]</span>.
Metapath coefficients were used to rank paths, using DrugMechDB as validation.
The method generally performed well, although interpretability was challenging when “hundreds, or thousands of paths ranked above the mechanistic path in DrugMechDB” <span class="citation" data-cites="KQNSDt1L">[<a href="#ref-KQNSDt1L" role="doc-biblioref">30</a>]</span>.
To address this issue, the study explored additional path filters, such as filtering for paths that traverse known drug targets, and dimensionality reduction by aggregating paths across intermediate nodes and summing the path weights.
Refinements to path-scoring techniques might also be helpful solutions in this context.</p>
<p>Hetnet connectivity search explains how two nodes are related in an unsupervised manner that captures the semantic richness of edge type and returns results in the form of both metapaths and paths.
Our open-source implementation, including for a query and visualization webserver, was designed with scalability and responsiveness in mind allowing in-browser exploration.</p>
<h2 class="page_break_before" id="findings">Findings</h2>
<p>Completing hetnet connectivity search involved advances on three fronts.
We implemented new software for efficient matrix-based operations on hetnets.
We developed strategies to efficiently calculate the desired connectivity score under the null.
We designed and developed a web interface for easy access to the connectivity search approach.</p>
<h3 id="hetmatpy-package">Hetmatpy Package</h3>
<p>We created the hetmatpy Python package,
available on <a href="https://github.com/hetio/hetmatpy">GitHub</a> and <a href="https://pypi.org/project/hetmatpy/">PyPI</a> under the permissive BSD-2-Clause Plus Patent License.
This package provides matrix-based utilities for hetnets.
Each metaedge is represented by a distinct adjacency matrix,
which can be either a dense Numpy array or sparse SciPy matrix (see <a href="#hetmat-architecture">HetMat architecture</a>).
Adjacency matrices are stored on disk and loaded in a lazy manner to help scale the software to hetnets that are too large to fit entirely in memory.</p>
<p>The primary focus of the package is to provide compute-optimized and memory-efficient implementations of path-counting algorithms.
Specifically, the package supports computing DWPCs,
which can be done efficiently using matrix multiplication
but require complex adjustments to avoid counting paths with duplicate nodes
(i.e. to filter walks that are not paths, see <a href="#dwpc-matrix-multiplication-algorithms">DWPC matrix multiplication algorithms</a>).
The package can reuse existing path count computations that span segments of a longer metapath.
The package also supports generating null distributions for DWPCs derived from permuted networks,
see <a href="#degree-grouping-of-node-pairs">Degree-grouping of node pairs</a>.
Since this approach generates too many permuted DWPC values to store on disk,
our implementation retains summary statistics for each degree group that allow computation of a <a href="#gamma-hurdle-distribution">gamma-hurdle distribution</a> from which null DWPC <em>p</em>-values can be generated.</p>
<h3 id="dwpc-null-distribution">DWPC null distribution</h3>
<p>To assess connectivity between a source and target node,
we use the DWPC metric.
The DWPC is similar to path count (number of paths between the source and target node along a given metapath),
except that it downweights paths through high-degree nodes.
Rather than using the raw DWPC for a source–metapath–target combination,
we transform the DWPC across all source–target node pairs for a metapath to yield a distribution that is more compact and amenable to modeling <span class="citation" data-cites="17mGFvxXu">[<a href="#ref-17mGFvxXu" role="doc-biblioref">31</a>]</span>.
<!--
Future work should consider a different method for transforming DWPCs:
scale by nonzero mean rather than mean.
https://github.com/hetio/hetmatpy/issues/11
--></p>
<p>Previously, we had no technique for detecting whether a DWPC value was exceptional.
One possibility is to evaluate the DWPCs for all pairs of nodes and select the top scores (e.g. the top 5% of DWPCs).
Another possibility is to pick a transformed DWPC score as a cutoff.
The shortcomings of these methods are twofold.
First, neither the percentile nor absolute value of a DWPC has inherent meaning.
To select transformed DWPCs greater than 3.5, or alternatively the top 1% of DWPCs, is arbitrary.
Second, comparing DWPCs between node pairs fails to account for the situation where high-degree node pairs are likely to score higher,
solely on due to their degree (<a href="#fig:degree-group-metrics">4</a>).</p>
<p>To address these shortcomings,
we developed a method to compute the right-tail <em>p</em>-value of a DWPC.
<em>p</em>-values have a broadly understood interpretation
— in our case, the probability that a DWPC equal to or greater than the observed DWPC could occur under a null model.
Our null model is based on DWPCs generated from permuted networks,
where edges have been randomized in a degree-preserving manner (see <a href="#permuted-hetnets">Permuted hetnets</a>).</p>
<p>By tailoring the null distribution for a DWPC to the degree of its source and target node (see <a href="#degree-grouping-of-node-pairs">Degree-grouping of node pairs</a>),
we account for degree effects when determining the significance of a DWPC.
To improve the accuracy of DWPC <em>p</em>-values,
we use fit a <a href="#gamma-hurdle-distribution">gamma-hurdle distribution</a> to the null DWPCs.
In rare cases, there are insufficient nonzero null DWPCs to fit the gamma portion of the null distribution.
In these cases, we fall back to an empirical calculation as described in <a href="#empirical-dwpc-p-values">Empirical DWPC p-values</a>.</p>
<h3 id="enriched-metapaths">Enriched metapaths</h3>
<p>For each of the 2,205 metapaths in Hetionet v1.0 with length ≤ 3,
we computed DWPCs for all node pairs and their corresponding null distributions, see <a href="#dwpc-and-null-distribution-computation">DWPC and null distribution computation</a>.
We store the most significant DWPCs as described in <a href="#prioritizing-enriched-metapaths-for-database-storage">Prioritizing enriched metapaths for database storage</a>,
which appear as the “precomputed” rows in the webapp metapath table (Figures <a href="#fig:webapp">3</a>B &amp; <a href="#fig:webapp-metapaths">2</a>).
DWPCs that are not retained by the database can be regenerated on the fly.
This design allows us to immediately provide users with the metapaths that are most enriched between two query nodes,
while still allowing on-demand access to the full metrics for all metapaths with length ≤ 3.</p>
<div id="fig:webapp-metapaths" class="fignos">
<figure>
<img src="https://github.com/greenelab/connectivity-search-manuscript/raw/b290b4ad435553c7126867e0720f3112b4692809/content/media/webapp/v3/b.metapaths-expanded.png" alt="Figure 2: Expanded metapath details from the connectivity search webapp. This is the expanded view of the metapath table in 3B showing enriched metapaths between Alzheimer’s disease and the circadian rhythm pathway." />
<figcaption aria-hidden="true"><span>Figure 2:</span> <strong>Expanded metapath details from the connectivity search webapp.</strong>
This is the expanded view of the <a href="https://het.io/search/?source=17287&amp;target=7607&amp;metapaths=DaGiGpPW%2CDdGiGpPW%2CDdGpPW%2CDlAeGpPW%2CDrDaGpPW%2CDrDuGpPW%2CDuGiGpPW&amp;complete=">metapath table</a> in <a href="#fig:webapp">3</a>B showing enriched metapaths between Alzheimer’s disease and the circadian rhythm pathway.</figcaption>
</figure>
</div>
<p>Figure <a href="#fig:webapp-metapaths">2</a> shows the information used to compute <em>p</em>-value for enriched metapaths.
The table includes the following columns:</p>
<ul>
<li><strong>path count</strong>:
The number of paths between the source and target node of the specified metapath</li>
<li><strong>adjusted <em>p</em>-value</strong>:
A measure of the significance of the DWPC that indicates whether more paths were observed than expected due to random chance.
Compares the DWPC to a null distribution of DWPCs generated from degree-preserving permuted networks.
<a href="#adjusting-dwpc-p-values">Bonferroni-adjusted</a> for the number of metapaths with the same source metanode, target metanode, and length.</li>
<li><strong><em>p</em>-value</strong>:
A measure of the significance of the DWPC that indicates whether more paths were observed than expected due to random chance.
Compares the DWPC to a null distribution of DWPCs generated from degree-preserving permuted networks.
Not adjusted for multiple comparisons
(i.e. when multiple metapaths are assessed for significant connectivity between the source and target node).</li>
<li><strong>DWPC</strong>:
Degree-Weighted Path Count —
Measures the extent of connectivity between the source and target node for the given metapath.
Like the path count, but with less weight given to paths along high-degree nodes.</li>
<li><strong>source degree</strong>:
The number of edges from the source node that are of the same type as the initial metaedge of the metapath.</li>
<li><strong>target degree</strong>:
The number of edges from the target node that are of the same type as the final metaedge of the metapath.</li>
<li><strong># DWPCs</strong>:
The number of DWPCs calculated on permuted networks used to generate a null distribution for the DWPC from the real network.
Permuted DWPCs are aggregated for all permuted node pairs with the same degrees as the source and target node.</li>
<li><strong># non-0 DWPCs</strong>:
The number of permuted DWPCs from the “# of DWPCs” column that were nonzero.
Nonzero DWPCs indicate at least one path between the source and target node existed in the permuted network.</li>
<li><strong>non-0 mean</strong>:
The mean of nonzero permuted DWPCs.
Used to generate the gamma-hurdle model of the null DWPC distribution.</li>
<li><strong>non-0 σ</strong>:
The standard deviation of nonzero permuted DWPCs.
Used to generate the gamma-hurdle model of the null DWPC distribution.</li>
<li><strong>Neo4j Actions</strong>:
A Cypher query that users can run in the <a href="https://neo4j.het.io/browser/">Neo4j browser</a> to show paths with the largest DWPCs for the metapath.</li>
</ul>
<h3 id="enriched-paths">Enriched paths</h3>
<p>In addition to knowing which metapaths are enriched between two query nodes,
it is helpful to see the specific paths that contribute highly to such enrichment.
Since the DWPC is a summation of a path metric (called the path degree product),
it is straightforward to calculate the proportion of a DWPC attributable to an individual path.
The webapp allows users to select a metapath to populate a table of the corresponding paths.
These paths are generated on the fly through a Cypher query to the Hetionet Neo4j database.</p>
<p>It is desirable to have a consolidated view of paths across multiple metapaths.
Therefore, we calculate a <em>path score</em> heuristic,
which can be used to compare the importance of paths between metapaths.
The path score equals the proportion of the DWPC contributed by a path multiplied by the magnitude of the DWPC’s <em>p</em>-value (-log<sub>10</sub>(<em>p</em>)).
To illustrate, the paths webapp panel includes the following information (Figure <a href="#fig:webapp">3</a>C):
<!-- https://github.com/greenelab/connectivity-search-frontend/blob/63bb4acbcebe4b346882754719071856b30b43ba/src/definitions.json --></p>
<ul>
<li><strong>path</strong>:
The sequence of edges in the network connecting the source node to the target node.
Duplicate nodes are not permitted in paths.</li>
<li><strong>path score</strong>:
A metric of how meaningful the path is in describing the connectivity between the source and target node.
The score combines the magnitude of the metapath’s p-value with the percentage of the DWPC contributed by the path.</li>
<li><strong>% of DWPC</strong>
The contribution of the path to the DWPC for its metapath.
This metric compares the importance of all paths of the same metapath from the source node to the target node.</li>
</ul>
<!--
### Comparison to Rephetio
  Omitted with some prior notes at
  https://github.com/dhimmel/connectivity-search-manuscript/commit/8f5a2ae4732249d4b6f33a6a142cc4043152c458
-->
<h3 id="hetio-website-and-connectivity-search-webapp">Hetio Website and Connectivity Search Webapp</h3>
<p>We revamped the website hosted at <a href="https://het.io" class="uri">https://het.io</a> to serve as a unified home for this study and the hetnet-related research that preceded it.
The website provides the connectivity search webapp running over the hetio network and several other interactive apps for prior projects.
It also includes high-level information on hetnets and Hetionet,
citation and contact details,
links to supporting studies and software,
downloads and exploration of Hetionet data,
and related media.</p>
<p>We created the connectivity search webapp available at <a href="https://het.io/search/" class="uri">https://het.io/search/</a>.
The tool is free to use, without any login or authentication.
The app allows users to quickly explore how any two nodes in Hetionet v1.0 might be related.
The workflow accepts one or more nodes as input and shows the user the most important metapaths and paths for a pair of query nodes.</p>
<p>The design guides the user through selecting a source and target node (Figure <a href="#fig:webapp">3</a>A).
The webapp returns metapaths, scored by whether they occurred more than expected based on network degree (Figure <a href="#fig:webapp">3</a>B).
Users can proceed by requesting the specific paths for each metapath,
which are placed in a unified table sorted according to their path score (Figure <a href="#fig:webapp">3</a>C).
Finally, the webapp produces publication-ready visualizations containing user-selected paths (Figure <a href="#fig:webapp">3</a>D).</p>
<div id="fig:webapp" class="fignos">
<figure>
<img src="https://github.com/greenelab/connectivity-search-manuscript/raw/b290b4ad435553c7126867e0720f3112b4692809/content/media/webapp/v3/webapp.png" class="white" style="width:100.0%" alt="Figure 3: Using the connectivity search webapp to explore the pathophysiology of Alzheimer’s disease. This figure shows an example user workflow for https://het.io/search/. A. The user selects two nodes. Here, the user is interested in Alzheimer’s disease, so selects this as the source node. The user limits the target node search to metanodes relating to gene function. The target node search box suggests nodes, sorted by the number of significant metapaths. When the user types in the target node box, the matches reorder based on search word similarity. Here, the user becomes interested in how the circadian rhythm might relate to Alzheimer’s disease. B. The webapp returns metapaths between Alzheimer’s disease and the circadian rhythm pathway. The user unchecks “precomputed only” to compute results for all metapaths with length ≤ 3, not just those that surpass the database inclusion threshold. The user sorts by adjusted p-value and selects 7 of the top 10 metapaths. C. Paths for the selected metapaths are ordered by their path score (limited to 100 paths for each metapath). The user selects 8 paths (1 from a subsequent page of results) to show in the graph visualization and highlights a single path involving ARNT2 for emphasis. D. A subgraph displays the previously selected paths. The user improves on the automated layout by repositioning nodes. Clicking an edge displays its properties, informing the user that association between Creutzfeldt–Jakob disease and NPAS2 was detected by genome-wide association study. " />
<figcaption aria-hidden="true"><span>Figure 3:</span> <strong>Using the connectivity search webapp to explore the pathophysiology of Alzheimer’s disease.</strong>
This figure shows an example user workflow for <a href="https://het.io/search/" class="uri">https://het.io/search/</a>.<br />
<strong>A.</strong>
The user selects two nodes.
Here, the user is interested in Alzheimer’s disease, so <a href="https://het.io/search/?source=17287">selects this</a> as the source node.
The user limits the target node search to metanodes relating to gene function.
The target node search box suggests nodes, sorted by the number of significant metapaths.
When the user types in the target node box, the matches reorder based on search word similarity.
Here, the user becomes interested in how the circadian rhythm might relate to Alzheimer’s disease.<br />
<strong>B.</strong>
The webapp returns metapaths between Alzheimer’s disease and the circadian rhythm pathway.
The user unchecks “precomputed only” to compute results for all metapaths with length ≤ 3, not just those that surpass the database inclusion threshold.
The user sorts by adjusted <em>p</em>-value and <a href="https://het.io/search/?source=17287&amp;target=7607&amp;metapaths=DaGiGpPW%2CDdGiGpPW%2CDdGpPW%2CDlAeGpPW%2CDrDaGpPW%2CDrDuGpPW%2CDuGiGpPW&amp;complete=">selects</a> 7 of the top 10 metapaths.<br />
<strong>C.</strong>
Paths for the selected metapaths are ordered by their path score (limited to 100 paths for each metapath).
The user selects 8 paths (1 from a subsequent page of results) to show in the graph visualization and highlights a single path involving <em>ARNT2</em> for emphasis.<br />
<strong>D.</strong>
A subgraph displays the previously selected paths.
The user improves on the automated layout by repositioning nodes.
Clicking an edge displays its properties, informing the user that association between Creutzfeldt–Jakob disease and <em>NPAS2</em> was detected by genome-wide association study.
<!--
  This example is also used at https://slides.com/dhimmel/rocky2019#/4
  More info at https://github.com/greenelab/connectivity-search-manuscript/issues/7
--></figcaption>
</figure>
</div>
<h2 class="page_break_before" id="discussion">Discussion</h2>
<p>In this study, we introduce a search engine for hetnet connectivity between two nodes that returns results in real time.
An interactive webapp helps users explore node connectivity by ranking metapaths and paths,
while visualizing multiple paths in a subgraph.</p>
<p>We made several methodological contributions to support this application.
We developed optimized algorithms for computing DWPCs using matrix multiplication.
In addition, we created a method for estimating a <em>p</em>-value for a DWPC,
using null DWPCs computed on permuted hetnets.
We implemented these advances in the open-source hetmatpy Python package and HetMat data structure to provide highly optimized computational infrastructure for representing and reasoning on hetnets using matrices.</p>
<p>This work lays the foundation for exciting future directions.
For many queries, a large number of paths are returned.
Interpretation of large lists is difficult.
Therefore, the dimensionality of results could be reduced by aggregating path scores across intermediate nodes or edges <span class="citation" data-cites="1693YErZj">[<a href="#ref-1693YErZj" role="doc-biblioref">32</a>]</span>.</p>
<p>Here, we computed all DWPCs for Hetionet metapaths with length ≤ 3.
Our search engine will therefore overlook important connectivity from longer metapaths.
However, it is infeasible to compute DWPCs for all longer metapaths.
One solution would be to only extend metapaths detected as informative.
For example, if a <em>CbGpPWpG</em> metapath is deemed informative,
it could be extended with additional metaedges like <em>CbGpPWpGaD</em>.
One unsupervised approach would be to use the distribution of DWPC <em>p</em>-values for a metapath to detect whether the paths still convey sufficient information, for example by requiring an enrichment of small <em>p</em>-values.
Were this method to fail, supervised alternatives could be explored,
such as the ability for DWPCs from a longer metapath to predict that of a shorter metapath or metaedge,
with care taken to prevent label leakage.
One final approach could learn from user interest and compute longer metapaths only when requested.</p>
<p>This work focuses on queries where the input is a node pair.
Equally interesting would be queries where the input is a set of nodes of the same type, optionally with weights.
The search would compute DWPCs for paths originating on the query nodes.
The simpler formulation would compute DWPCs for metapaths separately and compare to null distributions from permuted hetnets.
A more advanced formulation would combine scores across metapaths such that every node in the hetnet would receive a single score capturing its connectivity to the query set.
This approach would have similar utility to gene set enrichment analysis in that the user could provide a set of genes as input and receive a ranked list of nodes that characterize the function of the query genes.
However, it would excel in its versatility by returning results of any node type without requiring predefined gene sets to match against.
Some users might be interested in node set transformations where scores for one node type are converted to another node type.
This approach could take scores for human genes and convert them to side effects, diseases, pathways, and so on.</p>
<p>Our work is not without limitations.
The final application relies on multiple databases and cached computations specific to Hetionet v1.0.
Despite striving for a modular architecture,
generating an equivalent search webapp for a different hetnet would require adaptation due to the many data sources involved.
Furthermore, we would benefit from greater real-world evaluation of the connectivity search results to help identify situations where the method underperforms.
Despite these challenges, our study demonstrates one of the first public search engines for node connectivity on a biomedical knowledge graph, while contributing methods and software that we hope will inspire future work.</p>
<h2 class="page_break_before" id="methods">Methods</h2>
<h3 id="hetionet">Hetionet</h3>
<p>We used the hetionet knowledge graph to demonstrate connectivity search.
<a href="https://het.io/about/">Hetionet</a> is a knowledge graph of human biology, disease, and medicine,
integrating information from millions of studies and decades of research.
Hetionet v1.0 combines information from <a href="https://git.dhimmel.com/rephetio-manuscript/#tbl:resources">29 public databases</a>.
The network contains 47,031 nodes of <a href="https://git.dhimmel.com/rephetio-manuscript/#tbl:metanodes">11 types</a> (Table <a href="#tbl:metanodes">1</a>) and 2,250,197 edges of <a href="https://git.dhimmel.com/rephetio-manuscript/#tbl:metaedges">24 types</a> (Figure <a href="#fig:rephetio">1</a>A).</p>
<div id="tbl:metanodes" class="tablenos">
<table id="tbl:metanodes">
<caption><span>Table 1:</span> <strong>Node types in Hetionet</strong>,
including the abbreviation, number of nodes, and description for each of the 11 metanodes in Hetionet v1.0.
</caption>
<colgroup>
<col style="width: 25%" />
<col style="width: 25%" />
<col style="width: 25%" />
<col style="width: 25%" />
</colgroup>
<thead>
<tr class="header">
<th>Metanode</th>
<th>Abbr</th>
<th>Nodes</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>Anatomy</td>
<td>A</td>
<td>402</td>
<td>Anatomical structures, excluding structures that are known not to be found in humans. From <a href="http://uberon.github.io/">Uberon</a>.</td>
</tr>
<tr class="even">
<td>Biological Process</td>
<td>BP</td>
<td>11381</td>
<td>Larger processes or biological programs accomplished by multiple molecular activities. From <a href="http://geneontology.org/">Gene Ontology</a>.</td>
</tr>
<tr class="odd">
<td>Cellular Component</td>
<td>CC</td>
<td>1391</td>
<td>The locations relative to cellular structures in which a gene product performs a function. From <a href="http://geneontology.org/">Gene Ontology</a>.</td>
</tr>
<tr class="even">
<td>Compound</td>
<td>C</td>
<td>1552</td>
<td>Approved small-molecule compounds with documented chemical structures. From <a href="https://www.drugbank.ca/">DrugBank</a>.</td>
</tr>
<tr class="odd">
<td>Disease</td>
<td>D</td>
<td>137</td>
<td>Complex diseases, selected to be distinct and specific enough to be clinically relevant yet general enough to be well annotated. From <a href="http://disease-ontology.org/">Disease Ontology</a>.</td>
</tr>
<tr class="even">
<td>Gene</td>
<td>G</td>
<td>20945</td>
<td>Protein-coding human genes. From <a href="https://www.ncbi.nlm.nih.gov/gene">Entrez Gene</a>.</td>
</tr>
<tr class="odd">
<td>Molecular Function</td>
<td>MF</td>
<td>2884</td>
<td>Activities that occur at the molecular level, such as “catalysis” or “transport”. From <a href="http://geneontology.org/">Gene Ontology</a>.</td>
</tr>
<tr class="even">
<td>Pathway</td>
<td>PW</td>
<td>1822</td>
<td>A series of actions among molecules in a cell that leads to a certain product or change in the cell. From <a href="https://www.wikipathways.org/index.php/WikiPathways">WikiPathways</a>, <a href="https://reactome.org/">Reactome</a>, and Pathway Interaction Database.</td>
</tr>
<tr class="odd">
<td>Pharmacologic Class</td>
<td>PC</td>
<td>345</td>
<td>“Chemical/Ingredient”, “Mechanism of Action”, and “Physiologic Effect” FDA class types. From <a href="http://drugcentral.org/">DrugCentral</a>.</td>
</tr>
<tr class="even">
<td>Side Effect</td>
<td>SE</td>
<td>5734</td>
<td>Adverse drug reactions. From <a href="http://sideeffects.embl.de/">SIDER</a>/<a href="https://www.nlm.nih.gov/research/umls/">UMLS</a>.</td>
</tr>
<tr class="odd">
<td>Symptom</td>
<td>S</td>
<td>438</td>
<td>Signs and Symptoms (i.e. clinical abnormalities that can indicate a medical condition). From the <a href="https://www.nlm.nih.gov/mesh/meshhome.html">MeSH ontology</a>.</td>
</tr>
</tbody>
</table>
</div>
<p>One limitation that restricts the applicability of Hetionet is incompleteness.
In many cases, Hetionet v1.0 includes only a subset of the nodes from a given resource.
For example, the Disease Ontology contains over 9,000 diseases <span class="citation" data-cites="1FsruosUW">[<a href="#ref-1FsruosUW" role="doc-biblioref">33</a>]</span>,
while Hetionet includes only 137 diseases <span class="citation" data-cites="203l6hC3">[<a href="#ref-203l6hC3" role="doc-biblioref">34</a>]</span>.
Nodes were excluded to avoid redundant or overly specific nodes,
while ensuring a minimum level of connectivity for compounds and diseases.
See the <a href="https://git.dhimmel.com/rephetio-manuscript/#nodes">Project Rephetio methods</a> for more details <span class="citation" data-cites="O21tn8vf">[<a href="#ref-O21tn8vf" role="doc-biblioref">2</a>]</span>.
Nonetheless, Hetionet v1.0 remains one of the most comprehensive and integrative networks that consolidates biomedical knowledge into a manageable number of node and edge types <span class="citation" data-cites="yqoUY2f9">[<a href="#ref-yqoUY2f9" role="doc-biblioref">35</a>]</span>.
Other integrative resources, some still under development, include <a href="https://www.wikidata.org">Wikidata</a> <span class="citation" data-cites="HUodcdz1">[<a href="#ref-HUodcdz1" role="doc-biblioref">36</a>]</span>, <a href="https://skr3.nlm.nih.gov/SemMedDB/">SemMedDB</a> <span class="citation" data-cites="eKVkBSDD reekunFL uTOtIsTj">[<a href="#ref-eKVkBSDD" role="doc-biblioref">37</a>,<a href="#ref-reekunFL" role="doc-biblioref">38</a>,<a href="#ref-uTOtIsTj" role="doc-biblioref">39</a>]</span>, <a href="https://spoke.ucsf.edu/">SPOKE</a> <span class="citation" data-cites="1BHNoKOV1">[<a href="#ref-1BHNoKOV1" role="doc-biblioref">40</a>]</span>, and <a href="https://github.com/RTXteam/RTX-KG2">RTX-KG2c</a> <span class="citation" data-cites="bniVH7q2">[<a href="#ref-bniVH7q2" role="doc-biblioref">41</a>]</span>.</p>
<h3 id="hetmat-architecture">HetMat architecture</h3>
<p>At the core of the hetmatpy package is the HetMat data structure for storing and accessing the network.
HetMats are stored on disk as a directory, which by convention uses a <code>.hetmat</code> extension.
A HetMat directory stores a single heterogeneous network, whose data resides in the following files.</p>
<ol type="1">
<li>A <code>metagraph.json</code> file stores the schema, defining which types of nodes and edges comprise the hetnet.
This format is defined by the <a href="https://github.com/hetio/hetnetpy">hetnetpy</a> Python package.
Hetnetpy was originally developed with the name hetio during prior studies
<span class="citation" data-cites="WkPlH1ds O21tn8vf">[<a href="#ref-O21tn8vf" role="doc-biblioref">2</a>,<a href="#ref-WkPlH1ds" role="doc-biblioref">42</a>]</span>,
but we <a href="https://github.com/hetio/hetnetpy/issues/40">renamed</a> it to het<strong>net</strong>py for better disambiguation from het<strong>mat</strong>py.</li>
<li>A <code>nodes</code> directory containing one file per node type (metanode) that defines each node.
Currently, <code>.tsv</code> files in which each row represents a node are supported.</li>
<li>An <code>edges</code> directory containing one file per edge type (metadata) that encodes the adjacency matrix.
The matrix can be serialized using either the Numpy dense format (<code>.npy</code>) or SciPy sparse format (<code>.sparse.npz</code>).</li>
</ol>
<p>For node and edge files, compression is supported as detected from <code>.gz</code>, <code>.bz2</code>, <code>.zip</code>, and <code>.xz</code> extensions.
This structure of storing a hetnet supports selectively reading nodes and edges into memory.
For example, a certain computation may only require access to a subset of the node and edge types.
By only loading the required node and edge types, we reduce memory usage and read times.</p>
<p>Additional subdirectories, such as <code>path-counts</code> and <code>permutations</code>, store data generated from the HetMat.
By using consistent paths for generated data, we avoid recomputing data that already exist on disk.
A HetMat directory can be zipped for archiving and transfer.
Users can selectively include generated data in archives.
Since the primary application of HetMats is to generate computationally demanding measurements on hetnets, the ability to share HetMats with precomputed data is paramount.</p>
<p>The <a href="https://hetio.github.io/hetmatpy/reference/hetmatpy/hetmat/#hetmat"><code>HetMat</code></a> class implements the above logic.
A <code>hetmat_from_graph</code> function creates a HetMat object and directory on disk from the preexisting <code>hetnetpy.hetnet.Graph</code> format.</p>
<p>We converted Hetionet v1.0 to HetMat format and uploaded the <code>hetionet-v1.0.hetmat.zip</code> archive to the <a href="https://github.com/hetio/hetionet/tree/master/hetnet/matrix">Hetionet data repository</a>.</p>
<h3 id="dwpc-matrix-multiplication-algorithms">DWPC matrix multiplication algorithms</h3>
<p>Prior to this study, we used two implementations for computing DWPCs.
The first is a pure Python implementation available in the <a href="https://github.com/hetio/hetnetpy/blob/aa16e6a7092c039a6b175a73a35c006e53acee20/hetnetpy/pathtools.py#L8-L21"><code>hetnetpy.pathtools.DWPC</code></a> function <span class="citation" data-cites="WkPlH1ds">[<a href="#ref-WkPlH1ds" role="doc-biblioref">42</a>]</span>.
The second uses a Cypher query, prepared by <a href="https://github.com/hetio/hetnetpy/blob/aa16e6a7092c039a6b175a73a35c006e53acee20/hetnetpy/neo4j.py#L363-L393"><code>hetnetpy.neo4j.construct_dwpc_query</code></a>, that is executed by the Neo4j database <span class="citation" data-cites="O21tn8vf kNYoxBiQ">[<a href="#ref-O21tn8vf" role="doc-biblioref">2</a>,<a href="#ref-kNYoxBiQ" role="doc-biblioref">43</a>]</span>.
Both of these implementations require traversing all paths between the source and target node.
Hence, they are computationally cumbersome despite optimizations <span class="citation" data-cites="xWWaK51z">[<a href="#ref-xWWaK51z" role="doc-biblioref">44</a>]</span>.</p>
<p>Since our methods only require degree-weighted counts, not fully enumerated paths, adjacency matrix multiplication presents an alternative approach.
Multiplication alone, however, counts walks rather than paths, meaning paths traversing a single node multiple times are counted.
When computing network-based features to quantify the relationship between a source and target node,
we would like to exclude traversing duplicate nodes (i.e. paths, not trails or walks) <span class="citation" data-cites="w6rL3OYz">[<a href="#ref-w6rL3OYz" role="doc-biblioref">45</a>]</span>.
We developed a suite of algorithms to compute true path counts and DWPCs using matrix multiplication that benefits from the speed advantages of only counting paths.</p>
<p>Our implementation begins by categorizing a metapath according to the pattern of its repeated metanodes, allowing DWPC computation using a specialized order of operations.
For example, the metapath <em>DrDtCrC</em> is categorized as a set of disjoint repeats, while <em>DtCtDpC</em> is categorized as repeats of the form BABA.
Many complex repeat patterns can be represented piecewise as simpler patterns, allowing us to compute DWPC for most metapaths up to length 5 and many of length 6 and beyond without enumerating individual paths.
For example, disjoint groups of repeats like <em>DrDtCrC</em> can be computed as the matrix product of DWPC matrices for <em>DrD</em> and <em>CrC</em>.
Randomly inserted non-repeated metanodes (e.g. <em>G</em> in <em>DrDaGaDrD</em>) require no special treatment and are included in DWPC with matrix multiplication.</p>
<p>After metapath categorization, we segment metapaths according to their repeat pattern, following our order of operations.
By segmenting and computing recursively, we can efficiently evaluate DWPC on highly complex metapaths, using simple patterns as building blocks for higher-level patterns.
Finally, our specialized DWPC functions are applied to individual segments, the results are combined, and final corrections are made to ensure no repeated nodes are counted.
The recursive, segmented approach we developed also allowed us to implement a caching strategy that improved speed by avoiding duplicate DWPC computations.
In summary, the functionality we developed resulted in more than a 175-fold reduction in compute time, allowing us to compute millions of DWPC values across Hetionet <span class="citation" data-cites="15nwuvjrA">[<a href="#ref-15nwuvjrA" role="doc-biblioref">46</a>]</span>.</p>
<h4 id="details-of-matrix-dwpc-implementation">Details of matrix DWPC implementation</h4>
<p>DWPC computation requires us to remove all duplicate nodes from paths.
We used three repeat patterns as the building blocks for DWPC computation: short repeats (AAA), nested repeats (BAAB), and overlapping repeats (BABA).
Let <span class="math inline">\(\mathrm{D}(\mathit{XwXyZ})\)</span> denote the DWPC matrix for metapath <em>XwXyZ</em>.
Under this notation, <span class="math inline">\(\mathrm{D}(\mathit{XyZ})\)</span> is the degree-weighted (bi)adjacency matrix for metaedge <em>XyZ</em>.
Additionally, let <span class="math inline">\(\mathrm{diag}(A)\)</span> represent a diagonal matrix whose entries are the diagonal elements of <span class="math inline">\(A\)</span>.</p>
<p>For the case of short (&lt; 4) repeats for a single metanode, <em>XaXbX</em> (e.g. <em>GiGdG</em>), we simply subtract the main diagonal.</p>
<p><span class="math display">\[\mathrm{D}(\mathit{XaXbX}) = \mathrm{D}(\mathit{XaX}) \mathrm{D}(\mathit{XbX}) - \mathrm{diag}(\mathrm{D}(\mathit{XaX}) \mathrm{D}(\mathit{XbX}))\]</span></p>
<p>Nested repeats <em>XaYbYcX</em> (e.g. <em>CtDrDtC</em>), are treated recursively, with both inner (YY) and outer (XX) repeats treated as separate short repeats.</p>
<p><span class="math display">\[
\mathrm{D}(\mathit{XaYbYcX}) = \mathrm{D}(\mathit{XaY}) \mathrm{D}(\mathit{YbY}) \mathrm{D}(\mathit{YcX}) - \mathrm{diag}(\mathrm{D}(\mathit{XaY}) (\mathrm{D}(\mathit{YbY}) \mathrm{D}(\mathit{YcX}))
\]</span></p>
<p>Overlapping repeats <em>XaYbXcY</em> (e.g. <em>CtDtCtD</em>) require several corrections (<span class="math inline">\(\odot\)</span> denotes the Hadamard product).</p>
<p><span class="math display">\[\begin{align*}
\mathrm{D}(\mathit{XaYbXcY}) =\ &amp;\mathrm{D}(\mathit{XaY})\  \mathrm{D}(\mathit{YbX})\  \mathrm{D}(\mathit{XcY}) \\
&amp;- \mathrm{diag}(\mathrm{D}(\mathit{XaY})\  \mathrm{D}(\mathit{YbX}))\  \mathrm{D}(\mathit{XcY}) \\
&amp;- \mathrm{D}(\mathit{XaY})\  \mathrm{diag}(\mathrm{D}(\mathit{YbX})\  \mathrm{D}(\mathit{XcY})) \\
&amp;+ \mathrm{D}(\mathit{XaY})\  \odot \mathrm{D}(\mathit{YbX})^T\  \odot \mathrm{D}(\mathit{XcY})
\end{align*}\]</span></p>
<p>Most paths of length six—and many even longer paths—can be represented hierarchically using these patterns.
For example, a long metapath pattern of the form CBABACXYZ can be segmented as (C(BABA)C)XYZ using patterns for short and overlapping repeats and can be computed using the tools we developed.
In addition to these matrix routines—which advantageously count rather than enumerate paths—we implemented a general matrix method for any metapath type.
The general method is important for patterns such as long (≥ 4) repeats, or complex repeat patterns (e.g. of the form ABCABC), but it requires path enumeration and is therefore slower.
As an alternative approach for complex paths, we developed an approximate DWPC method that corrects repeats in disjoint simple patterns but only corrects the first repeat in complex patterns (e.g. ≥ length four repeat).
Mayers et al. developed an alternative approximation, which subtracts the main diagonal at every occurrence of the first repeated metanode <span class="citation" data-cites="IJWsN67g">[<a href="#ref-IJWsN67g" role="doc-biblioref">47</a>]</span>.
Our matrix methods were validated against the existing Python and Cypher implementations in the <code>hetnetpy</code> package that rely on explicit path enumeration.</p>
<!-- Discuss caching strategies, sequential versus recursive
Runtime comparison to show the speedup.
Rephetio computed a portion of in XX time -->
<!-- From [@vagelos-2017]:
> We reduced the time to compute DWPC over nearly 1200 metapaths from roughly four-and-a-half days to roughly one hour and thirty-seven minutes
175-fold, which underestimates since Rephetio did not compute the full DWPC matrix and benefited from concurrency. -->
<!-- references
  https://github.com/greenelab/connectivity-search-analyses/issues/20
  https://github.com/greenelab/connectivity-search-analyses/pull/70
  https://github.com/greenelab/connectivity-search-analyses/issues/53
  https://github.com/greenelab/connectivity-search-analyses/pull/67
  https://github.com/hetio/hetmatpy/blob/bc36aa9859c43a1a5fb22808cd6eb952ef9d497c/hetmatpy/degree_weight.py#L210-L239
  https://github.com/mmayers12/hetnet_ml
  https://nbviewer.jupyter.org/github/greenelab/connectivity-search-analyses/blob/042063fb559048e52b3dc2731b6d6c6836f698cf/7.rephetio-times.ipynb
  https://nbviewer.jupyter.org/github/greenelab/connectivity-search-analyses/blob/042063fb559048e52b3dc2731b6d6c6836f698cf/explore/recursive-chain/dwwc-chain-recursive.ipynb
  https://nbviewer.jupyter.org/github/greenelab/connectivity-search-analyses/blob/042063fb559048e52b3dc2731b6d6c6836f698cf/explore/cache-speeds.ipynb
-->
<h3 id="permuted-hetnets">Permuted hetnets</h3>
<p>In order to generate a null distribution for a DWPC, we rely on DWPCs computed from permuted hetnets.
We derive permuted hetnets from the unpermuted network using the XSwap algorithm <span class="citation" data-cites="iKOIEzQ9">[<a href="#ref-iKOIEzQ9" role="doc-biblioref">48</a>]</span>.
XSwap randomizes edges while preserving node degree.
Therefore, it’s ideal for generating null distributions that retain general degree effects,
but destroy the actual meaning of edges.
We adapt XSwap to hetnets by applying it separately to each metaedge <span class="citation" data-cites="O21tn8vf GE6jhWnt 83EvMSI6">[<a href="#ref-O21tn8vf" role="doc-biblioref">2</a>,<a href="#ref-GE6jhWnt" role="doc-biblioref">49</a>,<a href="#ref-83EvMSI6" role="doc-biblioref">50</a>]</span>.</p>
<p>Project Rephetio created 5 permuted hetnets <span class="citation" data-cites="O21tn8vf GE6jhWnt">[<a href="#ref-O21tn8vf" role="doc-biblioref">2</a>,<a href="#ref-GE6jhWnt" role="doc-biblioref">49</a>]</span>,
which were used to generate a null distribution of classifier performance for each metapath-based feature.
Here, we aim to create a null distribution for individual DWPCs, which requires vastly more permuted values to estimate with accuracy.
Therefore, we generated <a href="https://github.com/hetio/hetionet/tree/a95ae76581af604e91d744680aee3f888fa18887/hetnet/permuted/matrix">200 permuted hetnets</a>.
Permutations 001–005 were those generated by Project Rephetio,
while permutations 006–200 were generated by this study.
For the newly generated permutations, we attempted 10 times the number of swaps as edges for a given metaedge,
which is the default multiplier set by the hetnetpy <code>permute_graph</code> function.
<!-- Permuted hetmats generated in PRs
https://github.com/greenelab/connectivity-search-analyses/pull/107
https://github.com/greenelab/connectivity-search-analyses/pull/127
Permutation statistics in
https://github.com/greenelab/connectivity-search-analyses/blob/15f1925c0481d8e6bab8b0931f48f2fad388c68c/data/hetionet-v1.0.hetmat/permutations/stats.tsv
-->
More recently, we also developed the <code>xswap</code> Python <a href="https://github.com/greenelab/xswap">package</a>, whose optimized C/C++ implementation will enable future research to generate even larger sets of permuted networks <span class="citation" data-cites="83EvMSI6">[<a href="#ref-83EvMSI6" role="doc-biblioref">50</a>]</span>.</p>
<h3 id="degree-grouping-of-node-pairs">Degree-grouping of node pairs</h3>
<p>For each of the 200 permuted networks and each of the 2,205 metapaths, we computed the entire DWPC matrix (i.e. all source nodes × target nodes).
Therefore, for each actual DWPC value, we computed 200 permuted DWPC values.
Because permutation preserves only node degree, DWPC values among nodes with the same source and target degrees are equivalent to additional permutations.
We greatly increased the effective number of permutations by grouping DWPC values according to node degree, affording us a superior estimation of the DWPC null distribution.</p>
<p>We have applied this <em>degree-grouping</em> approach previously when calculating the prior probability of edge existence based on the source and target node degrees <span class="citation" data-cites="19OyKPQ4M 83EvMSI6">[<a href="#ref-83EvMSI6" role="doc-biblioref">50</a>,<a href="#ref-19OyKPQ4M" role="doc-biblioref">51</a>]</span>.
But here, we apply <em>degree-grouping</em> to null DWPCs.
The result is that the null distribution for a DWPC is based not only on permuted DWPCs for the corresponding source–metapath–target combination,
but instead on all permuted DWPCs for the source-degree–metapath–target-degree combination.</p>
<p>The “# DWPCs” column in Figure <a href="#fig:webapp-metapaths">2</a> illustrates how degree-grouping inflates the sample size of null DWPCs.
The <em>p</em>-value for the <em>DaGiGpPW</em> metapath relies on the minimum number of null DWPCs (200),
since no other disease besides Alzheimer’s had 196 <em>associates</em> edges (source degree) and no other pathway besides circadian rhythm had 201 <em>participates</em> edges (target degree).
However, for other metapaths with over 5,000 null DWPCs, degree-grouping increased the size of the null distribution by a factor of 25.
In general, source–target node pairs with lower degrees receive the largest sample size multiplier from degree-grouping.
This is convenient since low-degree nodes also tend to produce the highest proportion of zero DWPCs, by virtue of low connectivity.
Consequently, degree-grouping excels where it is most needed.</p>
<p>One final benefit of degree-grouping is that it reduces the disk space required to store null DWPC summary statistics.
For example, with 20,945 genes in Hetionet v1.0, there exist 438,693,025 gene pairs.
Gene nodes have 302 distinct degrees for <em>interacts</em> edges, resulting in 91,204 degree pairs.
This equates to an 4,810-fold reduction in the number of summary statistics that need to be stored to represent the null DWPC distribution for a metapath starting and ending with a <em>Gene–interacts–Gene</em> metaedge.</p>
<p>We store the following null DWPC summary statistics for each metapath–source-degree–target-degree combination:
total number of null DWPCs,
total number of nonzero null DWPCs,
sum of null DWPCs,
sum of squared null DWPCs,
and number of permuted hetnets.
These values are sufficient to estimate the <em>p</em>-value for a DWPC, by either <a href="https://github.com/hetio/hetmatpy/blob/bc36aa9859c43a1a5fb22808cd6eb952ef9d497c/hetmatpy/pipeline.py#L42-L63" title="hetmatpy.pipeline.add_gamma_hurdle_to_dgp_df source code">fitting</a> a gamma-hurdle null distribution or generating an empiric <em>p</em>-value.
Furthermore, these statistics are additive across permuted hetnets.
Their values are always a running total and can be updated incrementally as statistics from each additional permuted hetnet become available.</p>
<!--
Should note:
Permuted DWPCs were scaled by dividing by the mean of all unpermuted DWPCs for the metapath and then applying the inverse hyperbolic sine transformation.
Every degree pair for a given metapath has corresponding statistics that summarize its values across permuted hetnets.
-->
<p>Figure <a href="#fig:degree-group-metrics">4</a> shows how various aspects of DWPCs vary by degree group.
The rows display the following metrics of the DWPC distribution for all node-pairs in a given degree-group:</p>
<ul>
<li><strong># Nonzero DWPCs</strong>:
The number of nonzero DWPCs values (on average per network to enable comparison).</li>
<li><strong>% Nonzero DWPCs</strong>:
Of the total number of DWPCs, the percentage that is nonzero.
As node degrees increase, the chance of node pairs having at least one path, and hence a nonzero DWPC, greatly increases.</li>
<li><strong>Mean DWPC</strong>:
The average value of all DWPCs including zeros.</li>
<li><strong>Mean Nonzero DWPC</strong>:
The average value of nonzero DWPCs.</li>
<li><strong>Std Dev Nonzero DWPC</strong>:
The standard deviation of nonzero DWPCs.</li>
<li><strong>Gamma Model β</strong>:
The β parameter of the gamma model fit on nonzero DWPCs.
Note that the gamma model is only fit on permuted network DWPCs to estimate a null distribution for the unpermuted network DWPCs.
Since this parameter varies with source &amp; target node degree,
it is important to fit a separate gamma model for each degree group.</li>
</ul>
<div id="fig:degree-group-metrics" class="fignos">
<figure>
<img src="https://github.com/greenelab/connectivity-search-analyses/raw/15f1925c0481d8e6bab8b0931f48f2fad388c68c/explore/degree-group-analyses.png" alt="Figure 4: Path-based metrics vary by node degree and network permutation status. Each row shows a different metric of the DWPC distribution for the CbGpPWpG metapath — traversing Compound–binds–Gene–participates–Pathway–participates–Gene, selected for illustrative purposes. Metrics are computed for degree-groups, which is a specific pair of source degree (in this case, the source compound’s count of CbG edges) and target degree (in this case, the target gene’s count of GpPW edges). On the left, metrics are reported for the unpermuted hetnet and on the right for the 200 permuted hetnets. Hence, each cell on the right summarizes 200 times the number of DWPCs as the corresponding cell on the left. The colormap is row normalized, such that its intensity peaks for the maximum value of each metric across the unpermuted and permuted values. Gray indicates null values." />
<figcaption aria-hidden="true"><span>Figure 4:</span> <strong>Path-based metrics vary by node degree and network permutation status.</strong>
Each row shows a different metric of the DWPC distribution for the <em>CbGpPWpG</em> metapath — traversing Compound–binds–Gene–participates–Pathway–participates–Gene, selected for illustrative purposes.
Metrics are computed for degree-groups,
which is a specific pair of source degree (in this case, the source compound’s count of CbG edges)
and target degree (in this case, the target gene’s count of GpPW edges).
On the left, metrics are reported for the unpermuted hetnet and on the right for the 200 permuted hetnets.
Hence, each cell on the right summarizes 200 times the number of DWPCs as the corresponding cell on the left.
The colormap is row normalized, such that its intensity peaks for the maximum value of each metric across the unpermuted and permuted values.
Gray indicates null values.</figcaption>
</figure>
</div>
<h3 id="gamma-hurdle-distribution">Gamma-hurdle distribution</h3>
<p>We are interested in identifying source and target nodes whose connectivity exceeds what typically arises at random.
To identify such especially-connected nodes, we compare DWPC values to the distribution of permuted network DWPC values for the same source and target nodes.
While a single DWPC value is not actually a test statistic, we use a framework akin to classical hypothesis testing to identify outliers.</p>
<p>Two observations led us to the quasi-significance testing framework we developed.
First, a sizable fraction of permuted DWPC values is often zero, indicating that the source and target nodes are not connected along the metapath in the permuted network.
Second, we observed that non-zero DWPC values for any given source and target nodes are reasonably approximated as following a gamma distribution.
Motivated by these observations, we parametrized permuted DWPC values using a zero-inflated gamma distribution, which we termed the gamma-hurdle distribution.
We fit a gamma-hurdle distribution to each combination of source node, target node, and metapath.
Finally, we estimated the probability of observing a permuted DWPC value greater than DWPC computed in the unpermuted network, akin to a one-tailed p-value.
These quasi-significance scores (‘<em>p</em>-values’) allow us to identify outlier node pairs at the metapath level (see examples in Figure <a href="#fig:null-dwpc-distributions">5</a>).</p>
<div id="fig:null-dwpc-distributions" class="fignos">
<figure>
<img src="https://github.com/greenelab/connectivity-search-analyses/raw/9287986f331607cfdbc1ac197b52f36085723c6e/explore/gamma-hurdle/gamma-hurdle-distributions.png" alt="Figure 5: From null distribution to p-value for DWPCs. Null DWPC distributions are shown for three metapaths between Alzheimer’s disease and the circadian rhythm pathway, selected from Figure 2. For each metapath, null DWPCs are computed on 200 permuted hetnets and grouped according to source–target degree. Histograms show the null DWPCs for the degree group corresponding to Alzheimer’s disease and the circadian rhythm pathway (as noted in the plot titles by deg). The proportion of null DWPCs that were zero is calculated, forming the “hurdle” of the null distribution model. The nonzero null DWPCs are modeled using a gamma distribution, which can be fit solely from a sample mean and standard deviation. The mean of nonzero null DWPCs is denoted with a diamond, with the standard deviation plotted twice as a line in either direction. Actual DWPCs are compared to the gamma-hurdle null distribution to yield a p-value." />
<figcaption aria-hidden="true"><span>Figure 5:</span> <strong>From null distribution to <em>p</em>-value for DWPCs.</strong>
Null DWPC distributions are shown for three metapaths between Alzheimer’s disease and the circadian rhythm pathway, selected from Figure <a href="#fig:webapp-metapaths">2</a>.
For each metapath, null DWPCs are computed on 200 permuted hetnets and grouped according to source–target degree.
Histograms show the null DWPCs for the degree group corresponding to Alzheimer’s disease and the circadian rhythm pathway (as noted in the plot titles by deg).
The proportion of null DWPCs that were zero is calculated, forming the “hurdle” of the null distribution model.
The nonzero null DWPCs are modeled using a gamma distribution, which can be fit solely from a sample mean and standard deviation.
The mean of nonzero null DWPCs is denoted with a diamond, with the standard deviation plotted twice as a line in either direction.
Actual DWPCs are compared to the gamma-hurdle null distribution to yield a <em>p</em>-value.</figcaption>
</figure>
</div>
<h4 id="details-of-the-gamma-hurdle-distribution">Details of the gamma-hurdle distribution</h4>
<p>Let <em>X</em> be a gamma-hurdle random variable with parameters <em>λ</em>, <em>α</em>, and <em>β</em>.</p>
<p><span class="math display">\[
X \sim \Gamma_H(\lambda, \alpha, \beta)
\]</span></p>
<p>The gamma-hurdle distribution is defined over the support [0, ∞).
The probability of a draw, X, from the gamma-hurdle distribution is given as follows:</p>
<p><span class="math display">\[\begin{align*}
  P(X = 0) &amp;= 1 - \lambda \\
  P(X \in A; A \subseteq (0, \infty)) &amp;= \frac{\lambda \beta^\alpha}{\Gamma(\alpha)}  \int  _{x \in A} \bigg( x^{\alpha - 1} e^{-\beta x} \bigg)
\end{align*}\]</span></p>
<p>We estimate all three parameters using the method of moments (using Bessel’s correction to estimate the second moment).
As a validation of our method,
we <a href="https://github.com/greenelab/connectivity-search-analyses/blob/1c6827ce2544c17cef42bbccf098a312f2c44f97/explore/gamma-hurdle/parameter_estimates.ipynb">compared</a> our method of moments parameter estimates to approximate maximum likelihood estimates
(gamma distribution parameters do not have closed-form maximum likelihood estimates)
and found excellent concordance between the methods.
Let <em>N</em> be the number of permuted DWPC values, and <em>n</em> the number of nonzero values.</p>
<p><span class="math display">\[\begin{align*}
  \hat{\lambda} &amp;= \frac{n}{N} \\
  \hat{\alpha} &amp;= \frac{(n - 1) \sum x_i}{n \sum (x_i^2) - (\sum x_i)^2} \\
  \hat{\beta} &amp;= \frac{n - 1}{n} \frac{(\sum x_i)^2}{n \sum (x_i)^2 - (\sum x_i)^2}
\end{align*}\]</span></p>
<p>Finally, we compute a p-value for each DWPC value, <em>t</em>.</p>
<p><span class="math display">\[
p = P(X ≥ t) = \frac{\beta^\alpha}{\Gamma(\alpha)} \int_t^\infty x^{\alpha - 1} \exp(-\beta x) dx
\]</span></p>
<!-- references
  https://nbviewer.jupyter.org/github/greenelab/connectivity-search-analyses/blob/025bdf8d5e63725ca2482d61fd8e421bf0001f93/explore/gamma-hurdle/gamma-heatmaps.ipynb
  https://nbviewer.jupyter.org/github/greenelab/connectivity-search-analyses/blob/0d83ec6063001d7b5cfcfa6a9fe3765bbe109aea/explore/gamma-hurdle/gamma-fit.ipynb
  https://github.com/greenelab/connectivity-search-analyses/pull/157
  https://github.com/greenelab/connectivity-search-analyses/issues/123
-->
<h3 id="empirical-dwpc-p-values">Empirical DWPC p-values</h3>
<p>We <a href="https://github.com/hetio/hetmatpy/blob/bc36aa9859c43a1a5fb22808cd6eb952ef9d497c/hetmatpy/pipeline.py#L92-L113" title="hetmatpy.pipeline.calculate_empirical_p_value source code">calculate</a> an empirical p-value for special cases where the gamma-hurdle model cannot be applied.
These cases include when the observed DWPC is zero or when the null DWPC distribution is all zeroes or has only a single distinct nonzero value.
The empirical <em>p</em>-value (<em>p<sub>empiric</sub></em>) equals the proportion of null DPWCs ≥ the observed DWPC.</p>
<p>Since we do not store all null DWPC values,
we apply the following criteria to calculate <em>p<sub>empiric</sub></em> from summary statistics:</p>
<ol type="1">
<li>When the observed DWPC = 0 (no paths of the specified metapath existed between the source and target node),
<em>p<sub>empiric</sub></em> = 1.</li>
<li>When all null DWPCs are zero but the observed DWPC is positive,
<em>p<sub>empiric</sub></em> = 0.</li>
<li>When all nonzero null DWPCs have the same positive value (standard deviation = 0),
<em>p<sub>empiric</sub></em> = 0 if the observed DWPC &gt; the null DWPC else <em>p<sub>empiric</sub></em> = proportion of nonzero null DWPCs.</li>
</ol>
<h3 id="dwpc-and-null-distribution-computation">DWPC and null distribution computation</h3>
<p>We decided to compute DWPCs and their significance for all source–target node pairs for metapaths with length ≤ 3.
On Hetionet v1.0, there are 24 metapaths of length 1, 242 metapaths of length 2, and 1,939 metapaths of length 3.
The decision to stop at length 3 was one of practicality, as length 4 would have added 17,511 metapaths.</p>
<p>For each of the 2,205 <a href="https://github.com/greenelab/connectivity-search-analyses/raw/042063fb559048e52b3dc2731b6d6c6836f698cf/explore/bulk-pipeline/archives/metapath-dwpc-stats.tsv">metapaths</a>,
we <a href="https://github.com/greenelab/connectivity-search-analyses/blob/1c6827ce2544c17cef42bbccf098a312f2c44f97/explore/bulk-pipeline/bulk.ipynb">computed</a> the complete path count matrix and DWPC matrix.
In total, we computed 137,786,767,964 path counts (and the same number of DWPCs) on the unpermuted network,
of which 11.6% were nonzero.</p>
<p>The DWPC has a single parameter, called the damping exponent (<em>w</em>), which controls how much paths through high-degree nodes are downweighted <span class="citation" data-cites="WkPlH1ds">[<a href="#ref-WkPlH1ds" role="doc-biblioref">42</a>]</span>.
When <em>w</em> = 0, the DWPC is equivalent to the path count.
Previously, we found <em>w</em> = 0.4 was optimal for predicting disease-associated genes <span class="citation" data-cites="WkPlH1ds">[<a href="#ref-WkPlH1ds" role="doc-biblioref">42</a>]</span>.
Here, we use <em>w</em> = 0.5, since taking the square root of degrees has more intuitive appeal.</p>
<p>We selected data types for matrix values that would allow for high precision.
We used 64-bit unsigned integers for path counts and 64-bit floating-point numbers for DWPCs.
We <a href="https://github.com/greenelab/connectivity-search-analyses/pull/91">considered</a> using 16 bits or 32 bits per DWPC to reduce memory/storage requirements,
but decided against it in case certain applications required greater precision.</p>
<p>We used SciPy sparse for path count and DWPC matrices with density &lt; 0.7, serialized to disk with compression and a <code>.sparse.npz</code> extension.
This format minimizes the space on disk and load time for the entire matrix but does not offer read access to slices.
We used Numpy 2 dimensional arrays for DWPC matrices with density ≥ 0.7, serialized to disk using Numpy’s <code>.npy</code> format.
We bundled the path count and DWPC matrix files into HetMat archives by metapath length and deposited the archives to Zenodo <span class="citation" data-cites="pkJqqM2W">[<a href="#ref-pkJqqM2W" role="doc-biblioref">52</a>]</span>.
The archive for length 3 DWPCs was the largest at 131.7 GB.</p>
<p>We also generated null DWPC summary statistics for the 2,205 metapaths,
which are also available by metapath length from Zenodo as HetMat archives consisting of <code>.tsv.gz</code> files <span class="citation" data-cites="pkJqqM2W">[<a href="#ref-pkJqqM2W" role="doc-biblioref">52</a>]</span>.
Due to degree-grouping, null DWPC summary statistic archives are much smaller than the DWPC archives.
The archive for length 3 null DWPCs summary statistics was 733.1 MB.
However, the compute required to generate null DWPCs is far greater because there are multiple permuted hetnets (in our case 200).
As a result, computing and saving all DWPCs took 6 hours,
whereas computing and saving the null DWPC summary statistics took 361 hours.</p>
<p>Including null DWPCs and path counts, the Zenodo deposit totals 185.1 GB and contains the results of computing ~28 trillion DWPCs — 27,832,927,128,728 to be exact.</p>
<h3 id="adjusting-dwpc-p-values">Adjusting DWPC <em>p</em>-values</h3>
<p>When a user applies hetnet connectivity search to identify enriched metapaths between two nodes,
many metapaths are evaluated for significance.
Due to multiple testing of many DWPCs, low <em>p</em>-values are likely to arise by chance.
Therefore, we devised a multiple-testing correction.</p>
<p>For each combination of source metanode, target metanode, and length, we counted the number of metapaths.
For Disease…Pathway metapaths, there are 0 metapaths of length 1, 3 metapaths of length 2, and 24 metapaths of length 3.
We calculated adjusted p-values by applying a Bonferroni correction based on the number of metapaths of the same length between the source and target metanode.
Using Figure <a href="#fig:webapp-metapaths">2</a> as an example, the <a href="https://het.io/search/?source=17287&amp;target=7607&amp;metapaths=DdGpPW&amp;complete=">DdGpPW</a> p-value of 5.9% was adjusted to 17.8% (multiplied by a factor of 3).</p>
<p>Bonferroni controls family-wise error rate,
which corresponds here to incorrectly finding that <em>any</em> metapath of a given length is enriched.
As a result, our adjusted p-values are conservative.
We would prefer to adjust <em>p</em>-values for false discovery rate <span class="citation" data-cites="q6QyrHEZ">[<a href="#ref-q6QyrHEZ" role="doc-biblioref">53</a>]</span>,
but these methods often require access to all <em>p</em>-values at once (impractical here) and assume a uniform distribution of <em>p</em>-values when there is no signal (not the case here when most DWPCs are zero).</p>
<h3 id="prioritizing-enriched-metapaths-for-database-storage">Prioritizing enriched metapaths for database storage</h3>
<p>Storing DWPCs and their significance in the database (as part of the <code>PathCount</code> table in Figure <a href="#fig:database">6</a>) enables the connectivity search webapp to provide users with enriched metapaths between query nodes in real time.
However, storing ~15.9 billion rows (the total number of nonzero DWPCs) in the database’s <code>PathCount</code> table would exceed a reasonable disk quota.
<!--
# python code to estimate the total number of nonzero DWPCs
url ="https://github.com/greenelab/connectivity-search-analyses/raw/042063fb559048e52b3dc2731b6d6c6836f698cf/explore/bulk-pipeline/archives/metapath-dwpc-stats.tsv"
metapath_df = pandas.read_table(url)
(metapath_df.n_pairs * metapath_df.pc_density).round().sum()
-->
An alternative would be to store all DWPCs in the database whose adjusted p-value exceeded a universal threshold (e.g. <em>p</em> &lt; 5%).
But we estimated this would still be prohibitively expensive.
Therefore, we devised a metapath-specific threshold.
For metapaths with length 1, we stored all nonzero DWPCs,
assuming users always want to be informed about direct edges between the query nodes, regardless of significance.
For metapaths with length ≥ 2, we chose an adjusted <em>p</em>-value threshold of
5 × (<em>n<sub>source</sub></em> × <em>n<sub>target</sub></em>)<sup>−0.3</sup>,
where <em>n<sub>source</sub></em> and <em>n<sub>target</sub></em> are the node counts for the source and target metanodes (i.e. “Nodes” column in Table <a href="#tbl:metanodes">1</a>).
Notice that metapaths with large number of possible source–target pairs (large DWPC matrices) are penalized.
This decision is based on practicality since otherwise the majority of the database quota would be consumed by a minority of metapaths between plentiful metanodes (e.g. Gene…Gene metapaths).
Also, we assume that users will search nodes at a similar rate by metanode (e.g. they’re more likely to search for a specific disease than a specific gene).
The constants in the threshold formula help scale it.
The multiplier of 5 relaxes the threshold to saturate the available database capacity.
The −0.3 exponent applies the large DWPC-matrix penalty.</p>
<p>Users can still evaluate DWPCs that are not stored in the database, using either the webapp or API.
These are calculated on the fly, delegating DWPC computation to the Neo4j database.
Unchecking “precomputed only” on the webapp shows all possible metapaths for two query nodes.
For some node pairs, the on-the-fly computation is quick (less than a second).
Other times, computing DWPCs for all metapaths might take more than a minute.</p>
<!--
https://github.com/greenelab/connectivity-search-analyses/blob/cad458b29f508c66b4b14cdd641db6855426221b/explore/metapath-thresholds/metapath-thresholds.ipynb

https://github.com/greenelab/connectivity-search-backend/blob/af12f8cf2ad47d9a25ce8d1b7889390654eb3bb9/dj_hetmech_app/management/commands/populate_database.py#L139-L146
p-value threshold:

        p_threshold = 5 * row.n_pairs ** -0.3 / row.n_similar
-->
<h3 id="backend-database-api">Backend Database &amp; API</h3>
<p>We created a backend application using Python’s Django web framework.
The source code is available in the <a href="https://github.com/greenelab/connectivity-search-backend"><code>connectivity-search-backend</code></a> repository.
The primary role of the backend is to manage a relational database and provide an API for requesting data.</p>
<p>We define the database schema <a href="https://github.com/greenelab/connectivity-search-backend/blob/af12f8cf2ad47d9a25ce8d1b7889390654eb3bb9/dj_hetmech_app/models.py" title="models.py for the dj_hetmech_app">using</a> Django’s object-relational mapping framework (Figure <a href="#fig:database">6</a>).
We <a href="https://github.com/greenelab/connectivity-search-backend/blob/af12f8cf2ad47d9a25ce8d1b7889390654eb3bb9/dj_hetmech_app/management/commands/populate_database.py" title="dj_hetmech_app populate_database management command">import</a> the data into a PostgreSQL database.
Populating the database for all 2,205 metapaths up to length 3 was a prolonged operation, <a href="https://github.com/greenelab/connectivity-search-backend/pull/41#issuecomment-488054789">taking</a> over 3 days.
The majority of the time is spent populating the <code>DegreeGroupedPermutation</code> (37,905,389 rows) and <code>PathCount</code> (174,986,768 rows) tables.
To avoid redundancy, the database only stores a single orientation of a metapath.
For example, if rows are stored for the <em>GpPWpGaD</em> metapath,
they would not also be stored for the <em>DaGpPWpG</em> metapath.
The backend is responsible for checking both orientations of a metapath in the database and reversing metapaths on-the-fly before returning results.
The database is located at <code>search-db.het.io</code> with public read-only access <a href="https://github.com/greenelab/connectivity-search-backend#database">available</a>.</p>
<div id="fig:database" class="fignos">
<figure>
<img src="https://github.com/greenelab/connectivity-search-backend/raw/752b423a4b7b57575d66ce0b797b0a84c23267a6/media/models-schema.png" style="width:100.0%" alt="Figure 6: Schema for the connectivity search backend relational database models. Each Django model is represented as a table, whose rows list the model’s field names and types. Each model corresponds to a database table. Arrows denote foreign key relationships. The arrow labels indicate the foreign key field name followed by reverse relation names generated by Django (in parentheses)." />
<figcaption aria-hidden="true"><span>Figure 6:</span> <strong>Schema for the connectivity search backend relational database models.</strong>
Each Django model is represented as a table,
whose rows list the model’s field names and types.
Each model corresponds to a database table.
Arrows denote foreign key relationships.
The arrow labels indicate the foreign key field name followed by reverse relation names generated by Django (in parentheses).</figcaption>
</figure>
</div>
<p>We host a public API instance at <a href="https://search-api.het.io" class="uri">https://search-api.het.io</a>.
Version 1 of the API exposes several endpoints that are used by the connectivity search frontend including queries for
node details (<code>/v1/node</code>),
node lookup (<code>/v1/nodes</code>),
metapath information (<code>/v1/metapaths</code>),
and path information (<code>/v1/paths</code>).
The endpoints return JSON payloads.
Producing results for these queries relies on internal calls to the PostgreSQL relational database as well as the pre-existing Hetionet v1.0 Neo4j graph database.
They were designed to power the hetnet connectivity search webapp,
but are also available for general research use.</p>
<h3 id="frontend">Frontend</h3>
<h4 id="hetio-website">Hetio Website</h4>
<p>We created a static website to serve as the home for the Hetio organization using Jekyll hosted on GitHub Pages (Figure <a href="#fig:website">7</a>).
The source code is available in the <a href="https://github.com/hetio/het.io"><code>het.io</code></a> repository.
To make a change to the website, an author simply commits the changes (either directly or through a pull request) to the repository’s <code>gh-pages</code> branch, and GitHub automatically recompiles the website and hosts the resulting files at the provided custom domain URL.</p>
<div id="fig:website" class="fignos">
<figure>
<img src="https://github.com/greenelab/connectivity-search-manuscript/raw/d5766924e8c774accdc143bea352e49610ee0673/content/media/website/website-homepage.png" class="white" style="width:100.0%" alt="Figure 7: Homepage of the Hetio website. The redesigned homepage provides a succinct overview of what Hetionet consists of and what its purpose is." />
<figcaption aria-hidden="true"><span>Figure 7:</span> <strong>Homepage of the Hetio website.</strong>
The redesigned homepage provides a succinct overview of what Hetionet consists of and what its purpose is.</figcaption>
</figure>
</div>
<h4 id="webapps">Webapps</h4>
<p>We developed the <a href="https://het.io/search">connectivity search app</a> as a single-page, standalone application using React and associated tools.
The source code is available in the <a href="https://github.com/greenelab/connectivity-search-frontend"><code>connectivity-search-frontend</code></a> repository.</p>
<p>Since the rest of the overarching Hetio website was mostly noninteractive content, it was appropriate to construct the bulk of the website in simple static formats like Markdown and HTML using Jekyll, and leave React for implementing the sections of the site that required more complex behavior.</p>
<p>We used React’s own <code>create-react-app</code> command-line tool to generate a boilerplate for the app.
This simplified setup, testing, and building pipelines, bypassing time-consuming configuration of things like Webpack and linters.
Some configuration was necessary to produce non-hashed, consistently named output files like <code>index.js</code> that could be easily and reliably referenced by and embedded into the Hetio Jekyll website.</p>
<p>For authoring components, we used React’s traditional class syntax.
At the time of authoring the app, React Hooks were still nascent, and thus the simpler and less-verbose functional syntax was not viable.</p>
<p>While writing this application, we also elected to rewrite the preexisting <a href="https://het.io/repurpose/">Rephetio</a> and <a href="https://het.io/disease-genes/">disease-associated genes</a> apps in the same manner.
We created a custom package of React components and utility functions that could be shared across the multiple interactive apps on the website.
The package is located at and can be installed from the <a href="https://github.com/hetio/frontend-components"><code>frontend-components</code></a> repository.
The package consists of interface “components” (building blocks) like buttons and sortable/searchable/paginated tables as well as utility functions for formatting data and debugging.
Each of the interactive apps imports this package to reduce code repetition and to enforce a consistent style and behavior across the website.</p>
<p>For managing state in the connectivity search app, we used the Redux library.
Redux was chosen over vanilla React or other state management libraries since:</p>
<ol type="1">
<li>The state in this app was very “global”, meaning most of it was needed by a lot of different parts of the app.
Redux provides a convenient global “store” of state that is easily accessible to any component, avoiding the “prop-drilling” phenomenon.</li>
<li>The structure of the state is nested and complex.
Redux’s “reducer” approach makes it cleaner to modify this type of data.</li>
<li>Redux’s approach to immutable state that is updated by actions and pure functions makes the application easier to debug.
It is easy to get a clear timeline of how and when the state changed, and what changed it.</li>
</ol>
<p>To create the graph visualization at the bottom of the app, the D3 library was used.
D3 satisfied several core requirements:</p>
<ol type="1">
<li>SVG implementation for high-resolution, publication-ready figures.</li>
<li>Force-directed layout for untangling nodes.</li>
<li>Pinnable nodes and other physics customizations.</li>
<li>Customizable node and edge drag/hover/select behavior.</li>
<li>Intuitive pan/zoom view that worked on desktop and mobile.</li>
<li>Node and edge appearances that were completely customizable for alignment, text wrapping, color, outlines, fonts, arrowheads, and noncolliding coincident edges.</li>
</ol>
<h3 id="visual-design">Visual Design</h3>
<p>A limited palette of colors was chosen to represent the different types of nodes (metanodes) in the Hetionet knowledge graph.
These colors are listed and programmatically accessible in the <a href="https://github.com/hetio/hetionet"><code>hetionet</code></a> repository under <code>/describe/styles.json</code>.</p>
<p>At the time of developing connectivity search, Hetionet already had an established palette of colors (from Project Rephetio).
To avoid confusion, we were <a href="https://github.com/hetio/hetionet/pull/18">careful</a> to keep the general hue of each metanode color the same for backward compatibility,
e.g. genes stayed generally blue, diseases stayed generally brown.
In this way, this palette selection was more of a modernization/refresh.
For cohesiveness, accessibility, and aesthetic appeal,
we used the professionally curated Material Design palette as a source for the specific color values.</p>
<p>The palette is now used in all Hetio-related applications and materials.
This is not just to maintain a consistent look and feel across the Hetio organization, but to convey clear and precise meaning.
For example, the colors used in the metagraph in Figure <a href="#fig:rephetio">1</a>A are exactly the same colors, and thus represent the same types of entities, as in <em>any part of the connectivity search app</em> (Figure <a href="#fig:webapp">3</a>).</p>
<p>Colors in the palette are also used in the Hetio logo (seen in Figure <a href="#fig:website">7</a>) and other miscellaneous logos and iconography across the website, to establish an identifiable brand for the Hetio organization as a whole.</p>
<h3 id="real-time-open-science">Real-time open science</h3>
<p>This study was conducted entirely in the open via public GitHub repositories.
We used GitHub Issues for discussion, leaving a rich online history of the scholarly process.
Furthermore, most additions to the analyses were performed by pull request,
whereby a contributor proposes a set of changes.
This provides an opportunity for other contributors to review changes before they are officially accepted.
For example, in <a href="https://github.com/greenelab/connectivity-search-analyses/pull/156">greenelab/connectivity-search-analyses#156</a> @zietzm proposed a notebook to visualize parameters for null DWPC distributions.
After @zietzm addressed @dhimmel’s comments, the pull request was approved and merged into the project’s main branch.</p>
<p>The manuscript for this study was written using <a href="https://manubot.org/">Manubot</a>, which allows authors to collaboratively write manuscripts on GitHub <span class="citation" data-cites="YuJbg3zO">[<a href="#ref-YuJbg3zO" role="doc-biblioref">54</a>]</span>.
The Manubot-rendered manuscript is available at <a href="https://greenelab.github.io/connectivity-search-manuscript/" class="uri">https://greenelab.github.io/connectivity-search-manuscript/</a>.
We encourage readers with feedback or questions to comment publicly via <a href="https://github.com/greenelab/connectivity-search-manuscript/issues">GitHub Issues</a>.</p>
<h3 id="software-data-availability">Software &amp; data availability</h3>
<p><em>Hetio</em> is a superset/collection of hetnet-related research, tools, and datasets that, most notably, includes the Hetionet project itself and the connectivity search tool that are the focus of this article.
Most of the Hetio resources and projects can be found under the <a href="https://github.com/hetio">Hetio GitHub organization</a>, with others being available under the <a href="https://github.com/greenelab">Greene Lab GitHub organization</a>, one of the collaborating groups.
Information about Hetio is also displayed and disseminated at <a href="https://het.io" class="uri">https://het.io</a>, as noted in the <a href="#hetio-website">Hetio Website</a> section.</p>
<h4 id="availability-of-supporting-source-code-and-requirements">Availability of Supporting Source Code and Requirements</h4>
<ul>
<li>Project name: Hetnet Connectivity Search</li>
<li>Project homepage: <a href="https://het.io/search/" class="uri">https://het.io/search/</a></li>
<li>Operating systems: Platform independent</li>
<li>Programming language: Python, Javascript, Cypher</li>
<li>Other requirements:
Refer to specific repositories below for their respective dependency configuration files</li>
<li>License:
Refer to specific repositories below,
but generally software is released under BSD,
figures and documentation under CC BY,
and data under CC0.</li>
<li>RRID: <a href="https://scicrunch.org/resolver/RRID:SCR_023630">SCR_023630</a></li>
<li>biotools ID: <a href="https://bio.tools/connectivity-search">connectivity-search</a></li>
</ul>
<p>This study primarily involves the following GitHub repositories:
<!-- https://github.com/topics/hetnet-connectivity-search --></p>
<ul>
<li><a href="https://github.com/greenelab/connectivity-search-manuscript">greenelab/connectivity-search-manuscript</a>
<span class="citation" data-cites="3KJWwgoz">[<a href="#ref-3KJWwgoz" role="doc-biblioref">55</a>]</span>:
Source code for this manuscript.
Best place for general comments or questions.
CC BY 4.0 License.</li>
<li><a href="https://github.com/greenelab/connectivity-search-analyses">greenelab/connectivity-search-analyses</a>
<span class="citation" data-cites="eq4IX1LX">[<a href="#ref-eq4IX1LX" role="doc-biblioref">56</a>]</span>:
The initial project repository that contains research notebooks, dataset generation code, and exploratory data analyses.
The hetmatpy package was first developed as part of this repository until its <a href="https://github.com/hetio/hetmatpy/issues/1">relocation</a> in November 2018.
BSD 3-Clause License.</li>
<li><a href="https://github.com/greenelab/connectivity-search-backend">greenelab/connectivity-search-backend</a>
<span class="citation" data-cites="ASx5cMWI">[<a href="#ref-ASx5cMWI" role="doc-biblioref">57</a>]</span>:
Source code for the connectivity search database and API.
BSD 3-Clause License.</li>
<li><a href="https://github.com/greenelab/connectivity-search-frontend">greenelab/connectivity-search-frontend</a>
<span class="citation" data-cites="CadAZTf2">[<a href="#ref-CadAZTf2" role="doc-biblioref">58</a>]</span>:
Source code for the connectivity search webapp.
BSD 3-Clause License.</li>
<li><a href="https://github.com/hetio/hetmatpy">hetio/hetmatpy</a>
<span class="citation" data-cites="gHk6T3vn">[<a href="#ref-gHk6T3vn" role="doc-biblioref">59</a>]</span>:
Python package for matrix storage and operations on hetnets.
Released on <a href="https://pypi.org/project/hetmatpy/">PyPI</a>.
BSD 2-Clause Plus Patent License.
Registered at <a href="https://bio.tools/hetmatpy">biotools:hetmatpy</a> and <a href="https://scicrunch.org/resolver/RRID:SCR_023409">RRID:SCR_023409</a>.</li>
<li><a href="https://github.com/hetio/hetnetpy">hetio/hetnetpy</a>
<span class="citation" data-cites="srnBddlB">[<a href="#ref-srnBddlB" role="doc-biblioref">60</a>]</span>:
Preexisiting Python package for representing hetnets.
Dependency of hetmatpy.
Released on <a href="https://pypi.org/project/hetnetpy/">PyPI</a>.
Dual licensed under BSD 2-Clause Plus Patent License and CC0 1.0 (public domain dedication).</li>
<li><a href="https://github.com/hetio/hetionet">hetio/hetionet</a>
<span class="citation" data-cites="L9cBdipF">[<a href="#ref-L9cBdipF" role="doc-biblioref">61</a>]</span>:
Preexisiting data repository for Hetionet,
including the public Neo4j instance and HetMat archives.
CC0 1.0 License.</li>
<li><a href="https://github.com/hetio/het.io">hetio/het.io</a>
<span class="citation" data-cites="12IQq3W3u">[<a href="#ref-12IQq3W3u" role="doc-biblioref">62</a>]</span>:
Preexisiting source code for the <a href="https://het.io/" class="uri">https://het.io/</a> website.
CC BY 4.0 License.</li>
</ul>
<h4 id="data-availability">Data availability</h4>
<p>An archival copy of the code and supporting data is available via the GigaScience repository <span class="citation" data-cites="hHRphzV0">[<a href="#ref-hHRphzV0" role="doc-biblioref">63</a>]</span>.
The connectivity-search-analyses and hetionet repositories contain datasets related to this study.
Large datasets were compressed and tracked with <a href="https://git-lfs.github.com/">Git LFS</a> (Large File Storage).
GitHub LFS had a max file size of 2 GB.
Datasets exceeding this size, along with other essential datasets, are available from Zenodo <span class="citation" data-cites="pkJqqM2W">[<a href="#ref-pkJqqM2W" role="doc-biblioref">52</a>]</span>.</p>
<!-- link reference syntax citation key aliases -->
<h2 id="declarations">Declarations</h2>
<h3 id="list-of-abbreviations">List of abbreviations</h3>
<dl>
<dt>Arnt</dt>
<dd>
aryl hydrocarbon receptor nuclear translocator protein
</dd>
<dt>ARNT2</dt>
<dd>
aryl hydrocarbon receptor nuclear translocator 2
</dd>
<dt>DWPC</dt>
<dd>
degree-weighted path count
</dd>
<dt>LFS</dt>
<dd>
large file storage
</dd>
<dt>NPAS2</dt>
<dd>
neuronal PAS domain protein 2
</dd>
<dt>PAS</dt>
<dd>
Per-Arnt-Sim
</dd>
<dt>Per</dt>
<dd>
period circadian protein
</dd>
<dt>Sim</dt>
<dd>
single-minded protein
</dd>
</dl>
<h3 id="competing-interests">Competing Interests</h3>
<p>This work was supported, in part, by Pfizer Worldwide Research, Development, and Medical.</p>
<h3 id="funding">Funding</h3>
<p>DSH, BJH, DH, and DNN were funded by The Gordon and Betty Moore Foundation (GBMF4552).
DSH and CSG were funded by Pfizer Worldwide Research, Development, and Medical.
KK was funded by The Gordon and Betty Moore Foundation (GBMF4560).
DNN was funded by The National Institutes of Health (T32 HG000046).
CSG was funded by the National Human Genome Research Institute (R01 HG010067).
CSG was funded by the National Cancer Institute (R01 CA237170).
CSG was funded by the Gordon and Betty Moore Foundation (GBMF 4552).
CSG was funded by the Eunice Kennedy Shriver National Institute of Child Health and Human Development (R01 HD109765).
The funders had no role in the study design, data analysis and interpretation, or writing of the manuscript.</p>
<h3 id="authors-contributions">Authors’ contributions</h3>
<p>Author contributions are noted here according to <a href="https://credit.niso.org/">CRediT</a> (Contributor Roles Taxonomy).
Conceptualization by DSH, MZ, KK, BDS, and CSG.
Data curation by DSH, MZ, VR, and DNN.
Formal analysis by DSH, MZ, KK, and BJH.
Funding acquisition by DSH, BDS, and CSG.
Investigation by DSH, MZ, BJH, YH, MWN, and CSG.
Methodology by DSH, MZ, VR, KK, BJH, and CSG.
Project administration by DSH and CSG.
Resources by DSH, FSA, DH, MWN, and CSG.
Software by DSH, MZ, VR, BJH, FSA, and DH.
Supervision by DSH, BDS, and CSG.
Visualization by DSH, MZ, VR, and YH.
Writing – original draft by DSH, MZ, VR, DNN, and CSG.
Writing – review &amp; editing by DSH, MZ, VR, DH, BDS, and CSG.
Validation by BJH.</p>
<h2 class="page_break_before" id="references">References</h2>
<!-- Explicitly insert bibliography here -->
<div id="refs" class="references csl-bib-body" role="doc-bibliography">
<div id="ref-QWTAHxFj" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">1. </div><div class="csl-right-inline"><strong>Renaming ‘heterogeneous networks’ to a more concise and catchy term</strong> <div class="csl-block">Daniel Himmelstein, Casey Greene, Sergio Baranzini</div> <em>ThinkLab</em> (2015-08-16) <a href="https://doi.org/f3mn4v">https://doi.org/f3mn4v</a> <div class="csl-block">DOI: <a href="https://doi.org/10.15363/thinklab.d104">10.15363/thinklab.d104</a></div></div>
</div>
<div id="ref-O21tn8vf" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">2. </div><div class="csl-right-inline"><strong>Systematic integration of biomedical knowledge prioritizes drugs for repurposing</strong> <div class="csl-block">Daniel Scott Himmelstein, Antoine Lizee, Christine Hessler, Leo Brueggeman, Sabrina L Chen, Dexter Hadley, Ari Green, Pouya Khankhanian, Sergio E Baranzini</div> <em>eLife</em> (2017-09-22) <a href="https://doi.org/cdfk">https://doi.org/cdfk</a> <div class="csl-block">DOI: <a href="https://doi.org/10.7554/elife.26726">10.7554/elife.26726</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/28936969">28936969</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5640425">PMC5640425</a></div></div>
</div>
<div id="ref-5t7pfaPv" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">3. </div><div class="csl-right-inline"><strong>Announcing PharmacotherapyDB: the Open Catalog of Drug Therapies for Disease</strong> <div class="csl-block">Daniel Himmelstein</div> <em>ThinkLab</em> (2016-03-15) <a href="https://doi.org/f3mqtv">https://doi.org/f3mqtv</a> <div class="csl-block">DOI: <a href="https://doi.org/10.15363/thinklab.d182">10.15363/thinklab.d182</a></div></div>
</div>
<div id="ref-vvdoVrc6" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">4. </div><div class="csl-right-inline"><strong>Our hetnet edge prediction methodology: the modeling framework for Project Rephetio</strong> <div class="csl-block">Daniel Himmelstein</div> <em>ThinkLab</em> (2016-05-04) <a href="https://doi.org/f3qbmj">https://doi.org/f3qbmj</a> <div class="csl-block">DOI: <a href="https://doi.org/10.15363/thinklab.d210">10.15363/thinklab.d210</a></div></div>
</div>
<div id="ref-EfWvuSjX" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">5. </div><div class="csl-right-inline"><strong>The link-prediction problem for social networks</strong> <div class="csl-block">David Liben-Nowell, Jon Kleinberg</div> <em>Journal of the American Society for Information Science and Technology</em> (2007) <a href="https://doi.org/c56765">https://doi.org/c56765</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1002/asi.20591">10.1002/asi.20591</a></div></div>
</div>
<div id="ref-7iZtCGcX" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">6. </div><div class="csl-right-inline"><strong>Link prediction in complex networks: A survey</strong> <div class="csl-block">Linyuan Lü, Tao Zhou</div> <em>Physica A: Statistical Mechanics and its Applications</em> (2011-03) <a href="https://doi.org/dbs8g8">https://doi.org/dbs8g8</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1016/j.physa.2010.11.027">10.1016/j.physa.2010.11.027</a></div></div>
</div>
<div id="ref-otY29wFV" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">7. </div><div class="csl-right-inline"><strong>Heterogeneous network embedding for identifying symptom candidate genes</strong> <div class="csl-block">Kuo Yang, Ning Wang, Guangming Liu, Ruyu Wang, Jian Yu, Runshun Zhang, Jianxin Chen, Xuezhong Zhou</div> <em>Journal of the American Medical Informatics Association</em> (2018-10-23) <a href="https://doi.org/gfg6nr">https://doi.org/gfg6nr</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1093/jamia/ocy117">10.1093/jamia/ocy117</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/30357378">30357378</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7646926">PMC7646926</a></div></div>
</div>
<div id="ref-NnOS86ev" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">8. </div><div class="csl-right-inline"><strong>Large-scale structural and textual similarity-based mining of knowledge graph to predict drug–drug interactions</strong> <div class="csl-block">Ibrahim Abdelaziz, Achille Fokoue, Oktie Hassanzadeh, Ping Zhang, Mohammad Sadoghi</div> <em>Journal of Web Semantics</em> (2017-05) <a href="https://doi.org/gcrwk3">https://doi.org/gcrwk3</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1016/j.websem.2017.06.002">10.1016/j.websem.2017.06.002</a></div></div>
</div>
<div id="ref-GYLacVTV" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">9. </div><div class="csl-right-inline"><strong>SMR: Medical Knowledge Graph Embedding for Safe Medicine Recommendation</strong> <div class="csl-block">Fan Gong, Meng Wang, Haofen Wang, Sen Wang, Mengyue Liu</div> <em>Big Data Research</em> (2021-02) <a href="https://doi.org/gjqwnc">https://doi.org/gjqwnc</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1016/j.bdr.2020.100174">10.1016/j.bdr.2020.100174</a></div></div>
</div>
<div id="ref-3gzmoW2V" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">10. </div><div class="csl-right-inline"><strong>PyKEEN 1.0: A Python Library for Training and Evaluating Knowledge Graph Embeddings</strong> <div class="csl-block">Mehdi Ali, Max Berrendorf, Charles Tapley Hoyt, Laurent Vermue, Sahand Sharifzadeh, Volker Tresp, Jens Lehmann</div> <em>Journal of Machine Learning Research</em> (2021) <a href="http://jmlr.org/papers/v22/20-825.html">http://jmlr.org/papers/v22/20-825.html</a></div>
</div>
<div id="ref-zHcAMCig" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">11. </div><div class="csl-right-inline"><strong>Understanding the performance of knowledge graph embeddings in drug discovery</strong> <div class="csl-block">Stephen Bonner, Ian P Barrett, Cheng Ye, Rowan Swiers, Ola Engkvist, Charles Tapley Hoyt, William L Hamilton</div> <em>Artificial Intelligence in the Life Sciences</em> (2022-12) <a href="https://doi.org/gqxmj9">https://doi.org/gqxmj9</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1016/j.ailsci.2022.100036">10.1016/j.ailsci.2022.100036</a></div></div>
</div>
<div id="ref-U1Dfux5r" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">12. </div><div class="csl-right-inline"><strong>node2vec: Scalable Feature Learning for Networks</strong> <div class="csl-block">Aditya Grover, Jure Leskovec</div> <em>Proceedings of the 22nd ACM SIGKDD International Conference on Knowledge Discovery and Data Mining</em> (2016-08-13) <a href="https://doi.org/gftdzj">https://doi.org/gftdzj</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1145/2939672.2939754">10.1145/2939672.2939754</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/27853626">27853626</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5108654">PMC5108654</a></div></div>
</div>
<div id="ref-1AeZs6xaT" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">13. </div><div class="csl-right-inline"><strong>metapath2vec: Scalable Representation Learning for Heterogeneous Networks</strong> <div class="csl-block">Yuxiao Dong, Nitesh V Chawla, Ananthram Swami</div> <em>KDD '17: Proceedings of the 23rd ACM SIGKDD International Conference on Knowledge Discovery and Data Mining</em> (2017-08-04) <a href="https://doi.org/gfsqzn">https://doi.org/gfsqzn</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1145/3097983.3098036">10.1145/3097983.3098036</a></div></div>
</div>
<div id="ref-jyhFBVzQ" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">14. </div><div class="csl-right-inline"><strong>edge2vec: Representation learning using edge semantics for biomedical knowledge discovery</strong> <div class="csl-block">Zheng Gao, Gang Fu, Chunping Ouyang, Satoshi Tsutsui, Xiaozhong Liu, Jeremy Yang, Christopher Gessner, Brian Foote, David Wild, Ying Ding, Qi Yu</div> <em>BMC Bioinformatics</em> (2019-06-10) <a href="https://doi.org/ggpcsq">https://doi.org/ggpcsq</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1186/s12859-019-2914-2">10.1186/s12859-019-2914-2</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/31238875">31238875</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6593489">PMC6593489</a></div></div>
</div>
<div id="ref-xq3Nsf5M" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">15. </div><div class="csl-right-inline"><strong>Preclinical validation of therapeutic targets predicted by tensor factorization on heterogeneous graphs</strong> <div class="csl-block">Saee Paliwal, Alex de Giorgio, Daniel Neil, Jean-Baptiste Michel, Alix MB Lacoste</div> <em>Scientific Reports</em> (2020-10-26) <a href="https://doi.org/gg49">https://doi.org/gg49</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1038/s41598-020-74922-z">10.1038/s41598-020-74922-z</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/33106501">33106501</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7589557">PMC7589557</a></div></div>
</div>
<div id="ref-1EeE9qtFA" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">16. </div><div class="csl-right-inline"><strong>Data Fusion by Matrix Factorization</strong> <div class="csl-block">Marinka Zitnik, Blaz Zupan</div> <em>IEEE Transactions on Pattern Analysis and Machine Intelligence</em> (2015-01-01) <a href="https://doi.org/f3mqwz">https://doi.org/f3mqwz</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1109/tpami.2014.2343973">10.1109/tpami.2014.2343973</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/26353207">26353207</a></div></div>
</div>
<div id="ref-cnCLKeU9" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">17. </div><div class="csl-right-inline"><strong>Translating embeddings for modeling multi-relational data</strong> <div class="csl-block">Antoine Bordes, Nicolas Usunier, Alberto Garcia-Durán, Jason Weston, Oksana Yakhnenko</div> <em>Proceedings of the 26th International Conference on Neural Information Processing Systems</em> (2013-12-05) <a href="https://dl.acm.org/doi/10.5555/2999792.2999923">https://dl.acm.org/doi/10.5555/2999792.2999923</a></div>
</div>
<div id="ref-vhsYX2y" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">18. </div><div class="csl-right-inline"><strong>Integrating and formatting biomedical data as pre-calculated knowledge graph embeddings in the Bioteque</strong> <div class="csl-block">Adrià Fernández-Torras, Miquel Duran-Frigola, Martino Bertoni, Martina Locatelli, Patrick Aloy</div> <em>Nature Communications</em> (2022-09-09) <a href="https://doi.org/gqtdvj">https://doi.org/gqtdvj</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1038/s41467-022-33026-0">10.1038/s41467-022-33026-0</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/36085310">36085310</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC9463154">PMC9463154</a></div></div>
</div>
<div id="ref-taI1UUAE" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">19. </div><div class="csl-right-inline"><strong>Predicting gene-disease associations from the heterogeneous network using graph embedding</strong> <div class="csl-block">Xiaochan Wang, Yuchong Gong, Jing Yi, Wen Zhang</div> <em>2019 IEEE International Conference on Bioinformatics and Biomedicine (BIBM)</em> (2019-11) <a href="https://doi.org/ggmrpj">https://doi.org/ggmrpj</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1109/bibm47256.2019.8983134">10.1109/bibm47256.2019.8983134</a></div></div>
</div>
<div id="ref-17ioZ6znu" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">20. </div><div class="csl-right-inline"><strong>A Method to Learn Embedding of a Probabilistic Medical Knowledge Graph: Algorithm Development</strong> <div class="csl-block">Linfeng Li, Peng Wang, Yao Wang, Shenghui Wang, Jun Yan, Jinpeng Jiang, Buzhou Tang, Chengliang Wang, Yuting Liu</div> <em>JMIR Medical Informatics</em> (2020-05-21) <a href="https://doi.org/ghqszs">https://doi.org/ghqszs</a> <div class="csl-block">DOI: <a href="https://doi.org/10.2196/17645">10.2196/17645</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/32436854">32436854</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7273238">PMC7273238</a></div></div>
</div>
<div id="ref-1D9FTzRBg" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">21. </div><div class="csl-right-inline"><strong>Semantic Disease Gene Embeddings (SmuDGE): phenotype-based disease gene prioritization without phenotypes</strong> <div class="csl-block">Mona Alshahrani, Robert Hoehndorf</div> <em>Bioinformatics</em> (2018-09-01) <a href="https://doi.org/gd9k8n">https://doi.org/gd9k8n</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1093/bioinformatics/bty559">10.1093/bioinformatics/bty559</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/30423077">30423077</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6129260">PMC6129260</a></div></div>
</div>
<div id="ref-6PISrkV5" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">22. </div><div class="csl-right-inline"><strong>A network embedding model for pathogenic genes prediction by multi-path random walking on heterogeneous network</strong> <div class="csl-block">Bo Xu, Yu Liu, Shuo Yu, Lei Wang, Jie Dong, Hongfei Lin, Zhihao Yang, Jian Wang, Feng Xia</div> <em>BMC Medical Genomics</em> (2019-12) <a href="https://doi.org/ggmrpq">https://doi.org/ggmrpq</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1186/s12920-019-0627-z">10.1186/s12920-019-0627-z</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/31865919">31865919</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6927107">PMC6927107</a></div></div>
</div>
<div id="ref-19E33rJiu" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">23. </div><div class="csl-right-inline"><strong>Deep mining heterogeneous networks of biomedical linked data to predict novel drug–target associations</strong> <div class="csl-block">Nansu Zong, Hyeoneui Kim, Victoria Ngo, Olivier Harismendy</div> <em>Bioinformatics</em> (2017-04-18) <a href="https://doi.org/gbqjgx">https://doi.org/gbqjgx</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1093/bioinformatics/btx160">10.1093/bioinformatics/btx160</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/28430977">28430977</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5860112">PMC5860112</a></div></div>
</div>
<div id="ref-Dbie18xe" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">24. </div><div class="csl-right-inline"><strong>Explaining and Suggesting Relatedness in Knowledge Graphs</strong> <div class="csl-block">Giuseppe Pirrò</div> <em>The Semantic Web - ISWC 2015</em> (2015) <a href="https://doi.org/gg5nkd">https://doi.org/gg5nkd</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1007/978-3-319-25007-6_36">10.1007/978-3-319-25007-6_36</a></div></div>
</div>
<div id="ref-GPBbIfq9" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">25. </div><div class="csl-right-inline"><strong>FAIRY: A Framework for Understanding Relationships Between Users' Actions and their Social Feeds</strong> <div class="csl-block">Azin Ghazimatin, Rishiraj Saha Roy, Gerhard Weikum</div> <em>Proceedings of the Twelfth ACM International Conference on Web Search and Data Mining</em> (2019-01-30) <a href="https://doi.org/gf2wqj">https://doi.org/gf2wqj</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1145/3289600.3290990">10.1145/3289600.3290990</a></div></div>
</div>
<div id="ref-L3DO8oq1" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">26. </div><div class="csl-right-inline"><strong>Using Knowledge Graphs to Explain Entity Co-occurrence in Twitter</strong> <div class="csl-block">Yiwei Wang, Mark James Carman, Yuan-Fang Li</div> <em>Proceedings of the 2017 ACM on Conference on Information and Knowledge Management</em> (2017-11-06) <a href="https://doi.org/gg5nkf">https://doi.org/gg5nkf</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1145/3132847.3133161">10.1145/3132847.3133161</a></div></div>
</div>
<div id="ref-Hf24Hzrl" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">27. </div><div class="csl-right-inline"><strong>ESPRESSO: Explaining Relationships between Entity Sets</strong> <div class="csl-block">Stephan Seufert, Klaus Berberich, Srikanta J Bedathur, Sarath Kumar Kondreddi, Patrick Ernst, Gerhard Weikum</div> <em>Proceedings of the 25th ACM International on Conference on Information and Knowledge Management</em> (2016-10-24) <a href="https://doi.org/gcpx7w">https://doi.org/gcpx7w</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1145/2983323.2983778">10.1145/2983323.2983778</a></div></div>
</div>
<div id="ref-19H7j1WBd" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">28. </div><div class="csl-right-inline"><strong>MetaExp: Interactive Explanation and Exploration of Large Knowledge Graphs</strong> <div class="csl-block">Freya Behrens, Fatemeh Aghaei, Emmanuel Müller, Martin Preusse, Nikola Müller, Michael Hunger, Sebastian Bischoff, Pius Ladenburger, Julius Rückin, Laurenz Seidel, … Davide Mottin</div> <em>WWW '18: Companion Proceedings of the The Web Conference 2018</em> (2018-04-23) <a href="https://doi.org/gg2c7w">https://doi.org/gg2c7w</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1145/3184558.3186978">10.1145/3184558.3186978</a></div></div>
</div>
<div id="ref-Vz9QIDKV" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">29. </div><div class="csl-right-inline"><strong>Discovering Meta-Paths in Large Heterogeneous Information Networks</strong> <div class="csl-block">Changping Meng, Reynold Cheng, Silviu Maniu, Pierre Senellart, Wangda Zhang</div> <em>Proceedings of the 24th International Conference on World Wide Web</em> (2015-05-18) <a href="https://doi.org/gg2c7v">https://doi.org/gg2c7v</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1145/2736277.2741123">10.1145/2736277.2741123</a></div></div>
</div>
<div id="ref-KQNSDt1L" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">30. </div><div class="csl-right-inline"><strong>Design and application of a knowledge network for automatic prioritization of drug mechanisms</strong> <div class="csl-block">Michael Mayers, Roger Tu, Dylan Steinecke, Tong Shu Li, Núria Queralt-Rosinach, Andrew I Su</div> <em>Bioinformatics</em> (2022-04-06) <a href="https://doi.org/gptwsz">https://doi.org/gptwsz</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1093/bioinformatics/btac205">10.1093/bioinformatics/btac205</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/35561182">35561182</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC9113361">PMC9113361</a></div></div>
</div>
<div id="ref-17mGFvxXu" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">31. </div><div class="csl-right-inline"><strong>Transforming DWPCs for hetnet edge prediction</strong> <div class="csl-block">Daniel Himmelstein, Pouya Khankhanian, Antoine Lizee</div> <em>ThinkLab</em> (2016-04-01) <a href="https://doi.org/f3qbmd">https://doi.org/f3qbmd</a> <div class="csl-block">DOI: <a href="https://doi.org/10.15363/thinklab.d193">10.15363/thinklab.d193</a></div></div>
</div>
<div id="ref-1693YErZj" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">32. </div><div class="csl-right-inline"><strong>Decomposing the DWPC to assess intermediate node or edge contributions</strong> <div class="csl-block">Daniel Himmelstein</div> <em>ThinkLab</em> (2016-12-15) <a href="https://doi.org/gbr42h">https://doi.org/gbr42h</a> <div class="csl-block">DOI: <a href="https://doi.org/10.15363/thinklab.d228">10.15363/thinklab.d228</a></div></div>
</div>
<div id="ref-1FsruosUW" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">33. </div><div class="csl-right-inline"><strong>Human Disease Ontology 2018 update: classification, content and workflow expansion</strong> <div class="csl-block">Lynn M Schriml, Elvira Mitraka, James Munro, Becky Tauber, Mike Schor, Lance Nickle, Victor Felix, Linda Jeng, Cynthia Bearer, Richard Lichenstein, … Carol Greene</div> <em>Nucleic Acids Research</em> (2018-11-08) <a href="https://doi.org/ggx9wp">https://doi.org/ggx9wp</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1093/nar/gky1032">10.1093/nar/gky1032</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/30407550">30407550</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6323977">PMC6323977</a></div></div>
</div>
<div id="ref-203l6hC3" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">34. </div><div class="csl-right-inline"><strong>Unifying disease vocabularies</strong> <div class="csl-block">Daniel Himmelstein, Tong Shu Li</div> <em>ThinkLab</em> (2015-03-30) <a href="https://doi.org/f3mqv5">https://doi.org/f3mqv5</a> <div class="csl-block">DOI: <a href="https://doi.org/10.15363/thinklab.d44">10.15363/thinklab.d44</a></div></div>
</div>
<div id="ref-yqoUY2f9" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">35. </div><div class="csl-right-inline"><strong>A review of biomedical datasets relating to drug discovery: a knowledge graph perspective</strong> <div class="csl-block">Stephen Bonner, Ian P Barrett, Cheng Ye, Rowan Swiers, Ola Engkvist, Andreas Bender, Charles Tapley Hoyt, William L Hamilton</div> <em>Briefings in Bioinformatics</em> (2022-09-23) <a href="https://doi.org/gqv3s3">https://doi.org/gqv3s3</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1093/bib/bbac404">10.1093/bib/bbac404</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/36151740">36151740</a></div></div>
</div>
<div id="ref-HUodcdz1" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">36. </div><div class="csl-right-inline"><strong>Wikidata as a knowledge graph for the life sciences</strong> <div class="csl-block">Andra Waagmeester, Gregory Stupp, Sebastian Burgstaller-Muehlbacher, Benjamin M Good, Malachi Griffith, Obi L Griffith, Kristina Hanspers, Henning Hermjakob, Toby S Hudson, Kevin Hybiske, … Andrew I Su</div> <em>eLife</em> (2020-03-17) <a href="https://doi.org/ggqqc6">https://doi.org/ggqqc6</a> <div class="csl-block">DOI: <a href="https://doi.org/10.7554/elife.52614">10.7554/elife.52614</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/32180547">32180547</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7077981">PMC7077981</a></div></div>
</div>
<div id="ref-eKVkBSDD" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">37. </div><div class="csl-right-inline"><strong>SemMedDB: a PubMed-scale repository of biomedical semantic predications</strong> <div class="csl-block">H Kilicoglu, D Shin, M Fiszman, G Rosemblat, TC Rindflesch</div> <em>Bioinformatics</em> (2012-10-08) <a href="https://doi.org/f4hp3x">https://doi.org/f4hp3x</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1093/bioinformatics/bts591">10.1093/bioinformatics/bts591</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/23044550">23044550</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3509487">PMC3509487</a></div></div>
</div>
<div id="ref-reekunFL" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">38. </div><div class="csl-right-inline"><strong>Constructing Biomedical Knowledge Graph Based on SemMedDB and Linked Open Data</strong> <div class="csl-block">Qing Cong, Zhiyong Feng, Fang Li, Li Zhang, Guozheng Rao, Cui Tao</div> <em>2018 IEEE International Conference on Bioinformatics and Biomedicine (BIBM)</em> (2018-12) <a href="https://doi.org/ggzb26">https://doi.org/ggzb26</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1109/bibm.2018.8621568">10.1109/bibm.2018.8621568</a></div></div>
</div>
<div id="ref-uTOtIsTj" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">39. </div><div class="csl-right-inline"><strong>Time-resolved evaluation of compound repositioning predictions on a text-mined knowledge network</strong> <div class="csl-block">Michael Mayers, Tong Shu Li, Núria Queralt-Rosinach, Andrew I Su</div> <em>BMC Bioinformatics</em> (2019-12) <a href="https://doi.org/ggpcsr">https://doi.org/ggpcsr</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1186/s12859-019-3297-0">10.1186/s12859-019-3297-0</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/31829175">31829175</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6907279">PMC6907279</a></div></div>
</div>
<div id="ref-1BHNoKOV1" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">40. </div><div class="csl-right-inline"><strong>The scalable precision medicine open knowledge engine (SPOKE): a massive knowledge graph of biomedical information</strong> <div class="csl-block">John H Morris, Karthik Soman, Rabia E Akbas, Xiaoyuan Zhou, Brett Smith, Elaine C Meng, Conrad C Huang, Gabriel Cerono, Gundolf Schenk, Angela Rizk-Jackson, … Sergio E Baranzini</div> <em>Bioinformatics</em> (2023-02-01) <a href="https://doi.org/grvqrf">https://doi.org/grvqrf</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1093/bioinformatics/btad080">10.1093/bioinformatics/btad080</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/36759942">36759942</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC9940622">PMC9940622</a></div></div>
</div>
<div id="ref-bniVH7q2" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">41. </div><div class="csl-right-inline"><strong>RTX-KG2: a system for building a semantically standardized knowledge graph for translational biomedicine</strong> <div class="csl-block">EC Wood, Amy K Glen, Lindsey G Kvarfordt, Finn Womack, Liliana Acevedo, Timothy S Yoon, Chunyu Ma, Veronica Flores, Meghamala Sinha, Yodsawalai Chodpathumwan, … Stephen A Ramsey</div> <em>BMC Bioinformatics</em> (2022-09-29) <a href="https://doi.org/gqxmkb">https://doi.org/gqxmkb</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1186/s12859-022-04932-3">10.1186/s12859-022-04932-3</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/36175836">36175836</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC9520835">PMC9520835</a></div></div>
</div>
<div id="ref-WkPlH1ds" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">42. </div><div class="csl-right-inline"><strong>Heterogeneous Network Edge Prediction: A Data Integration Approach to Prioritize Disease-Associated Genes</strong> <div class="csl-block">Daniel S Himmelstein, Sergio E Baranzini</div> <em>PLOS Computational Biology</em> (2015-07-09) <a href="https://doi.org/98q">https://doi.org/98q</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1371/journal.pcbi.1004259">10.1371/journal.pcbi.1004259</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/26158728">26158728</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4497619">PMC4497619</a></div></div>
</div>
<div id="ref-kNYoxBiQ" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">43. </div><div class="csl-right-inline"><strong>Using the neo4j graph database for hetnets</strong> <div class="csl-block">Daniel Himmelstein</div> <em>ThinkLab</em> (2015-10-02) <a href="https://doi.org/f3mqvk">https://doi.org/f3mqvk</a> <div class="csl-block">DOI: <a href="https://doi.org/10.15363/thinklab.d112">10.15363/thinklab.d112</a></div></div>
</div>
<div id="ref-xWWaK51z" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">44. </div><div class="csl-right-inline"><strong>Estimating the complexity of hetnet traversal</strong> <div class="csl-block">Daniel Himmelstein, Antoine Lizee</div> <em>ThinkLab</em> (2016-03-22) <a href="https://doi.org/gbr42x">https://doi.org/gbr42x</a> <div class="csl-block">DOI: <a href="https://doi.org/10.15363/thinklab.d187">10.15363/thinklab.d187</a></div></div>
</div>
<div id="ref-w6rL3OYz" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">45. </div><div class="csl-right-inline"><strong>Path exclusion conditions</strong> <div class="csl-block">Daniel Himmelstein</div> <em>ThinkLab</em> (2015-12-08) <a href="https://doi.org/gg2rw2">https://doi.org/gg2rw2</a> <div class="csl-block">DOI: <a href="https://doi.org/10.15363/thinklab.d134">10.15363/thinklab.d134</a></div></div>
</div>
<div id="ref-15nwuvjrA" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">46. </div><div class="csl-right-inline"><strong>Vagelos Report Summer 2017</strong> <div class="csl-block">Michael Zietz</div> <em>figshare</em> (2017) <a href="https://doi.org/gbr3pf">https://doi.org/gbr3pf</a> <div class="csl-block">DOI: <a href="https://doi.org/10.6084/m9.figshare.5346577">10.6084/m9.figshare.5346577</a></div></div>
</div>
<div id="ref-IJWsN67g" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">47. </div><div class="csl-right-inline"><strong>GitHub - mmayers12/hetnet_ml: Software to quickly extract features from heterogeneous networks for machine learning.</strong> <div class="csl-block">GitHub</div> <a href="https://github.com/mmayers12/hetnet_ml">https://github.com/mmayers12/hetnet_ml</a></div>
</div>
<div id="ref-iKOIEzQ9" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">48. </div><div class="csl-right-inline"><strong>Randomization Techniques for Graphs</strong> <div class="csl-block">Sami Hanhijärvi, Gemma C Garriga, Kai Puolamäki</div> <em>Proceedings of the 2009 SIAM International Conference on Data Mining</em> (2009-04-30) <a href="https://doi.org/f3mn58">https://doi.org/f3mn58</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1137/1.9781611972795.67">10.1137/1.9781611972795.67</a></div></div>
</div>
<div id="ref-GE6jhWnt" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">49. </div><div class="csl-right-inline"><strong>Assessing the effectiveness of our hetnet permutations</strong> <div class="csl-block">Daniel Himmelstein</div> <em>ThinkLab</em> (2016-02-25) <a href="https://doi.org/f3mqt5">https://doi.org/f3mqt5</a> <div class="csl-block">DOI: <a href="https://doi.org/10.15363/thinklab.d178">10.15363/thinklab.d178</a></div></div>
</div>
<div id="ref-83EvMSI6" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">50. </div><div class="csl-right-inline"><strong>The probability of edge existence due to node degree: a baseline for network-based predictions</strong> <div class="csl-block">Michael Zietz, Daniel S Himmelstein, Kyle Kloster, Christopher Williams, Michael W Nagle, Casey S Greene</div> <em>bioRxiv</em> (2023-01-06) <a href="https://doi.org/grk9gd">https://doi.org/grk9gd</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1101/2023.01.05.522939">10.1101/2023.01.05.522939</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/36711569">36711569</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC9881952">PMC9881952</a></div></div>
</div>
<div id="ref-19OyKPQ4M" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">51. </div><div class="csl-right-inline"><strong>Network Edge Prediction: Estimating the prior</strong> <div class="csl-block">Antoine Lizee, Daniel Himmelstein</div> <em>ThinkLab</em> (2016-04-14) <a href="https://doi.org/f3qbmg">https://doi.org/f3qbmg</a> <div class="csl-block">DOI: <a href="https://doi.org/10.15363/thinklab.d201">10.15363/thinklab.d201</a></div></div>
</div>
<div id="ref-pkJqqM2W" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">52. </div><div class="csl-right-inline"><strong>Node connectivity measurements for Hetionet v1.0 metapaths</strong> <div class="csl-block">Daniel Himmelstein, Michael Zietz, Kyle Kloster, Michael Nagle, Blair Sullivan, Casey Greene</div> <em>Zenodo</em> (2018-11-06) <a href="https://doi.org/cww7">https://doi.org/cww7</a> <div class="csl-block">DOI: <a href="https://doi.org/10.5281/zenodo.1435833">10.5281/zenodo.1435833</a></div></div>
</div>
<div id="ref-q6QyrHEZ" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">53. </div><div class="csl-right-inline"><strong>A practical guide to methods controlling false discoveries in computational biology</strong> <div class="csl-block">Keegan Korthauer, Patrick K Kimes, Claire Duvallet, Alejandro Reyes, Ayshwarya Subramanian, Mingxiang Teng, Chinmay Shukla, Eric J Alm, Stephanie C Hicks</div> <em>Genome Biology</em> (2019-06-04) <a href="https://doi.org/gf3ncd">https://doi.org/gf3ncd</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1186/s13059-019-1716-1">10.1186/s13059-019-1716-1</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/31164141">31164141</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6547503">PMC6547503</a></div></div>
</div>
<div id="ref-YuJbg3zO" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">54. </div><div class="csl-right-inline"><strong>Open collaborative writing with Manubot</strong> <div class="csl-block">Daniel S Himmelstein, Vincent Rubinetti, David R Slochower, Dongbo Hu, Venkat S Malladi, Casey S Greene, Anthony Gitter</div> <em>PLOS Computational Biology</em> (2019-06-24) <a href="https://doi.org/c7np">https://doi.org/c7np</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1371/journal.pcbi.1007128">10.1371/journal.pcbi.1007128</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/31233491">31233491</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6611653">PMC6611653</a></div></div>
</div>
<div id="ref-3KJWwgoz" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">55. </div><div class="csl-right-inline"><strong>greenelab/connectivity-search-manuscript repository: Manuscript source code for Hetnet Connectivity Search</strong> <div class="csl-block">Daniel Himmelstein, Michael Zietz, Vincent Rubinetti, Kyle Kloster, Benjamin Heil, Faisal Alquaddoomi, Dongbo Hu, David Nicholson, Yun Hao, Blair Sullivan, … Casey Greene</div> <em>GitHub</em> (2023-03-05) <a href="https://github.com/greenelab/connectivity-search-manuscript">https://github.com/greenelab/connectivity-search-manuscript</a></div>
</div>
<div id="ref-eq4IX1LX" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">56. </div><div class="csl-right-inline"><strong>greenelab/connectivity-search-analyses repository: Hetnet connectivity search research notebooks</strong> <div class="csl-block">Daniel Himmelstein, Michael Zietz, Kyle Kloster, Benjamin Heil, Yun Hao, Gregory Way, Michael Nagle, Blair Sullivan, Casey Greene</div> <em>GitHub</em> (2022-03-20) <a href="https://github.com/greenelab/connectivity-search-analyses">https://github.com/greenelab/connectivity-search-analyses</a></div>
</div>
<div id="ref-ASx5cMWI" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">57. </div><div class="csl-right-inline"><strong>greenelab/connectivity-search-backend repository: Django backend for hetnet connectivity search</strong> <div class="csl-block">Daniel Himmelstein, Dongbo Hu, Faisal Alquaddoomi, Casey Greene</div> <em>GitHub</em> (2023-03-30) <a href="https://github.com/greenelab/connectivity-search-backend">https://github.com/greenelab/connectivity-search-backend</a></div>
</div>
<div id="ref-CadAZTf2" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">58. </div><div class="csl-right-inline"><strong>greenelab/connectivity-search-frontend repository: Frontend source code for Hetnet connectivity search</strong> <div class="csl-block">Vincent Rubinetti, Daniel Himmelstein, Dongbo Hu, Casey Greene</div> <em>GitHub</em> (2023-03-30) <a href="https://github.com/greenelab/connectivity-search-frontend">https://github.com/greenelab/connectivity-search-frontend</a></div>
</div>
<div id="ref-gHk6T3vn" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">59. </div><div class="csl-right-inline"><strong>hetio/hetmatpy repository: Python package for matrix storage and operations on hetnets</strong> <div class="csl-block">Daniel Himmelstein, Michael Zietz, Kyle Kloster, Benjamin Heil, Gregory Way, Blair Sullivan, Casey Greene</div> <em>GitHub</em> (2022-04) <a href="https://github.com/hetio/hetmatpy">https://github.com/hetio/hetmatpy</a></div>
</div>
<div id="ref-srnBddlB" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">60. </div><div class="csl-right-inline"><strong>hetio/hetnetpy repository: Hetnets in Python</strong> <div class="csl-block">Daniel Himmelstein, Benjamin Heil, Michael Zietz, Tong Shu Li, Kyle Kloster, Gregory Way, Casey Greene, Sergio Baranzini</div> <em>GitHub</em> (2021-06) <a href="https://github.com/hetio/hetnetpy">https://github.com/hetio/hetnetpy</a></div>
</div>
<div id="ref-L9cBdipF" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">61. </div><div class="csl-right-inline"><strong>hetio/hetionet: Data repository containing Hetionet downloads</strong> <div class="csl-block">Daniel Himmelstein, Antoine Lizee, Faisal Alquaddoomi, Vincent Rubinetti, Dongbo Hu, Casey Greene, Sergio Baranzini</div> <em>GitHub</em> (2023-04-04) <a href="https://github.com/greenelab/connectivity-search-manuscript">https://github.com/greenelab/connectivity-search-manuscript</a></div>
</div>
<div id="ref-12IQq3W3u" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">62. </div><div class="csl-right-inline"><strong>hetio/het.io repository: Source code for het.io website</strong> <div class="csl-block">Vincent Rubinetti, Daniel Himmelstein, Dongbo Hu, Sergio Baranzini, Casey Greene</div> <em>GitHub</em> (2021-12-20) <a href="https://github.com/hetio/het.io">https://github.com/hetio/het.io</a></div>
</div>
<div id="ref-hHRphzV0" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">63. </div><div class="csl-right-inline"><strong>Supporting data for "Hetnet connectivity search provides rapid insights into how two biomedical entities are related"</strong> <div class="csl-block">Himmelstein S Daniel, Zietz Michael, Rubinetti Vincent, Kloster Kyle, Heil J Benjamin, Alquaddoomi Faisal, Hu Dongbo, Nicholson N David, Hao Yun, Sullivan D Blair, … Greene S Casey</div> <em>GigaScience Database</em> (2023-05-12) <a href="https://doi.org/gr77bv">https://doi.org/gr77bv</a> <div class="csl-block">DOI: <a href="https://doi.org/10.5524/102389">10.5524/102389</a></div></div>
</div>
</div>
<!-- default theme -->

<style>
  /* import google fonts */
  @import url("https://fonts.googleapis.com/css?family=Open+Sans:400,600,700");
  @import url("https://fonts.googleapis.com/css?family=Source+Code+Pro");

  /* -------------------------------------------------- */
  /* global */
  /* -------------------------------------------------- */

  /* all elements */
  * {
    /* force sans-serif font unless specified otherwise */
    font-family: "Open Sans", "Helvetica", sans-serif;

    /* prevent text inflation on some mobile browsers */
    -webkit-text-size-adjust: none !important;
    -moz-text-size-adjust: none !important;
    -o-text-size-adjust: none !important;
    text-size-adjust: none !important;
  }

  @media only screen {
    /* "page" element */
    body {
      position: relative;
      box-sizing: border-box;
      font-size: 12pt;
      line-height: 1.5;
      max-width: 8.5in;
      margin: 20px auto;
      padding: 40px;
      border-radius: 5px;
      border: solid 1px #bdbdbd;
      box-shadow: 0 0 20px rgba(0, 0, 0, 0.05);
      background: #ffffff;
    }
  }

  /* when on screen < 8.5in wide */
  @media only screen and (max-width: 8.5in) {
    /* "page" element */
    body {
      padding: 20px;
      margin: 0;
      border-radius: 0;
      border: none;
      box-shadow: 0 0 20px rgba(0, 0, 0, 0.05) inset;
      background: none;
    }
  }

  /* -------------------------------------------------- */
  /* headings */
  /* -------------------------------------------------- */

  /* all headings */
  h1,
  h2,
  h3,
  h4,
  h5,
  h6 {
    margin: 20px 0;
    padding: 0;
    font-weight: bold;
  }

  /* biggest heading */
  h1 {
    margin: 40px 0;
    text-align: center;
  }

  /* second biggest heading */
  h2 {
    margin-top: 30px;
    padding-bottom: 5px;
    border-bottom: solid 1px #bdbdbd;
  }

  /* heading font sizes */
  h1 {
    font-size: 2em;
  }
  h2 {
    font-size: 1.5em;
  }
  h3 {
    font-size: 1.35em;
  }
  h4 {
    font-size: 1.25em;
  }
  h5 {
    font-size: 1.15em;
  }
  h6 {
    font-size: 1em;
  }

  /* -------------------------------------------------- */
  /* manuscript header */
  /* -------------------------------------------------- */

  /* manuscript title */
  header > h1 {
    margin: 0;
  }

  /* manuscript title caption text (ie "automatically generated on") */
  header + p {
    text-align: center;
    margin-top: 10px;
  }

  /* -------------------------------------------------- */
  /* text elements */
  /* -------------------------------------------------- */

  /* links */
  a {
    color: #2196f3;
    overflow-wrap: break-word;
  }

  /* superscripts and subscripts */
  sub,
  sup {
    /* prevent from affecting line height */
    line-height: 0;
  }

  /* unordered and ordered lists*/
  ul,
  ol {
    padding-left: 20px;
  }

  /* class for styling text semibold */
  .semibold {
    font-weight: 600;
  }

  /* class for styling elements horizontally left aligned */
  .left {
    display: block;
    text-align: left;
    margin-left: auto;
    margin-right: 0;
    justify-content: left;
  }

  /* class for styling elements horizontally centered */
  .center {
    display: block;
    text-align: center;
    margin-left: auto;
    margin-right: auto;
    justify-content: center;
  }

  /* class for styling elements horizontally right aligned */
  .right {
    display: block;
    text-align: right;
    margin-left: 0;
    margin-right: auto;
    justify-content: right;
  }

  /* -------------------------------------------------- */
  /* section elements */
  /* -------------------------------------------------- */

  /* horizontal divider line */
  hr {
    border: none;
    height: 1px;
    background: #bdbdbd;
  }

  /* paragraphs, horizontal dividers, figures, tables, code */
  p,
  hr,
  figure,
  table,
  pre {
    /* treat all as "paragraphs", with consistent vertical margins */
    margin-top: 20px;
    margin-bottom: 20px;
  }

  /* -------------------------------------------------- */
  /* figures */
  /* -------------------------------------------------- */

  /* figure */
  figure {
    max-width: 100%;
    margin-left: auto;
    margin-right: auto;
  }

  /* figure caption */
  figcaption {
    padding: 0;
    padding-top: 10px;
  }

  /* figure image element */
  figure > img,
  figure > svg {
    max-width: 100%;
    display: block;
    margin-left: auto;
    margin-right: auto;
  }

  /* figure auto-number */
  img + figcaption > span:first-of-type,
  svg + figcaption > span:first-of-type {
    font-weight: bold;
    margin-right: 5px;
  }

  /* -------------------------------------------------- */
  /* tables */
  /* -------------------------------------------------- */

  /* table */
  table {
    border-collapse: collapse;
    border-spacing: 0;
    width: 100%;
    margin-left: auto;
    margin-right: auto;
  }

  /* table cells */
  th,
  td {
    border: solid 1px #bdbdbd;
    padding: 10px;
    /* squash table if too wide for page by forcing line breaks */
    overflow-wrap: break-word;
    word-break: break-word;
  }

  /* header row and even rows */
  th,
  tr:nth-child(2n) {
    background-color: #fafafa;
  }

  /* odd rows */
  tr:nth-child(2n + 1) {
    background-color: #ffffff;
  }

  /* table caption */
  caption {
    text-align: left;
    padding: 0;
    padding-bottom: 10px;
  }

  /* table auto-number */
  table > caption > span:first-of-type {
    font-weight: bold;
    margin-right: 5px;
  }

  /* -------------------------------------------------- */
  /* code */
  /* -------------------------------------------------- */

  /* multi-line code block */
  pre {
    padding: 10px;
    background-color: #eeeeee;
    color: #000000;
    border-radius: 5px;
    break-inside: avoid;
    text-align: left;
  }

  /* inline code, ie code within normal text */
  :not(pre) > code {
    padding: 0 4px;
    background-color: #eeeeee;
    color: #000000;
    border-radius: 5px;
  }

  /* code text */
  /* apply all children, to reach syntax highlighting sub-elements */
  code,
  code * {
    /* force monospace font */
    font-family: "Source Code Pro", "Courier New", monospace;
  }

  /* -------------------------------------------------- */
  /* quotes */
  /* -------------------------------------------------- */

  /* quoted text */
  blockquote {
    margin: 0;
    padding: 0;
    border-left: 4px solid #bdbdbd;
    padding-left: 16px;
    break-inside: avoid;
  }

  /* -------------------------------------------------- */
  /* banners */
  /* -------------------------------------------------- */

  /* info banners */
  .banner {
    box-sizing: border-box;
    display: block;
    position: relative;
    width: 100%;
    margin-top: 20px;
    margin-bottom: 20px;
    padding: 20px;
    text-align: center;
  }

  /* paragraph in banner */
  .banner > p {
    margin: 0;
  }

  /* -------------------------------------------------- */
  /* highlight colors */
  /* -------------------------------------------------- */

  .white {
    background: #ffffff;
  }
  .lightgrey {
    background: #eeeeee;
  }
  .grey {
    background: #757575;
  }
  .darkgrey {
    background: #424242;
  }
  .black {
    background: #000000;
  }
  .lightred {
    background: #ffcdd2;
  }
  .lightyellow {
    background: #ffecb3;
  }
  .lightgreen {
    background: #dcedc8;
  }
  .lightblue {
    background: #e3f2fd;
  }
  .lightpurple {
    background: #f3e5f5;
  }
  .red {
    background: #f44336;
  }
  .orange {
    background: #ff9800;
  }
  .yellow {
    background: #ffeb3b;
  }
  .green {
    background: #4caf50;
  }
  .blue {
    background: #2196f3;
  }
  .purple {
    background: #9c27b0;
  }
  .white,
  .lightgrey,
  .lightred,
  .lightyellow,
  .lightgreen,
  .lightblue,
  .lightpurple,
  .orange,
  .yellow,
  .white a,
  .lightgrey a,
  .lightred a,
  .lightyellow a,
  .lightgreen a,
  .lightblue a,
  .lightpurple a,
  .orange a,
  .yellow a {
    color: #000000;
  }
  .grey,
  .darkgrey,
  .black,
  .red,
  .green,
  .blue,
  .purple,
  .grey a,
  .darkgrey a,
  .black a,
  .red a,
  .green a,
  .blue a,
  .purple a {
    color: #ffffff;
  }

  /* -------------------------------------------------- */
  /* buttons */
  /* -------------------------------------------------- */

  /* class for styling links like buttons */
  .button {
    display: inline-flex;
    justify-content: center;
    align-items: center;
    margin: 5px;
    padding: 10px 20px;
    font-size: 0.75em;
    font-weight: 600;
    text-transform: uppercase;
    text-decoration: none;
    letter-spacing: 1px;
    background: none;
    color: #2196f3;
    border: solid 1px #bdbdbd;
    border-radius: 5px;
  }

  /* buttons when hovered */
  .button:hover:not([disabled]),
  .icon_button:hover:not([disabled]) {
    cursor: pointer;
    background: #f5f5f5;
  }

  /* buttons when disabled */
  .button[disabled],
  .icon_button[disabled] {
    opacity: 0.35;
    pointer-events: none;
  }

  /* class for styling buttons containg only single icon */
  .icon_button {
    display: inline-flex;
    justify-content: center;
    align-items: center;
    text-decoration: none;
    margin: 0;
    padding: 0;
    background: none;
    border-radius: 5px;
    border: none;
    width: 20px;
    height: 20px;
    min-width: 20px;
    min-height: 20px;
  }

  /* icon button inner svg image */
  .icon_button > svg {
    height: 16px;
  }

  /* -------------------------------------------------- */
  /* icons */
  /* -------------------------------------------------- */

  /* class for styling icons inline with text */
  .inline_icon {
    height: 1em;
    position: relative;
    top: 0.125em;
  }

  /* -------------------------------------------------- */
  /* references */
  /* -------------------------------------------------- */

  .csl-entry {
    margin-top: 15px;
    margin-bottom: 15px;
  }

  /* -------------------------------------------------- */
  /* print control */
  /* -------------------------------------------------- */

  @media print {
    @page {
      /* suggested printing margin */
      margin: 0.5in;
    }

    /* document and "page" elements */
    html,
    body {
      margin: 0;
      padding: 0;
      width: 100%;
      height: 100%;
    }

    /* "page" element */
    body {
      font-size: 11pt !important;
      line-height: 1.35;
    }

    /* all headings */
    h1,
    h2,
    h3,
    h4,
    h5,
    h6 {
      margin: 15px 0;
    }

    /* figures and tables */
    figure,
    table {
      font-size: 0.85em;
    }

    /* table cells */
    th,
    td {
      padding: 5px;
    }

    /* shrink font awesome icons */
    i.fas,
    i.fab,
    i.far,
    i.fal {
      transform: scale(0.85);
    }

    /* decrease banner margins */
    .banner {
      margin-top: 15px;
      margin-bottom: 15px;
      padding: 15px;
    }

    /* class for centering an element vertically on its own page */
    .page_center {
      margin: auto;
      width: 100%;
      height: 100%;
      display: flex;
      align-items: center;
      vertical-align: middle;
      break-before: page;
      break-after: page;
    }

    /* always insert a page break before the element */
    .page_break_before {
      break-before: page;
    }

    /* always insert a page break after the element */
    .page_break_after {
      break-after: page;
    }

    /* avoid page break before the element */
    .page_break_before_avoid {
      break-before: avoid;
    }

    /* avoid page break after the element */
    .page_break_after_avoid {
      break-after: avoid;
    }

    /* avoid page break inside the element */
    .page_break_inside_avoid {
      break-inside: avoid;
    }
  }

  /* -------------------------------------------------- */
  /* override pandoc css quirks */
  /* -------------------------------------------------- */

  .sourceCode {
    /* prevent unsightly overflow in wide code blocks */
    overflow: auto !important;
  }

  div.sourceCode {
    /* prevent background fill on top-most code block  container */
    background: none !important;
  }

  .sourceCode * {
    /* force consistent line spacing */
    line-height: 1.5 !important;
  }

  div.sourceCode {
    /* style code block margins same as <pre> element */
    margin-top: 20px;
    margin-bottom: 20px;
  }

  /* -------------------------------------------------- */
  /* tablenos */
  /* -------------------------------------------------- */

  /* tablenos wrapper */
  .tablenos {
    width: 100%;
    margin: 20px 0;
  }

  .tablenos > table {
    /* move margins from table to table_wrapper to allow margin collapsing */
    margin: 0;
  }

  @media only screen {
    /* tablenos wrapper */
    .tablenos {
      /* show scrollbar on tables if necessary to prevent overflow */
      overflow-x: auto !important;
    }

    .tablenos th,
    .tablenos td {
      overflow-wrap: unset !important;
      word-break: unset !important;
    }

    /* table in wrapper */
    .tablenos table,
    .tablenos table * {
      /* don't break table words */
      overflow-wrap: normal !important;
    }
  }
</style>
<!-- 
    Plugin Core

    Functions needed for and shared across all first-party plugins.
-->

<script>
  // get element that is target of hash (from link element or url)
  function getHashTarget(link) {
    const hash = link ? link.hash : window.location.hash;
    const id = hash.slice(1);
    let target = document.querySelector(`[id="${id}"]`);
    if (!target) return;

    // if figure or table, modify target to get expected element
    if (id.indexOf("fig:") === 0) target = target.querySelector("figure");
    if (id.indexOf("tbl:") === 0) target = target.querySelector("table");

    return target;
  }

  // get position/dimensions of element or viewport
  function getRectInView(element) {
    let rect = {};
    rect.left = 0;
    rect.top = 0;
    rect.right = document.documentElement.clientWidth;
    rect.bottom = document.documentElement.clientHeight;
    let style = {};

    if (element instanceof HTMLElement) {
      rect = element.getBoundingClientRect();
      style = window.getComputedStyle(element);
    }

    const margin = {};
    margin.left = parseFloat(style.marginLeftWidth) || 0;
    margin.top = parseFloat(style.marginTopWidth) || 0;
    margin.right = parseFloat(style.marginRightWidth) || 0;
    margin.bottom = parseFloat(style.marginBottomWidth) || 0;

    const border = {};
    border.left = parseFloat(style.borderLeftWidth) || 0;
    border.top = parseFloat(style.borderTopWidth) || 0;
    border.right = parseFloat(style.borderRightWidth) || 0;
    border.bottom = parseFloat(style.borderBottomWidth) || 0;

    const newRect = {};
    newRect.left = rect.left + margin.left + border.left;
    newRect.top = rect.top + margin.top + border.top;
    newRect.right = rect.right + margin.right + border.right;
    newRect.bottom = rect.bottom + margin.bottom + border.bottom;
    newRect.width = newRect.right - newRect.left;
    newRect.height = newRect.bottom - newRect.top;

    return newRect;
  }

  // get position of element relative to page
  function getRectInPage(element) {
    const rect = getRectInView(element);
    const body = getRectInView(document.body);

    const newRect = {};
    newRect.left = rect.left - body.left;
    newRect.top = rect.top - body.top;
    newRect.right = rect.right - body.left;
    newRect.bottom = rect.bottom - body.top;
    newRect.width = rect.width;
    newRect.height = rect.height;

    return newRect;
  }

  // get closest element before specified element that matches query
  function firstBefore(element, query) {
    while (element && element !== document.body && !element.matches(query))
      element = element.previousElementSibling || element.parentNode;

    return element;
  }

  // check if element is part of collapsed heading
  function isCollapsed(element) {
    while (element && element !== document.body) {
      if (element.dataset.collapsed === "true") return true;
      element = element.parentNode;
    }
    return false;
  }

  // expand any collapsed parent containers of element if necessary
  function expandElement(element) {
    if (isCollapsed(element)) {
      // accordion plugin
      const heading = firstBefore(element, "h2");
      if (heading) heading.click();
      // details/summary HTML element
      const summary = firstBefore(element, "summary");
      if (summary) summary.click();
    }
  }

  // scroll to and focus element
  function goToElement(element, offset) {
    // expand accordion section if collapsed
    expandElement(element);
    const y =
      getRectInView(element).top -
      getRectInView(document.documentElement).top -
      (offset || 0);

    // trigger any function listening for "onscroll" event
    window.dispatchEvent(new Event("scroll"));
    window.scrollTo(0, y);
    document.activeElement.blur();
    element.focus();
  }

  // get list of elements after a start element up to element matching query
  function nextUntil(element, query, exclude) {
    const elements = [];
    while (((element = element.nextElementSibling), element)) {
      if (element.matches(query)) break;
      if (!element.matches(exclude)) elements.push(element);
    }
    return elements;
  }
</script>
<!--
  Accordion Plugin

  Allows sections of content under h2 headings to be collapsible.
-->

<script type="module">
  // whether to always start expanded ('false'), always start collapsed
  // ('true'), or start collapsed when screen small ('auto')
  const startCollapsed = "auto";

  // start script
  function start() {
    // run through each <h2> heading
    const headings = document.querySelectorAll("h2");
    for (const heading of headings) {
      addArrow(heading);

      // start expanded/collapsed based on option
      if (
        startCollapsed === "true" ||
        (startCollapsed === "auto" && isSmallScreen()) ||
        heading.dataset.collapsed === "true"
      )
        collapseHeading(heading);
      else expandElement(heading);
    }

    // attach hash change listener to window
    window.addEventListener("hashchange", onHashChange);
  }

  // when hash (eg manuscript.html#introduction) changes
  function onHashChange() {
    const target = getHashTarget();
    if (target) goToElement(target);
  }

  // add arrow to heading
  function addArrow(heading) {
    // add arrow button
    const arrow = document.createElement("button");
    arrow.innerHTML = document.querySelector(".icon_angle_down").innerHTML;
    arrow.classList.add("icon_button", "accordion_arrow");
    heading.insertBefore(arrow, heading.firstChild);

    // attach click listener to heading and button
    heading.addEventListener("click", onHeadingClick);
    arrow.addEventListener("click", onArrowClick);
  }

  // determine if on mobile-like device with small screen
  function isSmallScreen() {
    return Math.min(window.innerWidth, window.innerHeight) < 480;
  }

  // when <h2> heading is clicked
  function onHeadingClick(event) {
    // only collapse if <h2> itself is target of click (eg, user did
    // not click on anchor within <h2>)
    if (event.target === this) toggleCollapse(this);
  }

  // when arrow button is clicked
  function onArrowClick() {
    toggleCollapse(this.parentNode);
  }

  // collapse section if expanded, expand if collapsed
  function toggleCollapse(heading) {
    if (heading.dataset.collapsed === "false") collapseHeading(heading);
    else expandElement(heading);
  }

  // elements to exclude from collapse, such as table of contents panel,
  // hypothesis panel, etc
  const exclude = "#toc_panel, div.annotator-frame, #lightbox_overlay";

  // collapse section
  function collapseHeading(heading) {
    heading.setAttribute("data-collapsed", "true");
    const children = getChildren(heading);
    for (const child of children) child.setAttribute("data-collapsed", "true");
  }

  // expand section
  function expandElement(heading) {
    heading.setAttribute("data-collapsed", "false");
    const children = getChildren(heading);
    for (const child of children) child.setAttribute("data-collapsed", "false");
  }

  // get list of elements between this <h2> and next <h2> or <h1>
  // ("children" of the <h2> section)
  function getChildren(heading) {
    return nextUntil(heading, "h2, h1", exclude);
  }

  // start script when document is finished loading
  window.addEventListener("load", start);
</script>

<!-- angle down icon -->

<template class="icon_angle_down">
  <!-- modified from: https://fontawesome.com/icons/angle-down -->
  <svg width="16" height="16" viewBox="0 0 448 512">
    <path
      fill="currentColor"
      d="M207.029 381.476L12.686 187.132c-9.373-9.373-9.373-24.569 0-33.941l22.667-22.667c9.357-9.357 24.522-9.375 33.901-.04L224 284.505l154.745-154.021c9.379-9.335 24.544-9.317 33.901.04l22.667 22.667c9.373 9.373 9.373 24.569 0 33.941L240.971 381.476c-9.373 9.372-24.569 9.372-33.942 0z"
    ></path>
  </svg>
</template>

<style>
  @media only screen {
    /* accordion arrow button */
    .accordion_arrow {
      margin-right: 10px;
    }

    /* arrow icon when <h2> data-collapsed attribute true */
    h2[data-collapsed="true"] > .accordion_arrow > svg {
      transform: rotate(-90deg);
    }

    /* all elements (except <h2>'s) when data-collapsed attribute true */
    *:not(h2)[data-collapsed="true"] {
      display: none;
    }

    /* accordion arrow button when hovered and <h2>'s when hovered */
    .accordion_arrow:hover,
    h2[data-collapsed="true"]:hover,
    h2[data-collapsed="false"]:hover {
      cursor: pointer;
    }
  }

  /* always hide accordion arrow button on print */
  @media only print {
    .accordion_arrow {
      display: none;
    }
  }
</style>
<!--
  Anchors Plugin

  Adds an anchor next to each of a certain type of element that provides a
  human-readable url to that specific item/position in the document (e.g.
  "manuscript.html#abstract"). It also makes it such that scrolling out of view
  of a target removes its identifier from the url.
-->

<script type="module">
  // which types of elements to add anchors next to, in "document.querySelector"
  // format
  const typesQuery =
    'h1, h2, h3, div[id^="fig:"], div[id^="tbl:"], span[id^="eq:"]';

  // start script
  function start() {
    // add anchor to each element of specified types
    const elements = document.querySelectorAll(typesQuery);
    for (const element of elements) addAnchor(element);

    // attach scroll listener to window
    window.addEventListener("scroll", onScroll);
  }

  // when window is scrolled
  function onScroll() {
    // if url has hash and user has scrolled out of view of hash
    // target, remove hash from url
    const tolerance = 100;
    const target = getHashTarget();
    if (target) {
      if (
        target.getBoundingClientRect().top > window.innerHeight + tolerance ||
        target.getBoundingClientRect().bottom < 0 - tolerance
      )
        history.pushState(null, null, " ");
    }
  }

  // add anchor to element
  function addAnchor(element) {
    let addTo; // element to add anchor button to

    // if figure or table, modify withId and addTo to get expected
    // elements
    if (element.id.indexOf("fig:") === 0) {
      addTo = element.querySelector("figcaption");
    } else if (element.id.indexOf("tbl:") === 0) {
      addTo = element.querySelector("caption");
    } else if (element.id.indexOf("eq:") === 0) {
      addTo = element.querySelector(".eqnos-number");
    }

    addTo = addTo || element;
    const id = element.id || null;

    // do not add anchor if element doesn't have assigned id.
    // id is generated by pandoc and is assumed to be unique and
    // human-readable
    if (!id) return;

    // create anchor button
    const anchor = document.createElement("a");
    anchor.innerHTML = document.querySelector(".icon_link").innerHTML;
    anchor.title = "Link to this part of the document";
    anchor.classList.add("icon_button", "anchor");
    anchor.dataset.ignore = "true";
    anchor.href = "#" + id;
    addTo.appendChild(anchor);
  }

  // start script when document is finished loading
  window.addEventListener("load", start);
</script>

<!-- link icon -->

<template class="icon_link">
  <!-- modified from: https://fontawesome.com/icons/link -->
  <svg width="16" height="16" viewBox="0 0 512 512">
    <path
      fill="currentColor"
      d="M326.612 185.391c59.747 59.809 58.927 155.698.36 214.59-.11.12-.24.25-.36.37l-67.2 67.2c-59.27 59.27-155.699 59.262-214.96 0-59.27-59.26-59.27-155.7 0-214.96l37.106-37.106c9.84-9.84 26.786-3.3 27.294 10.606.648 17.722 3.826 35.527 9.69 52.721 1.986 5.822.567 12.262-3.783 16.612l-13.087 13.087c-28.026 28.026-28.905 73.66-1.155 101.96 28.024 28.579 74.086 28.749 102.325.51l67.2-67.19c28.191-28.191 28.073-73.757 0-101.83-3.701-3.694-7.429-6.564-10.341-8.569a16.037 16.037 0 0 1-6.947-12.606c-.396-10.567 3.348-21.456 11.698-29.806l21.054-21.055c5.521-5.521 14.182-6.199 20.584-1.731a152.482 152.482 0 0 1 20.522 17.197zM467.547 44.449c-59.261-59.262-155.69-59.27-214.96 0l-67.2 67.2c-.12.12-.25.25-.36.37-58.566 58.892-59.387 154.781.36 214.59a152.454 152.454 0 0 0 20.521 17.196c6.402 4.468 15.064 3.789 20.584-1.731l21.054-21.055c8.35-8.35 12.094-19.239 11.698-29.806a16.037 16.037 0 0 0-6.947-12.606c-2.912-2.005-6.64-4.875-10.341-8.569-28.073-28.073-28.191-73.639 0-101.83l67.2-67.19c28.239-28.239 74.3-28.069 102.325.51 27.75 28.3 26.872 73.934-1.155 101.96l-13.087 13.087c-4.35 4.35-5.769 10.79-3.783 16.612 5.864 17.194 9.042 34.999 9.69 52.721.509 13.906 17.454 20.446 27.294 10.606l37.106-37.106c59.271-59.259 59.271-155.699.001-214.959z"
    ></path>
  </svg>
</template>

<style>
  @media only screen {
    /* anchor button */
    .anchor {
      opacity: 0;
      margin-left: 5px;
    }

    /* anchor buttons within <h2>'s */
    h2 .anchor {
      margin-left: 10px;
    }

    /* anchor buttons when hovered/focused and anything containing an anchor button when hovered */
    *:hover > .anchor,
    .anchor:hover,
    .anchor:focus {
      opacity: 1;
    }

    /* anchor button when hovered */
    .anchor:hover {
      cursor: pointer;
    }
  }

  /* always show anchor button on devices with no mouse/hover ability */
  @media (hover: none) {
    .anchor {
      opacity: 1;
    }
  }

  /* always hide anchor button on print */
  @media only print {
    .anchor {
      display: none;
    }
  }
</style>
<!-- 
    Attributes Plugin

    Allows arbitrary HTML attributes to be attached to (almost) any element.
    Place an HTML comment inside or next to the desired element with the content:
    $attribute="value"
-->

<script type="module">
  // start script
  function start() {
    // get list of comments in document
    const comments = findComments();

    for (const comment of comments)
      if (comment.parentElement)
        addAttributes(comment.parentElement, comment.nodeValue.trim());
  }

  // add html attributes to specified element based on string of
  // html attributes and values
  function addAttributes(element, text) {
    // regex's for finding attribute/value pairs in the format of
    // attribute="value" or attribute='value
    const regex2 = /\$([a-zA-Z\-]+)?=\"(.+?)\"/;
    const regex1 = /\$([a-zA-Z\-]+)?=\'(.+?)\'/;

    // loop through attribute/value pairs
    let match;
    while ((match = text.match(regex2) || text.match(regex1))) {
      // get attribute and value from regex capture groups
      let attribute = match[1];
      let value = match[2];

      // remove from string
      text = text.substring(match.index + match[0].length);

      if (!attribute || !value) break;

      // set attribute of parent element
      try {
        element.setAttribute(attribute, value);
      } catch (error) {
        console.log(error);
      }

      // special case for colspan
      if (attribute === "colspan") removeTableCells(element, value);
    }
  }

  // get list of comment elements in document
  function findComments() {
    const comments = [];

    // iterate over comment nodes in document
    function acceptNode(node) {
      return NodeFilter.FILTER_ACCEPT;
    }
    const iterator = document.createNodeIterator(
      document.body,
      NodeFilter.SHOW_COMMENT,
      acceptNode
    );
    let node;
    while ((node = iterator.nextNode())) comments.push(node);

    return comments;
  }

  // remove certain number of cells after specified cell
  function removeTableCells(cell, number) {
    number = parseInt(number);
    if (!number) return;

    // remove elements
    for (; number > 1; number--) {
      if (cell.nextElementSibling) cell.nextElementSibling.remove();
    }
  }

  // start script on DOMContentLoaded instead of load to ensure this plugins
  // runs before other plugins
  window.addEventListener("DOMContentLoaded", start);
</script>
<!--
  Jump to First Plugin

  Adds a button next to each reference entry, figure, and table that jumps the
  page to the first occurrence of a link to that item in the manuscript.
-->

<script type="module">
  // whether to add buttons next to reference entries
  const references = "true";
  // whether to add buttons next to figures
  const figures = "true";
  // whether to add buttons next to tables
  const tables = "true";

  // start script
  function start() {
    if (references !== "false")
      makeButtons(`div[id^="ref-"]`, ".csl-left-margin", "reference");
    if (figures !== "false")
      makeButtons(`div[id^="fig:"]`, "figcaption", "figure");
    if (tables !== "false") makeButtons(`div[id^="tbl:"]`, "caption", "table");
  }

  // when jump button clicked
  function onButtonClick() {
    const first = getFirstOccurrence(this.dataset.id);
    if (!first) return;

    // update url hash so navigating "back" in history will return user to button
    window.location.hash = this.dataset.id;
    // scroll to link
    const timeout = function () {
      goToElement(first, window.innerHeight * 0.5);
    };
    window.setTimeout(timeout, 0);
  }

  // get first occurrence of link to item in document
  function getFirstOccurrence(id) {
    let query = "a";
    query += '[href="#' + id + '"]';
    // exclude buttons, anchor links, toc links, etc
    query += ":not(.button):not(.icon_button):not(.anchor):not(.toc_link)";
    return document.querySelector(query);
  }

  // add button next to each reference entry, figure, or table
  function makeButtons(query, containerQuery, subject) {
    const elements = document.querySelectorAll(query);
    for (const element of elements) {
      const id = element.id;
      const buttonContainer = element.querySelector(containerQuery);
      const first = getFirstOccurrence(id);

      // if can't find link to reference or place to put button, ignore
      if (!first || !buttonContainer) continue;

      // make jump button
      let button = document.createElement("button");
      button.classList.add("icon_button", "jump_arrow");
      button.title = `Jump to the first occurrence of this ${subject} in the document`;
      const icon = document.querySelector(".icon_angle_double_up");
      button.innerHTML = icon.innerHTML;
      button.dataset.id = id;
      button.dataset.ignore = "true";
      button.addEventListener("click", onButtonClick);
      buttonContainer.prepend(button);
    }
  }

  // start script when document is finished loading
  window.addEventListener("load", start);
</script>

<!-- angle double up icon -->

<template class="icon_angle_double_up">
  <!-- modified from: https://fontawesome.com/icons/angle-double-up -->
  <svg width="16" height="16" viewBox="0 0 320 512">
    <path
      fill="currentColor"
      d="M177 255.7l136 136c9.4 9.4 9.4 24.6 0 33.9l-22.6 22.6c-9.4 9.4-24.6 9.4-33.9 0L160 351.9l-96.4 96.4c-9.4 9.4-24.6 9.4-33.9 0L7 425.7c-9.4-9.4-9.4-24.6 0-33.9l136-136c9.4-9.5 24.6-9.5 34-.1zm-34-192L7 199.7c-9.4 9.4-9.4 24.6 0 33.9l22.6 22.6c9.4 9.4 24.6 9.4 33.9 0l96.4-96.4 96.4 96.4c9.4 9.4 24.6 9.4 33.9 0l22.6-22.6c9.4-9.4 9.4-24.6 0-33.9l-136-136c-9.2-9.4-24.4-9.4-33.8 0z"
    ></path>
  </svg>
</template>

<style>
  @media only screen {
    /* jump button */
    .jump_arrow {
      position: relative;
      top: 0.125em;
      margin-right: 5px;
    }
  }

  /* always hide jump button on print */
  @media only print {
    .jump_arrow {
      display: none;
    }
  }
</style>
<!-- 
    Lightbox Plugin

    Makes it such that when a user clicks on an image, the image fills the
    screen and the user can pan/drag/zoom the image and navigate between other
    images in the document.
-->

<script type="module">
  // list of possible zoom/scale factors
  const zooms =
    "0.1, 0.25, 0.333333, 0.5, 0.666666, 0.75, 1, 1.25, 1.5, 1.75, 2, 2.5, 3, 3.5, 4, 5, 6, 7, 8";
  // whether to fit image to view ('fit'), display at 100% and shrink if
  // necessary ('shrink'), or always display at 100% ('100')
  const defaultZoom = "fit";
  // whether to zoom in/out toward center of view ('true') or mouse ('false')
  const centerZoom = "false";

  // start script
  function start() {
    // run through each <img> element
    const imgs = document.querySelectorAll("figure > img");
    let count = 1;
    for (const img of imgs) {
      img.classList.add("lightbox_document_img");
      img.dataset.number = count;
      img.dataset.total = imgs.length;
      img.addEventListener("click", openLightbox);
      count++;
    }

    // attach mouse and key listeners to window
    window.addEventListener("mousemove", onWindowMouseMove);
    window.addEventListener("keyup", onKeyUp);
  }

  // when mouse is moved anywhere in window
  function onWindowMouseMove(event) {
    window.mouseX = event.clientX;
    window.mouseY = event.clientY;
  }

  // when key pressed
  function onKeyUp(event) {
    if (!event || !event.key) return;

    switch (event.key) {
      // trigger click of prev button
      case "ArrowLeft":
        const prevButton = document.getElementById("lightbox_prev_button");
        if (prevButton) prevButton.click();
        break;
      // trigger click of next button
      case "ArrowRight":
        const nextButton = document.getElementById("lightbox_next_button");
        if (nextButton) nextButton.click();
        break;
      // close on esc
      case "Escape":
        closeLightbox();
        break;
    }
  }

  // open lightbox
  function openLightbox() {
    const lightbox = makeLightbox(this);
    if (!lightbox) return;

    blurBody(lightbox);
    document.body.appendChild(lightbox);
  }

  // make lightbox
  function makeLightbox(img) {
    // delete lightbox if it exists, start fresh
    closeLightbox();

    // create screen overlay containing lightbox
    const overlay = document.createElement("div");
    overlay.id = "lightbox_overlay";

    // create image info boxes
    const numberInfo = document.createElement("div");
    const zoomInfo = document.createElement("div");
    numberInfo.id = "lightbox_number_info";
    zoomInfo.id = "lightbox_zoom_info";

    // create container for image
    const imageContainer = document.createElement("div");
    imageContainer.id = "lightbox_image_container";
    const lightboxImg = makeLightboxImg(
      img,
      imageContainer,
      numberInfo,
      zoomInfo
    );
    imageContainer.appendChild(lightboxImg);

    // create bottom container for caption and navigation buttons
    const bottomContainer = document.createElement("div");
    bottomContainer.id = "lightbox_bottom_container";
    const caption = makeCaption(img);
    const prevButton = makePrevButton(img);
    const nextButton = makeNextButton(img);
    bottomContainer.appendChild(prevButton);
    bottomContainer.appendChild(caption);
    bottomContainer.appendChild(nextButton);

    // attach top middle and bottom to overlay
    overlay.appendChild(numberInfo);
    overlay.appendChild(zoomInfo);
    overlay.appendChild(imageContainer);
    overlay.appendChild(bottomContainer);

    return overlay;
  }

  // make <img> object that is intuitively draggable and zoomable
  function makeLightboxImg(sourceImg, container, numberInfoBox, zoomInfoBox) {
    // create copy of source <img>
    const img = sourceImg.cloneNode(true);
    img.classList.remove("lightbox_document_img");
    img.removeAttribute("id");
    img.removeAttribute("width");
    img.removeAttribute("height");
    img.style.position = "unset";
    img.style.margin = "0";
    img.style.padding = "0";
    img.style.width = "";
    img.style.height = "";
    img.style.minWidth = "";
    img.style.minHeight = "";
    img.style.maxWidth = "";
    img.style.maxHeight = "";
    img.id = "lightbox_img";

    // build sorted list of zoomSteps
    const zoomSteps = zooms.split(/[^0-9.]/).map((step) => parseFloat(step));
    zoomSteps.sort((a, b) => a - b);

    // <img> object property variables
    let zoom = 1;
    let translateX = 0;
    let translateY = 0;
    let clickMouseX = undefined;
    let clickMouseY = undefined;
    let clickTranslateX = undefined;
    let clickTranslateY = undefined;

    updateNumberInfo();

    // update image numbers displayed in info box
    function updateNumberInfo() {
      numberInfoBox.innerHTML =
        sourceImg.dataset.number + " of " + sourceImg.dataset.total;
    }

    // update zoom displayed in info box
    function updateZoomInfo() {
      let zoomInfo = zoom * 100;
      if (!Number.isInteger(zoomInfo)) zoomInfo = zoomInfo.toFixed(2);
      zoomInfoBox.innerHTML = zoomInfo + "%";
    }

    // move to closest zoom step above current zoom
    const zoomIn = function () {
      for (const zoomStep of zoomSteps) {
        if (zoomStep > zoom) {
          zoom = zoomStep;
          break;
        }
      }
      updateTransform();
    };

    // move to closest zoom step above current zoom
    const zoomOut = function () {
      zoomSteps.reverse();
      for (const zoomStep of zoomSteps) {
        if (zoomStep < zoom) {
          zoom = zoomStep;
          break;
        }
      }
      zoomSteps.reverse();

      updateTransform();
    };

    // update display of <img> based on scale/translate properties
    const updateTransform = function () {
      // set transform
      img.style.transform =
        "translate(" +
        (translateX || 0) +
        "px," +
        (translateY || 0) +
        "px) scale(" +
        (zoom || 1) +
        ")";

      // get new width/height after scale
      const rect = img.getBoundingClientRect();
      // limit translate
      translateX = Math.max(translateX, -rect.width / 2);
      translateX = Math.min(translateX, rect.width / 2);
      translateY = Math.max(translateY, -rect.height / 2);
      translateY = Math.min(translateY, rect.height / 2);

      // set transform
      img.style.transform =
        "translate(" +
        (translateX || 0) +
        "px," +
        (translateY || 0) +
        "px) scale(" +
        (zoom || 1) +
        ")";

      updateZoomInfo();
    };

    // fit <img> to container
    const fit = function () {
      // no x/y offset, 100% zoom by default
      translateX = 0;
      translateY = 0;
      zoom = 1;

      // widths of <img> and container
      const imgWidth = img.naturalWidth;
      const imgHeight = img.naturalHeight;
      const containerWidth = parseFloat(
        window.getComputedStyle(container).width
      );
      const containerHeight = parseFloat(
        window.getComputedStyle(container).height
      );

      // how much zooming is needed to fit <img> to container
      const xRatio = imgWidth / containerWidth;
      const yRatio = imgHeight / containerHeight;
      const maxRatio = Math.max(xRatio, yRatio);
      const newZoom = 1 / maxRatio;

      // fit <img> to container according to option
      if (defaultZoom === "shrink") {
        if (maxRatio > 1) zoom = newZoom;
      } else if (defaultZoom === "fit") zoom = newZoom;

      updateTransform();
    };

    // when mouse wheel is rolled anywhere in container
    const onContainerWheel = function (event) {
      if (!event) return;

      // let ctrl + mouse wheel to zoom behave as normal
      if (event.ctrlKey) return;

      // prevent normal scroll behavior
      event.preventDefault();
      event.stopPropagation();

      // point around which to scale img
      const viewRect = container.getBoundingClientRect();
      const viewX = (viewRect.left + viewRect.right) / 2;
      const viewY = (viewRect.top + viewRect.bottom) / 2;
      const originX = centerZoom === "true" ? viewX : mouseX;
      const originY = centerZoom === "true" ? viewY : mouseY;

      // get point on image under origin
      const oldRect = img.getBoundingClientRect();
      const oldPercentX = (originX - oldRect.left) / oldRect.width;
      const oldPercentY = (originY - oldRect.top) / oldRect.height;

      // increment/decrement zoom
      if (event.deltaY < 0) zoomIn();
      if (event.deltaY > 0) zoomOut();

      // get offset between previous image point and origin
      const newRect = img.getBoundingClientRect();
      const offsetX = originX - (newRect.left + newRect.width * oldPercentX);
      const offsetY = originY - (newRect.top + newRect.height * oldPercentY);

      // translate image to keep image point under origin
      translateX += offsetX;
      translateY += offsetY;

      // perform translate
      updateTransform();
    };

    // when container is clicked
    function onContainerClick(event) {
      // if container itself is target of click, and not other
      // element above it
      if (event.target === this) closeLightbox();
    }

    // when mouse button is pressed on image
    const onImageMouseDown = function (event) {
      // store original mouse position relative to image
      clickMouseX = window.mouseX;
      clickMouseY = window.mouseY;
      clickTranslateX = translateX;
      clickTranslateY = translateY;
      event.stopPropagation();
      event.preventDefault();
    };

    // when mouse button is released anywhere in window
    const onWindowMouseUp = function (event) {
      // reset original mouse position
      clickMouseX = undefined;
      clickMouseY = undefined;
      clickTranslateX = undefined;
      clickTranslateY = undefined;

      // remove global listener if lightbox removed from document
      if (!document.body.contains(container))
        window.removeEventListener("mouseup", onWindowMouseUp);
    };

    // when mouse is moved anywhere in window
    const onWindowMouseMove = function (event) {
      if (
        clickMouseX === undefined ||
        clickMouseY === undefined ||
        clickTranslateX === undefined ||
        clickTranslateY === undefined
      )
        return;

      // offset image based on original and current mouse position
      translateX = clickTranslateX + window.mouseX - clickMouseX;
      translateY = clickTranslateY + window.mouseY - clickMouseY;
      updateTransform();
      event.preventDefault();

      // remove global listener if lightbox removed from document
      if (!document.body.contains(container))
        window.removeEventListener("mousemove", onWindowMouseMove);
    };

    // when window is resized
    const onWindowResize = function (event) {
      fit();

      // remove global listener if lightbox removed from document
      if (!document.body.contains(container))
        window.removeEventListener("resize", onWindowResize);
    };

    // attach the necessary event listeners
    img.addEventListener("dblclick", fit);
    img.addEventListener("mousedown", onImageMouseDown);
    container.addEventListener("wheel", onContainerWheel);
    container.addEventListener("mousedown", onContainerClick);
    container.addEventListener("touchstart", onContainerClick);
    window.addEventListener("mouseup", onWindowMouseUp);
    window.addEventListener("mousemove", onWindowMouseMove);
    window.addEventListener("resize", onWindowResize);

    // run fit() after lightbox atttached to document and <img> Loaded
    // so needed container and img dimensions available
    img.addEventListener("load", fit);

    return img;
  }

  // make caption
  function makeCaption(img) {
    const caption = document.createElement("div");
    caption.id = "lightbox_caption";
    const captionSource = img.nextElementSibling;
    if (captionSource.tagName.toLowerCase() === "figcaption") {
      const captionCopy = makeCopy(captionSource);
      caption.innerHTML = captionCopy.innerHTML;
    }

    caption.addEventListener("touchstart", function (event) {
      event.stopPropagation();
    });

    return caption;
  }

  // make carbon copy of html dom element
  function makeCopy(source) {
    const sourceCopy = source.cloneNode(true);

    // delete elements marked with ignore (eg anchor and jump buttons)
    const deleteFromCopy = sourceCopy.querySelectorAll('[data-ignore="true"]');
    for (const element of deleteFromCopy) element.remove();

    // delete certain element attributes
    const attributes = [
      "id",
      "data-collapsed",
      "data-selected",
      "data-highlighted",
      "data-glow",
    ];
    for (const attribute of attributes) {
      sourceCopy.removeAttribute(attribute);
      const elements = sourceCopy.querySelectorAll("[" + attribute + "]");
      for (const element of elements) element.removeAttribute(attribute);
    }

    return sourceCopy;
  }

  // make button to jump to previous image in document
  function makePrevButton(img) {
    const prevButton = document.createElement("button");
    prevButton.id = "lightbox_prev_button";
    prevButton.title = "Jump to the previous image in the document [←]";
    prevButton.classList.add("icon_button", "lightbox_button");
    prevButton.innerHTML = document.querySelector(".icon_caret_left").innerHTML;

    // attach click listeners to button
    prevButton.addEventListener("click", function () {
      getPrevImg(img).click();
    });

    return prevButton;
  }

  // make button to jump to next image in document
  function makeNextButton(img) {
    const nextButton = document.createElement("button");
    nextButton.id = "lightbox_next_button";
    nextButton.title = "Jump to the next image in the document [→]";
    nextButton.classList.add("icon_button", "lightbox_button");
    nextButton.innerHTML = document.querySelector(
      ".icon_caret_right"
    ).innerHTML;

    // attach click listeners to button
    nextButton.addEventListener("click", function () {
      getNextImg(img).click();
    });

    return nextButton;
  }

  // get previous image in document
  function getPrevImg(img) {
    const imgs = document.querySelectorAll(".lightbox_document_img");

    // find index of provided img
    let index;
    for (index = 0; index < imgs.length; index++) {
      if (imgs[index] === img) break;
    }

    // wrap index to other side if < 1
    if (index - 1 >= 0) index--;
    else index = imgs.length - 1;
    return imgs[index];
  }

  // get next image in document
  function getNextImg(img) {
    const imgs = document.querySelectorAll(".lightbox_document_img");

    // find index of provided img
    let index;
    for (index = 0; index < imgs.length; index++) {
      if (imgs[index] === img) break;
    }

    // wrap index to other side if > total
    if (index + 1 <= imgs.length - 1) index++;
    else index = 0;
    return imgs[index];
  }

  // close lightbox
  function closeLightbox() {
    focusBody();

    const lightbox = document.getElementById("lightbox_overlay");
    if (lightbox) lightbox.remove();
  }

  // make all elements behind lightbox non-focusable
  function blurBody(overlay) {
    const all = document.querySelectorAll("*");
    for (const element of all) element.tabIndex = -1;
    document.body.classList.add("body_no_scroll");
  }

  // make all elements focusable again
  function focusBody() {
    const all = document.querySelectorAll("*");
    for (const element of all) element.removeAttribute("tabIndex");
    document.body.classList.remove("body_no_scroll");
  }

  // start script when document is finished loading
  window.addEventListener("load", start);
</script>

<!-- caret left icon -->

<template class="icon_caret_left">
  <!-- modified from: https://fontawesome.com/icons/caret-left -->
  <svg width="16" height="16" viewBox="0 0 192 512">
    <path
      fill="currentColor"
      d="M192 127.338v257.324c0 17.818-21.543 26.741-34.142 14.142L29.196 270.142c-7.81-7.81-7.81-20.474 0-28.284l128.662-128.662c12.599-12.6 34.142-3.676 34.142 14.142z"
    ></path>
  </svg>
</template>

<!-- caret right icon -->

<template class="icon_caret_right">
  <!-- modified from: https://fontawesome.com/icons/caret-right -->
  <svg width="16" height="16" viewBox="0 0 192 512">
    <path
      fill="currentColor"
      d="M0 384.662V127.338c0-17.818 21.543-26.741 34.142-14.142l128.662 128.662c7.81 7.81 7.81 20.474 0 28.284L34.142 398.804C21.543 411.404 0 402.48 0 384.662z"
    ></path>
  </svg>
</template>

<style>
  @media only screen {
    /* regular <img> in document when hovered */
    img.lightbox_document_img:hover {
      cursor: pointer;
    }

    .body_no_scroll {
      overflow: hidden !important;
    }

    /* screen overlay */
    #lightbox_overlay {
      display: flex;
      flex-direction: column;
      position: fixed;
      left: 0;
      top: 0;
      right: 0;
      bottom: 0;
      background: rgba(0, 0, 0, 0.75);
      z-index: 3;
    }

    /* middle area containing lightbox image */
    #lightbox_image_container {
      flex-grow: 1;
      display: flex;
      justify-content: center;
      align-items: center;
      overflow: hidden;
      position: relative;
      padding: 20px;
    }

    /* bottom area containing caption */
    #lightbox_bottom_container {
      display: flex;
      justify-content: center;
      align-items: center;
      height: 100px;
      min-height: 100px;
      max-height: 100px;
      background: rgba(0, 0, 0, 0.5);
    }

    /* image number info text box */
    #lightbox_number_info {
      position: absolute;
      color: #ffffff;
      font-weight: 600;
      left: 2px;
      top: 0;
      z-index: 4;
    }

    /* zoom info text box */
    #lightbox_zoom_info {
      position: absolute;
      color: #ffffff;
      font-weight: 600;
      right: 2px;
      top: 0;
      z-index: 4;
    }

    /* copy of image caption */
    #lightbox_caption {
      box-sizing: border-box;
      display: inline-block;
      width: 100%;
      max-height: 100%;
      padding: 10px 0;
      text-align: center;
      overflow-y: auto;
      color: #ffffff;
    }

    /* navigation previous/next button */
    .lightbox_button {
      width: 100px;
      height: 100%;
      min-width: 100px;
      min-height: 100%;
      color: #ffffff;
    }

    /* navigation previous/next button when hovered */
    .lightbox_button:hover {
      background: none !important;
    }

    /* navigation button icon */
    .lightbox_button > svg {
      height: 25px;
    }

    /* figure auto-number */
    #lightbox_caption > span:first-of-type {
      font-weight: bold;
      margin-right: 5px;
    }

    /* lightbox image when hovered */
    #lightbox_img:hover {
      cursor: grab;
    }

    /* lightbox image when grabbed */
    #lightbox_img:active {
      cursor: grabbing;
    }
  }

  /* when on screen < 480px wide */
  @media only screen and (max-width: 480px) {
    /* make navigation buttons skinnier on small screens to make more room for caption text */
    .lightbox_button {
      width: 50px;
      min-width: 50px;
    }
  }

  /* always hide lightbox on print */
  @media only print {
    #lightbox_overlay {
      display: none;
    }
  }
</style>
<!-- 
  Link Highlight Plugin

  Makes it such that when a user hovers or focuses a link, other links that have
  the same target will be highlighted. It also makes it such that when clicking
  a link, the target of the link (eg reference, figure, table) is briefly
  highlighted.
-->

<script type="module">
  // whether to also highlight links that go to external urls
  const externalLinks = "false";
  // whether user must click off to unhighlight instead of just
  // un-hovering
  const clickUnhighlight = "false";
  // whether to also highlight links that are unique
  const highlightUnique = "true";

  // start script
  function start() {
    const links = getLinks();
    for (const link of links) {
      // attach mouse and focus listeners to link
      link.addEventListener("mouseenter", onLinkFocus);
      link.addEventListener("focus", onLinkFocus);
      link.addEventListener("mouseleave", onLinkUnhover);
    }

    // attach click and hash change listeners to window
    window.addEventListener("click", onClick);
    window.addEventListener("touchstart", onClick);
    window.addEventListener("hashchange", onHashChange);

    // run hash change on window load in case user has navigated
    // directly to hash
    onHashChange();
  }

  // when link is focused (tabbed to) or hovered
  function onLinkFocus() {
    highlight(this);
  }

  // when link is unhovered
  function onLinkUnhover() {
    if (clickUnhighlight !== "true") unhighlightAll();
  }

  // when the mouse is clicked anywhere in window
  function onClick(event) {
    unhighlightAll();
  }

  // when hash (eg manuscript.html#introduction) changes
  function onHashChange() {
    const target = getHashTarget();
    if (target) glowElement(target);
  }

  // start glow sequence on an element
  function glowElement(element) {
    const startGlow = function () {
      onGlowEnd();
      element.dataset.glow = "true";
      element.addEventListener("animationend", onGlowEnd);
    };
    const onGlowEnd = function () {
      element.removeAttribute("data-glow");
      element.removeEventListener("animationend", onGlowEnd);
    };
    startGlow();
  }

  // highlight link and all others with same target
  function highlight(link) {
    // force unhighlight all to start fresh
    unhighlightAll();

    // get links with same target
    if (!link) return;
    const sameLinks = getSameLinks(link);

    // if link unique and option is off, exit and don't highlight
    if (sameLinks.length <= 1 && highlightUnique !== "true") return;

    // highlight all same links, and "select" (special highlight) this
    // one
    for (const sameLink of sameLinks) {
      if (sameLink === link) sameLink.setAttribute("data-selected", "true");
      else sameLink.setAttribute("data-highlighted", "true");
    }
  }

  // unhighlight all links
  function unhighlightAll() {
    const links = getLinks();
    for (const link of links) {
      link.setAttribute("data-selected", "false");
      link.setAttribute("data-highlighted", "false");
    }
  }

  // get links with same target
  function getSameLinks(link) {
    const results = [];
    const links = getLinks();
    for (const otherLink of links) {
      if (otherLink.getAttribute("href") === link.getAttribute("href"))
        results.push(otherLink);
    }
    return results;
  }

  // get all links of types we wish to handle
  function getLinks() {
    let query = "a";
    if (externalLinks !== "true") query += '[href^="#"]';
    // exclude buttons, anchor links, toc links, etc
    query += ":not(.button):not(.icon_button):not(.anchor):not(.toc_link)";
    return document.querySelectorAll(query);
  }

  // start script when document is finished loading
  window.addEventListener("load", start);
</script>

<style>
  @media only screen {
    /* anything with data-highlighted attribute true */
    [data-highlighted="true"] {
      background: #ffeb3b;
    }

    /* anything with data-selected attribute true */
    [data-selected="true"] {
      background: #ff8a65 !important;
    }

    /* animation definition for glow */
    @keyframes highlight_glow {
      0% {
        background: none;
      }
      10% {
        background: #bbdefb;
      }
      100% {
        background: none;
      }
    }

    /* anything with data-glow attribute true */
    [data-glow="true"] {
      animation: highlight_glow 2s;
    }
  }
</style>
<!--
  Table of Contents Plugin

  Provides a "table of contents" (toc) panel on the side of the document that
  allows the user to conveniently navigate between sections of the document.
-->

<script type="module">
  // which types of elements to add links for, in "document.querySelector" format
  const typesQuery = "h1, h2, h3";
  // whether toc starts open. use 'true' or 'false', or 'auto' to
  // use 'true' behavior when screen wide enough and 'false' when not
  const startOpen = "false";
  // whether toc closes when clicking on toc link. use 'true' or
  // 'false', or 'auto' to use 'false' behavior when screen wide
  // enough and 'true' when not
  const clickClose = "auto";
  // if list item is more than this many characters, text will be
  // truncated
  const charLimit = "50";
  // whether or not to show bullets next to each toc item
  const bullets = "false";

  // start script
  function start() {
    // make toc panel and populate with entries (links to document
    // sections)
    const panel = makePanel();
    if (!panel) return;
    makeEntries(panel);
    // attach panel to document after making entries, so 'toc' heading
    // in panel isn't included in toc
    document.body.insertBefore(panel, document.body.firstChild);

    // initial panel state
    if (startOpen === "true" || (startOpen === "auto" && !isSmallScreen()))
      openPanel();
    else closePanel();

    // attach click, scroll, and hash change listeners to window
    window.addEventListener("click", onClick);
    window.addEventListener("scroll", onScroll);
    window.addEventListener("hashchange", onScroll);
    window.addEventListener("keyup", onKeyUp);
    onScroll();

    // add class to push document body down out of way of toc button
    document.body.classList.add("toc_body_nudge");
  }

  // determine if screen wide enough to fit toc panel
  function isSmallScreen() {
    // in default theme:
    // 816px = 8.5in = width of "page" (<body>) element
    // 260px = min width of toc panel (*2 for both sides of <body>)
    return window.innerWidth < 816 + 260 * 2;
  }

  // when mouse is clicked anywhere in window
  function onClick() {
    if (isSmallScreen()) closePanel();
  }

  // when window is scrolled or hash changed
  function onScroll() {
    highlightViewed();
  }

  // when key pressed
  function onKeyUp(event) {
    if (!event || !event.key) return;

    // close on esc
    if (event.key === "Escape") closePanel();
  }

  // find entry of currently viewed document section in toc and highlight
  function highlightViewed() {
    const firstId = getFirstInView(typesQuery);

    // get toc entries (links), unhighlight all, then highlight viewed
    const list = document.getElementById("toc_list");
    if (!firstId || !list) return;
    const links = list.querySelectorAll("a");
    for (const link of links) link.dataset.viewing = "false";
    const link = list.querySelector('a[href="#' + firstId + '"]');
    if (!link) return;
    link.dataset.viewing = "true";
  }

  // get first or previous toc listed element in top half of view
  function getFirstInView(query) {
    // get all elements matching query and with id
    const elements = document.querySelectorAll(query);
    const elementsWithIds = [];
    for (const element of elements) {
      if (element.id) elementsWithIds.push(element);
    }

    // get first or previous element in top half of view
    for (let i = 0; i < elementsWithIds.length; i++) {
      const element = elementsWithIds[i];
      const prevElement = elementsWithIds[Math.max(0, i - 1)];
      if (element.getBoundingClientRect().top >= 0) {
        if (element.getBoundingClientRect().top < window.innerHeight / 2)
          return element.id;
        else return prevElement.id;
      }
    }
  }

  // make panel
  function makePanel() {
    // create panel
    const panel = document.createElement("div");
    panel.id = "toc_panel";
    if (bullets === "true") panel.dataset.bullets = "true";

    // create header
    const header = document.createElement("div");
    header.id = "toc_header";

    // create toc button
    const button = document.createElement("button");
    button.id = "toc_button";
    button.innerHTML = document.querySelector(".icon_th_list").innerHTML;
    button.title = "Table of Contents";
    button.classList.add("icon_button");

    // create header text
    const text = document.createElement("h4");
    text.innerHTML = "Table of Contents";

    // create container for toc list
    const list = document.createElement("div");
    list.id = "toc_list";

    // attach click listeners
    panel.addEventListener("click", onPanelClick);
    header.addEventListener("click", onHeaderClick);
    button.addEventListener("click", onButtonClick);

    // attach elements
    header.appendChild(button);
    header.appendChild(text);
    panel.appendChild(header);
    panel.appendChild(list);

    return panel;
  }

  // create toc entries (links) to each element of the specified types
  function makeEntries(panel) {
    const elements = document.querySelectorAll(typesQuery);
    for (const element of elements) {
      // do not add link if element doesn't have assigned id
      if (!element.id) continue;

      // create link/list item
      const link = document.createElement("a");
      link.classList.add("toc_link");
      switch (element.tagName.toLowerCase()) {
        case "h1":
          link.dataset.level = "1";
          break;
        case "h2":
          link.dataset.level = "2";
          break;
        case "h3":
          link.dataset.level = "3";
          break;
        case "h4":
          link.dataset.level = "4";
          break;
      }
      link.title = element.innerText;
      let text = element.innerText;
      if (text.length > charLimit) text = text.slice(0, charLimit) + "...";
      link.innerHTML = text;
      link.href = "#" + element.id;
      link.addEventListener("click", onLinkClick);

      // attach link
      panel.querySelector("#toc_list").appendChild(link);
    }
  }

  // when panel is clicked
  function onPanelClick(event) {
    // stop click from propagating to window/document and closing panel
    event.stopPropagation();
  }

  // when header itself is clicked
  function onHeaderClick(event) {
    togglePanel();
  }

  // when button is clicked
  function onButtonClick(event) {
    togglePanel();
    // stop header underneath button from also being clicked
    event.stopPropagation();
  }

  // when link is clicked
  function onLinkClick(event) {
    if (clickClose === "true" || (clickClose === "auto" && isSmallScreen()))
      closePanel();
    else openPanel();
  }

  // open panel if closed, close if opened
  function togglePanel() {
    const panel = document.getElementById("toc_panel");
    if (!panel) return;

    if (panel.dataset.open === "true") closePanel();
    else openPanel();
  }

  // open panel
  function openPanel() {
    const panel = document.getElementById("toc_panel");
    if (panel) panel.dataset.open = "true";
  }

  // close panel
  function closePanel() {
    const panel = document.getElementById("toc_panel");
    if (panel) panel.dataset.open = "false";
  }

  // start script when document is finished loading
  window.addEventListener("load", start);
</script>

<!-- th list icon -->

<template class="icon_th_list">
  <!-- modified from: https://fontawesome.com/icons/th-list -->
  <svg width="16" height="16" viewBox="0 0 512 512" tabindex="-1">
    <path
      fill="currentColor"
      d="M96 96c0 26.51-21.49 48-48 48S0 122.51 0 96s21.49-48 48-48 48 21.49 48 48zM48 208c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48-21.49-48-48-48zm0 160c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48-21.49-48-48-48zm96-236h352c8.837 0 16-7.163 16-16V76c0-8.837-7.163-16-16-16H144c-8.837 0-16 7.163-16 16v40c0 8.837 7.163 16 16 16zm0 160h352c8.837 0 16-7.163 16-16v-40c0-8.837-7.163-16-16-16H144c-8.837 0-16 7.163-16 16v40c0 8.837 7.163 16 16 16zm0 160h352c8.837 0 16-7.163 16-16v-40c0-8.837-7.163-16-16-16H144c-8.837 0-16 7.163-16 16v40c0 8.837 7.163 16 16 16z"
      tabindex="-1"
    ></path>
  </svg>
</template>

<style>
  @media only screen {
    /* toc panel */
    #toc_panel {
      box-sizing: border-box;
      position: fixed;
      top: 0;
      left: 0;
      background: #ffffff;
      box-shadow: 0 0 20px rgba(0, 0, 0, 0.05);
      z-index: 2;
    }

    /* toc panel when closed */
    #toc_panel[data-open="false"] {
      min-width: 60px;
      width: 60px;
      height: 60px;
      border-right: solid 1px #bdbdbd;
      border-bottom: solid 1px #bdbdbd;
    }

    /* toc panel when open */
    #toc_panel[data-open="true"] {
      min-width: 260px;
      max-width: 480px;
      /* keep panel edge consistent distance away from "page" edge */
      width: calc(((100vw - 8.5in) / 2) - 30px - 40px);
      bottom: 0;
      border-right: solid 1px #bdbdbd;
    }

    /* toc panel header */
    #toc_header {
      box-sizing: border-box;
      display: flex;
      flex-direction: row;
      align-items: center;
      height: 60px;
      margin: 0;
      padding: 20px;
    }

    /* toc panel header when hovered */
    #toc_header:hover {
      cursor: pointer;
    }

    /* toc panel header when panel open */
    #toc_panel[data-open="true"] > #toc_header {
      border-bottom: solid 1px #bdbdbd;
    }

    /* toc open/close header button */
    #toc_button {
      margin-right: 20px;
    }

    /* hide toc list and header text when closed */
    #toc_panel[data-open="false"] > #toc_header > *:not(#toc_button),
    #toc_panel[data-open="false"] > #toc_list {
      display: none;
    }

    /* toc list of entries */
    #toc_list {
      box-sizing: border-box;
      width: 100%;
      padding: 20px;
      position: absolute;
      top: calc(60px + 1px);
      bottom: 0;
      overflow: auto;
    }

    /* toc entry, link to section in document */
    .toc_link {
      display: block;
      padding: 5px;
      position: relative;
      font-weight: 600;
      text-decoration: none;
    }

    /* toc entry when hovered or when "viewed" */
    .toc_link:hover,
    .toc_link[data-viewing="true"] {
      background: #f5f5f5;
    }

    /* toc entry, level 1 indentation */
    .toc_link[data-level="1"] {
      margin-left: 0;
    }

    /* toc entry, level 2 indentation */
    .toc_link[data-level="2"] {
      margin-left: 20px;
    }

    /* toc entry, level 3 indentation */
    .toc_link[data-level="3"] {
      margin-left: 40px;
    }

    /* toc entry, level 4 indentation */
    .toc_link[data-level="4"] {
      margin-left: 60px;
    }

    /* toc entry bullets */
    #toc_panel[data-bullets="true"] .toc_link[data-level]:before {
      position: absolute;
      left: -15px;
      top: -1px;
      font-size: 1.5em;
    }

    /* toc entry, level 2 bullet */
    #toc_panel[data-bullets="true"] .toc_link[data-level="2"]:before {
      content: "\2022";
    }

    /* toc entry, level 3 bullet */
    #toc_panel[data-bullets="true"] .toc_link[data-level="3"]:before {
      content: "\25AB";
    }

    /* toc entry, level 4 bullet */
    #toc_panel[data-bullets="true"] .toc_link[data-level="4"]:before {
      content: "-";
    }
  }

  /* when on screen < 8.5in wide */
  @media only screen and (max-width: 8.5in) {
    /* push <body> ("page") element down to make room for toc icon */
    .toc_body_nudge {
      padding-top: 60px;
    }

    /* toc icon when panel closed and not hovered */
    #toc_panel[data-open="false"]:not(:hover) {
      background: rgba(255, 255, 255, 0.75);
    }
  }

  /* always hide toc panel on print */
  @media only print {
    #toc_panel {
      display: none;
    }
  }
</style>
<!-- 
  Tooltips Plugin

  Makes it such that when the user hovers or focuses a link to a citation or
  figure, a tooltip appears with a preview of the reference content, along with
  arrows to navigate between instances of the same reference in the document.
-->

<script type="module">
  // whether user must click off to close tooltip instead of just un-hovering
  const clickClose = "false";
  // delay (in ms) between opening and closing tooltip
  const delay = "100";

  // start script
  function start() {
    const links = getLinks();
    for (const link of links) {
      // attach hover and focus listeners to link
      link.addEventListener("mouseover", onLinkHover);
      link.addEventListener("mouseleave", onLinkUnhover);
      link.addEventListener("focus", onLinkFocus);
      link.addEventListener("touchend", onLinkTouch);
    }

    // attach mouse, key, and resize listeners to window
    window.addEventListener("mousedown", onClick);
    window.addEventListener("touchstart", onClick);
    window.addEventListener("keyup", onKeyUp);
    window.addEventListener("resize", onResize);
  }

  // when link is hovered
  function onLinkHover() {
    // function to open tooltip
    const delayOpenTooltip = function () {
      openTooltip(this);
    }.bind(this);

    // run open function after delay
    this.openTooltipTimer = window.setTimeout(delayOpenTooltip, delay);
  }

  // when mouse leaves link
  function onLinkUnhover() {
    // cancel opening tooltip
    window.clearTimeout(this.openTooltipTimer);

    // don't close on unhover if option specifies
    if (clickClose === "true") return;

    // function to close tooltip
    const delayCloseTooltip = function () {
      // if tooltip open and if mouse isn't over tooltip, close
      const tooltip = document.getElementById("tooltip");
      if (tooltip && !tooltip.matches(":hover")) closeTooltip();
    };

    // run close function after delay
    this.closeTooltipTimer = window.setTimeout(delayCloseTooltip, delay);
  }

  // when link is focused (tabbed to)
  function onLinkFocus(event) {
    openTooltip(this);
  }

  // when link is touched on touch screen
  function onLinkTouch(event) {
    // attempt to force hover state on first tap always, and trigger
    // regular link click (and navigation) on second tap
    if (event.target === document.activeElement) event.target.click();
    else {
      document.activeElement.blur();
      event.target.focus();
    }
    if (event.cancelable) event.preventDefault();
    event.stopPropagation();
    return false;
  }

  // when mouse is clicked anywhere in window
  function onClick(event) {
    closeTooltip();
  }

  // when key pressed
  function onKeyUp(event) {
    if (!event || !event.key) return;

    switch (event.key) {
      // trigger click of prev button
      case "ArrowLeft":
        const prevButton = document.getElementById("tooltip_prev_button");
        if (prevButton) prevButton.click();
        break;
      // trigger click of next button
      case "ArrowRight":
        const nextButton = document.getElementById("tooltip_next_button");
        if (nextButton) nextButton.click();
        break;
      // close on esc
      case "Escape":
        closeTooltip();
        break;
    }
  }

  // when window is resized or zoomed
  function onResize() {
    closeTooltip();
  }

  // get all links of types we wish to handle
  function getLinks() {
    const queries = [];
    // exclude buttons, anchor links, toc links, etc
    const exclude =
      ":not(.button):not(.icon_button):not(.anchor):not(.toc_link)";
    queries.push('a[href^="#ref-"]' + exclude); // citation links
    queries.push('a[href^="#fig:"]' + exclude); // figure links
    const query = queries.join(", ");
    return document.querySelectorAll(query);
  }

  // get links with same target, get index of link in set, get total
  // same links
  function getSameLinks(link) {
    const sameLinks = [];
    const links = getLinks();
    for (const otherLink of links) {
      if (otherLink.getAttribute("href") === link.getAttribute("href"))
        sameLinks.push(otherLink);
    }

    return {
      elements: sameLinks,
      index: sameLinks.indexOf(link),
      total: sameLinks.length,
    };
  }

  // open tooltip
  function openTooltip(link) {
    // delete tooltip if it exists, start fresh
    closeTooltip();

    // make tooltip element
    const tooltip = makeTooltip(link);

    // if source couldn't be found and tooltip not made, exit
    if (!tooltip) return;

    // make navbar elements
    const navBar = makeNavBar(link);
    if (navBar) tooltip.firstElementChild.appendChild(navBar);

    // attach tooltip to page
    document.body.appendChild(tooltip);

    // position tooltip
    const position = function () {
      positionTooltip(link);
    };
    position();

    // if tooltip contains images, position again after they've loaded
    const imgs = tooltip.querySelectorAll("img");
    for (const img of imgs) img.addEventListener("load", position);
  }

  // close (delete) tooltip
  function closeTooltip() {
    const tooltip = document.getElementById("tooltip");
    if (tooltip) tooltip.remove();
  }

  // make tooltip
  function makeTooltip(link) {
    // get target element that link points to
    const source = getSource(link);

    // if source can't be found, exit
    if (!source) return;

    // create new tooltip
    const tooltip = document.createElement("div");
    tooltip.id = "tooltip";
    const tooltipContent = document.createElement("div");
    tooltipContent.id = "tooltip_content";
    tooltip.appendChild(tooltipContent);

    // make copy of source node and put in tooltip
    const sourceCopy = makeCopy(source);
    tooltipContent.appendChild(sourceCopy);

    // attach mouse event listeners
    tooltip.addEventListener("click", onTooltipClick);
    tooltip.addEventListener("mousedown", onTooltipClick);
    tooltip.addEventListener("touchstart", onTooltipClick);
    tooltip.addEventListener("mouseleave", onTooltipUnhover);

    // (for interaction with lightbox plugin)
    // transfer click on tooltip copied img to original img
    const sourceImg = source.querySelector("img");
    const sourceCopyImg = sourceCopy.querySelector("img");
    if (sourceImg && sourceCopyImg) {
      const clickImg = function () {
        sourceImg.click();
        closeTooltip();
      };
      sourceCopyImg.addEventListener("click", clickImg);
    }

    return tooltip;
  }

  // make carbon copy of html dom element
  function makeCopy(source) {
    const sourceCopy = source.cloneNode(true);

    // delete elements marked with ignore (eg anchor and jump buttons)
    const deleteFromCopy = sourceCopy.querySelectorAll('[data-ignore="true"]');
    for (const element of deleteFromCopy) element.remove();

    // delete certain element attributes
    const attributes = [
      "id",
      "data-collapsed",
      "data-selected",
      "data-highlighted",
      "data-glow",
      "class",
    ];
    for (const attribute of attributes) {
      sourceCopy.removeAttribute(attribute);
      const elements = sourceCopy.querySelectorAll("[" + attribute + "]");
      for (const element of elements) element.removeAttribute(attribute);
    }

    return sourceCopy;
  }

  // when tooltip is clicked
  function onTooltipClick(event) {
    // when user clicks on tooltip, stop click from transferring
    // outside of tooltip (eg, click off to close tooltip, or eg click
    // off to unhighlight same refs)
    event.stopPropagation();
  }

  // when tooltip is unhovered
  function onTooltipUnhover(event) {
    if (clickClose === "true") return;

    // make sure new mouse/touch/focus no longer over tooltip or any
    // element within it
    const tooltip = document.getElementById("tooltip");
    if (!tooltip) return;
    if (this.contains(event.relatedTarget)) return;

    closeTooltip();
  }

  // make nav bar to go betwen prev/next instances of same reference
  function makeNavBar(link) {
    // find other links to the same source
    const sameLinks = getSameLinks(link);

    // don't show nav bar when singular reference
    if (sameLinks.total <= 1) return;

    // find prev/next links with same target
    const prevLink = getPrevLink(link, sameLinks);
    const nextLink = getNextLink(link, sameLinks);

    // create nav bar
    const navBar = document.createElement("div");
    navBar.id = "tooltip_nav_bar";
    const text = sameLinks.index + 1 + " of " + sameLinks.total;

    // create nav bar prev/next buttons
    const prevButton = document.createElement("button");
    const nextButton = document.createElement("button");
    prevButton.id = "tooltip_prev_button";
    nextButton.id = "tooltip_next_button";
    prevButton.title =
      "Jump to the previous occurence of this item in the document [←]";
    nextButton.title =
      "Jump to the next occurence of this item in the document [→]";
    prevButton.classList.add("icon_button");
    nextButton.classList.add("icon_button");
    prevButton.innerHTML = document.querySelector(".icon_caret_left").innerHTML;
    nextButton.innerHTML =
      document.querySelector(".icon_caret_right").innerHTML;
    navBar.appendChild(prevButton);
    navBar.appendChild(document.createTextNode(text));
    navBar.appendChild(nextButton);

    // attach click listeners to buttons
    prevButton.addEventListener("click", function () {
      onPrevNextClick(link, prevLink);
    });
    nextButton.addEventListener("click", function () {
      onPrevNextClick(link, nextLink);
    });

    return navBar;
  }

  // get previous link with same target
  function getPrevLink(link, sameLinks) {
    if (!sameLinks) sameLinks = getSameLinks(link);
    // wrap index to other side if < 1
    let index;
    if (sameLinks.index - 1 >= 0) index = sameLinks.index - 1;
    else index = sameLinks.total - 1;
    return sameLinks.elements[index];
  }

  // get next link with same target
  function getNextLink(link, sameLinks) {
    if (!sameLinks) sameLinks = getSameLinks(link);
    // wrap index to other side if > total
    let index;
    if (sameLinks.index + 1 <= sameLinks.total - 1) index = sameLinks.index + 1;
    else index = 0;
    return sameLinks.elements[index];
  }

  // get element that is target of link or url hash
  function getSource(link) {
    const hash = link ? link.hash : window.location.hash;
    const id = hash.slice(1);
    let target = document.querySelector('[id="' + id + '"]');
    if (!target) return;

    // if ref or figure, modify target to get expected element
    if (id.indexOf("ref-") === 0) target = target.querySelector(":nth-child(2)");
    else if (id.indexOf("fig:") === 0) target = target.querySelector("figure");

    return target;
  }

  // when prev/next arrow button is clicked
  function onPrevNextClick(link, prevNextLink) {
    if (link && prevNextLink)
      goToElement(prevNextLink, window.innerHeight * 0.5);
  }

  // scroll to and focus element
  function goToElement(element, offset) {
    // expand accordion section if collapsed
    expandElement(element);
    const y =
      getRectInView(element).top -
      getRectInView(document.documentElement).top -
      (offset || 0);
    // trigger any function listening for "onscroll" event
    window.dispatchEvent(new Event("scroll"));
    window.scrollTo(0, y);
    document.activeElement.blur();
    element.focus();
  }

  // determine position to place tooltip based on link position in
  // viewport and tooltip size
  function positionTooltip(link, left, top) {
    const tooltipElement = document.getElementById("tooltip");
    if (!tooltipElement) return;

    // get convenient vars for position/dimensions of
    // link/tooltip/page/view
    link = getRectInPage(link);
    const tooltip = getRectInPage(tooltipElement);
    const view = getRectInPage();

    // horizontal positioning
    if (left)
      // use explicit value
      left = left;
    else if (link.left + tooltip.width < view.right)
      // fit tooltip to right of link
      left = link.left;
    else if (link.right - tooltip.width > view.left)
      // fit tooltip to left of link
      left = link.right - tooltip.width;
    // center tooltip in view
    else left = (view.right - view.left) / 2 - tooltip.width / 2;

    // vertical positioning
    if (top)
      // use explicit value
      top = top;
    else if (link.top - tooltip.height > view.top)
      // fit tooltip above link
      top = link.top - tooltip.height;
    else if (link.bottom + tooltip.height < view.bottom)
      // fit tooltip below link
      top = link.bottom;
    else {
      // center tooltip in view
      top = view.top + view.height / 2 - tooltip.height / 2;
      // nudge off of link to left/right if possible
      if (link.right + tooltip.width < view.right) left = link.right;
      else if (link.left - tooltip.width > view.left)
        left = link.left - tooltip.width;
    }

    tooltipElement.style.left = left + "px";
    tooltipElement.style.top = top + "px";
  }

  // start script when document is finished loading
  window.addEventListener("load", start);
</script>

<!-- caret left icon -->

<template class="icon_caret_left">
  <!-- modified from: https://fontawesome.com/icons/caret-left -->
  <svg width="16" height="16" viewBox="0 0 192 512">
    <path
      fill="currentColor"
      d="M192 127.338v257.324c0 17.818-21.543 26.741-34.142 14.142L29.196 270.142c-7.81-7.81-7.81-20.474 0-28.284l128.662-128.662c12.599-12.6 34.142-3.676 34.142 14.142z"
    ></path>
  </svg>
</template>

<!-- caret right icon -->

<template class="icon_caret_right">
  <!-- modified from: https://fontawesome.com/icons/caret-right -->
  <svg width="16" height="16" viewBox="0 0 192 512">
    <path
      fill="currentColor"
      d="M0 384.662V127.338c0-17.818 21.543-26.741 34.142-14.142l128.662 128.662c7.81 7.81 7.81 20.474 0 28.284L34.142 398.804C21.543 411.404 0 402.48 0 384.662z"
    ></path>
  </svg>
</template>

<style>
  @media only screen {
    /* tooltip container */
    #tooltip {
      position: absolute;
      width: 50%;
      min-width: 240px;
      max-width: 75%;
      z-index: 1;
    }

    /* tooltip content */
    #tooltip_content {
      margin-bottom: 5px;
      padding: 20px;
      border-radius: 5px;
      border: solid 1px #bdbdbd;
      box-shadow: 0 0 20px rgba(0, 0, 0, 0.05);
      background: #ffffff;
      overflow-wrap: break-word;
    }

    /* tooltip copy of paragraphs and figures */
    #tooltip_content > p,
    #tooltip_content > figure {
      margin: 0;
      max-height: 320px;
      overflow-y: auto;
    }

    /* tooltip copy of <img> */
    #tooltip_content > figure > img,
    #tooltip_content > figure > svg {
      max-height: 260px;
    }

    /* navigation bar */
    #tooltip_nav_bar {
      margin-top: 10px;
      text-align: center;
    }

    /* navigation bar previous/next buton */
    #tooltip_nav_bar > .icon_button {
      position: relative;
      top: 3px;
    }

    /* navigation bar previous button */
    #tooltip_nav_bar > .icon_button:first-of-type {
      margin-right: 5px;
    }

    /* navigation bar next button */
    #tooltip_nav_bar > .icon_button:last-of-type {
      margin-left: 5px;
    }
  }

  /* always hide tooltip on print */
  @media only print {
    #tooltip {
      display: none;
    }
  }
</style>
<!--
  Analytics Plugin (third-party) 
  
  Copy and paste code from Google Analytics or similar service here.
-->
<!-- 
  Annotations Plugin

  Allows public annotation of the  manuscript. See https://web.hypothes.is/.
-->

<script type="module">
  // configuration
  window.hypothesisConfig = function () {
    return {
      branding: {
        accentColor: "#2196f3",
        appBackgroundColor: "#f8f8f8",
        ctaBackgroundColor: "#f8f8f8",
        ctaTextColor: "#000000",
        selectionFontFamily: "Open Sans, Helvetica, sans serif",
        annotationFontFamily: "Open Sans, Helvetica, sans serif",
      },
    };
  };

  // hypothesis client script
  const embed = "https://hypothes.is/embed.js";
  // hypothesis annotation count query url
  const query = "https://api.hypothes.is/api/search?limit=0&url=";

  // start script
  function start() {
    const button = makeButton();
    document.body.insertBefore(button, document.body.firstChild);
    insertCount(button);
  }

  // make button
  function makeButton() {
    // create button
    const button = document.createElement("button");
    button.id = "hypothesis_button";
    button.innerHTML = document.querySelector(".icon_hypothesis").innerHTML;
    button.title = "Hypothesis annotations";
    button.classList.add("icon_button");

    function onClick(event) {
      onButtonClick(event, button);
    }

    // attach click listeners
    button.addEventListener("click", onClick);

    return button;
  }

  // insert annotations count
  async function insertCount(button) {
    // get annotation count from Hypothesis based on url
    let count = "-";
    try {
      const canonical = document.querySelector('link[rel="canonical"]');
      const location = window.location;
      const url = encodeURIComponent((canonical || location).href);
      const response = await fetch(query + url);
      const json = await response.json();
      count = json.total || "-";
    } catch (error) {
      console.log(error);
    }

    // put count into button
    const counter = document.createElement("span");
    counter.id = "hypothesis_count";
    counter.innerHTML = count;
    button.title = "View " + count + " Hypothesis annotations";
    button.append(counter);
  }

  // when button is clicked
  function onButtonClick(event, button) {
    const script = document.createElement("script");
    script.src = embed;
    document.body.append(script);
    button.remove();
  }

  // start script when document is finished loading
  window.addEventListener("load", start);
</script>

<!-- hypothesis icon -->

<template class="icon_hypothesis">
  <!-- modified from: https://simpleicons.org/icons/hypothesis.svg / https://git.io/Jf1VB -->
  <svg width="16" height="16" viewBox="0 0 24 24" tabindex="-1">
    <path
      fill="currentColor"
      d="M3.43 0C2.5 0 1.72 .768 1.72 1.72V18.86C1.72 19.8 2.5 20.57 3.43 20.57H9.38L12 24L14.62 20.57H20.57C21.5 20.57 22.29 19.8 22.29 18.86V1.72C22.29 .77 21.5 0 20.57 0H3.43M5.14 3.43H7.72V9.43S8.58 7.72 10.28 7.72C12 7.72 13.74 8.57 13.74 11.24V17.14H11.16V12C11.16 10.61 10.28 10.07 9.43 10.29C8.57 10.5 7.72 11.41 7.72 13.29V17.14H5.14V3.43M18 13.72C18.95 13.72 19.72 14.5 19.72 15.42A1.71 1.71 0 0 1 18 17.13A1.71 1.71 0 0 1 16.29 15.42C16.29 14.5 17.05 13.71 18 13.71Z"
      tabindex="-1"
    ></path>
  </svg>
</template>

<style>
  /* hypothesis activation button */
  #hypothesis_button {
    box-sizing: border-box;
    position: fixed;
    top: 0;
    right: 0;
    width: 60px;
    height: 60px;
    background: #ffffff;
    border-radius: 0;
    border-left: solid 1px #bdbdbd;
    border-bottom: solid 1px #bdbdbd;
    box-shadow: 0 0 20px rgba(0, 0, 0, 0.05);
    z-index: 2;
  }

  /* hypothesis button svg */
  #hypothesis_button > svg {
    position: relative;
    top: -4px;
  }

  /* hypothesis annotation count */
  #hypothesis_count {
    position: absolute;
    left: 0;
    right: 0;
    bottom: 5px;
  }

  /* side panel */
  .annotator-frame {
    width: 280px !important;
  }

  /* match highlight color to rest of theme */
  .annotator-highlights-always-on .annotator-hl {
    background-color: #ffeb3b !important;
  }

  /* match focused color to rest of theme */
  .annotator-hl.annotator-hl-focused {
    background-color: #ff8a65 !important;
  }

  /* match bucket bar color to rest of theme */
  .annotator-bucket-bar {
    background: #f5f5f5 !important;
  }

  /* always hide button, toolbar, and tooltip on print */
  @media only print {
    #hypothesis_button {
      display: none;
    }

    .annotator-frame {
      display: none !important;
    }

    hypothesis-adder {
      display: none !important;
    }
  }
</style>
<!-- 
  Mathjax Plugin (third-party) 

  Allows the proper rendering of math/equations written in LaTeX.
  See https://www.mathjax.org/.
-->

<script type="text/x-mathjax-config">
  // configuration
  MathJax.Hub.Config({
    "CommonHTML": { linebreaks: { automatic: true } },
    "HTML-CSS": { linebreaks: { automatic: true } },
    "SVG": { linebreaks: { automatic: true } },
    "fast-preview": { disabled: true }
  });
</script>

<script
  src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js?config=TeX-MML-AM_CHTML"
  integrity="sha512-0doc9hKxR3PYwso42RD1p5ySZpzzuDiOwMrdCEh2WdJZCjcmFKc/wEnL+z8fBQrnHoiNWbo+3fiGkOYXBdQp4A=="
  crossorigin="anonymous"
></script>

<style>
  /* mathjax containers */
  .math.display > span:not(.MathJax_Preview) {
    /* turn inline element (no dimensions) into block (allows fixed width and thus scrolling) */
    display: flex !important;
    overflow-x: auto !important;
    overflow-y: hidden !important;
    justify-content: center;
    align-items: center;
    margin: 0 !important;
  }

  /* right click menu */
  .MathJax_Menu {
    border-radius: 5px !important;
    border: solid 1px #bdbdbd !important;
    box-shadow: none !important;
  }

  /* equation auto-number */
  span[id^="eq:"] > span.math.display + span {
    font-weight: 600;
  }

  /* equation */
  span[id^="eq:"] > span.math.display > span {
    /* nudge to make room for equation auto-number and anchor */
    margin-right: 60px !important;
  }
</style>
</body>
</html>
