\emph{A DOI-citable version of this manuscript is available at \url{https://doi.org/10.1101/2023.01.05.522941}}.

This manuscript
(\href{https://greenelab.github.io/connectivity-search-manuscript/v/79b83d77cd74c8914bba55bf1beeb99ace4c9d2a/}{permalink})
was automatically generated
from \href{https://github.com/greenelab/connectivity-search-manuscript/tree/79b83d77cd74c8914bba55bf1beeb99ace4c9d2a}{greenelab/connectivity-search-manuscript@79b83d7}
on June 1, 2023.

\hypertarget{authors}{%
\subsection{Authors}\label{authors}}

\begin{itemize}
\item
  \textbf{Daniel S. Himmelstein}
  \includegraphics[width=0.16667in,height=0.16667in]{images/orcid.svg}
  \href{https://orcid.org/0000-0002-3012-7446}{0000-0002-3012-7446}
  · \includegraphics[width=0.16667in,height=0.16667in]{images/github.svg}
  \href{https://github.com/dhimmel}{dhimmel}
  · \includegraphics[width=0.16667in,height=0.16667in]{images/twitter.svg}
  \href{https://twitter.com/dhimmel}{dhimmel}
  Department of Systems Pharmacology and Translational Therapeutics, University of Pennsylvania, Philadelphia, Pennsylvania, United States of America; Related Sciences
  · Funded by The Gordon and Betty Moore Foundation (GBMF4552); Pfizer Worldwide Research, Development, and Medical
\item
  \textbf{Michael Zietz}
  \includegraphics[width=0.16667in,height=0.16667in]{images/orcid.svg}
  \href{https://orcid.org/0000-0003-0539-630X}{0000-0003-0539-630X}
  · \includegraphics[width=0.16667in,height=0.16667in]{images/github.svg}
  \href{https://github.com/zietzm}{zietzm}
  · \includegraphics[width=0.16667in,height=0.16667in]{images/twitter.svg}
  \href{https://twitter.com/ZietzMichael}{ZietzMichael}
  Department of Systems Pharmacology and Translational Therapeutics, University of Pennsylvania, Philadelphia, Pennsylvania, United States of America; Department of Biomedical Informatics, Columbia University, New York, New York, United States of America
\item
  \textbf{Vincent Rubinetti}
  \includegraphics[width=0.16667in,height=0.16667in]{images/orcid.svg}
  \href{https://orcid.org/0000-0002-4655-3773}{0000-0002-4655-3773}
  · \includegraphics[width=0.16667in,height=0.16667in]{images/github.svg}
  \href{https://github.com/vincerubinetti}{vincerubinetti}
  Department of Systems Pharmacology and Translational Therapeutics, University of Pennsylvania, Philadelphia, Pennsylvania, United States of America; Center for Health AI, University of Colorado School of Medicine, Aurora, Colorado, United States of America
\item
  \textbf{Kyle Kloster}
  \includegraphics[width=0.16667in,height=0.16667in]{images/orcid.svg}
  \href{https://orcid.org/0000-0001-5678-7197}{0000-0001-5678-7197}
  · \includegraphics[width=0.16667in,height=0.16667in]{images/github.svg}
  \href{https://github.com/kkloste}{kkloste}
  · \includegraphics[width=0.16667in,height=0.16667in]{images/twitter.svg}
  \href{https://twitter.com/kylekloster}{kylekloster}
  Carbon, Inc.; Department of Computer Science, North Carolina State University, Raleigh, North Carolina, United States of America
  · Funded by The Gordon and Betty Moore Foundation (GBMF4560)
\item
  \textbf{Benjamin J. Heil}
  \includegraphics[width=0.16667in,height=0.16667in]{images/orcid.svg}
  \href{https://orcid.org/0000-0002-2811-1031}{0000-0002-2811-1031}
  · \includegraphics[width=0.16667in,height=0.16667in]{images/github.svg}
  \href{https://github.com/ben-heil}{ben-heil}
  · \includegraphics[width=0.16667in,height=0.16667in]{images/twitter.svg}
  \href{https://twitter.com/autobencoder}{autobencoder}
  Genomics and Computational Biology Graduate Group, Perelman School of Medicine, University of Pennsylvania
  · Funded by The Gordon and Betty Moore Foundation (GBMF4552)
\item
  \textbf{Faisal Alquaddoomi}
  \includegraphics[width=0.16667in,height=0.16667in]{images/orcid.svg}
  \href{https://orcid.org/0000-0003-4297-8747}{0000-0003-4297-8747}
  · \includegraphics[width=0.16667in,height=0.16667in]{images/github.svg}
  \href{https://github.com/falquaddoomi}{falquaddoomi}
  Department of Biochemistry and Molecular Genetics, University of Colorado School of Medicine, Aurora, Colorado, United States of America; Center for Health AI, University of Colorado School of Medicine, Aurora, Colorado, United States of America
\item
  \textbf{Dongbo Hu}
  \includegraphics[width=0.16667in,height=0.16667in]{images/orcid.svg}
  \href{https://orcid.org/0000-0001-7870-8242}{0000-0001-7870-8242}
  · \includegraphics[width=0.16667in,height=0.16667in]{images/github.svg}
  \href{https://github.com/dongbohu}{dongbohu}
  Department of Pathology, Perelman School of Medicine University of Pennsylvania, Philadelphia PA, USA
  · Funded by The Gordon and Betty Moore Foundation (GBMF4552)
\item
  \textbf{David N. Nicholson}
  \includegraphics[width=0.16667in,height=0.16667in]{images/orcid.svg}
  \href{https://orcid.org/0000-0003-0002-5761}{0000-0003-0002-5761}
  · \includegraphics[width=0.16667in,height=0.16667in]{images/github.svg}
  \href{https://github.com/danich1}{danich1}
  Department of Systems Pharmacology and Translational Therapeutics, Perelman School of Medicine University of Pennsylvania, Philadelphia PA, USA
  · Funded by The Gordon and Betty Moore Foundation (GBMF4552); The National Institutes of Health (T32 HG000046)
\item
  \textbf{Yun Hao}
  \includegraphics[width=0.16667in,height=0.16667in]{images/orcid.svg}
  \href{https://orcid.org/0000-0002-1684-0085}{0000-0002-1684-0085}
  · \includegraphics[width=0.16667in,height=0.16667in]{images/github.svg}
  \href{https://github.com/yhao-compbio}{yhao-compbio}
  · \includegraphics[width=0.16667in,height=0.16667in]{images/twitter.svg}
  \href{https://twitter.com/YhaoC}{YhaoC}
  Genomics and Computational Biology Graduate Group, Perelman School of Medicine, University of Pennsylvania, Philadelphia PA, USA
\item
  \textbf{Blair D. Sullivan}
  \includegraphics[width=0.16667in,height=0.16667in]{images/orcid.svg}
  \href{https://orcid.org/0000-0001-7720-6208}{0000-0001-7720-6208}
  · \includegraphics[width=0.16667in,height=0.16667in]{images/github.svg}
  \href{https://github.com/bdsullivan}{bdsullivan}
  · \includegraphics[width=0.16667in,height=0.16667in]{images/twitter.svg}
  \href{https://twitter.com/BlairDSullivan}{BlairDSullivan}
  School of Computing, University of Utah, Salt Lake City, Utah, USA
\item
  \textbf{Michael W. Nagle}
  \includegraphics[width=0.16667in,height=0.16667in]{images/orcid.svg}
  \href{https://orcid.org/0000-0002-4677-7582}{0000-0002-4677-7582}
  · \includegraphics[width=0.16667in,height=0.16667in]{images/github.svg}
  \href{https://github.com/naglem}{naglem}
  · \includegraphics[width=0.16667in,height=0.16667in]{images/twitter.svg}
  \href{https://twitter.com/mikenagle84}{mikenagle84}
  Integrative Biology, Internal Medicine Research Unit, Worldwide Research, Development, and Medicine, Pfizer Inc, Cambridge, Massachusetts, United States of America; Neurogenomics, Translational Sciences, Neurology Business Group, Eisai Inc, Cambridge, Massachusetts, United States of America
\item
  \textbf{Casey S. Greene}
  \textsuperscript{\protect\hyperlink{correspondence}{✉}}
  \includegraphics[width=0.16667in,height=0.16667in]{images/orcid.svg}
  \href{https://orcid.org/0000-0001-8713-9213}{0000-0001-8713-9213}
  · \includegraphics[width=0.16667in,height=0.16667in]{images/github.svg}
  \href{https://github.com/cgreene}{cgreene}
  · \includegraphics[width=0.16667in,height=0.16667in]{images/twitter.svg}
  \href{https://twitter.com/GreeneScientist}{GreeneScientist}
  Department of Systems Pharmacology and Translational Therapeutics, University of Pennsylvania, Philadelphia, Pennsylvania, United States of America; Department of Biochemistry and Molecular Genetics, University of Colorado School of Medicine, Aurora, Colorado, United States of America; Center for Health AI, University of Colorado School of Medicine, Aurora, Colorado, United States of America
  · Funded by the National Human Genome Research Institute (R01 HG010067); the National Cancer Institute (R01 CA237170); the Gordon and Betty Moore Foundation (GBMF 4552); Pfizer Worldwide Research, Development, and Medical; the Eunice Kennedy Shriver National Institute of Child Health and Human Development (R01 HD109765)
\end{itemize}

\leavevmode\vadjust pre{\hypertarget{correspondence}{}}%
✉ --- Correspondence possible via \href{https://github.com/greenelab/connectivity-search-manuscript/issues}{GitHub Issues}
or email to
Casey S. Greene \textless casey.s.greene@cuanschutz.edu\textgreater.

\hypertarget{abstract}{%
\subsection{Abstract}\label{abstract}}

\textbf{Background}
Hetnets, short for ``heterogeneous networks'', contain multiple node and relationship types and offer a way to encode biomedical knowledge.
One such example, Hetionet connects 11 types of nodes
--- including genes, diseases, drugs, pathways, and anatomical structures
--- with over 2 million edges of 24 types.
Previous work has demonstrated that supervised machine learning methods applied to such networks can identify drug repurposing opportunities.
However, a training set of known relationships does not exist for many types of node pairs,
even when it would be useful to examine how nodes of those types are meaningfully connected.
For example, users may be curious not only how metformin is related to breast cancer,
but also how a given gene might be involved in insomnia.

\textbf{Findings}
We developed a new procedure, termed hetnet connectivity search, that proposes important paths between any two nodes without requiring a supervised gold standard.
The algorithm behind connectivity search identifies types of paths that occur more frequently than would be expected by chance (based on node degree alone).
Several optimizations were required to precompute significant instances of node connectivity at the scale of large knowledge graphs.

\textbf{Conclusion}
We implemented the method on Hetionet and provide an online interface at \url{https://het.io/search}.
We provide an open source implementation of these methods in our new Python package named \href{https://github.com/hetio/hetmatpy}{hetmatpy}.

\hypertarget{introduction}{%
\subsection{Introduction}\label{introduction}}

A~\emph{network}~(also known as a~\href{https://en.wikipedia.org/wiki/Graph_(discrete_mathematics)}{graph}) is a conceptual representation of a group of entities --- called~\emph{nodes}~--- and the relationships between them --- called~\emph{edges}.
Typically, a network has only one type of node and one type of edge.
However, in many cases, it is necessary to be able to distinguish between different types of entities and relationships.

A~\emph{hetnet}~(short for~\textbf{het}erogeneous information~\textbf{net}work {[}\protect\hyperlink{ref-QWTAHxFj}{1}{]}) is a network where nodes and edges have type.
The ability to differentiate between different types of entities and relationships allows a hetnet to describe more complex data accurately.
Hetnets are particularly useful in biomedicine,
where it is important to capture the conceptual distinctions between various entities,
such as genes and diseases, and linkages, such as upregulation and binding.

The types of nodes and edges in a hetnet are defined by a schema, referred to as a metagraph.
The metagraph consists of metanodes (types of nodes) and metaedges (types of edges).
Note that the prefix~\emph{meta} refers to the type (e.g.~compound),
as opposed to a specific node/edge/path itself (e.g.~acetaminophen).

One such network is Hetionet, which provides a foundation for building hetnet applications.
It unifies data from several different, disparate sources into a single, comprehensive, accessible, common-format network.
The database is publicly accessible without login at \url{https://neo4j.het.io}.
The Neo4j graph database enables querying Hetionet using the Cypher language,
which was designed to interact with networks where nodes and edges have both types and properties.

The initial application of Hetionet, named Project Rephetio, focused on drug repurposing {[}\protect\hyperlink{ref-O21tn8vf}{2}{]}.
The authors \href{https://het.io/repurpose/}{predicted} the probability of drug efficacy for 209,168 compound--disease pairs.
A supervised machine learning approach identified types of paths that occur more or less frequently between known treatments than non-treatments (Figure \ref{fig:rephetio}B).
To train the model, the authors created \href{https://doi.org/10.6084/m9.figshare.3103054}{PharmacotherapyDB},
a physician-curated catalog of 755 disease-modifying treatments {[}\protect\hyperlink{ref-5t7pfaPv}{3}{]}.

\begin{figure}
\hypertarget{fig:rephetio}{%
\centering
\includegraphics[width=1\textwidth,height=\textheight]{https://github.com/greenelab/connectivity-search-manuscript/raw/f98c3470a8bf8f40f5f6aed2794c6ea66b93b14b/content/media/rephetio/metagraph-and-features.png}
\caption{\textbf{A. Hetionet v1.0 metagraph.}
The types of nodes and edges in Hetionet.\\
\textbf{B. Supervised machine learning approach from Project Rephetio.}
This figure visualizes the feature matrix used by Project Rephetio to make supervised predictions.
Each row represents a compound--disease pair.
The bottom half of rows correspond to known treatments (i.e.~positives),
while the top half correspond to non-treatments
(i.e.~negatives under a \emph{closed-world assumption}, not known to be treatments in PharmacotherapyDB).
Here, an equal number of treatments and non-treatments are shown, but in reality the problem is heavily imbalanced.
Project Rephetio scaled models to assume a positive prevalence of 0.36\% {[}\protect\hyperlink{ref-O21tn8vf}{2},\protect\hyperlink{ref-vvdoVrc6}{4}{]}.
Each column represents a metapath, labeled with its abbreviation.\\
Feature values are degree-weighted path counts (abbreviated DWPCs, transformed and standardized),
which assess the connectivity along the specified metapath between the specific compound and disease.
Green colored values indicate above-average connectivity,
whereas blue values indicate below average connectivity.
In general, positives have greater connectivity for the selected metapaths than negatives.
Rephetio used a logistic regression model to learn the effect of each type of connectivity (feature) on the likelihood that a compound treats a disease.
The model predicts whether a compound--disease pair is a treatment based on its features,
but requires supervision in the form of known treatments.
}\label{fig:rephetio}
}
\end{figure}

Project Rephetio successfully predicted treatments,
including those under investigation by clinical trail.
However, two challenges limit the applicability of Rephetio.
First, Rephetio required known labels (i.e.~treatment status) to train a model.
Hence, the approach cannot be applied to domains where training labels do not exist.
Second, the DWPC metric used to assess connectivity is sensitive to node degree.
The Rephetio approach was incapable of detecting whether a high DWPC score indicated meaningful connectivity above the level expected by the background network degrees.
Here we develop Hetnet connectivity search, which defines a null distribution for DWPCs that accounts for degree and enables detecting meaningful hetnet connectivity without training labels.

Existing research into methods for determining whether two nodes are related primarily focuses on homogeneous networks (without type).
Early approaches detected related nodes by measuring neighborhood overlap or path similarity between two nodes {[}\protect\hyperlink{ref-EfWvuSjX}{5},\protect\hyperlink{ref-7iZtCGcX}{6}{]}.
These approaches predicted node relatedness with success.
However, they are challenging to scale as a network grows in size or semantic richness (i.e.~type) {[}\protect\hyperlink{ref-EfWvuSjX}{5}{]}.

More recently, focus has shifted to graph embeddings to determine if two nodes are related,
specifically in the context of knowledge graphs, which are often semantically rich and include type
{[}\protect\hyperlink{ref-otY29wFV}{7},\protect\hyperlink{ref-NnOS86ev}{8},\protect\hyperlink{ref-GYLacVTV}{9},\protect\hyperlink{ref-3gzmoW2V}{10},\protect\hyperlink{ref-zHcAMCig}{11}{]}.
These types of methods involve mapping nodes and sometimes edges to dense vectors via
neural network models {[}\protect\hyperlink{ref-U1Dfux5r}{12},\protect\hyperlink{ref-1AeZs6xaT}{13},\protect\hyperlink{ref-jyhFBVzQ}{14}{]},
matrix factorization {[}\protect\hyperlink{ref-xq3Nsf5M}{15},\protect\hyperlink{ref-1EeE9qtFA}{16}{]},
or translational distance models {[}\protect\hyperlink{ref-cnCLKeU9}{17}{]}.
Bioteque creates node embeddings from the bipartite network of DWPCs for a given metapath {[}\protect\hyperlink{ref-vhsYX2y}{18}{]}.
Once these dense vectors have been produced, quantitative scores that measure node relatedness can be generated via a machine learning model {[}\protect\hyperlink{ref-NnOS86ev}{8},\protect\hyperlink{ref-taI1UUAE}{19},\protect\hyperlink{ref-17ioZ6znu}{20}{]} or by selected similarity metrics {[}\protect\hyperlink{ref-otY29wFV}{7},\protect\hyperlink{ref-GYLacVTV}{9},\protect\hyperlink{ref-1D9FTzRBg}{21},\protect\hyperlink{ref-6PISrkV5}{22},\protect\hyperlink{ref-19E33rJiu}{23}{]}.
These approaches have been quite successful in determining node relatedness.
Yet, they only state \emph{whether} two nodes are related and fail to explain \emph{why} two nodes are related.

Explaining why two nodes are related is a non-trivial task because approaches must output more than a simple similarity score.
The first group of approaches output a list of ranked paths that are most relevant between two nodes {[}\protect\hyperlink{ref-Dbie18xe}{24},\protect\hyperlink{ref-GPBbIfq9}{25},\protect\hyperlink{ref-L3DO8oq1}{26}{]}.
For example, the FAIRY framework explains for why items appear on a user's social media feed based on a network of users and content classes (e.g.~categories, user posts, songs) {[}\protect\hyperlink{ref-GPBbIfq9}{25}{]}.
ESPRESSO explains how two sets of nodes are related by returning subgraphs {[}\protect\hyperlink{ref-Hf24Hzrl}{27}{]}.
Other approaches such as MetaExp return important metapaths rather than paths,
but require some form of supervision {[}\protect\hyperlink{ref-19H7j1WBd}{28},\protect\hyperlink{ref-Vz9QIDKV}{29}{]}.

MechRepoNet is a hetnet containing 250,035 nodes across 9 metanodes and 9,652,116 edges across 68 metaedges {[}\protect\hyperlink{ref-KQNSDt1L}{30}{]}.
The study trained a model using DWPCs as features to predict \emph{Compound--treats--Disease} relationships,
which was able to select 89 metapaths with positive regression coefficients.
The authors also created \href{https://sulab.github.io/DrugMechDB/}{DrugMechDB} with a curated set of paths capturing known mechanisms of action for 123 compound--disease pairs {[}\protect\hyperlink{ref-KQNSDt1L}{30}{]}.
Metapath coefficients were used to rank paths, using DrugMechDB as validation.
The method generally performed well, although interpretability was challenging when ``hundreds, or thousands of paths ranked above the mechanistic path in DrugMechDB'' {[}\protect\hyperlink{ref-KQNSDt1L}{30}{]}.
To address this issue, the study explored additional path filters, such as filtering for paths that traverse known drug targets, and dimensionality reduction by aggregating paths across intermediate nodes and summing the path weights.
Refinements to path scoring techniques might also be helpful solutions in this context.

Hetnet connectivity search explains how two nodes are related in an unsupervised manner that captures the semantic richness of edge type and returns results in the form of both metapaths and paths.
Our open source implementation, including for a query and visualization webserver, was designed with scalability and responsiveness in mind allowing in-browser exploration.

\hypertarget{findings}{%
\subsection{Findings}\label{findings}}

Completing hetnet connectivity search involved advances on three fronts.
We implemented new software for efficient matrix-based operations on hetnets.
We developed strategies to efficiently calculate the desired connectivity score under the null.
We designed and developed a web interface for easy access to the connectivity search approach.

\hypertarget{hetmatpy-package}{%
\subsubsection{Hetmatpy Package}\label{hetmatpy-package}}

We created the hetmatpy Python package,
available on \href{https://github.com/hetio/hetmatpy}{GitHub} and \href{https://pypi.org/project/hetmatpy/}{PyPI} under the permissive BSD-2-Clause Plus Patent License.
This package provides matrix-based utilities for hetnets.
Each metaedge is represented by a distinct adjacency matrix,
which can be either a dense Numpy array or sparse SciPy matrix (see \protect\hyperlink{hetmat-architecture}{HetMat architecture}).
Adjacency matrices are stored on disk and loaded in a lazy manner to help scale the software to hetnets that are too large to fit entirely in memory.

The primary focus of the package is to provide compute-optimized and memory-efficient implementations of path counting algorithms.
Specifically, the package supports computing \emph{degree-weighted} path counts (DWPCs),
which can be done efficiently using matrix multiplication
but require complex adjustments to avoid counting paths with duplicate nodes
(i.e.~to filter walks that are not paths, see \protect\hyperlink{dwpc-matrix-multiplication-algorithms}{DWPC matrix multiplication algorithms}).
The package can reuse existing path count computations that span segments of a longer metapath.
The package also supports generating null distributions for DWPCs derived from permuted networks,
see \protect\hyperlink{degree-grouping-of-node-pairs}{Degree-grouping of node pairs}.
Since this approach generates too many permuted DWPC values to store on disk,
our implementation retains summary statistics for each degree-group that allow computation of a \protect\hyperlink{gamma-hurdle-distribution}{Gamma-hurdle distribution} from which null DWPC \emph{p}-values can be generated.

\hypertarget{dwpc-null-distribution}{%
\subsubsection{DWPC null distribution}\label{dwpc-null-distribution}}

To assess connectivity between a source and target node,
we use the DWPC (degree-weighted path count) metric.
The DWPC is similar to path count (number of paths between the source and target node along a given metapath),
except that it downweights paths through high degree nodes.
Rather than using the raw DWPC for a source-metapath-target combination,
we transform the DWPC across all source-target node pairs for a metapath to yield a distribution that is more compact and amenable to modeling {[}\protect\hyperlink{ref-17mGFvxXu}{31}{]}.

Previously, we had no technique for detecting whether a DWPC value was exceptional.
One possibility is to evaluate the DWPCs for all pairs of nodes and select the top scores (e.g.~the top 5\% of DWPCs).
Another possibility is to pick a transformed DWPC score as a cutoff.
The shortcomings of these methods are twofold.
First, neither the percentile nor absolute value of a DWPC has inherent meaning.
To select transformed DWPCs greater than 3.5, or alternatively the top 1\% of DWPCs, is arbitrary.
Second, comparing DWPCs between node pairs fails to account for the situation where high-degree node pairs are likely to score higher,
solely on due to their degree (\ref{fig:degree-group-metrics}).

To address these shortcomings,
we developed a method to compute the right-tail \emph{p}-value of a DWPC.
\emph{p}-values have a broadly understood interpretation
--- in our case, the probability that a DWPC equal to or greater than the observed DWPC could occur under a null model.
Our null model is based on DWPCs generated from permuted networks,
where edges have been randomized in a degree-preserving manner (see \protect\hyperlink{permuted-hetnets}{Permuted hetnets}).

By tailoring the null distribution for a DWPC to the degree of its source and target node (see \protect\hyperlink{degree-grouping-of-node-pairs}{Degree-grouping of node pairs}),
we account for degree effects when determining the significance of a DWPC.
To improve the accuracy of DWPC \emph{p}-values,
we use fit a \protect\hyperlink{gamma-hurdle-distribution}{gamma-hurdle distribution} to the null DWPCs.
In rare cases, there are insufficient nonzero null DWPCs to fit the gamma portion of the null distribution.
In these cases, we fallback to an empirical calculation as described in \protect\hyperlink{empirical-dwpc-p-values}{Empirical DWPC p-values}.

\hypertarget{enriched-metapaths}{%
\subsubsection{Enriched metapaths}\label{enriched-metapaths}}

For each of the 2,205 metapaths in Hetionet v1.0 with length ≤ 3,
we computed DWPCs for all node pairs and their corresponding null distributions, see \protect\hyperlink{dwpc-and-null-distribution-computation}{DWPC and null distribution computation}.
We store the most significant DWPCs as described in \protect\hyperlink{prioritizing-enriched-metapaths-for-database-storage}{Prioritizing enriched metapaths for database storage},
which appear as the ``precomputed'' rows in the webapp metapath table (Figure \ref{fig:webapp}B \& \ref{fig:webapp-metapaths}).
DWPCs that are not retained by the database can be regenerated on the fly.
This design allows us to immediately provide users with the metapaths that are most enriched between two query nodes,
while still allowing on-demand access to the full metrics for all metapaths with length ≤ 3.

\begin{figure}
\hypertarget{fig:webapp-metapaths}{%
\centering
\includegraphics{https://github.com/greenelab/connectivity-search-manuscript/raw/b290b4ad435553c7126867e0720f3112b4692809/content/media/webapp/v3/b.metapaths-expanded.png}
\caption{\textbf{Expanded metapath details from the connectivity search webapp.}
This is the expanded view of the \href{https://het.io/search/?source=17287\&target=7607\&metapaths=DaGiGpPW\%2CDdGiGpPW\%2CDdGpPW\%2CDlAeGpPW\%2CDrDaGpPW\%2CDrDuGpPW\%2CDuGiGpPW\&complete=}{metapath table} in \ref{fig:webapp}B showing enriched metapaths between Alzheimer's disease and the circadian rhythm pathway.}\label{fig:webapp-metapaths}
}
\end{figure}

Figure \ref{fig:webapp-metapaths} shows the information used to compute \emph{p}-value for enriched metapaths.
The table includes the following columns:

\begin{itemize}
\tightlist
\item
  \textbf{path count}:
  The number of paths between the source and target node of the specified metapath
\item
  \textbf{adjusted \emph{p}-value}:
  A measure of the significance of the DWPC that indicates whether more paths were observed than expected due to random chance.
  Compares the DWPC to a null distribution of DWPCs generated from degree-preserving permuted networks.
  \protect\hyperlink{adjusting-dwpc-p-values}{Bonferroni-adjusted} for the number of metapaths with the same source metanode, target metanode, and length.
\item
  \textbf{\emph{p}-value}:
  A measure of the significance of the DWPC that indicates whether more paths were observed than expected due to random chance.
  Compares the DWPC to a null distribution of DWPCs generated from degree-preserving permuted networks.
  Not adjusted for multiple comparisons
  (i.e.~when multiple metapaths are assessed for significant connectivity between the source and target node).
\item
  \textbf{DWPC}:
  Degree-Weighted Path Count ---
  Measures the extent of connectivity between the source and target node for the given metapath.
  Like the path count, but with less weight given to paths along high-degree nodes.
\item
  \textbf{source degree}:
  The number of edges from the source node that are of the same type as the initial metaedge of the metapath.
\item
  \textbf{target degree}:
  The number of edges from the target node that are of the same type as the final metaedge of the metapath.
\item
  \textbf{\# DWPCs}:
  The number of DWPCs calculated on permuted networks used to generate a null distribution for the DWPC from the real network.
  Permuted DWPCs are aggregated for all permuted node pairs with the same degrees as the source and target node.
\item
  \textbf{\# non-0 DWPCs}:
  The number of permuted DWPCs from `\# of DWPCs' column that were nonzero.
  Nonzero DWPCs indicate at least one path between the source and target node existed in the permuted network.
\item
  \textbf{non-0 mean}:
  The mean of nonzero permuted DWPCs.
  Used to generate the gamma-hurdle model of the null DWPC distribution.
\item
  \textbf{non-0 σ}:
  The standard deviation of nonzero permuted DWPCs.
  Used to generate the gamma-hurdle model of the null DWPC distribution.
\item
  \textbf{Neo4j Actions}:
  A Cypher query that users can run in the \href{https://neo4j.het.io/browser/}{Neo4j browser} to show paths with the largest DWPCs for the metapath.
\end{itemize}

\hypertarget{enriched-paths}{%
\subsubsection{Enriched paths}\label{enriched-paths}}

In addition to knowing which metapaths are enriched between two query nodes,
it is helpful to see the specific paths that contribute highly to such enrichment.
Since the DWPC is a summation of a path metric (called the path degree product),
it is straightforward to calculate the proportion of a DWPC attributable to an individual path.
The webapp allows users to select a metapath to populate a table of the corresponding paths.
These paths are generated on-the-fly through a Cypher query to the Hetionet Neo4j database.

It is desirable to have a consolidated view of paths across multiple metapaths.
Therefore, we calculate a \emph{path score} heuristic,
which can be used to compare the importance of paths between metapaths.
The path score equals the proportion of the DWPC contributed by a path multiplied by the magnitude of the DWPC's \emph{p}-value (-log\textsubscript{10}(\emph{p})).
To illustrate, the paths webapp panel includes the following information (Figure \ref{fig:webapp}C):

\begin{itemize}
\tightlist
\item
  \textbf{path}:
  The sequence of edges in the network connecting the source node to the target node.
  Duplicate nodes are not permitted in paths.
\item
  \textbf{path score}:
  A metric of how meaningful the path is in describing the connectivity between the source and target node.
  The score combines the magnitude of the metapath's p-value with the percent of the DWPC contributed by the path.
\item
  \textbf{\% of DWPC}
  The contribution of the path to the DWPC for its metapath.
  This metric compares the importance of all paths of the same metapath from the source node to the target node.
\end{itemize}

\hypertarget{hetio-website-and-connectivity-search-webapp}{%
\subsubsection{Hetio Website and Connectivity Search Webapp}\label{hetio-website-and-connectivity-search-webapp}}

We revamped the website hosted at \url{https://het.io} to serve as a unified home for this study and the hetnet-related research that preceded it.
The website provides the connectivity search webapp running over the hetio network and several other interactive apps for prior projects.
It also includes high-level information on hetnets and Hetionet,
citation and contact details,
links to supporting studies and software,
downloads and exploration of Hetionet data,
and related media.

We created the connectivity search webapp available at \url{https://het.io/search/}.
The tool is free to use, without any login or authentication.
The app allows users to quickly explore how any two nodes in Hetionet v1.0 might be related.
The workflow accepts one or more nodes as input and shows the user the most important metapaths and paths for a pair of query nodes.

The design guides the user through selecting a source and target node (Figure \ref{fig:webapp}A).
The webapp returns metapaths, scored by whether they occurred more than expected based on network degree (Figure \ref{fig:webapp}B).
Users can proceed by requesting the specific paths for each metapath,
which are placed in a unified table sorted according to their path score (Figure \ref{fig:webapp}C).
Finally, the webapp produces publication-ready visualizations containing user-selected paths (Figure \ref{fig:webapp}D).

\begin{figure}
\hypertarget{fig:webapp}{%
\centering
\includegraphics[width=1\textwidth,height=\textheight]{https://github.com/greenelab/connectivity-search-manuscript/raw/b290b4ad435553c7126867e0720f3112b4692809/content/media/webapp/v3/webapp.png}
\caption{\textbf{Using the connectivity search webapp to explore the pathophysiology of Alzheimer's disease.}
This figure shows an example user workflow for \url{https://het.io/search/}.\\
\textbf{A.}
The user selects two nodes.
Here, the user is interested in Alzheimer's disease, so \href{https://het.io/search/?source=17287}{selects this} as the source node.
The user limits the target node search to metanodes relating to gene function.
The target node search box suggests nodes, sorted by the number of significant metapaths.
When the user types in the target node box, the matches reorder based on search word similarity.
Here, the user becomes interested in how the circadian rhythm might relate to Alzheimer's disease.\\
\textbf{B.}
The webapp returns metapaths between Alzheimer's disease and the circadian rhythm pathway.
The user unchecks ``precomputed only'' to compute results for all metapaths with length ≤ 3, not just those that surpass the database inclusion threshold.
The user sorts by adjusted \emph{p}-value and \href{https://het.io/search/?source=17287\&target=7607\&metapaths=DaGiGpPW\%2CDdGiGpPW\%2CDdGpPW\%2CDlAeGpPW\%2CDrDaGpPW\%2CDrDuGpPW\%2CDuGiGpPW\&complete=}{selects} 7 of the top 10 metapaths.\\
\textbf{C.}
Paths for the selected metapaths are ordered by their path score (limited to 100 paths for each metapath).
The user selects 8 paths (1 from a subsequent page of results) to show in the graph visualization and highlights a single path involving \emph{ARNT2} for emphasis.\\
\textbf{D.}
A subgraph displays the previously selected paths.
The user improves on the automated layout by repositioning nodes.
Clicking an edge displays its properties, informing the user that association between Creutzfeldt-Jakob disease and \emph{NPAS2} was detected by GWAS.
}\label{fig:webapp}
}
\end{figure}

\hypertarget{discussion}{%
\subsection{Discussion}\label{discussion}}

In this study, we introduce a search engine for hetnet connectivity between two nodes that returns results in realtime.
An interactive webapp helps users explore node connectivity by ranking metapaths and paths,
while visualizing multiple paths in a subgraph.

We made several methodological contributions to support this application.
We developed optimized algorithms for computing DWPCs using matrix multiplication.
In addition, we created a method for estimating a \emph{p}-value for a DWPC,
using null DWPCs computed on permuted hetnets.
We implemented these advances in the open-source hetmatpy Python package and HetMat data structure to provide highly-optimized computational infrastructure for representing and reasoning on hetnets using matrices.

This work lays the foundation for exciting future directions.
For many queries, a large number of paths are returned.
Interpretation of large lists is difficult.
Therefore, the dimensionality of results could be reduced by aggregating path scores across intermediate nodes or edges {[}\protect\hyperlink{ref-1693YErZj}{32}{]}.

Here, we computed all DWPCs for Hetionet metapaths with length ≤ 3.
Our search engine will therefore overlook important connectivity from longer metapaths.
However, it is infeasible to compute DWPCs for all longer metapaths.
One solution would be to only extend metapaths detected as informative.
For example, if a \emph{CbGpPWpG} metapath is deemed informative,
it could be extended with additional metaedges like \emph{CbGpPWpGaD}.
One unsupervised approach would be to use the distribution of DWPC \emph{p}-values for a metapath to detect whether the paths still convey sufficient information, for example by requiring an enrichment of small \emph{p}-values.
Were this method to fail, supervised alternatives could be explored,
such as the ability for DWPCs from a longer metapath to predict that of a shorter metapath or metaedge,
with care taken to prevent label leakage.
One final approach could learn from user interest and compute longer metapaths only when requested.

This work focuses on queries where the input is a node pair.
Equally interesting would be queries where the input is a set of nodes of the same type, optionally with weights.
The search would compute DWPCs for paths originating on the query nodes.
The simpler formulation would compute DWPCs for metapaths separately and compare to null distributions from permuted hetnets.
A more advanced formulation would combine scores across metapaths such that every node in the hetnet would receive a single score capturing its connectivity to the query set.
This approach would have similar utility to gene set enrichment analysis in that the user could provide a set of genes as input and receive a ranked list of nodes that characterize the function of the query genes.
However, it would excel in its versatility by returning results of any node type without requiring pre-defined gene sets to match against.
Some users might be interested in node set transformations where scores for one node type are converted to another node type.
This approach could take scores for human genes and convert them to side effects, diseases, pathways, etcetera.

Our work is not without limitations.
The final application relies on multiple databases and cached computations specific to Hetionet v1.0.
Despite striving for a modular architecture,
generating an equivalent search webapp for a different hetnet would require adaptation due to the many data sources involved.
Furthermore, we would benefit from greater real-world evaluation of the connectivity search results to help identify situations where the method underperforms.
Despite these challenges, our study demonstrates one of the first public search engines for node connectivity on a biomedical knowledge graph, while contributing methods and software that we hope will inspire future work.

\hypertarget{methods}{%
\subsection{Methods}\label{methods}}

\hypertarget{hetionet}{%
\subsubsection{Hetionet}\label{hetionet}}

We used the hetionet knowledge graph to demonstrate connectivity search.
\href{https://het.io/about/}{Hetionet} is a knowledge graph of human biology, disease, and medicine,
integrating information from millions of studies and decades of research.
Hetionet v1.0 combines information from~\href{https://git.dhimmel.com/rephetio-manuscript/\#tbl:resources}{29 public databases}.
The network contains 47,031 nodes of~\href{https://git.dhimmel.com/rephetio-manuscript/\#tbl:metanodes}{11 types}~(Table \ref{tbl:metanodes}) and 2,250,197 edges of~\href{https://git.dhimmel.com/rephetio-manuscript/\#tbl:metaedges}{24 types} (Figure \ref{fig:rephetio}A).

\begin{longtable}[]{@{}
  >{\raggedright\arraybackslash}p{(\columnwidth - 6\tabcolsep) * \real{0.2500}}
  >{\raggedright\arraybackslash}p{(\columnwidth - 6\tabcolsep) * \real{0.2500}}
  >{\raggedright\arraybackslash}p{(\columnwidth - 6\tabcolsep) * \real{0.2500}}
  >{\raggedright\arraybackslash}p{(\columnwidth - 6\tabcolsep) * \real{0.2500}}@{}}
\caption{\textbf{Node types in Hetionet}
The abbreviation, number of nodes, and description for each of the 11 metanodes in Hetionet v1.0.
\label{tbl:metanodes}}\label{tbl:metanodes}\tabularnewline
\toprule()
\begin{minipage}[b]{\linewidth}\raggedright
Metanode
\end{minipage} & \begin{minipage}[b]{\linewidth}\raggedright
Abbr
\end{minipage} & \begin{minipage}[b]{\linewidth}\raggedright
Nodes
\end{minipage} & \begin{minipage}[b]{\linewidth}\raggedright
Description
\end{minipage} \\
\midrule()
\endfirsthead
\toprule()
\begin{minipage}[b]{\linewidth}\raggedright
Metanode
\end{minipage} & \begin{minipage}[b]{\linewidth}\raggedright
Abbr
\end{minipage} & \begin{minipage}[b]{\linewidth}\raggedright
Nodes
\end{minipage} & \begin{minipage}[b]{\linewidth}\raggedright
Description
\end{minipage} \\
\midrule()
\endhead
Anatomy & A & 402 & Anatomical structures, excluding structures that are known not to be found in humans. From~\href{http://uberon.github.io/}{Uberon}. \\
Biological Process & BP & 11381 & Larger processes or biological programs accomplished by multiple molecular activities. From~\href{http://geneontology.org/}{Gene Ontology}. \\
Cellular Component & CC & 1391 & The locations relative to cellular structures in which a gene product performs a function. From~\href{http://geneontology.org/}{Gene Ontology}. \\
Compound & C & 1552 & Approved small molecule compounds with documented chemical structures. From~\href{https://www.drugbank.ca/}{DrugBank}. \\
Disease & D & 137 & Complex diseases, selected to be distinct and specific enough to be clinically relevant yet general enough to be well annotated. From~\href{http://disease-ontology.org/}{Disease Ontology}. \\
Gene & G & 20945 & Protein-coding human genes. From~\href{https://www.ncbi.nlm.nih.gov/gene}{Entrez Gene}. \\
Molecular Function & MF & 2884 & Activities that occur at the molecular level, such as ``catalysis'' or ``transport''. From~\href{http://geneontology.org/}{Gene Ontology}. \\
Pathway & PW & 1822 & A series of actions among molecules in a cell that leads to a certain product or change in the cell. From~\href{https://www.wikipathways.org/index.php/WikiPathways}{WikiPathways},~\href{https://reactome.org/}{Reactome}, and Pathway Interaction Database. \\
Pharmacologic Class & PC & 345 & ``Chemical/Ingredient'', ``Mechanism of Action'', and ``Physiologic Effect'' FDA class types. From~\href{http://drugcentral.org/}{DrugCentral}. \\
Side Effect & SE & 5734 & Adverse drug reactions. From~\href{http://sideeffects.embl.de/}{SIDER}/\href{https://www.nlm.nih.gov/research/umls/}{UMLS}. \\
Symptom & S & 438 & Signs and Symptoms (i.e.~clinical abnormalities that can indicate a medical condition). From the~\href{https://www.nlm.nih.gov/mesh/meshhome.html}{MeSH ontology}. \\
\bottomrule()
\end{longtable}

One limitation that restricts the applicability of Hetionet is incompleteness.
In many cases, Hetionet v1.0 includes only a subset of the nodes from a given resource.
For example, the Disease Ontology contains over 9,000 diseases {[}\protect\hyperlink{ref-1FsruosUW}{33}{]},
while Hetionet includes only 137 diseases {[}\protect\hyperlink{ref-203l6hC3}{34}{]}.
Nodes were excluded to avoid redundant or overly specific nodes,
while ensuring a minimum level of connectivity for compounds and diseases.
See the \href{https://git.dhimmel.com/rephetio-manuscript/\#nodes}{Project Rephetio methods} for more details {[}\protect\hyperlink{ref-O21tn8vf}{2}{]}.
Nonetheless, Hetionet v1.0 remains one of the most comprehensive and integrative networks that consolidates biomedical knowledge into a manageable number of node and edge types {[}\protect\hyperlink{ref-yqoUY2f9}{35}{]}.
Other integrative resources, some still under development, include \href{https://www.wikidata.org}{Wikidata} {[}\protect\hyperlink{ref-HUodcdz1}{36}{]}, \href{https://skr3.nlm.nih.gov/SemMedDB/}{SemMedDB} {[}\protect\hyperlink{ref-eKVkBSDD}{37},\protect\hyperlink{ref-reekunFL}{38},\protect\hyperlink{ref-uTOtIsTj}{39}{]}, \href{https://spoke.ucsf.edu/}{SPOKE} {[}\protect\hyperlink{ref-1BHNoKOV1}{40}{]}, and \href{https://github.com/RTXteam/RTX-KG2}{RTX-KG2c} {[}\protect\hyperlink{ref-bniVH7q2}{41}{]}.

\hypertarget{hetmat-architecture}{%
\subsubsection{HetMat architecture}\label{hetmat-architecture}}

At the core of the hetmatpy package is the HetMat data structure for storing and accessing the network.
HetMats are stored on disk as a directory, which by convention uses a \texttt{.hetmat} extension.
A HetMat directory stores a single heterogeneous network, whose data resides in the following files.

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\tightlist
\item
  A \texttt{metagraph.json} file stores the schema, defining which types of nodes and edges comprise the hetnet.
  This format is defined by the \href{https://github.com/hetio/hetnetpy}{hetnetpy} Python package.
  Hetnetpy was originally developed with the name hetio during prior studies
  {[}\protect\hyperlink{ref-O21tn8vf}{2},\protect\hyperlink{ref-WkPlH1ds}{42}{]},
  but we \href{https://github.com/hetio/hetnetpy/issues/40}{renamed} it to het\textbf{net}py for better disambiguation from het\textbf{mat}py.
\item
  A \texttt{nodes} directory containing one file per node type (metanode) that defines each node.
  Currently, \texttt{.tsv} files where each row represents a node are supported.
\item
  An \texttt{edges} directory containing one file per edge type (metadata) that encodes the adjacency matrix.
  The matrix can be serialized using either the Numpy dense format (\texttt{.npy}) or SciPy sparse format (\texttt{.sparse.npz}).
\end{enumerate}

For node and edge files, compression is supported as detected from \texttt{.gz}, \texttt{.bz2}, \texttt{.zip}, and \texttt{.xz} extensions.
This structure of storing a hetnet supports selectively reading nodes and edges into memory.
For example, a certain computation may only require access to a subset of the node and edge types.
By only loading the required node and edge types, we reduce memory usage and read times.

Additional subdirectories, such as \texttt{path-counts} and \texttt{permutations}, store data generated from the HetMat.
By using consistent paths for generated data, we avoid recomputing data that already exists on disk.
A HetMat directory can be zipped for archiving and transfer.
Users can selectively include generated data in archives.
Since the primary application of HetMats is to generate computationally demanding measurements on hetnets, the ability to share HetMats with precomputed data is paramount.

The \href{https://hetio.github.io/hetmatpy/reference/hetmatpy/hetmat/\#hetmat}{\texttt{HetMat}} class implements the above logic.
A \texttt{hetmat\_from\_graph} function creates a HetMat object and directory on disk from the pre-existing \texttt{hetnetpy.hetnet.Graph} format.

We converted Hetionet v1.0 to HetMat format and uploaded the \texttt{hetionet-v1.0.hetmat.zip} archive to the \href{https://github.com/hetio/hetionet/tree/master/hetnet/matrix}{Hetionet data repository}.

\hypertarget{dwpc-matrix-multiplication-algorithms}{%
\subsubsection{DWPC matrix multiplication algorithms}\label{dwpc-matrix-multiplication-algorithms}}

Prior to this study, we used two implementations for computing DWPCs.
The first is a pure Python implementation available in the \href{https://github.com/hetio/hetnetpy/blob/aa16e6a7092c039a6b175a73a35c006e53acee20/hetnetpy/pathtools.py\#L8-L21}{\texttt{hetnetpy.pathtools.DWPC}} function {[}\protect\hyperlink{ref-WkPlH1ds}{42}{]}.
The second uses a Cypher query, prepared by \href{https://github.com/hetio/hetnetpy/blob/aa16e6a7092c039a6b175a73a35c006e53acee20/hetnetpy/neo4j.py\#L363-L393}{\texttt{hetnetpy.neo4j.construct\_dwpc\_query}}, that is executed by the Neo4j database {[}\protect\hyperlink{ref-O21tn8vf}{2},\protect\hyperlink{ref-kNYoxBiQ}{43}{]}.
Both of these implementations require traversing all paths between the source and target node.
Hence, they are computationally cumbersome despite optimizations {[}\protect\hyperlink{ref-xWWaK51z}{44}{]}.

Since our methods only require degree-weighted counts, not fully enumerated paths, adjacency matrix multiplication presents an alternative approach.
Multiplication alone, however, counts walks rather than paths, meaning paths traversing a single node multiple times are counted.
When computing network-based features to quantify the relationship between a source and target node,
we would like to exclude traversing duplicate nodes (i.e.~paths, not trails nor walks) {[}\protect\hyperlink{ref-w6rL3OYz}{45}{]}.
We developed a suite of algorithms to compute true path counts and DWPCs using matrix multiplication that benefits from the speed advantages of only counting paths.

Our implementation begins by categorizing a metapath according to the pattern of its repeated metanodes, allowing DWPC computation using a specialized order of operations.
For example, the metapath \emph{DrDtCrC} is categorized as a set of disjoint repeats, while \emph{DtCtDpC} is categorized as repeats of the form BABA.
Many complex repeat patterns can be represented piecewise as simpler patterns, allowing us to compute DWPC for most metapaths up to length 5 and many of length 6 and beyond without enumerating individual paths.
For example, disjoint groups of repeats like \emph{DrDtCrC} can be computed as the matrix product of DWPC matrices for \emph{DrD} and \emph{CrC}.
Randomly-inserted non-repeated metanodes (e.g.~\emph{G} in \emph{DrDaGaDrD}) require no special treatment and are included in DWPC with matrix multiplication.

After metapath categorization, we segment metapaths according to their repeat pattern, following our order of operations.
By segmenting and computing recursively, we can efficiently evaluate DWPC on highly complex metapaths, using simple patterns as building-blocks for higher-level patterns.
Finally, our specialized DWPC functions are applied to individual segments, the results are combined, and final corrections are made to ensure no repeated nodes are counted.
The recursive, segmented approach we developed also allowed us to implement a caching strategy that improved speed by avoiding duplicate DWPC computations.
In summary, the functionality we developed resulted in more than a 175-fold reduction in compute time, allowing us to compute millions of DWPC values across Hetionet {[}\protect\hyperlink{ref-15nwuvjrA}{46}{]}.

\hypertarget{details-of-matrix-dwpc-implementation}{%
\paragraph{Details of matrix DWPC implementation}\label{details-of-matrix-dwpc-implementation}}

DWPC computation requires us to remove all duplicate nodes from paths.
We used three repeat patterns as the building blocks for DWPC computation: short repeats (AAA), nested repeats (BAAB), and overlapping repeats (BABA).
Let \(\mathrm{D}(\mathit{XwXyZ})\) denote the DWPC matrix for metapath \emph{XwXyZ}.
Under this notation, \(\mathrm{D}(\mathit{XyZ})\) is the degree-weighted (bi)adjacency matrix for metaedge \emph{XyZ}.
Additionally, let \(\mathrm{diag}(A)\) represent a diagonal matrix whose entries are the diagonal elements of \(A\).

For the case of short (\textless{} 4) repeats for a single metanode, \emph{XaXbX} (e.g.~\emph{GiGdG}), we simply subtract the main diagonal.

\[\mathrm{D}(\mathit{XaXbX}) = \mathrm{D}(\mathit{XaX}) \mathrm{D}(\mathit{XbX}) - \mathrm{diag}(\mathrm{D}(\mathit{XaX}) \mathrm{D}(\mathit{XbX}))\]

Nested repeats \emph{XaYbYcX} (e.g.~\emph{CtDrDtC}), are treated recursively, with both inner (YY) and outer (XX) repeats treated as separate short repeats.

\[
\mathrm{D}(\mathit{XaYbYcX}) = \mathrm{D}(\mathit{XaY}) \mathrm{D}(\mathit{YbY}) \mathrm{D}(\mathit{YcX}) - \mathrm{diag}(\mathrm{D}(\mathit{XaY}) (\mathrm{D}(\mathit{YbY}) \mathrm{D}(\mathit{YcX}))
\]

Overlapping repeats \emph{XaYbXcY} (e.g.~\emph{CtDtCtD}) require several corrections (\(\odot\) denotes the Hadamard product).

\begin{align}
\mathrm{D}(\mathit{XaYbXcY}) =\ &\mathrm{D}(\mathit{XaY})\  \mathrm{D}(\mathit{YbX})\  \mathrm{D}(\mathit{XcY}) \\
&- \mathrm{diag}(\mathrm{D}(\mathit{XaY})\  \mathrm{D}(\mathit{YbX}))\  \mathrm{D}(\mathit{XcY}) \\
&- \mathrm{D}(\mathit{XaY})\  \mathrm{diag}(\mathrm{D}(\mathit{YbX})\  \mathrm{D}(\mathit{XcY})) \\
&+ \mathrm{D}(\mathit{XaY})\  \odot \mathrm{D}(\mathit{YbX})^T\  \odot \mathrm{D}(\mathit{XcY})
\end{align}

Most paths of length six---and many even longer paths---can be represented hierarchically using these patterns.
For example, a long metapath pattern of the form CBABACXYZ can be segmented as (C(BABA)C)XYZ using patterns for short and overlapping repeats and can be computed using the tools we developed.
In addition to these matrix routines---which advantageously count rather than enumerate paths---we implemented a general matrix method for any metapath type.
The general method is important for patterns such as long (≥ 4) repeats, or complex repeat patterns (e.g.~of the form ABCABC), but it requires path enumeration and is therefore slower.
As an alternative approach for complex paths, we developed an approximate DWPC method that corrects repeats in disjoint simple patterns but only corrects the first repeat in complex patterns (e.g.~≥ length four repeat).
Mayers et al.~developed an alternative approximation, which subtracts the main diagonal at every occurrence of the first repeated metanode {[}\protect\hyperlink{ref-IJWsN67g}{47}{]}.
Our matrix methods were validated against the existing Python and Cypher implementations in the \texttt{hetnetpy} package that rely on explicit path enumeration.

\hypertarget{permuted-hetnets}{%
\subsubsection{Permuted hetnets}\label{permuted-hetnets}}

In order to generate a null distribution for a DWPC, we rely on DWPCs computed from permuted hetnets.
We derive permuted hetnets from the unpermuted network using the XSwap algorithm {[}\protect\hyperlink{ref-iKOIEzQ9}{48}{]}.
XSwap randomizes edges while preserving node degree.
Therefore, it's ideal for generating null distributions that retain general degree effects,
but destroy the actual meaning of edges.
We adapt XSwap to hetnets by applying it separately to each metaedge {[}\protect\hyperlink{ref-O21tn8vf}{2},\protect\hyperlink{ref-GE6jhWnt}{49},\protect\hyperlink{ref-83EvMSI6}{50}{]}.

Project Rephetio created 5 permuted hetnets {[}\protect\hyperlink{ref-O21tn8vf}{2},\protect\hyperlink{ref-GE6jhWnt}{49}{]},
which were used to generate a null distribution of classifier performance for each metapath-based feature.
Here, we aim to create a null distribution for individual DWPCs, which requires vastly more permuted values to estimate with accuracy.
Therefore, we generated \href{https://github.com/hetio/hetionet/tree/a95ae76581af604e91d744680aee3f888fa18887/hetnet/permuted/matrix}{200 permuted hetnets}.
Permutations 001--005 are those generated by Project Rephetio,
while permutations 006--200 were generated by this study.
For the newly generated permutations, we attempted 10 times the number of swaps as edges for a given metaedge,
which is the default multiplier set by the hetnetpy \texttt{permute\_graph} function.
More recently, we also developed the \texttt{xswap} Python \href{https://github.com/greenelab/xswap}{package}, whose optimized C/C++ implementation will enable future research to generate even larger sets of permuted networks {[}\protect\hyperlink{ref-83EvMSI6}{50}{]}.

\hypertarget{degree-grouping-of-node-pairs}{%
\subsubsection{Degree-grouping of node pairs}\label{degree-grouping-of-node-pairs}}

For each of the 200 permuted networks and each of the 2,205 metapaths, we computed the entire DWPC matrix (i.e.~all source nodes × target nodes).
Therefore, for each actual DWPC value, we computed 200 permuted DWPC values.
Because permutation preserves only node degree, DWPC values among nodes with the same source and target degrees are equivalent to additional permutations.
We greatly increased the effective number of permutations by grouping DWPC values according to node degree, affording us a superior estimation of the DWPC null distribution.

We have applied this \emph{degree-grouping} approach previously when calculating the prior probability of edge existence based on the source and target node degrees {[}\protect\hyperlink{ref-83EvMSI6}{50},\protect\hyperlink{ref-19OyKPQ4M}{51}{]}.
But here, we apply \emph{degree-grouping} to null DWPCs.
The result is that the null distribution for a DWPC is based not only on permuted DWPCs for the corresponding source--metapath--target combination,
but instead on all permuted DWPCs for the source-degree--metapath--target-degree combination.

The ``\# DWPCs'' column in Figure \ref{fig:webapp-metapaths} illustrates how degree-grouping inflates the sample size of null DWPCs.
The \emph{p}-value for the \emph{DaGiGpPW} metapath relies on the minimum number of null DWPCs (200),
since no other disease besides Alzheimer's had 196 \emph{associates} edges (source degree) and no other pathway besides circadian rhythm had 201 \emph{participates} edges (target degree).
However, for other metapaths with over 5,000 null DWPCs, degree-grouping increased the size of the null distribution by a factor of 25.
In general, source--target node pairs with lower degrees receive the largest sample size multiplier from degree-grouping.
This is convenient since low-degree nodes also tend to produce the highest proportion of zero DWPCs, by virtue of low connectivity.
Consequently, degree grouping excels where it is most needed.

One final benefit of degree-grouping is that it reduces the disk space required to store null DWPC summary statistics.
For example, with 20,945 genes in Hetionet v1.0, there exist 438,693,025 gene pairs.
Gene nodes have 302 distinct degrees for \emph{interacts} edges, resulting in 91,204 degree pairs.
This equates to an 4,810-fold reduction in the number of summary statistics that need to be stored to represent the null DWPC distribution for a metapath starting and ending with a \emph{Gene--interacts--Gene} metaedge.

We store the following null DWPC summary statistics for each metapath--source-degree--target-degree combination:
total number of null DWPCs,
total number of nonzero null DWPCs,
sum of null DWPCs,
sum of squared null DWPCs,
and number of permuted hetnets.
These values are sufficient to estimate the \emph{p}-value for a DWPC, by either \href{https://github.com/hetio/hetmatpy/blob/bc36aa9859c43a1a5fb22808cd6eb952ef9d497c/hetmatpy/pipeline.py\#L42-L63}{fitting} a gamma-hurdle null distribution or generating an empiric \emph{p}-value.
Furthermore, these statistics are additive across permuted hetnets.
Their values are always a running total and can be updated incrementally as statistics from each additional permuted hetnet become available.

Figure \ref{fig:degree-group-metrics} shows how various aspects of DWPCs vary by degree group.
The rows display the following metrics of the DWPC distribution for all node-pairs in a given degree-group:

\begin{itemize}
\tightlist
\item
  \textbf{\# Nonzero DWPCs}:
  The number of nonzero DWPCs values (on average per network to enable comparison).
\item
  \textbf{\% Nonzero DWPCs}:
  Of the total number of DWPCs, the percent that is nonzero.
  As node degrees increase, the chance of node pairs having at least one path, and hence a nonzero DWPC, greatly increases.
\item
  \textbf{Mean DWPC}:
  The average value of all DWPCs including zeros.
\item
  \textbf{Mean Nonzero DWPC}:
  The average value of nonzero DWPCs.
\item
  \textbf{Std Dev Nonzero DWPC}:
  The standard deviation of nonzero DWPCs.
\item
  \textbf{Gamma Model β}:
  The β parameter of the gamma model fit on nonzero DWPCs.
  Note that the gamma model is only fit on permuted network DWPCs to estimate a null distribution for the unpermuted network DWPCs.
  Since this parameter varies with source \& target node degree,
  it is important to fit a separate gamma model for each degree group.
\end{itemize}

\begin{figure}
\hypertarget{fig:degree-group-metrics}{%
\centering
\includegraphics{https://github.com/greenelab/connectivity-search-analyses/raw/15f1925c0481d8e6bab8b0931f48f2fad388c68c/explore/degree-group-analyses.png}
\caption{\textbf{Path-based metrics vary by node degree and network permutation status.}
Each row shows a different metric of the DWPC distribution for the \emph{CbGpPWpG} metapath --- traversing Compound--binds--Gene--participates--Pathway--participates--Gene, selected for illustrative purposes.
Metrics are computed for degree-groups,
which is a specific pair of source degree (in this case, the source compound's count of CbG edges)
and target degree (in this case, the target gene's count of GpPW edges).
On the left, metrics are reported for the unpermuted hetnet and on the right for the 200 permuted hetnets.
Hence, each cell on right summarizes 200 times the number of DWPCs as the corresponding cell on the left.
The colormap is row normalized, such that its intensity peaks for the maximum value of each metric across the unpermuted and permuted values.
Gray indicates null values.}\label{fig:degree-group-metrics}
}
\end{figure}

\hypertarget{gamma-hurdle-distribution}{%
\subsubsection{Gamma-hurdle distribution}\label{gamma-hurdle-distribution}}

We are interested in identifying source and target nodes whose connectivity exceeds what typically arises at random.
To identify such especially-connected nodes, we compare DWPC values to the distribution of permuted network DWPC values for the same source and target nodes.
While a single DWPC value is not actually a test statistic, we use a framework akin to classical hypothesis testing to identify outliers.

Two observations led us to the quasi-significance testing framework we developed.
First, a sizable fraction of permuted DWPC values is often zero, indicating that the source and target nodes are not connected along the metapath in the permuted network.
Second, we observed that non-zero DWPC values for any given source and target nodes are reasonably approximated as following a gamma distribution.
Motivated by these observations, we parametrized permuted DWPC values using a zero-inflated gamma distribution, which we termed the gamma-hurdle distribution.
We fit a gamma-hurdle distribution to each combination of source node, target node, and metapath.
Finally, we estimated the probability of observing a permuted DWPC value greater than DWPC computed in the unpermuted network, akin to a one-tailed p-value.
These quasi-significance scores (`\emph{p}-values') allow us to identify outlier node pairs at the metapath level (see examples in Figure \ref{fig:null-dwpc-distributions}).

\begin{figure}
\hypertarget{fig:null-dwpc-distributions}{%
\centering
\includegraphics{https://github.com/greenelab/connectivity-search-analyses/raw/9287986f331607cfdbc1ac197b52f36085723c6e/explore/gamma-hurdle/gamma-hurdle-distributions.png}
\caption{\textbf{From null distribution to \emph{p}-value for DWPCs.}
Null DWPC distributions are shown for three metapaths between Alzheimer's disease and the circadian rhythm pathway, selected from Figure \ref{fig:webapp-metapaths}.
For each metapath, null DWPCs are computed on 200 permuted hetnets and grouped according to source--target degree.
Histograms show the null DWPCs for the degree group corresponding to Alzheimer's disease and the circadian rhythm pathway (as noted in the plot titles by deg.)
The proportion of null DWPCs that were zero is calculated, forming the ``hurdle'' of the null distribution model.
The nonzero null DWPCs are modeled using a gamma distribution, which can be fit solely from a sample mean and standard deviation.
The mean of nonzero null DWPCs is denoted with a diamond, with the standard deviation plotted twice as a line in either direction.
Actual DWPCs are compared to the gamma-hurdle null distribution to yield a \emph{p}-value.}\label{fig:null-dwpc-distributions}
}
\end{figure}

\hypertarget{details-of-the-gamma-hurdle-distribution}{%
\paragraph{Details of the gamma-hurdle distribution}\label{details-of-the-gamma-hurdle-distribution}}

Let \emph{X} be a gamma-hurdle random variable with parameters \emph{λ}, \emph{α}, and \emph{β}.

\[
X \sim \Gamma_H(\lambda, \alpha, \beta)
\]

The gamma-hurdle distribution is defined over the support {[}0, ∞).
The probability of a draw, X, from the gamma-hurdle distribution is given as follows:

\begin{align}
  P(X = 0) &= 1 - \lambda \\
  P(X \in A; A \subseteq (0, \infty)) &= \frac{\lambda \beta^\alpha}{\Gamma(\alpha)}  \int  _{x \in A} \bigg( x^{\alpha - 1} e^{-\beta x} \bigg)
\end{align}

We estimate all three parameters using the method of moments (using Bessel's correction to estimate the second moment).
As a validation of our method,
we \href{https://github.com/greenelab/connectivity-search-analyses/blob/1c6827ce2544c17cef42bbccf098a312f2c44f97/explore/gamma-hurdle/parameter_estimates.ipynb}{compared} our method of moments parameter estimates to approximate maximum likelihood estimates
(gamma distribution parameters do not have closed-form maximum likelihood estimates)
and found excellent concordance between the methods.
Let \emph{N} be the number of permuted DWPC values, and \emph{n} the number of nonzero values.

\begin{align}
  \hat{\lambda} &= \frac{n}{N} \\
  \hat{\alpha} &= \frac{(n - 1) \sum x_i}{n \sum (x_i^2) - (\sum x_i)^2} \\
  \hat{\beta} &= \frac{n - 1}{n} \frac{(\sum x_i)^2}{n \sum (x_i)^2 - (\sum x_i)^2}
\end{align}

Finally, we compute a p-value for each DWPC value, \emph{t}.

\[
p = P(X ≥ t) = \frac{\beta^\alpha}{\Gamma(\alpha)} \int_t^\infty x^{\alpha - 1} \exp(-\beta x) dx
\]

\hypertarget{empirical-dwpc-p-values}{%
\subsubsection{Empirical DWPC p-values}\label{empirical-dwpc-p-values}}

We \href{https://github.com/hetio/hetmatpy/blob/bc36aa9859c43a1a5fb22808cd6eb952ef9d497c/hetmatpy/pipeline.py\#L92-L113}{calculate} an empirical p-value for special cases where the gamma-hurdle model cannot be applied.
These cases include when the observed DWPC is zero or when the null DWPC distribution is all zeroes or has only a single distinct nonzero value.
The empirical \emph{p}-value (\emph{p\textsubscript{empiric}}) equals the proportion of null DPWCs ≥ the observed DWPC.

Since we don't store all null DWPC values,
we apply the following criteria to calculate \emph{p\textsubscript{empiric}} from summary statistics:

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\tightlist
\item
  When the observed DWPC = 0 (no paths of the specified metapath existed between the source and target node),
  \emph{p\textsubscript{empiric}} = 1.
\item
  When all null DWPCs are zero but the observed DWPC is positive,
  \emph{p\textsubscript{empiric}} = 0.
\item
  When all nonzero null DWPCs have the same positive value (standard deviation = 0),
  \emph{p\textsubscript{empiric}} = 0 if the observed DWPC \textgreater{} the null DWPC else \emph{p\textsubscript{empiric}} = proportion of nonzero null DWPCs.
\end{enumerate}

\hypertarget{dwpc-and-null-distribution-computation}{%
\subsubsection{DWPC and null distribution computation}\label{dwpc-and-null-distribution-computation}}

We decided to compute DWPCs and their significance for all source--target node pairs for metapaths with length ≤ 3.
On Hetionet v1.0, there are 24 metapaths of length 1, 242 metapaths of length 2, and 1,939 metapaths of length 3.
The decision to stop at length 3 was one of practicality, as length 4 would have added 17,511 metapaths.

For each of the 2,205 \href{https://github.com/greenelab/connectivity-search-analyses/raw/042063fb559048e52b3dc2731b6d6c6836f698cf/explore/bulk-pipeline/archives/metapath-dwpc-stats.tsv}{metapaths},
we \href{https://github.com/greenelab/connectivity-search-analyses/blob/1c6827ce2544c17cef42bbccf098a312f2c44f97/explore/bulk-pipeline/bulk.ipynb}{computed} the complete path count matrix and DWPC matrix.
In total, we computed 137,786,767,964 path counts (and the same number of DWPCs) on the unpermuted network,
of which 11.6\% were nonzero.

The DWPC has a single parameter, called the damping exponent (\emph{w}), which controls how much paths through high-degree nodes are downweighted {[}\protect\hyperlink{ref-WkPlH1ds}{42}{]}.
When \emph{w} = 0, the DWPC is equivalent to the path count.
Previously, we found \emph{w} = 0.4 was optimal for predicting disease-associated genes {[}\protect\hyperlink{ref-WkPlH1ds}{42}{]}.
Here, we use \emph{w} = 0.5, since taking the square root of degrees has more intuitive appeal.

We selected data types for matrix values that would allow for high precision.
We used 64-bit unsigned integers for path counts and 64-bit floating-point numbers for DWPCs.
We \href{https://github.com/greenelab/connectivity-search-analyses/pull/91}{considered} using 16-bits or 32-bits per DWPC to reduce memory/storage requirements,
but decided against it in case certain applications required greater precision.

We used SciPy sparse for path count and DWPC matrices with density \textless{} 0.7, serialized to disk with compression and a \texttt{.sparse.npz} extension.
This format minimizes the space on disk and load time for the entire matrix but does not offer read access to slices.
We used Numpy 2D arrays for DWPC matrices with density ≥ 0.7, serialized to disk using Numpy's \texttt{.npy} format.
We bundled the path count and DWPC matrix files into HetMat archives by metapath length and deposited the archives to Zenodo {[}\protect\hyperlink{ref-pkJqqM2W}{52}{]}.
The archive for length 3 DWPCs was the largest at 131.7 GB.

We also generated null DWPC summary statistics for the 2,205 metapaths,
which are also available by metapath length from Zenodo as HetMat archives consisting of \texttt{.tsv.gz} files {[}\protect\hyperlink{ref-pkJqqM2W}{52}{]}.
Due to degree grouping, null DWPCs summary statistic archives are much smaller than the DWPC archives.
The archive for length 3 null DWPCs summary statistics was 733.1 MB.
However, the compute required to generate null DWPCs is far greater because there are multiple permuted hetnets (in our case 200).
As a result, computing and saving all DWPCs took 6 hours,
whereas computing and saving the null DWPC summary statistics took 361 hours.

Including null DWPCs and path counts, the Zenodo deposit totals 185.1 GB and contains the results of computing \textasciitilde28 trillion DWPCs --- 27,832,927,128,728 to be exact.

\hypertarget{adjusting-dwpc-p-values}{%
\subsubsection{\texorpdfstring{Adjusting DWPC \emph{p}-values}{Adjusting DWPC p-values}}\label{adjusting-dwpc-p-values}}

When a user applies hetnet connectivity search to identify enriched metapaths between two nodes,
many metapaths are evaluated for significance.
Due to multiple testing of many DWPCs, low \emph{p}-values are likely to arise by chance.
Therefore, we devised a multiple-testing correction.

For each combination of source metanode, target metanode, and length, we counted the number of metapaths.
For Disease\ldots Pathway metapaths, there are 0 metapaths of length 1, 3 metapaths of length 2, and 24 metapaths of length 3.
We calculated adjusted p-values by applying a Bonferroni correction based on the number of metapaths of the same length between the source and target metanode.
Using Figure \ref{fig:webapp-metapaths} as an example, the \href{https://het.io/search/?source=17287\&target=7607\&metapaths=DdGpPW\&complete=}{DdGpPW} p-value of 5.9\% was adjusted to 17.8\% (multiplied by a factor of 3).

Bonferroni controls family-wise error rate,
which corresponds here to incorrectly finding that \emph{any} metapath of a given length is enriched.
As a result, our adjusted p-values are conservative.
We would prefer to adjust \emph{p}-values for false discovery rate {[}\protect\hyperlink{ref-q6QyrHEZ}{53}{]},
but these methods often require access to all \emph{p}-values at once (impractical here) and assume a uniform distribution of \emph{p}-values when there is no signal (not the case here when most DWPCs are zero).

\hypertarget{prioritizing-enriched-metapaths-for-database-storage}{%
\subsubsection{Prioritizing enriched metapaths for database storage}\label{prioritizing-enriched-metapaths-for-database-storage}}

Storing DWPCs and their significance in the database (as part of the \texttt{PathCount} table in Figure \ref{fig:database}) enables the connectivity search webapp to provide users with enriched metapaths between query nodes in real time.
However, storing \textasciitilde15.9 billion rows (the total number of nonzero DWPCs) in the database's \texttt{PathCount} table would exceed a reasonable disk quota.
An alternative would be to store all DWPCs in the database whose adjusted p-value exceeded a universal threshold (e.g.~\emph{p} \textless{} 5\%).
But we estimated this would still be prohibitively expensive.
Therefore, we devised a metapath-specific threshold.
For metapaths with length 1, we stored all nonzero DWPCs,
assuming users always want to be informed about direct edges between the query nodes, regardless of significance.
For metapaths with length ≥ 2, we chose an adjusted \emph{p}-value threshold of
5 × (\emph{n\textsubscript{source}} × \emph{n\textsubscript{target}})\textsuperscript{−0.3},
where \emph{n\textsubscript{source}} and \emph{n\textsubscript{target}} are the node counts for the source and target metanodes (i.e.~``Nodes'' column in Table \ref{tbl:metanodes}).
Notice that metapaths with large number of possible source--target pairs (large DWPC matrices) are penalized.
This decision is based on practicality since otherwise the majority of the database quota would be consumed by a minority of metapaths between plentiful metanodes (e.g.~Gene\ldots Gene metapaths).
Also, we assume that users will search nodes at a similar rate by metanode (e.g.~they're more likely to search for a specific disease than a specific gene).
The constants in the threshold formula help scale it.
The multiplier of 5 relaxes the threshold to saturate the available database capacity.
The −0.3 exponent applies the large DWPC-matrix penalty.

Users can still evaluate DWPCs that are not stored in the database, using either the webapp or API.
These are calculated on the fly, delegating DWPC computation to the Neo4j database.
Unchecking ``precomputed only'' on the webapp shows all possible metapaths for two query nodes.
For some node pairs, the on-the-fly computation is quick (less than a second).
Other times, computing DWPCs for all metapaths might take more than a minute.

\hypertarget{backend-database-api}{%
\subsubsection{Backend Database \& API}\label{backend-database-api}}

We created a backend application using Python's Django web framework.
The source code is available in the \href{https://github.com/greenelab/connectivity-search-backend}{\texttt{connectivity-search-backend}} repository.
The primary role of the backend is to manage a relational database and provide an API for requesting data.

We define the database schema \href{https://github.com/greenelab/connectivity-search-backend/blob/af12f8cf2ad47d9a25ce8d1b7889390654eb3bb9/dj_hetmech_app/models.py}{using} Django's object-relational mapping framework (Figure \ref{fig:database}).
We \href{https://github.com/greenelab/connectivity-search-backend/blob/af12f8cf2ad47d9a25ce8d1b7889390654eb3bb9/dj_hetmech_app/management/commands/populate_database.py}{import} the data into a PostgreSQL database.
Populating the database for all 2,205 metapaths up to length 3 was a prolonged operation, \href{https://github.com/greenelab/connectivity-search-backend/pull/41\#issuecomment-488054789}{taking} over 3 days.
The majority of the time is spent populating the \texttt{DegreeGroupedPermutation} (37,905,389 rows) and \texttt{PathCount} (174,986,768 rows) tables.
To avoid redundancy, the database only stores a single orientation of a metapath.
For example, if rows are stored for the \emph{GpPWpGaD} metapath,
they would not also be stored for the \emph{DaGpPWpG} metapath.
The backend is responsible for checking both orientations of a metapath in the database and reversing metapaths on-the-fly before returning results.
The database is located at \texttt{search-db.het.io} with public read-only access \href{https://github.com/greenelab/connectivity-search-backend\#database}{available}.

\begin{figure}
\hypertarget{fig:database}{%
\centering
\includegraphics[width=1\textwidth,height=\textheight]{https://github.com/greenelab/connectivity-search-backend/raw/752b423a4b7b57575d66ce0b797b0a84c23267a6/media/models-schema.png}
\caption{\textbf{Schema for the connectivity search backend relational database models.}
Each Django model is represented as a table,
whose rows list the model's field names and types.
Each model corresponds to a database table.
Arrows denote foreign key relationships.
The arrow labels indicate the foreign key field name followed by reverse relation names generated by Django (in parentheses).}\label{fig:database}
}
\end{figure}

We host a public API instance at \url{https://search-api.het.io}.
Version 1 of the API exposes several endpoints that are used by the connectivity search frontend including queries for
node details (\texttt{/v1/node}),
node lookup (\texttt{/v1/nodes}),
metapath information (\texttt{/v1/metapaths}),
and path information (\texttt{/v1/paths}).
The endpoints return JSON payloads.
Producing results for these queries relies on internal calls to the PostgreSQL relational database as well as the pre-existing Hetionet v1.0 Neo4j graph database.
They were designed to power the hetnet connectivity search webapp,
but are also available for general research use.

\hypertarget{frontend}{%
\subsubsection{Frontend}\label{frontend}}

\hypertarget{hetio-website}{%
\paragraph{Hetio Website}\label{hetio-website}}

We created a static website to serve as the home for the Hetio organization using Jekyll hosted on GitHub Pages (Figure \ref{fig:website}).
The source code is available in the \href{https://github.com/hetio/het.io}{\texttt{het.io}} repository.
To make a change to the website, an author simply commits the changes (either directly or through a pull request) to the repository's \texttt{gh-pages} branch, and GitHub automatically re-compiles the website and hosts the resulting files at the provided custom domain URL.

\begin{figure}
\hypertarget{fig:website}{%
\centering
\includegraphics[width=1\textwidth,height=\textheight]{https://github.com/greenelab/connectivity-search-manuscript/raw/d5766924e8c774accdc143bea352e49610ee0673/content/media/website/website-homepage.png}
\caption{\textbf{Homepage of the Hetio website.}
The redesigned homepage provides a succinct overview of what Hetionet consists of and what its purpose is.}\label{fig:website}
}
\end{figure}

\hypertarget{webapps}{%
\paragraph{Webapps}\label{webapps}}

We developed the \href{https://het.io/search}{connectivity search app} as a single-page, standalone application using React and associated tools.
The source code is available in the \href{https://github.com/greenelab/connectivity-search-frontend}{\texttt{connectivity-search-frontend}} repository.

Since the rest of the overarching Hetio website was mostly non-interactive content, it was appropriate to construct the bulk of the website in simple static formats like Markdown and HTML using Jekyll, and leave React for implementing the sections of the site that required more complex behavior.

We used React's own \texttt{create-react-app} command line tool to generate a boilerplate for the app.
This simplified setup, testing, and building pipelines, bypassing time-consuming configuration of things like Webpack and linters.
Some configuration was necessary to produce non-hashed, consistently named output files like \texttt{index.js} that could be easily and reliably referenced by and embedded into the Hetio Jekyll website.

For authoring components, we used React's traditional class syntax.
At the time of authoring the app, React Hooks were still nascent, thus the simpler and less-verbose functional syntax was not viable.

While writing this application, we also elected to re-write the pre-existing \href{https://het.io/repurpose/}{Rephetio} and \href{https://het.io/disease-genes/}{disease-associated genes} apps in the same manner.
We created a custom package of React components and utility functions that could be shared across the multiple interactive apps on the website.
The package is located at and can be installed from the \href{https://github.com/hetio/frontend-components}{\texttt{frontend-components}} repository.
The package consists of interface ``components'' (building blocks) like buttons and sortable/searchable/paginated tables as well as utility functions for formatting data and debugging.
Each of the interactive apps imports this package to reduce code repetition and to enforce a consistent style and behavior across the website.

For managing state in the connectivity search app, we used the Redux library.
Redux was chosen over vanilla React or other state management libraries since:

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\tightlist
\item
  The state in this app was very ``global'', meaning most of it was needed by a lot of different parts of the app.
  Redux provides a convenient global ``store'' of state that is easily accessible to any component, avoiding the ``prop-drilling'' phenomenon.
\item
  The structure of the state is nested and complex.
  Redux's ``reducer'' approach makes it cleaner to modify this type of data.
\item
  Redux's approach to immutable state that is updated by actions and pure functions makes the application easier to debug.
  It is easy to get a clear timeline of how and when the state changed, and what changed it.
\end{enumerate}

To create the graph visualization at the bottom of the app, the D3 library was used.
D3 satisfied several core requirements:

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\tightlist
\item
  SVG implementation for high-resolution, publication-ready figures.
\item
  Force-directed layout for untangling nodes.
\item
  Pinnable nodes and other physics customizations.
\item
  Customizable node and edge drag/hover/select behavior.
\item
  Intuitive pan/zoom view that worked on desktop and mobile.
\item
  Node and edge appearances that were completely customizable for alignment, text wrapping, color, outlines, fonts, arrowheads, and non-colliding coincident edges.
\end{enumerate}

\hypertarget{visual-design}{%
\subsubsection{Visual Design}\label{visual-design}}

A limited palette of colors was chosen to represent the different types of nodes (metanodes) in the Hetionet knowledge graph.
These colors are listed and programmatically accessible in the \href{https://github.com/hetio/hetionet}{\texttt{hetionet}} repository under \texttt{/describe/styles.json}.

At the time of developing connectivity search, Hetionet already had an established palette of colors (from Project Rephetio).
To avoid confusion, we were \href{https://github.com/hetio/hetionet/pull/18}{careful} to keep the general hue of each metanode color the same for backwards compatibility,
e.g.~genes stayed generally blue, diseases stayed generally brown.
In this way, this palette selection was more of a modernization/refresh.
For cohesiveness, accessibility, and aesthetic appeal,
we used the professionally-curated Material Design palette as a source for the specific color values.

The palette is now used in all Hetio-related applications and materials.
This is not just to maintain a consistent look and feel across the Hetio organization, but to convey clear and precise meaning.
For example, the colors used in the metagraph in Figure \ref{fig:rephetio}A are exactly the same colors, and thus represent the same types of entities, as in \emph{any part of the connectivity search app} (Figure \ref{fig:webapp}).

Colors in the palette are also used in the Hetio logo (seen in Figure \ref{fig:website}) and other miscellaneous logos and iconography across the website, to establish an identifiable brand for the Hetio organization as a whole.

\hypertarget{realtime-open-science}{%
\subsubsection{Realtime open science}\label{realtime-open-science}}

This study was conducted entirely in the open via public GitHub repositories.
We used GitHub Issues for discussion, leaving a rich online history of the scholarly process.
Furthermore, most additions to the analyses were performed by pull request,
whereby a contributor proposes a set of changes.
This provides an opportunity for other contributors to review changes before they are officially accepted.
For example, in \href{https://github.com/greenelab/connectivity-search-analyses/pull/156}{greenelab/connectivity-search-analyses\#156} @zietzm proposed a notebook to visualize parameters for null DWPC distributions.
After @zietzm addressed @dhimmel's comments, the pull request was approved and merged into the project's main branch.

The manuscript for this study was written using \href{https://manubot.org/}{Manubot}, which allows authors to collaboratively write manuscripts on GitHub {[}\protect\hyperlink{ref-YuJbg3zO}{54}{]}.
The Manubot-rendered manuscript is available at \url{https://greenelab.github.io/connectivity-search-manuscript/}.
We encourage readers with feedback or questions to comment publicly via \href{https://github.com/greenelab/connectivity-search-manuscript/issues}{GitHub Issues}.

\hypertarget{software-data-availability}{%
\subsubsection{Software \& data availability}\label{software-data-availability}}

\emph{Hetio} is a superset/collection of hetnet-related research, tools, and datasets that, most notably, includes the Hetionet project itself and the connectivity search tool that are the focus of this manuscript.
Most of the Hetio resources and projects can be found under the \href{https://github.com/hetio}{Hetio GitHub organization}, with others being available under the \href{https://github.com/greenelab}{Greene Lab GitHub organization}, one of the collaborating groups.
Information about Hetio is also displayed and disseminated at \url{https://het.io}, as noted in the \protect\hyperlink{hetio-website}{Hetio Website} section.

\hypertarget{availability-of-supporting-source-code-and-requirements}{%
\paragraph{Availability of Supporting Source Code and Requirements}\label{availability-of-supporting-source-code-and-requirements}}

\begin{itemize}
\tightlist
\item
  Project name: Hetnet Connectivity Search
\item
  Project home page: \url{https://het.io/search/}
\item
  Operating systems: Platform independent
\item
  Programming language: Python, Javascript, Cypher
\item
  Other requirements:
  refer to specific repositories below for their respective dependency configuration files
\item
  License:
  refer to specific repositories below,
  but generally software is released under BSD,
  figures and documentation under CC BY,
  and data under CC0.
\item
  RRID: \href{https://scicrunch.org/resolver/RRID:SCR_023630}{SCR\_023630}
\item
  biotools ID: \href{https://bio.tools/connectivity-search}{connectivity-search}
\end{itemize}

This study primarily involves the following GitHub repositories:

\begin{itemize}
\tightlist
\item
  \href{https://github.com/greenelab/connectivity-search-manuscript}{greenelab/connectivity-search-manuscript}
  {[}\protect\hyperlink{ref-3KJWwgoz}{55}{]}:
  Source code for this manuscript.
  Best place for general comments or questions.
  CC BY 4.0 License.
\item
  \href{https://github.com/greenelab/connectivity-search-analyses}{greenelab/connectivity-search-analyses}
  {[}\protect\hyperlink{ref-eq4IX1LX}{56}{]}:
  The initial project repository that contains research notebooks, dataset generation code, and exploratory data analyses.
  The hetmatpy package was first developed as part of this repository until its \href{https://github.com/hetio/hetmatpy/issues/1}{relocation} in November 2018.
  BSD 3-Clause License.
\item
  \href{https://github.com/greenelab/connectivity-search-backend}{greenelab/connectivity-search-backend}
  {[}\protect\hyperlink{ref-ASx5cMWI}{57}{]}:
  Source code for the connectivity search database and API.
  BSD 3-Clause License.
\item
  \href{https://github.com/greenelab/connectivity-search-frontend}{greenelab/connectivity-search-frontend}
  {[}\protect\hyperlink{ref-CadAZTf2}{58}{]}:
  Source code for the connectivity search webapp.
  BSD 3-Clause License.
\item
  \href{https://github.com/hetio/hetmatpy}{hetio/hetmatpy}
  {[}\protect\hyperlink{ref-gHk6T3vn}{59}{]}:
  Python package for matrix storage and operations on hetnets.
  Released on \href{https://pypi.org/project/hetmatpy/}{PyPI}.
  BSD 2-Clause Plus Patent License.
  Registered at \href{https://bio.tools/hetmatpy}{biotools:hetmatpy} and \href{https://scicrunch.org/resolver/RRID:SCR_023409}{RRID:SCR\_023409}.
\item
  \href{https://github.com/hetio/hetnetpy}{hetio/hetnetpy}
  {[}\protect\hyperlink{ref-srnBddlB}{60}{]}:
  Preexisiting python package for representing hetnets.
  Dependency of hetmatpy.
  Released on \href{https://pypi.org/project/hetnetpy/}{PyPI}.
  Dual licensed under BSD 2-Clause Plus Patent License and CC0 1.0 (public domain dedication).
\item
  \href{https://github.com/hetio/hetionet}{hetio/hetionet}
  {[}\protect\hyperlink{ref-L9cBdipF}{61}{]}:
  Preexisiting data repository for Hetionet,
  including the public Neo4j instance and HetMat archives.
  CC0 1.0 License.
\item
  \href{https://github.com/hetio/het.io}{hetio/het.io}
  {[}\protect\hyperlink{ref-12IQq3W3u}{62}{]}:
  Preexisiting source code for the \url{https://het.io/} website.
  CC BY 4.0 License.
\end{itemize}

\hypertarget{data-availability}{%
\paragraph{Data availability}\label{data-availability}}

An archival copy of the code and supporting data is available via the GigaScience repository {[}\protect\hyperlink{ref-hHRphzV0}{63}{]}.
The connectivity-search-analyses and hetionet repositories contain datasets related to this study.
Large datasets were compressed and tracked with \href{https://git-lfs.github.com/}{Git LFS} (Large File Storage).
GitHub LFS had a max file size of 2 GB.
Datasets exceeding this size, along with other essential datasets, are available from Zenodo {[}\protect\hyperlink{ref-pkJqqM2W}{52}{]}.

\hypertarget{declarations}{%
\subsection{Declarations}\label{declarations}}

\hypertarget{list-of-abbreviations}{%
\subsubsection{List of abbreviations}\label{list-of-abbreviations}}

\begin{description}
\tightlist
\item[Arnt]
aryl hydrocarbon receptor nuclear translocator protein
\item[ARNT2]
aryl hydrocarbon receptor nuclear translocator 2
\item[DWPC]
degree-weighted path count
\item[LFS]
large file storage
\item[NPAS2]
neuronal PAS domain protein 2
\item[PAS]
Per-Arnt-Sim
\item[Per]
period circadian protein
\item[Sim]
single-minded protein
\end{description}

\hypertarget{competing-interests}{%
\subsubsection{Competing Interests}\label{competing-interests}}

This work was supported, in part, by Pfizer Worldwide Research, Development, and Medical.

\hypertarget{funding}{%
\subsubsection{Funding}\label{funding}}

DSH, BJH, DH, and DNN were funded by The Gordon and Betty Moore Foundation (GBMF4552).
DSH and CSG were funded by Pfizer Worldwide Research, Development, and Medical.
KK was funded by The Gordon and Betty Moore Foundation (GBMF4560).
DNN was funded by The National Institutes of Health (T32 HG000046).
CSG was funded by the National Human Genome Research Institute (R01 HG010067).
CSG was funded by the National Cancer Institute (R01 CA237170).
CSG was funded by the Gordon and Betty Moore Foundation (GBMF 4552).
CSG was funded by the Eunice Kennedy Shriver National Institute of Child Health and Human Development (R01 HD109765).
The funders had no role in the study design, data analysis and interpretation, or writing of the manuscript.

\hypertarget{authors-contributions}{%
\subsubsection{Authors' contributions}\label{authors-contributions}}

Author contributions are noted here according to \href{https://credit.niso.org/}{CRediT} (Contributor Roles Taxonomy).
Conceptualization by DSH, MZ, KK, BDS, and CSG.
Data curation by DSH, MZ, VR, and DNN.
Formal analysis by DSH, MZ, KK, and BJH.
Funding acquisition by DSH, BDS, and CSG.
Investigation by DSH, MZ, BJH, YH, MWN, and CSG.
Methodology by DSH, MZ, VR, KK, BJH, and CSG.
Project administration by DSH and CSG.
Resources by DSH, FSA, DH, MWN, and CSG.
Software by DSH, MZ, VR, BJH, FSA, and DH.
Supervision by DSH, BDS, and CSG.
Visualization by DSH, MZ, VR, and YH.
Writing -- original draft by DSH, MZ, VR, DNN, and CSG.
Writing -- review \& editing by DSH, MZ, VR, DH, BDS, and CSG.
Validation by BJH.

\hypertarget{references}{%
\subsection{References}\label{references}}

\hypertarget{refs}{}
\begin{CSLReferences}{0}{0}
\leavevmode\vadjust pre{\hypertarget{ref-QWTAHxFj}{}}%
\CSLLeftMargin{1. }%
\CSLRightInline{\textbf{Renaming `heterogeneous networks' to a more concise and catchy term}
\CSLBlock{Daniel Himmelstein, Casey Greene, Sergio Baranzini} \emph{ThinkLab} (2015-08-16) \url{https://doi.org/f3mn4v}
\CSLBlock{DOI: \href{https://doi.org/10.15363/thinklab.d104}{10.15363/thinklab.d104}}}

\leavevmode\vadjust pre{\hypertarget{ref-O21tn8vf}{}}%
\CSLLeftMargin{2. }%
\CSLRightInline{\textbf{Systematic integration of biomedical knowledge prioritizes drugs for repurposing}
\CSLBlock{Daniel Scott Himmelstein, Antoine Lizee, Christine Hessler, Leo Brueggeman, Sabrina L Chen, Dexter Hadley, Ari Green, Pouya Khankhanian, Sergio E Baranzini} \emph{eLife} (2017-09-22) \url{https://doi.org/cdfk}
\CSLBlock{DOI: \href{https://doi.org/10.7554/elife.26726}{10.7554/elife.26726} · PMID: \href{https://www.ncbi.nlm.nih.gov/pubmed/28936969}{28936969} · PMCID: \href{https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5640425}{PMC5640425}}}

\leavevmode\vadjust pre{\hypertarget{ref-5t7pfaPv}{}}%
\CSLLeftMargin{3. }%
\CSLRightInline{\textbf{Announcing PharmacotherapyDB: the Open Catalog of Drug Therapies for Disease}
\CSLBlock{Daniel Himmelstein} \emph{ThinkLab} (2016-03-15) \url{https://doi.org/f3mqtv}
\CSLBlock{DOI: \href{https://doi.org/10.15363/thinklab.d182}{10.15363/thinklab.d182}}}

\leavevmode\vadjust pre{\hypertarget{ref-vvdoVrc6}{}}%
\CSLLeftMargin{4. }%
\CSLRightInline{\textbf{Our hetnet edge prediction methodology: the modeling framework for Project Rephetio}
\CSLBlock{Daniel Himmelstein} \emph{ThinkLab} (2016-05-04) \url{https://doi.org/f3qbmj}
\CSLBlock{DOI: \href{https://doi.org/10.15363/thinklab.d210}{10.15363/thinklab.d210}}}

\leavevmode\vadjust pre{\hypertarget{ref-EfWvuSjX}{}}%
\CSLLeftMargin{5. }%
\CSLRightInline{\textbf{The link-prediction problem for social networks}
\CSLBlock{David Liben-Nowell, Jon Kleinberg} \emph{Journal of the American Society for Information Science and Technology} (2007) \url{https://doi.org/c56765}
\CSLBlock{DOI: \href{https://doi.org/10.1002/asi.20591}{10.1002/asi.20591}}}

\leavevmode\vadjust pre{\hypertarget{ref-7iZtCGcX}{}}%
\CSLLeftMargin{6. }%
\CSLRightInline{\textbf{Link prediction in complex networks: A survey}
\CSLBlock{Linyuan Lü, Tao Zhou} \emph{Physica A: Statistical Mechanics and its Applications} (2011-03) \url{https://doi.org/dbs8g8}
\CSLBlock{DOI: \href{https://doi.org/10.1016/j.physa.2010.11.027}{10.1016/j.physa.2010.11.027}}}

\leavevmode\vadjust pre{\hypertarget{ref-otY29wFV}{}}%
\CSLLeftMargin{7. }%
\CSLRightInline{\textbf{Heterogeneous network embedding for identifying symptom candidate genes}
\CSLBlock{Kuo Yang, Ning Wang, Guangming Liu, Ruyu Wang, Jian Yu, Runshun Zhang, Jianxin Chen, Xuezhong Zhou} \emph{Journal of the American Medical Informatics Association} (2018-10-23) \url{https://doi.org/gfg6nr}
\CSLBlock{DOI: \href{https://doi.org/10.1093/jamia/ocy117}{10.1093/jamia/ocy117} · PMID: \href{https://www.ncbi.nlm.nih.gov/pubmed/30357378}{30357378} · PMCID: \href{https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7646926}{PMC7646926}}}

\leavevmode\vadjust pre{\hypertarget{ref-NnOS86ev}{}}%
\CSLLeftMargin{8. }%
\CSLRightInline{\textbf{Large-scale structural and textual similarity-based mining of knowledge graph to predict drug--drug interactions}
\CSLBlock{Ibrahim Abdelaziz, Achille Fokoue, Oktie Hassanzadeh, Ping Zhang, Mohammad Sadoghi} \emph{Journal of Web Semantics} (2017-05) \url{https://doi.org/gcrwk3}
\CSLBlock{DOI: \href{https://doi.org/10.1016/j.websem.2017.06.002}{10.1016/j.websem.2017.06.002}}}

\leavevmode\vadjust pre{\hypertarget{ref-GYLacVTV}{}}%
\CSLLeftMargin{9. }%
\CSLRightInline{\textbf{SMR: Medical Knowledge Graph Embedding for Safe Medicine Recommendation}
\CSLBlock{Fan Gong, Meng Wang, Haofen Wang, Sen Wang, Mengyue Liu} \emph{Big Data Research} (2021-02) \url{https://doi.org/gjqwnc}
\CSLBlock{DOI: \href{https://doi.org/10.1016/j.bdr.2020.100174}{10.1016/j.bdr.2020.100174}}}

\leavevmode\vadjust pre{\hypertarget{ref-3gzmoW2V}{}}%
\CSLLeftMargin{10. }%
\CSLRightInline{\textbf{PyKEEN 1.0: A Python Library for Training and Evaluating Knowledge Graph Embeddings}
\CSLBlock{Mehdi Ali, Max Berrendorf, Charles Tapley Hoyt, Laurent Vermue, Sahand Sharifzadeh, Volker Tresp, Jens Lehmann} \emph{Journal of Machine Learning Research} (2021) \url{http://jmlr.org/papers/v22/20-825.html}}

\leavevmode\vadjust pre{\hypertarget{ref-zHcAMCig}{}}%
\CSLLeftMargin{11. }%
\CSLRightInline{\textbf{Understanding the performance of knowledge graph embeddings in drug discovery}
\CSLBlock{Stephen Bonner, Ian P Barrett, Cheng Ye, Rowan Swiers, Ola Engkvist, Charles Tapley Hoyt, William L Hamilton} \emph{Artificial Intelligence in the Life Sciences} (2022-12) \url{https://doi.org/gqxmj9}
\CSLBlock{DOI: \href{https://doi.org/10.1016/j.ailsci.2022.100036}{10.1016/j.ailsci.2022.100036}}}

\leavevmode\vadjust pre{\hypertarget{ref-U1Dfux5r}{}}%
\CSLLeftMargin{12. }%
\CSLRightInline{\textbf{node2vec: Scalable Feature Learning for Networks}
\CSLBlock{Aditya Grover, Jure Leskovec} \emph{Proceedings of the 22nd ACM SIGKDD International Conference on Knowledge Discovery and Data Mining} (2016-08-13) \url{https://doi.org/gftdzj}
\CSLBlock{DOI: \href{https://doi.org/10.1145/2939672.2939754}{10.1145/2939672.2939754} · PMID: \href{https://www.ncbi.nlm.nih.gov/pubmed/27853626}{27853626} · PMCID: \href{https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5108654}{PMC5108654}}}

\leavevmode\vadjust pre{\hypertarget{ref-1AeZs6xaT}{}}%
\CSLLeftMargin{13. }%
\CSLRightInline{\textbf{metapath2vec: Scalable Representation Learning for Heterogeneous Networks}
\CSLBlock{Yuxiao Dong, Nitesh V Chawla, Ananthram Swami} \emph{KDD \textquotesingle17: Proceedings of the 23rd ACM SIGKDD International Conference on Knowledge Discovery and Data Mining} (2017-08-04) \url{https://doi.org/gfsqzn}
\CSLBlock{DOI: \href{https://doi.org/10.1145/3097983.3098036}{10.1145/3097983.3098036}}}

\leavevmode\vadjust pre{\hypertarget{ref-jyhFBVzQ}{}}%
\CSLLeftMargin{14. }%
\CSLRightInline{\textbf{edge2vec: Representation learning using edge semantics for biomedical knowledge discovery}
\CSLBlock{Zheng Gao, Gang Fu, Chunping Ouyang, Satoshi Tsutsui, Xiaozhong Liu, Jeremy Yang, Christopher Gessner, Brian Foote, David Wild, Ying Ding, Qi Yu} \emph{BMC Bioinformatics} (2019-06-10) \url{https://doi.org/ggpcsq}
\CSLBlock{DOI: \href{https://doi.org/10.1186/s12859-019-2914-2}{10.1186/s12859-019-2914-2} · PMID: \href{https://www.ncbi.nlm.nih.gov/pubmed/31238875}{31238875} · PMCID: \href{https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6593489}{PMC6593489}}}

\leavevmode\vadjust pre{\hypertarget{ref-xq3Nsf5M}{}}%
\CSLLeftMargin{15. }%
\CSLRightInline{\textbf{Preclinical validation of therapeutic targets predicted by tensor factorization on heterogeneous graphs}
\CSLBlock{Saee Paliwal, Alex de Giorgio, Daniel Neil, Jean-Baptiste Michel, Alix MB Lacoste} \emph{Scientific Reports} (2020-10-26) \url{https://doi.org/gg49}
\CSLBlock{DOI: \href{https://doi.org/10.1038/s41598-020-74922-z}{10.1038/s41598-020-74922-z} · PMID: \href{https://www.ncbi.nlm.nih.gov/pubmed/33106501}{33106501} · PMCID: \href{https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7589557}{PMC7589557}}}

\leavevmode\vadjust pre{\hypertarget{ref-1EeE9qtFA}{}}%
\CSLLeftMargin{16. }%
\CSLRightInline{\textbf{Data Fusion by Matrix Factorization}
\CSLBlock{Marinka Zitnik, Blaz Zupan} \emph{IEEE Transactions on Pattern Analysis and Machine Intelligence} (2015-01-01) \url{https://doi.org/f3mqwz}
\CSLBlock{DOI: \href{https://doi.org/10.1109/tpami.2014.2343973}{10.1109/tpami.2014.2343973} · PMID: \href{https://www.ncbi.nlm.nih.gov/pubmed/26353207}{26353207}}}

\leavevmode\vadjust pre{\hypertarget{ref-cnCLKeU9}{}}%
\CSLLeftMargin{17. }%
\CSLRightInline{\textbf{Translating embeddings for modeling multi-relational data}
\CSLBlock{Antoine Bordes, Nicolas Usunier, Alberto Garcia-Durán, Jason Weston, Oksana Yakhnenko} \emph{Proceedings of the 26th International Conference on Neural Information Processing Systems} (2013-12-05) \url{https://dl.acm.org/doi/10.5555/2999792.2999923}}

\leavevmode\vadjust pre{\hypertarget{ref-vhsYX2y}{}}%
\CSLLeftMargin{18. }%
\CSLRightInline{\textbf{Integrating and formatting biomedical data as pre-calculated knowledge graph embeddings in the Bioteque}
\CSLBlock{Adrià Fernández-Torras, Miquel Duran-Frigola, Martino Bertoni, Martina Locatelli, Patrick Aloy} \emph{Nature Communications} (2022-09-09) \url{https://doi.org/gqtdvj}
\CSLBlock{DOI: \href{https://doi.org/10.1038/s41467-022-33026-0}{10.1038/s41467-022-33026-0} · PMID: \href{https://www.ncbi.nlm.nih.gov/pubmed/36085310}{36085310} · PMCID: \href{https://www.ncbi.nlm.nih.gov/pmc/articles/PMC9463154}{PMC9463154}}}

\leavevmode\vadjust pre{\hypertarget{ref-taI1UUAE}{}}%
\CSLLeftMargin{19. }%
\CSLRightInline{\textbf{Predicting gene-disease associations from the heterogeneous network using graph embedding}
\CSLBlock{Xiaochan Wang, Yuchong Gong, Jing Yi, Wen Zhang} \emph{2019 IEEE International Conference on Bioinformatics and Biomedicine (BIBM)} (2019-11) \url{https://doi.org/ggmrpj}
\CSLBlock{DOI: \href{https://doi.org/10.1109/bibm47256.2019.8983134}{10.1109/bibm47256.2019.8983134}}}

\leavevmode\vadjust pre{\hypertarget{ref-17ioZ6znu}{}}%
\CSLLeftMargin{20. }%
\CSLRightInline{\textbf{A Method to Learn Embedding of a Probabilistic Medical Knowledge Graph: Algorithm Development}
\CSLBlock{Linfeng Li, Peng Wang, Yao Wang, Shenghui Wang, Jun Yan, Jinpeng Jiang, Buzhou Tang, Chengliang Wang, Yuting Liu} \emph{JMIR Medical Informatics} (2020-05-21) \url{https://doi.org/ghqszs}
\CSLBlock{DOI: \href{https://doi.org/10.2196/17645}{10.2196/17645} · PMID: \href{https://www.ncbi.nlm.nih.gov/pubmed/32436854}{32436854} · PMCID: \href{https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7273238}{PMC7273238}}}

\leavevmode\vadjust pre{\hypertarget{ref-1D9FTzRBg}{}}%
\CSLLeftMargin{21. }%
\CSLRightInline{\textbf{Semantic Disease Gene Embeddings (SmuDGE): phenotype-based disease gene prioritization without phenotypes}
\CSLBlock{Mona Alshahrani, Robert Hoehndorf} \emph{Bioinformatics} (2018-09-01) \url{https://doi.org/gd9k8n}
\CSLBlock{DOI: \href{https://doi.org/10.1093/bioinformatics/bty559}{10.1093/bioinformatics/bty559} · PMID: \href{https://www.ncbi.nlm.nih.gov/pubmed/30423077}{30423077} · PMCID: \href{https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6129260}{PMC6129260}}}

\leavevmode\vadjust pre{\hypertarget{ref-6PISrkV5}{}}%
\CSLLeftMargin{22. }%
\CSLRightInline{\textbf{A network embedding model for pathogenic genes prediction by multi-path random walking on heterogeneous network}
\CSLBlock{Bo Xu, Yu Liu, Shuo Yu, Lei Wang, Jie Dong, Hongfei Lin, Zhihao Yang, Jian Wang, Feng Xia} \emph{BMC Medical Genomics} (2019-12) \url{https://doi.org/ggmrpq}
\CSLBlock{DOI: \href{https://doi.org/10.1186/s12920-019-0627-z}{10.1186/s12920-019-0627-z} · PMID: \href{https://www.ncbi.nlm.nih.gov/pubmed/31865919}{31865919} · PMCID: \href{https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6927107}{PMC6927107}}}

\leavevmode\vadjust pre{\hypertarget{ref-19E33rJiu}{}}%
\CSLLeftMargin{23. }%
\CSLRightInline{\textbf{Deep mining heterogeneous networks of biomedical linked data to predict novel drug--target associations}
\CSLBlock{Nansu Zong, Hyeoneui Kim, Victoria Ngo, Olivier Harismendy} \emph{Bioinformatics} (2017-04-18) \url{https://doi.org/gbqjgx}
\CSLBlock{DOI: \href{https://doi.org/10.1093/bioinformatics/btx160}{10.1093/bioinformatics/btx160} · PMID: \href{https://www.ncbi.nlm.nih.gov/pubmed/28430977}{28430977} · PMCID: \href{https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5860112}{PMC5860112}}}

\leavevmode\vadjust pre{\hypertarget{ref-Dbie18xe}{}}%
\CSLLeftMargin{24. }%
\CSLRightInline{\textbf{Explaining and Suggesting Relatedness in Knowledge Graphs}
\CSLBlock{Giuseppe Pirrò} \emph{The Semantic Web - ISWC 2015} (2015) \url{https://doi.org/gg5nkd}
\CSLBlock{DOI: \href{https://doi.org/10.1007/978-3-319-25007-6_36}{10.1007/978-3-319-25007-6\_36}}}

\leavevmode\vadjust pre{\hypertarget{ref-GPBbIfq9}{}}%
\CSLLeftMargin{25. }%
\CSLRightInline{\textbf{FAIRY: A Framework for Understanding Relationships Between Users\textquotesingle{} Actions and their Social Feeds}
\CSLBlock{Azin Ghazimatin, Rishiraj Saha Roy, Gerhard Weikum} \emph{Proceedings of the Twelfth ACM International Conference on Web Search and Data Mining} (2019-01-30) \url{https://doi.org/gf2wqj}
\CSLBlock{DOI: \href{https://doi.org/10.1145/3289600.3290990}{10.1145/3289600.3290990}}}

\leavevmode\vadjust pre{\hypertarget{ref-L3DO8oq1}{}}%
\CSLLeftMargin{26. }%
\CSLRightInline{\textbf{Using Knowledge Graphs to Explain Entity Co-occurrence in Twitter}
\CSLBlock{Yiwei Wang, Mark James Carman, Yuan-Fang Li} \emph{Proceedings of the 2017 ACM on Conference on Information and Knowledge Management} (2017-11-06) \url{https://doi.org/gg5nkf}
\CSLBlock{DOI: \href{https://doi.org/10.1145/3132847.3133161}{10.1145/3132847.3133161}}}

\leavevmode\vadjust pre{\hypertarget{ref-Hf24Hzrl}{}}%
\CSLLeftMargin{27. }%
\CSLRightInline{\textbf{ESPRESSO: Explaining Relationships between Entity Sets}
\CSLBlock{Stephan Seufert, Klaus Berberich, Srikanta J Bedathur, Sarath Kumar Kondreddi, Patrick Ernst, Gerhard Weikum} \emph{Proceedings of the 25th ACM International on Conference on Information and Knowledge Management} (2016-10-24) \url{https://doi.org/gcpx7w}
\CSLBlock{DOI: \href{https://doi.org/10.1145/2983323.2983778}{10.1145/2983323.2983778}}}

\leavevmode\vadjust pre{\hypertarget{ref-19H7j1WBd}{}}%
\CSLLeftMargin{28. }%
\CSLRightInline{\textbf{MetaExp: Interactive Explanation and Exploration of Large Knowledge Graphs}
\CSLBlock{Freya Behrens, Fatemeh Aghaei, Emmanuel Müller, Martin Preusse, Nikola Müller, Michael Hunger, Sebastian Bischoff, Pius Ladenburger, Julius Rückin, Laurenz Seidel, \ldots{} Davide Mottin} \emph{WWW \textquotesingle18: Companion Proceedings of the The Web Conference 2018} (2018-04-23) \url{https://doi.org/gg2c7w}
\CSLBlock{DOI: \href{https://doi.org/10.1145/3184558.3186978}{10.1145/3184558.3186978}}}

\leavevmode\vadjust pre{\hypertarget{ref-Vz9QIDKV}{}}%
\CSLLeftMargin{29. }%
\CSLRightInline{\textbf{Discovering Meta-Paths in Large Heterogeneous Information Networks}
\CSLBlock{Changping Meng, Reynold Cheng, Silviu Maniu, Pierre Senellart, Wangda Zhang} \emph{Proceedings of the 24th International Conference on World Wide Web} (2015-05-18) \url{https://doi.org/gg2c7v}
\CSLBlock{DOI: \href{https://doi.org/10.1145/2736277.2741123}{10.1145/2736277.2741123}}}

\leavevmode\vadjust pre{\hypertarget{ref-KQNSDt1L}{}}%
\CSLLeftMargin{30. }%
\CSLRightInline{\textbf{Design and application of a knowledge network for automatic prioritization of drug mechanisms}
\CSLBlock{Michael Mayers, Roger Tu, Dylan Steinecke, Tong Shu Li, Núria Queralt-Rosinach, Andrew I Su} \emph{Bioinformatics} (2022-04-06) \url{https://doi.org/gptwsz}
\CSLBlock{DOI: \href{https://doi.org/10.1093/bioinformatics/btac205}{10.1093/bioinformatics/btac205} · PMID: \href{https://www.ncbi.nlm.nih.gov/pubmed/35561182}{35561182} · PMCID: \href{https://www.ncbi.nlm.nih.gov/pmc/articles/PMC9113361}{PMC9113361}}}

\leavevmode\vadjust pre{\hypertarget{ref-17mGFvxXu}{}}%
\CSLLeftMargin{31. }%
\CSLRightInline{\textbf{Transforming DWPCs for hetnet edge prediction}
\CSLBlock{Daniel Himmelstein, Pouya Khankhanian, Antoine Lizee} \emph{ThinkLab} (2016-04-01) \url{https://doi.org/f3qbmd}
\CSLBlock{DOI: \href{https://doi.org/10.15363/thinklab.d193}{10.15363/thinklab.d193}}}

\leavevmode\vadjust pre{\hypertarget{ref-1693YErZj}{}}%
\CSLLeftMargin{32. }%
\CSLRightInline{\textbf{Decomposing the DWPC to assess intermediate node or edge contributions}
\CSLBlock{Daniel Himmelstein} \emph{ThinkLab} (2016-12-15) \url{https://doi.org/gbr42h}
\CSLBlock{DOI: \href{https://doi.org/10.15363/thinklab.d228}{10.15363/thinklab.d228}}}

\leavevmode\vadjust pre{\hypertarget{ref-1FsruosUW}{}}%
\CSLLeftMargin{33. }%
\CSLRightInline{\textbf{Human Disease Ontology 2018 update: classification, content and workflow expansion}
\CSLBlock{Lynn M Schriml, Elvira Mitraka, James Munro, Becky Tauber, Mike Schor, Lance Nickle, Victor Felix, Linda Jeng, Cynthia Bearer, Richard Lichenstein, \ldots{} Carol Greene} \emph{Nucleic Acids Research} (2018-11-08) \url{https://doi.org/ggx9wp}
\CSLBlock{DOI: \href{https://doi.org/10.1093/nar/gky1032}{10.1093/nar/gky1032} · PMID: \href{https://www.ncbi.nlm.nih.gov/pubmed/30407550}{30407550} · PMCID: \href{https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6323977}{PMC6323977}}}

\leavevmode\vadjust pre{\hypertarget{ref-203l6hC3}{}}%
\CSLLeftMargin{34. }%
\CSLRightInline{\textbf{Unifying disease vocabularies}
\CSLBlock{Daniel Himmelstein, Tong Shu Li} \emph{ThinkLab} (2015-03-30) \url{https://doi.org/f3mqv5}
\CSLBlock{DOI: \href{https://doi.org/10.15363/thinklab.d44}{10.15363/thinklab.d44}}}

\leavevmode\vadjust pre{\hypertarget{ref-yqoUY2f9}{}}%
\CSLLeftMargin{35. }%
\CSLRightInline{\textbf{A review of biomedical datasets relating to drug discovery: a knowledge graph perspective}
\CSLBlock{Stephen Bonner, Ian P Barrett, Cheng Ye, Rowan Swiers, Ola Engkvist, Andreas Bender, Charles Tapley Hoyt, William L Hamilton} \emph{Briefings in Bioinformatics} (2022-09-23) \url{https://doi.org/gqv3s3}
\CSLBlock{DOI: \href{https://doi.org/10.1093/bib/bbac404}{10.1093/bib/bbac404} · PMID: \href{https://www.ncbi.nlm.nih.gov/pubmed/36151740}{36151740}}}

\leavevmode\vadjust pre{\hypertarget{ref-HUodcdz1}{}}%
\CSLLeftMargin{36. }%
\CSLRightInline{\textbf{Wikidata as a knowledge graph for the life sciences}
\CSLBlock{Andra Waagmeester, Gregory Stupp, Sebastian Burgstaller-Muehlbacher, Benjamin M Good, Malachi Griffith, Obi L Griffith, Kristina Hanspers, Henning Hermjakob, Toby S Hudson, Kevin Hybiske, \ldots{} Andrew I Su} \emph{eLife} (2020-03-17) \url{https://doi.org/ggqqc6}
\CSLBlock{DOI: \href{https://doi.org/10.7554/elife.52614}{10.7554/elife.52614} · PMID: \href{https://www.ncbi.nlm.nih.gov/pubmed/32180547}{32180547} · PMCID: \href{https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7077981}{PMC7077981}}}

\leavevmode\vadjust pre{\hypertarget{ref-eKVkBSDD}{}}%
\CSLLeftMargin{37. }%
\CSLRightInline{\textbf{SemMedDB: a PubMed-scale repository of biomedical semantic predications}
\CSLBlock{H Kilicoglu, D Shin, M Fiszman, G Rosemblat, TC Rindflesch} \emph{Bioinformatics} (2012-10-08) \url{https://doi.org/f4hp3x}
\CSLBlock{DOI: \href{https://doi.org/10.1093/bioinformatics/bts591}{10.1093/bioinformatics/bts591} · PMID: \href{https://www.ncbi.nlm.nih.gov/pubmed/23044550}{23044550} · PMCID: \href{https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3509487}{PMC3509487}}}

\leavevmode\vadjust pre{\hypertarget{ref-reekunFL}{}}%
\CSLLeftMargin{38. }%
\CSLRightInline{\textbf{Constructing Biomedical Knowledge Graph Based on SemMedDB and Linked Open Data}
\CSLBlock{Qing Cong, Zhiyong Feng, Fang Li, Li Zhang, Guozheng Rao, Cui Tao} \emph{2018 IEEE International Conference on Bioinformatics and Biomedicine (BIBM)} (2018-12) \url{https://doi.org/ggzb26}
\CSLBlock{DOI: \href{https://doi.org/10.1109/bibm.2018.8621568}{10.1109/bibm.2018.8621568}}}

\leavevmode\vadjust pre{\hypertarget{ref-uTOtIsTj}{}}%
\CSLLeftMargin{39. }%
\CSLRightInline{\textbf{Time-resolved evaluation of compound repositioning predictions on a text-mined knowledge network}
\CSLBlock{Michael Mayers, Tong Shu Li, Núria Queralt-Rosinach, Andrew I Su} \emph{BMC Bioinformatics} (2019-12) \url{https://doi.org/ggpcsr}
\CSLBlock{DOI: \href{https://doi.org/10.1186/s12859-019-3297-0}{10.1186/s12859-019-3297-0} · PMID: \href{https://www.ncbi.nlm.nih.gov/pubmed/31829175}{31829175} · PMCID: \href{https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6907279}{PMC6907279}}}

\leavevmode\vadjust pre{\hypertarget{ref-1BHNoKOV1}{}}%
\CSLLeftMargin{40. }%
\CSLRightInline{\textbf{The scalable precision medicine open knowledge engine (SPOKE): a massive knowledge graph of biomedical information}
\CSLBlock{John H Morris, Karthik Soman, Rabia E Akbas, Xiaoyuan Zhou, Brett Smith, Elaine C Meng, Conrad C Huang, Gabriel Cerono, Gundolf Schenk, Angela Rizk-Jackson, \ldots{} Sergio E Baranzini} \emph{Bioinformatics} (2023-02-01) \url{https://doi.org/grvqrf}
\CSLBlock{DOI: \href{https://doi.org/10.1093/bioinformatics/btad080}{10.1093/bioinformatics/btad080} · PMID: \href{https://www.ncbi.nlm.nih.gov/pubmed/36759942}{36759942} · PMCID: \href{https://www.ncbi.nlm.nih.gov/pmc/articles/PMC9940622}{PMC9940622}}}

\leavevmode\vadjust pre{\hypertarget{ref-bniVH7q2}{}}%
\CSLLeftMargin{41. }%
\CSLRightInline{\textbf{RTX-KG2: a system for building a semantically standardized knowledge graph for translational biomedicine}
\CSLBlock{EC Wood, Amy K Glen, Lindsey G Kvarfordt, Finn Womack, Liliana Acevedo, Timothy S Yoon, Chunyu Ma, Veronica Flores, Meghamala Sinha, Yodsawalai Chodpathumwan, \ldots{} Stephen A Ramsey} \emph{BMC Bioinformatics} (2022-09-29) \url{https://doi.org/gqxmkb}
\CSLBlock{DOI: \href{https://doi.org/10.1186/s12859-022-04932-3}{10.1186/s12859-022-04932-3} · PMID: \href{https://www.ncbi.nlm.nih.gov/pubmed/36175836}{36175836} · PMCID: \href{https://www.ncbi.nlm.nih.gov/pmc/articles/PMC9520835}{PMC9520835}}}

\leavevmode\vadjust pre{\hypertarget{ref-WkPlH1ds}{}}%
\CSLLeftMargin{42. }%
\CSLRightInline{\textbf{Heterogeneous Network Edge Prediction: A Data Integration Approach to Prioritize Disease-Associated Genes}
\CSLBlock{Daniel S Himmelstein, Sergio E Baranzini} \emph{PLOS Computational Biology} (2015-07-09) \url{https://doi.org/98q}
\CSLBlock{DOI: \href{https://doi.org/10.1371/journal.pcbi.1004259}{10.1371/journal.pcbi.1004259} · PMID: \href{https://www.ncbi.nlm.nih.gov/pubmed/26158728}{26158728} · PMCID: \href{https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4497619}{PMC4497619}}}

\leavevmode\vadjust pre{\hypertarget{ref-kNYoxBiQ}{}}%
\CSLLeftMargin{43. }%
\CSLRightInline{\textbf{Using the neo4j graph database for hetnets}
\CSLBlock{Daniel Himmelstein} \emph{ThinkLab} (2015-10-02) \url{https://doi.org/f3mqvk}
\CSLBlock{DOI: \href{https://doi.org/10.15363/thinklab.d112}{10.15363/thinklab.d112}}}

\leavevmode\vadjust pre{\hypertarget{ref-xWWaK51z}{}}%
\CSLLeftMargin{44. }%
\CSLRightInline{\textbf{Estimating the complexity of hetnet traversal}
\CSLBlock{Daniel Himmelstein, Antoine Lizee} \emph{ThinkLab} (2016-03-22) \url{https://doi.org/gbr42x}
\CSLBlock{DOI: \href{https://doi.org/10.15363/thinklab.d187}{10.15363/thinklab.d187}}}

\leavevmode\vadjust pre{\hypertarget{ref-w6rL3OYz}{}}%
\CSLLeftMargin{45. }%
\CSLRightInline{\textbf{Path exclusion conditions}
\CSLBlock{Daniel Himmelstein} \emph{ThinkLab} (2015-12-08) \url{https://doi.org/gg2rw2}
\CSLBlock{DOI: \href{https://doi.org/10.15363/thinklab.d134}{10.15363/thinklab.d134}}}

\leavevmode\vadjust pre{\hypertarget{ref-15nwuvjrA}{}}%
\CSLLeftMargin{46. }%
\CSLRightInline{\textbf{Vagelos Report Summer 2017}
\CSLBlock{Michael Zietz} \emph{figshare} (2017) \url{https://doi.org/gbr3pf}
\CSLBlock{DOI: \href{https://doi.org/10.6084/m9.figshare.5346577}{10.6084/m9.figshare.5346577}}}

\leavevmode\vadjust pre{\hypertarget{ref-IJWsN67g}{}}%
\CSLLeftMargin{47. }%
\CSLRightInline{\textbf{GitHub - mmayers12/hetnet\_ml: Software to quickly extract features from heterogeneous networks for machine learning.}
\CSLBlock{GitHub} \url{https://github.com/mmayers12/hetnet_ml}}

\leavevmode\vadjust pre{\hypertarget{ref-iKOIEzQ9}{}}%
\CSLLeftMargin{48. }%
\CSLRightInline{\textbf{Randomization Techniques for Graphs}
\CSLBlock{Sami Hanhijärvi, Gemma C Garriga, Kai Puolamäki} \emph{Proceedings of the 2009 SIAM International Conference on Data Mining} (2009-04-30) \url{https://doi.org/f3mn58}
\CSLBlock{DOI: \href{https://doi.org/10.1137/1.9781611972795.67}{10.1137/1.9781611972795.67}}}

\leavevmode\vadjust pre{\hypertarget{ref-GE6jhWnt}{}}%
\CSLLeftMargin{49. }%
\CSLRightInline{\textbf{Assessing the effectiveness of our hetnet permutations}
\CSLBlock{Daniel Himmelstein} \emph{ThinkLab} (2016-02-25) \url{https://doi.org/f3mqt5}
\CSLBlock{DOI: \href{https://doi.org/10.15363/thinklab.d178}{10.15363/thinklab.d178}}}

\leavevmode\vadjust pre{\hypertarget{ref-83EvMSI6}{}}%
\CSLLeftMargin{50. }%
\CSLRightInline{\textbf{The probability of edge existence due to node degree: a baseline for network-based predictions}
\CSLBlock{Michael Zietz, Daniel S Himmelstein, Kyle Kloster, Christopher Williams, Michael W Nagle, Casey S Greene} \emph{bioRxiv} (2023-01-06) \url{https://doi.org/grk9gd}
\CSLBlock{DOI: \href{https://doi.org/10.1101/2023.01.05.522939}{10.1101/2023.01.05.522939} · PMID: \href{https://www.ncbi.nlm.nih.gov/pubmed/36711569}{36711569} · PMCID: \href{https://www.ncbi.nlm.nih.gov/pmc/articles/PMC9881952}{PMC9881952}}}

\leavevmode\vadjust pre{\hypertarget{ref-19OyKPQ4M}{}}%
\CSLLeftMargin{51. }%
\CSLRightInline{\textbf{Network Edge Prediction: Estimating the prior}
\CSLBlock{Antoine Lizee, Daniel Himmelstein} \emph{ThinkLab} (2016-04-14) \url{https://doi.org/f3qbmg}
\CSLBlock{DOI: \href{https://doi.org/10.15363/thinklab.d201}{10.15363/thinklab.d201}}}

\leavevmode\vadjust pre{\hypertarget{ref-pkJqqM2W}{}}%
\CSLLeftMargin{52. }%
\CSLRightInline{\textbf{Node connectivity measurements for Hetionet v1.0 metapaths}
\CSLBlock{Daniel Himmelstein, Michael Zietz, Kyle Kloster, Michael Nagle, Blair Sullivan, Casey Greene} \emph{Zenodo} (2018-11-06) \url{https://doi.org/cww7}
\CSLBlock{DOI: \href{https://doi.org/10.5281/zenodo.1435833}{10.5281/zenodo.1435833}}}

\leavevmode\vadjust pre{\hypertarget{ref-q6QyrHEZ}{}}%
\CSLLeftMargin{53. }%
\CSLRightInline{\textbf{A practical guide to methods controlling false discoveries in computational biology}
\CSLBlock{Keegan Korthauer, Patrick K Kimes, Claire Duvallet, Alejandro Reyes, Ayshwarya Subramanian, Mingxiang Teng, Chinmay Shukla, Eric J Alm, Stephanie C Hicks} \emph{Genome Biology} (2019-06-04) \url{https://doi.org/gf3ncd}
\CSLBlock{DOI: \href{https://doi.org/10.1186/s13059-019-1716-1}{10.1186/s13059-019-1716-1} · PMID: \href{https://www.ncbi.nlm.nih.gov/pubmed/31164141}{31164141} · PMCID: \href{https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6547503}{PMC6547503}}}

\leavevmode\vadjust pre{\hypertarget{ref-YuJbg3zO}{}}%
\CSLLeftMargin{54. }%
\CSLRightInline{\textbf{Open collaborative writing with Manubot}
\CSLBlock{Daniel S Himmelstein, Vincent Rubinetti, David R Slochower, Dongbo Hu, Venkat S Malladi, Casey S Greene, Anthony Gitter} \emph{PLOS Computational Biology} (2019-06-24) \url{https://doi.org/c7np}
\CSLBlock{DOI: \href{https://doi.org/10.1371/journal.pcbi.1007128}{10.1371/journal.pcbi.1007128} · PMID: \href{https://www.ncbi.nlm.nih.gov/pubmed/31233491}{31233491} · PMCID: \href{https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6611653}{PMC6611653}}}

\leavevmode\vadjust pre{\hypertarget{ref-3KJWwgoz}{}}%
\CSLLeftMargin{55. }%
\CSLRightInline{\textbf{greenelab/connectivity-search-manuscript repository: Manuscript source code for Hetnet Connectivity Search}
\CSLBlock{Daniel Himmelstein, Michael Zietz, Vincent Rubinetti, Kyle Kloster, Benjamin Heil, Faisal Alquaddoomi, Dongbo Hu, David Nicholson, Yun Hao, Blair Sullivan, \ldots{} Casey Greene} \emph{GitHub} (2023-03-05) \url{https://github.com/greenelab/connectivity-search-manuscript}}

\leavevmode\vadjust pre{\hypertarget{ref-eq4IX1LX}{}}%
\CSLLeftMargin{56. }%
\CSLRightInline{\textbf{greenelab/connectivity-search-analyses repository: Hetnet connectivity search research notebooks}
\CSLBlock{Daniel Himmelstein, Michael Zietz, Kyle Kloster, Benjamin Heil, Yun Hao, Gregory Way, Michael Nagle, Blair Sullivan, Casey Greene} \emph{GitHub} (2022-03-20) \url{https://github.com/greenelab/connectivity-search-analyses}}

\leavevmode\vadjust pre{\hypertarget{ref-ASx5cMWI}{}}%
\CSLLeftMargin{57. }%
\CSLRightInline{\textbf{greenelab/connectivity-search-backend repository: Django backend for hetnet connectivity search}
\CSLBlock{Daniel Himmelstein, Dongbo Hu, Faisal Alquaddoomi, Casey Greene} \emph{GitHub} (2023-03-30) \url{https://github.com/greenelab/connectivity-search-backend}}

\leavevmode\vadjust pre{\hypertarget{ref-CadAZTf2}{}}%
\CSLLeftMargin{58. }%
\CSLRightInline{\textbf{greenelab/connectivity-search-frontend repository: Frontend source code for Hetnet connectivity search}
\CSLBlock{Vincent Rubinetti, Daniel Himmelstein, Dongbo Hu, Casey Greene} \emph{GitHub} (2023-03-30) \url{https://github.com/greenelab/connectivity-search-frontend}}

\leavevmode\vadjust pre{\hypertarget{ref-gHk6T3vn}{}}%
\CSLLeftMargin{59. }%
\CSLRightInline{\textbf{hetio/hetmatpy repository: Python package for matrix storage and operations on hetnets}
\CSLBlock{Daniel Himmelstein, Michael Zietz, Kyle Kloster, Benjamin Heil, Gregory Way, Blair Sullivan, Casey Greene} \emph{GitHub} (2022-04) \url{https://github.com/hetio/hetmatpy}}

\leavevmode\vadjust pre{\hypertarget{ref-srnBddlB}{}}%
\CSLLeftMargin{60. }%
\CSLRightInline{\textbf{hetio/hetnetpy repository: Hetnets in Python}
\CSLBlock{Daniel Himmelstein, Benjamin Heil, Michael Zietz, Tong Shu Li, Kyle Kloster, Gregory Way, Casey Greene, Sergio Baranzini} \emph{GitHub} (2021-06) \url{https://github.com/hetio/hetnetpy}}

\leavevmode\vadjust pre{\hypertarget{ref-L9cBdipF}{}}%
\CSLLeftMargin{61. }%
\CSLRightInline{\textbf{hetio/hetionet: Data repository containing Hetionet downloads}
\CSLBlock{Daniel Himmelstein, Antoine Lizee, Faisal Alquaddoomi, Vincent Rubinetti, Dongbo Hu, Casey Greene, Sergio Baranzini} \emph{GitHub} (2023-04-04) \url{https://github.com/greenelab/connectivity-search-manuscript}}

\leavevmode\vadjust pre{\hypertarget{ref-12IQq3W3u}{}}%
\CSLLeftMargin{62. }%
\CSLRightInline{\textbf{hetio/het.io repository: Source code for het.io website}
\CSLBlock{Vincent Rubinetti, Daniel Himmelstein, Dongbo Hu, Sergio Baranzini, Casey Greene} \emph{GitHub} (2021-12-20) \url{https://github.com/hetio/het.io}}

\leavevmode\vadjust pre{\hypertarget{ref-hHRphzV0}{}}%
\CSLLeftMargin{63. }%
\CSLRightInline{\textbf{Supporting data for "Hetnet connectivity search provides rapid insights into how two biomedical entities are related"}
\CSLBlock{Himmelstein S Daniel, Zietz Michael, Rubinetti Vincent, Kloster Kyle, Heil J Benjamin, Alquaddoomi Faisal, Hu Dongbo, Nicholson N David, Hao Yun, Sullivan D Blair, \ldots{} Greene S Casey} \emph{GigaScience Database} (2023-05-12) \url{https://doi.org/gr77bv}
\CSLBlock{DOI: \href{https://doi.org/10.5524/102389}{10.5524/102389}}}

\end{CSLReferences}
