## Introduction {.page_break_before}

A *network* (also known as a [graph](https://en.wikipedia.org/wiki/Graph_(discrete_mathematics))) is a conceptual representation of a group of entities — called *nodes* — and the relationships between them — called *edges*.
Typically, a network has only one type of node and one type of edge.
But in many cases, it is necessary to be able to distinguish between different types of entities and relationships.

### Hetnets

A *hetnet* (short for **het**erogeneous information **net**work [@doi:10.15363/thinklab.d104]) is a network where nodes and edges have type.
The ability to differentiate between different types of entities and relationships allows a hetnet to accurately describe more complex data.
Hetnets are particularly useful in biomedicine, where it is important to capture the conceptual distinctions between various concepts,
such as genes and diseases, or upregulation and binding.

The types of nodes and edges in a hetnet are defined by a schema, referred to as a metagraph.
The metagraph consists of metanodes (types of nodes) and metaedges (types of edges).
Note that the prefix *meta* is used to refer to type (e.g. compound),
as opposed to a specific node/edge/path itself (e.g. acetaminophen).

### Hetionet

[Hetionet](https://het.io/about/) is a knowledge graph of human biology, disease, and medicine, integrating information from millions of studies and decades of research.
Hetionet v1.0 combines information from [29 public databases](https://git.dhimmel.com/rephetio-manuscript/#tbl:resources).
The network contains 47,031 nodes of [11 types](https://git.dhimmel.com/rephetio-manuscript/#tbl:metanodes) (Table @tbl:metanodes) and 2,250,197 edges of [24 types](https://git.dhimmel.com/rephetio-manuscript/#tbl:metaedges) (Figure {@fig:rephetio}A).

| Metanode | Abbr | Nodes | Description |
| --- | --- | --- | --- |
| Anatomy | A | 402 | Anatomical structures, excluding structures that are known not to be found in humans. From [Uberon](http://uberon.github.io/). |
| Biological Process | BP | 11381 | Larger processes or biological programs accomplished by multiple molecular activities. From [Gene Ontology](http://geneontology.org/). |
| Cellular Component | CC | 1391 | The locations relative to cellular structures in which a gene product performs a function. From [Gene Ontology](http://geneontology.org/). |
| Compound | C | 1552 | Approved small molecule compounds with documented chemical structures. From [DrugBank](https://www.drugbank.ca/). |
| Disease | D | 137 | Complex diseases, selected to be distinct and specific enough to be clinically relevant yet general enough to be well annotated. From [Disease Ontology](http://disease-ontology.org/). |
| Gene | G | 20945 | Protein-coding human genes. From [Entrez Gene](https://www.ncbi.nlm.nih.gov/gene). |
| Molecular Function | MF | 2884 | Activities that occur at the molecular level, such as "catalysis" or "transport". From [Gene Ontology](http://geneontology.org/). |
| Pathway | PW | 1822 | A series of actions among molecules in a cell that leads to a certain product or change in the cell. From [WikiPathways](https://www.wikipathways.org/index.php/WikiPathways), [Reactome](https://reactome.org/), and Pathway Interaction Database. |
| Pharmacologic Class | PC | 345 | "Chemical/Ingredient", "Mechanism of Action", and "Physiologic Effect" FDA class types. From [DrugCentral](http://drugcentral.org/). |
| Side Effect | SE | 5734 | Adverse drug reactions. From [SIDER](http://sideeffects.embl.de/)/[UMLS](https://www.nlm.nih.gov/research/umls/). |
| Symptom | S | 438 | Signs and Symptoms (i.e. clinical abnormalities that can indicate a medical condition). From the [MeSH ontology](https://www.nlm.nih.gov/mesh/meshhome.html). |

Table: **Node types in Hetionet**
The abbreviation, number of nodes, and description for each of the 11 metanodes in Hetionet v1.0.
{#tbl:metanodes}

Hetionet provides a foundation for building hetnet applications.
It unifies data from several different, disparate sources into a single, comprehensive, accessible, common-format network.
The database is publicly accessible without login at <https://neo4j.het.io>.
The Neo4j graph database enables querying Hetionet using the Cypher language,
which was designed to interact with networks where nodes and edges have both types and properties.

One limitation that restricts the applicability of Hetionet is incompleteness.
In many cases, Hetionet v1.0 includes only a subset of the nodes from a given resource.
For example, the Disease Ontology contains over 9,000 diseases [@doi:10.1093/nar/gky1032], while Hetionet includes only 137 diseases [@doi:10.15363/thinklab.d44].
Nodes were excluded to avoid redundant or overly specific nodes, while ensuring a minimum level of connectivity for compounds and diseases.
See the [Project Rephetio methods](https://git.dhimmel.com/rephetio-manuscript/#nodes) for more details [@rephetio].
Nonetheless, Hetionet v1.0 remains one of the most comprehensive and integrative networks that consolidates biomedical knowledge into a manageable number of node and edge types.
Other integrative resources, some still under development, include [Wikidata](https://www.wikidata.org) [@doi:10.7554/eLife.52614], [SemMedDB](https://skr3.nlm.nih.gov/SemMedDB/) [@doi:10.1093/bioinformatics/bts591; @doi:10.1109/BIBM.2018.8621568; @doi:10.1186/s12859-019-3297-0], [SPOKE](https://spoke.ucsf.edu/), and [DRKG](https://github.com/gnn4dr/DRKG).

### Rephetio

Project Rephetio is the name of the [study](https://git.dhimmel.com/rephetio-manuscript/) that created Hetionet and applied it repurpose drugs [@rephetio].
This project [predicted](https://het.io/repurpose/) the probability of drug efficacy for 209,168 compound–disease pairs.
The approach learned which types of paths occur more or less frequently between known treatments than non-treatments (Figure {@fig:rephetio}B).
To train the model, Rephetio created [PharmacotherapyDB](https://doi.org/10.6084/m9.figshare.3103054), a physician-curated catalog of 755 disease-modifying treatments [@doi:10.15363/thinklab.d182].

![
**A. Hetionet v1.0 metagraph.**
The types of nodes and edges in Hetionet.
\
**B. Supervised machine learning approach from Project Rephetio.**
This figure visualizes the feature matrix used by Project Rephetio to make supervised predictions.
Each row represents a compound–disease pair.
The top half of rows correspond to known treatments (i.e. positives), while the bottom half correspond to non-treatments (i.e. negatives, not known to be treatments in PharmacotherapyDB).
Here, an equal number of treatments and non-treatments are shown, but in reality the problem is heavily imbalanced.
Project Rephetio scaled models to assume a positive prevalence of 0.36% [@rephetio; @doi:10.15363/thinklab.d210].
Each column represents a metapath, labeled with its abbreviation.
\
Feature values are DWPCs (transformed and standardized), which assess the connectivity along the specified metapath between the specific compound and disease.
Green colored values indicate above-average connectivity,
whereas blue values indicate below average connectivity.
In general, positives have greater connectivity for the selected metapaths than negatives.
Rephetio used a logistic regression model to learn the effect of each type of connectivity (feature) on the likelihood that a compound treats a disease.
The model predicts whether a compound–disease pair is a treatment based on its features, but requires supervision in the form of known treatments.
<!-- info on this figure's creation in https://github.com/greenelab/connectivity-search-manuscript/issues/11 -->
](https://github.com/greenelab/connectivity-search-manuscript/raw/f98c3470a8bf8f40f5f6aed2794c6ea66b93b14b/content/media/rephetio/metagraph-and-features.png){#fig:rephetio width="100%" .white}

### Related Works

Copious amount of network studies are focused on determining whether two nodes are related.
Typically, this is accomplished by predicting whether two nodes share a link [@doi:10.1016/j.websem.2017.06.002] or by ranking node pairs based on a custom scoring function [@arxiv:1710.05980].
With the success of graph embedding methods [@arxiv:1607.00653; @doi:10.1007/978-3-319-55753-3_12; @doi:10.1186/s12859-019-2914-2; @doi:10.1145/3097983.3098036] majority of works are utilize graph embedding approaches to accomplish the above tasks.
<div color=red>
When diving deeper into these papers see if they need a gold standard for training. I'd imagine evaluating, but doesn't hurt to double check
</div>
Caveat to these studies is that they only state two nodes are related and they fail to provide an explanation as to how these nodes are related..

There have been attempts to provide an explanation as to why two nodes are related.
These studies used Learn to Rank algorithms to rank metapaths that best explain how two nodes are related [@doi:10.1145/3132847.3133161; @doi:10.1145/3289600.3290990; @doi:10.1007/978-3-319-25007-6_36].
<div color="red">
Provide a simple explanation for these studies/group studies together.
Next make sure you explicitly say that these algorithms can be used for hetetogeneous networks.
</div>
Another study used subsgraphs to provide an explanation on why two nodes are related [@doi:10.1007/978-3-319-25007-6_36].
These networks have been centered on social networks and have yet to enter the biological domain.


### Unsupervised connectivity search

## Results {.page_break_before}

### Connectivity Search Webapp

We created the connectivity search webapp available at <https://het.io/search/>.
The tool is free to use, without any login or authentication.
The purpose is let users quickly explore how any two nodes in Hetionet v1.0 might be related.
The workflow is based around showing the user the most important metapaths and paths for a pair of query nodes.

The design guides the user through selecting a source and target node (Figure {@fig:webapp}A).
The webapp returns metapaths, scored by whether they occurred more than expected based on network degree (Figure {@fig:webapp}B).
Users can proceed by requesting the specific paths for each metapath, which are placed in a unified table sorted according to their path score (Figure {@fig:webapp}C).
Finally, the webapp produces publication-ready visualizations containing user-selected paths (Figure {@fig:webapp}D).

![
**Using the connectivity search webapp to explore the pathophysiology of Alzheimer's disease.**
This figure shows an example user workflow for <https://het.io/search/>.
\
**A.**
The user selects two nodes.
Here, the user is interested in Alzheimer's disease, so [selects this](https://het.io/search/?source=17287) as the source node.
The user limits the target node search to metanodes relating to gene function.
The target node search box suggests nodes, sorted by the number of significant metapaths.
When the user types in the target node box, the matches reorder based on search word similarity.
Here, the user becomes interested in how the circadian rhythm might relate to Alzheimer's disease.
\
**B.**
The webapp returns metapaths between Alzheimer's disease and the circadian rhythm pathway.
The user unchecks "precomputed only" to compute results for all metapaths with length ≤ 3, not just those that surpass the database inclusion threshold.
The user sorts by adjusted _p_-value and [selects][alzheimer-metapaths] 7 of the top 10 metapaths.
\
**C.**
Paths for the selected metapaths are ordered by their path score.
The user selects 8 paths (1 from a subsequent page of results) to show in the graph visualization and highlights a single path involving _ARNT2_ for emphasis.
\
**D.**
A subgraph displays the previously selected paths.
The user improves on the automated layout by repositioning nodes.
Clicking an edge displays its properties, informing the user that association between Creutzfeldt-Jakob disease and _NPAS2_ was detected by GWAS.
<!--
  This example is also used at https://slides.com/dhimmel/rocky2019#/4
  More info at https://github.com/greenelab/connectivity-search-manuscript/issues/7
-->
](https://github.com/greenelab/connectivity-search-manuscript/raw/b290b4ad435553c7126867e0720f3112b4692809/content/media/webapp/v3/webapp.png){#fig:webapp width="100%" .white}

[alzheimer-metapaths]: https://het.io/search/?source=17287&target=7607&metapaths=DaGiGpPW%2CDdGiGpPW%2CDdGpPW%2CDlAeGpPW%2CDrDaGpPW%2CDrDuGpPW%2CDuGiGpPW&complete=

### Hetmatpy Package

We created the hetmatpy Python package,
available on [GitHub](https://github.com/hetio/hetmatpy) and [PyPI](https://pypi.org/project/hetmatpy/) under the permissive BSD-2-Clause Plus Patent License.
This package provides a matrix-based utilities for hetnets.

### DWPC null distribution

To assess connectivity between a source and target node, we use the DWPC (degree-weighted path count) metric.
The DWPC is similar to path count (number of paths between the source and target node along a given metapath), except that it downweights paths through high degree nodes.
Rather than using the raw DWPC for a source-metapath-target combination, we transform the DWPC across all source-target node pairs for a metapath to yield a distribution that is more compact and amenable to modeling [@doi:10.15363/thinklab.d193].

Previously, we had no technique for detecting whether a DWPC value was exceptional.
One possibility is to evaluate the DWPCs for all pairs of nodes and select the top scores (e.g. the top 5% of DWPCs).
Another possibility is to pick a transformed DWPC score as a cutoff.
The shortcomings of these methods are twofold.
First, neither the percentile nor absolute value of a DWPC has inherent meaning.
To select transformed DWPCs greater than 6, or alternatively the top 1% of DWPCs, is arbitrary.
Second, comparing DWPCs between node pairs fails to account for the situation where high-degree node pairs are likely to score higher, solely on account of their degree (TODO: figure).

To address these shortcomings, we developed a method to compute the right-tail _p_-value of a DWPC.
_p_-values have a broadly understood interpretation --- in our case, the probability that a DWPC equal to or greater than the observed DWPC could occur under a null model.
By tailoring the null distribution for a DWPC to the degree of its source and target node, we account for degree effects when determining the significance of a DWPC.


### Enriched metapaths

![
**Expanded metapath details from the connectivity search webapp.**
This is the expanded view of the [metapath table][alzheimer-metapaths] in {@fig:webapp}B.
](https://github.com/greenelab/connectivity-search-manuscript/raw/b290b4ad435553c7126867e0720f3112b4692809/content/media/webapp/v3/b.metapaths-expanded.png){#fig:webapp-metapaths width="100%}

Figure @fig:webapp-metapaths shows the information used to compute _p_-value for enriched metapaths.
The table includes the following columns:

- **path count**:
  The number of paths between the source and target node of the specified metapath
- **adjusted _p_-value**:
  A measure of the significance of the DWPC that indicates whether more paths were observed than expected due to random chance.
  Compares the DWPC to a null distribution of DWPCs generated from degree-preserving permuted networks.
  Bonferroni-adjusted for the number of metapaths with the same source metanode, target metanode, and length.
- **_p_-value**:
  A measure of the significance of the DWPC that indicates whether more paths were observed than expected due to random chance.
  Compares the DWPC to a null distribution of DWPCs generated from degree-preserving permuted networks.
  Not adjusted for multiple comparisons
  (i.e. when multiple metapaths are assessed for significant connectivity between the source and target node).
- **DWPC**:
  Degree-Weighted Path Count —
  Measures the extent of connectivity between the source and target node for the given metapath.
  Like the path count, but with less weight given to paths along high-degree nodes.
- **source degree**:
  The number of edges from the source node that are of the same type as the initial metaedge of the metapath.
- **target degree**:
  The number of edges from the target node that are of the same type as the final metaedge of the metapath.
- **# DWPCs**:
  The number of DWPCs calculated on permuted networks used to generate a null distribution for the DWPC from the real network.
  Permuted DWPCs are aggregated for all permuted node pairs with the same degrees as the source and target node.
- **# non-0 DWPCs**:
  The number of permuted DWPCs from '# of DWPCs' column that were nonzero.
  Nonzero DWPCs indicate at least one path between the source and target node existed in the permuted network.
- **non-0 mean**:
  The mean of nonzero permuted DWPCs.
  Used to generate the gamma-hurdle model of the null DWPC distribution.
- **non-0 σ**:
  The standard deviation of nonzero permuted DWPCs.
  Used to generate the gamma-hurdle model of the null DWPC distribution.
- **Neo4j Actions**:
  A Cypher query that users can run in the [Neo4j browser](https://neo4j.het.io/browser/) to show paths with the largest DWPCs for the metapath.

### Enriched paths

### Comparison to Rephetio

### Detecting Mechanisms of Action for Indications

Assess ability to predict paths in <https://github.com/SuLab/DrugMechDB>

### Use cases

## Discussion {.page_break_before}

STUB: Contributions of this work:

- search engine for hetnet connectivity between two nodes, realtime results
- interactive webapp and user interface for displaying metapaths, paths, and subgraphs.
- optimized methods for computing DWPCs using matrix multiplication
- method for estimating _p_-values for a DWPC, based on null DWPCs computed from permuted hetnets.
- the hetmatpy Python package and HetMat data structure that provide a highly-optimized computational infrastructure to make this possible.

STUB: Future work:

- node set transformations
- [improved DWPC scaling](https://github.com/hetio/hetmatpy/issues/11)
- longer metapaths
- auto-detection of informative metapaths

## Methods {.page_break_before}

### The HetMat awakens

At the core of the hetmatpy package is the HetMat data structure for storing and accessing the network.
HetMats are stored on disk as a directory, which by convention uses a `.hetmat` extension.
A HetMat directory stores a single heterogeneous network, whose data resides in the following files.

1. A `metagraph.json` file stores the schema, defining which types of nodes and edges comprise the hetnet.
   This format is defined by the [hetnetpy](https://github.com/hetio/hetnetpy) Python package.
   Hetnetpy was originally developed with the name hetio during prior studies
   [@hetio-dag; @rephetio],
   but we [renamed](https://github.com/hetio/hetnetpy/issues/40) it to het**net**py for better disambiguation from het**mat**py.
2. A `nodes` directory containing one file per node type (metanode) that defines each node.
   Currently, `.tsv` files where each row represents a node are supported.
3. An `edges` directory containing one file per edge type (metadata) that encodes the adjacency matrix.
   The matrix can be serialized using either the Numpy dense format (`.npy`) or SciPy sparse format (`.sparse.npz`).

For node and edge files, compression is supported as detected from `.gz`, `.bz2`, `.zip`, and `.xz` extensions.
This structure of storing a hetnet supports selectively reading nodes and edges into memory.
For example, a certain computation may only require access to a subset of the node and edge types.
By only loading the required node and edge types, we reduce memory usage and read times.

Additional subdirectories, such as `path-counts` and `permutations`, store data generated from the HetMat.
By using consistent paths for generated data, we avoid recomputing data that already exists on disk.
A HetMat directory can be zipped for archiving and transfer.
Users can selectively include generated data in archives.
Since the primary application of HetMats is to generate computationally demanding measurements on hetnets, the ability to share HetMats with precomputed data is paramount.

The [`HetMat`](https://hetio.github.io/hetmatpy/reference/hetmatpy/hetmat/#hetmat) class implements the above logic.
A `hetmat_from_graph` function creates a HetMat object and directory on disk from the pre-existing `hetnetpy.hetnet.Graph` format.

We converted Hetionet v1.0 to HetMat format and uploaded the `hetionet-v1.0.hetmat.zip` archive to the [Hetionet data repository](https://github.com/hetio/hetionet/tree/master/hetnet/matrix).

### DWPC matrix multiplication algorithms

Prior to this study, we used two implementations for computing DWPCs.
The first is a pure Python implementation available in the [`hetnetpy.pathtools.DWPC`](https://github.com/hetio/hetnetpy/blob/aa16e6a7092c039a6b175a73a35c006e53acee20/hetnetpy/pathtools.py#L8-L21) function [@hetio-dag].
The second uses a Cypher query, prepared by [`hetnetpy.neo4j.construct_dwpc_query`](https://github.com/hetio/hetnetpy/blob/aa16e6a7092c039a6b175a73a35c006e53acee20/hetnetpy/neo4j.py#L363-L393), that is executed by the Neo4j database [@rephetio; @doi:10.15363/thinklab.d112].
Both of these implementations require traversing all paths between the source and target node.
Hence, they are computationally cumbersome despite optimizations [@doi:10.15363/thinklab.d187].

Since our methods only require degree-weighted counts, not fully enumerated paths, adjacency matrix multiplication presents an alternative approach.
Multiplication alone, however, counts walks rather than paths, meaning paths traversing a single node multiple times are counted.
When computing network-based features to quantify the relationship between a source and target node,
we would like to exclude traversing duplicate nodes (i.e. paths, not trails nor walks) [@doi:10.15363/thinklab.d134].
To benefit from the speed advantages of only counting paths, we developed a suite of algorithms to compute true path counts and DWPCs using matrix multiplication.

Our implementation begins by categorizing a metapath according to the pattern of its repeated metanodes, allowing DWPC computation using a specialized order of operations.
For example, the metapath _DrDtCrC_  is categorized as a set of disjoint repeats, while _DtCtDpC_ is categorized as repeats of the form BABA.
Many complex repeat patterns can be represented piecewise as simpler patterns, allowing us to compute DWPC for most metapaths up to length 5 and many of length 6 and beyond without enumerating individual paths.
For example, disjoint groups of repeats like _DrDtCrC_ can be computed as the matrix product of DWPC matrices for _DrD_ and _CrC_.
Randomly-inserted non-repeated metanodes (e.g. _G_ in _DrDaGaDrD_) require no special treatment, and are included in DWPC with a simple matrix multiplication.

After metapath categorization, we segment metapaths according to their repeat pattern, following our order of operations.
By segmenting and computing recursively, we can evalute DWPC efficiently on highly complex metapaths, using simple patterns as building-blocks for higher-level patterns.
Finally, our specialized DWPC functions are applied to individual segments, the results are combined, and final corrections are made to ensure no repeated nodes are counted.
The recursive, segmented approach we developed allowed us additionally to implement a caching strategy that improved speed by avoiding duplicate DWPC computations.
In summary, the functionality we developed resulted in greater than a 175-fold reduction in compute time, allowing us to compute millions of DWPC values across Hetionet [@vagelos-2017].

#### Details of matrix DWPC implementation

DWPC computation requires us to remove all duplicate nodes from paths.
We used three repeat patterns as the building blocks for DWPC computation: short repeats (AAA), nested repeats (BAAB), and overlapping repeats (BABA).
Let $\mathrm{D}(\mathit{XwXyZ})$ denote the DWPC matrix for metapath _XwXyZ_.
Under this notation, $\mathrm{D}(\mathit{XyZ})$ is the degree-weighted (bi)adjacency matrix for metaedge _XyZ_.
Additionally, let $\mathrm{diag}(A)$ represent a diagonal matrix whose entries are the diagonal elements of $A$.

For the case of short (< 4) repeats for a single metanode, _XaXbX_ (e.g. _GiGdG_), we simply subtract the main diagonal.

$$\mathrm{D}(\mathit{XaXbX}) = \mathrm{D}(\mathit{XaX}) \mathrm{D}(\mathit{XbX}) - \mathrm{diag}(\mathrm{D}(\mathit{XaX}) \mathrm{D}(\mathit{XbX}))$$

Nested repeats _XaYbYcX_ (e.g. _CtDrDtC_), are treated recursively, with both inner (YY) and outer (XX) repeats treated as separate short repeats.

$$
\mathrm{D}(\mathit{XaYbYcX}) = \mathrm{D}(\mathit{XaY}) \mathrm{D}(\mathit{YbY}) \mathrm{D}(\mathit{YcX}) - \mathrm{diag}(\mathrm{D}(\mathit{XaY}) (\mathrm{D}(\mathit{YbY}) \mathrm{D}(\mathit{YcX}))
$$

Overlapping repeats _XaYbXcY_ (e.g. _CtDtCtD_) require several corrections ($\odot$ denotes the Hadamard product).

\begin{align}
\mathrm{D}(\mathit{XaYbXcY}) =\ &\mathrm{D}(\mathit{XaY})\  \mathrm{D}(\mathit{YbX})\  \mathrm{D}(\mathit{XcY}) \\
&- \mathrm{diag}(\mathrm{D}(\mathit{XaY})\  \mathrm{D}(\mathit{YbX}))\  \mathrm{D}(\mathit{XcY}) \\
&- \mathrm{D}(\mathit{XaY})\  \mathrm{diag}(\mathrm{D}(\mathit{YbX})\  \mathrm{D}(\mathit{XcY})) \\
&+ \mathrm{D}(\mathit{XaY})\  \odot \mathrm{D}(\mathit{YbX})^T\  \odot \mathrm{D}(\mathit{XcY})
\end{align}

Most paths of length six---and many even longer paths---can be represented hierarchically using these patterns.
For example, a long metapath pattern of the form CBABACXYZ can be segmented as (C(BABA)C)XYZ using patterns for short and overlapping repeats and can be computed using the tools we developed.
In addition to these matrix routines---which advantageously count rather than enumerate paths---we implemented a general matrix method for any metapath type.
The general method is important for patterns such as long (≥ 4) repeats, or complex repeat patterns (e.g. of the form ABCABC), but it requires path enumeration and is therefore slower.
As an alternative approach for complex paths, we developed an approximate DWPC method that corrects repeats in disjoint simple patterns but only corrects the first repeat in complex patterns (e.g. ≥ length four repeat).
Mayers et al. developed an alternative approximation, which subtracts the main diagonal at every occurrence of the first repeated metanode [@url:https://github.com/mmayers12/hetnet_ml].
All our matrix methods were validated against existing implementations involving explicit path enumeration to ensure consistent results.


<!-- Discuss caching strategies, sequential versus recursive
Runtime comparison to show the speedup.
Rephetio computed a portion of in XX time -->


<!-- From [@vagelos-2017]:
> We reduced the time to compute DWPC over nearly 1200 metapaths from roughly four-and-a-half days to roughly one hour and thirty-seven minutes
175-fold, which underestimates since Rephetio did not compute the full DWPC matrix and benefited from concurrency. -->

<!-- references
  https://github.com/greenelab/hetmech/issues/20
  https://github.com/greenelab/hetmech/pull/70
  https://github.com/greenelab/hetmech/issues/53
  https://github.com/greenelab/hetmech/pull/67
  https://github.com/hetio/hetmatpy/blob/bc36aa9859c43a1a5fb22808cd6eb952ef9d497c/hetmatpy/degree_weight.py#L210-L239
  https://github.com/mmayers12/hetnet_ml
  https://nbviewer.jupyter.org/github/greenelab/hetmech/blob/042063fb559048e52b3dc2731b6d6c6836f698cf/7.rephetio-times.ipynb
  https://nbviewer.jupyter.org/github/greenelab/hetmech/blob/042063fb559048e52b3dc2731b6d6c6836f698cf/explore/recursive-chain/dwwc-chain-recursive.ipynb
  https://nbviewer.jupyter.org/github/greenelab/hetmech/blob/042063fb559048e52b3dc2731b6d6c6836f698cf/explore/cache-speeds.ipynb
-->

### Permuted hetnets

In order to generate a null distribution for a DWPC, we rely on DWPCs computed from permuted hetnets.
We derive permuted hetnets from the unpermuted network using the XSwap algorithm [@doi:10.1137/1.9781611972795.67].
XSwap randomizes edges while preserving node degree.
Therefore, it's ideal for generating null distributions that retain general degree effects,
but destroy the actual meaning of edges.
We adapt XSwap to hetnets by applying it separately to each metaedge [@rephetio; @doi:10.15363/thinklab.d178; @xswap].

Project Rephetio created 5 permuted hetnets [@rephetio; @doi:10.15363/thinklab.d178],
which were used to generate a null distribution of classifier performance for each metapath-based feature.
Here, we aim to create a null distribution for individual DWPCs, which requires vastly more permuted values to estimate with accuracy.
Therefore, we generated 200 permuted hetnets ([archive](https://github.com/hetio/hetionet/tree/a95ae76581af604e91d744680aee3f888fa18887/hetnet/permuted/matrix)).
<!-- Permuted hetmats generated in PR https://github.com/greenelab/hetmech/pull/127 -->
More recently, we also developed the `xswap` Python [package](https://github.com/greenelab/xswap), whose optimized C/C++ implementation will enable future research to generate even larger sets of permuted networks [@xswap].

### Degree-grouping of node pairs

For each of the 200 permuted networks and each of the 2,205 metapaths, we computed the entire DWPC matrix (i.e. all source nodes × target nodes).
Therefore, for each actual DWPC value, we computed 200 permuted DWPC values.
Because permutation preserves only node degree, DWPC values among nodes with the same source and target degrees are equivalent to additional permutations.
We greatly increased the effective number of permutations by grouping DWPC values according to node degree, affording us a superior estimation of the DWPC null distribution.

We have applied this _degree-grouping_ approach previously when calculating the prior probability of edge existence based on the source and target node degrees [@doi:10.15363/thinklab.d201; @xswap].
But here, we apply _degree-grouping_ to null DWPCs.
The result is that the null distribution for a DWPC is based not only on permuted DWPCs for the corresponding source--metapath--target combination,
but instead on all permuted DWPCs for the source-degree--metapath--target-degree combination.

The "# DWPCs" column in Figure @fig:webapp-metapaths illustrates how degree-grouping inflates the sample size of null DWPCs.
The _p_-value for the _DaGiGpPW_ metapath relies on the minimum number of null DWPCs (200),
since no other disease besides Alzheimer's had 196 _associates_ edges (source degree) and no other pathway besides circadian rhythm had 201 _participates_ edges (target degree).
However, for other metapaths with over 5,000 null DWPCs, degree-grouping increased the size of the null distribution by a factor of 25.
In general, source--target node pairs with lower degrees receive the largest sample size multiplier from degree-grouping.
This is convenient since low degree nodes also tend to produce the highest proportion of zero DWPCs, by virtue of low connectivity.
Consequently, degree-grouping excels where it is needed most.

One final benefit of degree-grouping is that reduces the disk space required to store null DWPC summary statistics.
For example, with 20,945 genes in Hetionet v1.0, there exists 438,693,025 gene pairs.
Gene nodes have 302 distinct degrees for _interacts_ edges, resulting in 91,204 degree pairs.
This equates to an 4,810-fold reduction in the number of summary statistics that need to be stored to represent the null DWPC distribution for a metapath starting and ending with a _Gene--interacts--Gene_ metaedge.

We store the following null DWPC summary statistics for each metapath--source-degree--target-degree combination:
total number of null DWPCs,
total number of nonzero null DWPCs,
sum of null DWPCs,
sum of squared null DWPCs,
and number of permuted hetnets.
These values are sufficient to estimate the _p_-value for a DWPC, by either [fitting](https://github.com/hetio/hetmatpy/blob/bc36aa9859c43a1a5fb22808cd6eb952ef9d497c/hetmatpy/pipeline.py#L42-L63 "hetmatpy.pipeline.add_gamma_hurdle_to_dgp_df source code") a gamma-hurdle null distribution or generating an empiric _p_-value.
Furthermore, these statistics are additive across permuted hetnets.
Their values are always a running total and can be updated incrementally as statistics from each additional permuted hetnet become available.

<!--
Should note:
Permuted DWPCs were scaled by dividing by the mean of all unpermuted DWPCs for the metapath and then applying the inverse hyperbolic sine transformation.
Every degree pair for a given metapath has corresponding statistics that summarize its values across permuted hetnets.

TODO: somewhere link to notebook https://github.com/greenelab/hetmech/blob/042063fb559048e52b3dc2731b6d6c6836f698cf/explore/bulk-pipeline/bulk.ipynb
-->

### Gamma-hurdle distribution

We are interested in identifying source and target nodes whose connectivity exceeds what typically arises at random.
To identify such especially-connected nodes, we compare DWPC values to the distribution of permuted network DWPC values for the same source and target nodes.
While a single DWPC value is not actually a test statistic, we use a framework akin to classical hypothesis testing to identify outliers.

Two observations led us to the quasi significance testing framework we developed.
First, a sizable fraction of permuted DWPC values are often zero, indicating that the source and target nodes are not connected along the metapath in the permuted network.
Second, we observed that non-zero DWPC values for any given source and target nodes are reasonably approximated as following a gamma distribution.
Motivated by these observations, we parametrized permuted DWPC values using a zero-inflated gamma distribution, which we termed the gamma-hurdle distribution.
We fit a gamma-hurdle distribution to each combination of source node, target node, and metapath.
Finally, we estimate the probability of observing a permuted DWPC value greater than DWPC computed in the unpermuted network, akin to a one-tailed p-value.
These quasi significance scores ('_p_-values') allow us to identify outlier node pairs at the metapath level (see examples in Figure @fig:null-dwpc-distributions).

![
**From null distribution to _p_-value for DWPCs.**
Null DWPC distributions are shown for 3 metapaths between Alzheimer's disease and the circadian rhythm pathway, selected from Figure @fig:fig:webapp-metapaths.
For each metapath, null DWPCs are computed on 200 permuted hetnets and grouped according to source--target degree.
Histograms show the null DWPCs for the degree group corresponding to Alzheimer's disease and the circadian rhythm pathway (as noted in the plot titles by deg.)
The proportion of null DWPCs that were zero is calculated, forming the "hurdle" of the null distribution model.
The nonzero null DWPCs are modeled using a gamma distribution, which can be fit solely from a sample mean and standard deviation.
The mean of nonzero null DWPCs is denoted with a diamond, with the standard deviation plotted twice as a line in either direction.
Actual DWPCs are compared to the gamma-hurdle null distribution to yield a _p_-value.
](https://github.com/greenelab/hetmech/raw/9287986f331607cfdbc1ac197b52f36085723c6e/explore/gamma-hurdle/gamma-hurdle-distributions.png){#fig:null-dwpc-distributions}

#### Details of the gamma-hurdle distribution

Let _X_ be a gamma-hurdle random variable with parameters _λ_, _α_, and _β_.

$$
X \sim \Gamma_H(\lambda, \alpha, \beta)
$$

The probability of a draw from the distribution is

\begin{align}
  P(X = 0) &= 1 - \lambda \\
  P(X \in A; A \subseteq (0, \infty)) &= \frac{\lambda \beta^\alpha}{\Gamma(\alpha)}  \int  _{x \in A} \bigg( x^{\alpha - 1} e^{-\beta x} \bigg)
\end{align}

We estimate all three parameters using the method of moments (using Bessel's correction to estimate the second moment).
As a validation of our method,
we [compared](https://nbviewer.jupyter.org/github/greenelab/hetmech/blob/025bdf8d5e63725ca2482d61fd8e421bf0001f93/explore/gamma-hurdle/parameter_estimates.ipynb) our method of moments parameter estimates to approximate maximum likelihood estimates
(gamma distribution parameters do not have closed-form maximum likelihood estimates)
and found excellent concordance between the methods.
Let *N* be the number of permuted DWPC values, and *n* the number of nonzero values.

\begin{align}
  \hat{\lambda} &= \frac{n}{N} \\
  \hat{\alpha} &= \frac{(n - 1) \sum x_i}{n \sum (x_i^2) - (\sum x_i)^2} \\
  \hat{\beta} &= \frac{n - 1}{n} \frac{(\sum x_i)^2}{n \sum (x_i)^2 - (\sum x_i)^2}
\end{align}

Finally, we compute a p-value for each DWPC value, *t*.

$$
p = P(X ≥ t) = \frac{\beta^\alpha}{\Gamma(\alpha)} \int_t^\infty x^{\alpha - 1} \exp(-\beta x) dx
$$

<!-- references
  https://nbviewer.jupyter.org/github/greenelab/hetmech/blob/025bdf8d5e63725ca2482d61fd8e421bf0001f93/explore/gamma-hurdle/gamma-heatmaps.ipynb
  https://nbviewer.jupyter.org/github/greenelab/hetmech/blob/0d83ec6063001d7b5cfcfa6a9fe3765bbe109aea/explore/gamma-hurdle/gamma-fit.ipynb
  https://github.com/greenelab/hetmech/pull/157
  https://github.com/greenelab/hetmech/issues/123
-->

### Empirical DWPC p-values

We [calculate](https://github.com/hetio/hetmatpy/blob/bc36aa9859c43a1a5fb22808cd6eb952ef9d497c/hetmatpy/pipeline.py#L92-L113 "hetmatpy.pipeline.calculate_empirical_p_value source code") an empirical p-value for special cases where the gamma-hurdle model cannot be applied.
These cases include when the observed DWPC is zero or when the null DWPC distribution is all zeroes or has only a single distinct nonzero value.
The empirical _p_-value (_p~empiric~_) equals the proportion of null DPWCs ≥ the observed DWPC.

Since we don't store all null DWPC values,
we apply the following criteria to calculate _p~empiric~_ from summary statistics:

1. When the observed DWPC = 0 (no paths of the specified metapath existed between the source and target node),
   _p~empiric~_ = 1.
2. When all null DWPCs are zero but the observed DWPC is positive,
   _p~empiric~_ = 0.
3. When all nonzero null DWPCs have the same positive value (standard deviation = 0),
   _p~empiric~_ = 0 if the observed DWPC > the null DWPC else _p~empiric~_ = proportion of nonzero null DWPCs.

### DWPC and null distribution computation

We decided to compute DWPCs and their significance for all source--target node pairs for metapaths with length ≤ 3.
On Hetionet v1.0, there are 24 metapaths of length 1, 242 metapaths of length 2, and 1,939 metapaths of length 3.
The decision to stop at length 3 was one of practicality, as length 4 would have added 17,511 metapaths.

For each of the 2,205 [metapaths](https://github.com/greenelab/hetmech/raw/042063fb559048e52b3dc2731b6d6c6836f698cf/explore/bulk-pipeline/archives/metapath-dwpc-stats.tsv),
we computed the complete path count matrix and DWPC matrix
([notebook](https://nbviewer.jupyter.org/github/greenelab/hetmech/blob/042063fb559048e52b3dc2731b6d6c6836f698cf/explore/bulk-pipeline/bulk.ipynb)).
In total, we computed 137,786,767,964 path counts (and the same number of DWPCs) on the unpermuted network,
of which 11.6% were nonzero.

The DWPC has a single parameter, called the damping exponent (_w_), which controls how much paths through high-degree nodes are downweighted [@hetio-dag].
When _w_ = 0, the DWPC is equivalent to the path count.
Previously, we found _w_ = 0.4 was optimal for predicting disease-associated genes [@hetio-dag, figure S3A].
Here, we use _w_ = 0.5, since taking the square root of degrees has more intuitive appeal.

We selected data types for matrix values that would allow for high precision.
We used 64-bit unsigned integers for path counts and 64-bit floating-point numbers for DWPCs.
We [considered](https://github.com/greenelab/hetmech/pull/91) using 16-bits or 32-bits per DWPC to reduce memory/storage size,
but decided against it in case certain applications required greater precision.

We used SciPy sparse for path count and DWPC matrices with density < 0.7, serialized to disk with compression and a `.sparse.npz` extension.
This format minimizes the space on disk and load time for the entire matrix, but does not offer read access to slices.
We used Numpy 2D arrays for DWPC matrices with density ≥ 0.7, serialized to disk using Numpy's `.npy` format.
We bundled the path count and DWPC matrix files into HetMat archives by metapath length and deposited the archives to Zenodo [@doi:10.5281/zenodo.1435834].
The archive for length 3 DWPCs was the largest at 131.7 GB.

We also generated null DWPC summary statistics for the 2,205 metapaths,
which are also available by metapath length from Zenodo as HetMat archives consisting of `.tsv.gz` files [@doi:10.5281/zenodo.1435834].
Due to degree-grouping, null DWPCs summary statistic archives are much smaller than the DWPC archives.
The archive for length 3 null DWPCs summary statistics was 733.1 MB.
However, the compute required to generate null DWPCs is far greater, because there are multiple permuted hetnets (in our case 200).
As a result, computing and saving all DWPCs took 6 hours,
whereas computing and saving the null DWPC summary statistics took 361 hours.

Including null DWPCs and path counts, the Zenodo deposit totals 185.1 GB and contains the results of computing ~28 trillion DWPCs --- 27,832,927,128,728 to be exact.

### Adjusting DWPC _p_-values

When a user applies hetnet connecitivity search to identify enriched metapaths between two nodes,
many metapaths are evaluated for significance.
Due to multiple testing of many DWPCs, low _p_-values are likely to arise by chance.
Therefore, we devised a multiple testing correction.

For each combination of source metanode, target metanode, and length, we counted the number of metapaths.
For Disease...Pathway metapaths, there are 0 metapaths of length 1, 3 metapaths of length 2, and 24 metapaths of length 3.
We calculated adjusted p-values by applying a Bonferroni correction based on the number of metapaths of the same length between the source and target metanode.
Using Figure @fig:webapp-metapaths as an example, the [DdGpPW](https://het.io/search/?source=17287&target=7607&metapaths=DdGpPW&complete=) p-value of 5.9% was adjusted to 17.8% (multiplied by a factor of 3).

Bonferroni controls familywise error rate,
which corresponds here to incorrectly finding that _any_ metapath of a given length is enriched.
As a result, our adjusted p-values are conservative.
We would prefer to adjust _p_-values for false discovery rate [@doi:10.1186/s13059-019-1716-1],
but these methods often require access to all _p_-values at once (impractical here) and assume a uniform distribution of _p_-values when there is no signal (not the case here when most DWPCs are zero).

### Prioritizing enriched metapaths for database storage

Storing DWPCs and their significance in the database (as part of the `PathCount` table in Figure @fig:database) enables the connectivity search webapp to provide users with enriched metapaths between query nodes in realtime.
However, storing ~15.9 billion rows (the total number of nonzero DWPCs) in the database's `PathCount` table would exceed a reasonable disk quota.
<!--
# python code to estimate the total number of nonzero DWPCs
url ="https://github.com/greenelab/hetmech/raw/042063fb559048e52b3dc2731b6d6c6836f698cf/explore/bulk-pipeline/archives/metapath-dwpc-stats.tsv"
metapath_df = pandas.read_table(url)
(metapath_df.n_pairs * metapath_df.pc_density).round().sum()
-->
An alternative would be to store all DWPCs in the database whose adjusted p-value exceeded a universal threshold (e.g. _p_ < 5%).
But we estimated this would still be prohibitively expensive.
Therefore, we devised a metapath-specific threshold.
For metapaths with length 1, we stored all nonzero DWPCs,
assuming users always want to be informed about direct edges between the query nodes, regardless of significance.
For metapaths with length ≥ 2, we chose an adjusted _p_-value threshold of
5 × (*n~source~* × *n~target~*)^−0.3^,
where *n~source~* and *n~target~* are the node counts for the source and target metanodes (i.e. "Nodes" column in Table @tbl:metanodes).
Notice that metapaths with large number of possible source--target pairs (large DWPC matrices) are penalized.
This decision is based on practicality, since otherwise the majority of the database quota would be consumed by a minority of metapaths between plentiful metanodes (e.g. Gene...Gene metapaths).
Also we assume that users will search nodes at a similar rate by metanode (e.g. they're more likely to search for a specific disease than a specific gene).
The constants in the threshold formula help scale it.
The multiplier of 5 relaxes the threshold to saturate the available database capacity.
The −0.3 exponent applies the large DWPC-matrix penalty.

Users can still evaluate DWPCs that are not stored in the database, using either the webapp or API.
These are calculated on-the-fly, delegating DWPC computation to the Neo4j database.
Unchecking "precomputed only" on the webapp shows all possible metapaths for two query nodes.
For some node pairs, the on-the-fly computation is quick (less than a second).
Othertimes, computing DWPCs for all metapaths might take more than a minute.

<!--
https://github.com/greenelab/hetmech/blob/cad458b29f508c66b4b14cdd641db6855426221b/explore/metapath-thresholds/metapath-thresholds.ipynb

https://github.com/greenelab/connectivity-search-backend/blob/af12f8cf2ad47d9a25ce8d1b7889390654eb3bb9/dj_hetmech_app/management/commands/populate_database.py#L139-L146
p-value threshold:

        p_threshold = 5 * row.n_pairs ** -0.3 / row.n_similar
-->

### Backend Database & API

We created a backend application using Python's Django web framework.
The source code is available in the [`connectivity-search-backend`](https://github.com/greenelab/connectivity-search-backend) repository.
The primary role of the backend is manage a relational database and provide an API for requesting data.

We define the database schema [using](https://github.com/greenelab/connectivity-search-backend/blob/af12f8cf2ad47d9a25ce8d1b7889390654eb3bb9/dj_hetmech_app/models.py "models.py for the dj_hetmech_app") Django's object-relational mapping framework (Figure @fig:database).
We [import](https://github.com/greenelab/connectivity-search-backend/blob/af12f8cf2ad47d9a25ce8d1b7889390654eb3bb9/dj_hetmech_app/management/commands/populate_database.py "dj_hetmech_app populate_database management command") the data into a PostgreSQL database.
Populating the database for all 2,205 metapaths up to length 3 was a prolonged operation, [taking](https://github.com/greenelab/connectivity-search-backend/pull/41#issuecomment-488054789) over 3 days.
The majority of the time is spent populating the `DegreeGroupedPermutation` (37,905,389 rows) and `PathCount` (174,986,768 rows) tables.

![
**Schema for the connectivity search backend relational database models.**
Each Django model is represented as a table,
whose rows list the model's field names and types.
Each model corresponds to a database table.
Arrows denote foreign key relationships.
The arrow labels indicate the foreign key field name followed by reverse relation names generated by Django (in parentheses).
](https://github.com/greenelab/connectivity-search-backend/raw/752b423a4b7b57575d66ce0b797b0a84c23267a6/media/models-schema.png){#fig:database width="100%"}

We host a public API instance at <https://search-api.het.io>.
Version 1 of the API exposes several endpoints that are used by the connectivity search frontend including queries for
node details (`/v1/node`),
node lookup (`/v1/nodes`),
metapath information (`/v1/metapaths`),
and path information (`/v1/paths`).
The endpoints return JSON payloads.
Producing results for these queries relies on internal calls to the PostgreSQL relational database as well as the pre-existing Hetionet v1.0 Neo4j graph database.
They were designed to power the hetnet connectivity search webapp,
but are also available for general research use.

Todo:

> The database only stores a single orientation of a metapath. For example, if GpPpGaD is stored between the given source and target node, DaGpPpG would not also be stored. Therefore, both orientations of a metapath are searched against the PathCount table.

### Webapp & Frontend

### Realtime open science

### Software & data availability


[@hetio-dag]: doi:10.1371/journal.pcbi.1004259
[@rephetio]: doi:10.7554/eLife.26726
[@vagelos-2017]: doi:10.6084/m9.figshare.5346577
[@xswap]: https://greenelab.github.io/xswap-manuscript/
